// @verdocs/event-tracker/bundles/verdocs-event-tracker.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core')) :
    typeof define === 'function' && define.amd ? define('@verdocs/event-tracker', ['exports', '@angular/core'], factory) :
    (global = global || self, factory((global.verdocs = global.verdocs || {}, global.verdocs['event-tracker'] = {}), global.ng.core));
}(this, function (exports, core) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var EventTrackerService = /** @class */ (function () {
        function EventTrackerService() {
        }
        /**
         * @param {?} event
         * @return {?}
         */
        EventTrackerService.prototype.createEvent = /**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            if (navigator && navigator.userAgent) {
                this.analytics = window['analytics'];
                if (event.action) {
                    this.analytics.track(event.action, {
                        category: event.category,
                        label: event.label,
                        value: event.value
                    });
                }
            }
        };
        EventTrackerService.decorators = [
            { type: core.Injectable }
        ];
        /** @nocollapse */
        EventTrackerService.ctorParameters = function () { return []; };
        return EventTrackerService;
    }());
    if (false) {
        /**
         * @type {?}
         * @private
         */
        EventTrackerService.prototype.analytics;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var EventTrackerDirective = /** @class */ (function () {
        function EventTrackerDirective() {
        }
        /**
         * @param {?} $event
         * @return {?}
         */
        EventTrackerDirective.prototype.onClick = /**
         * @param {?} $event
         * @return {?}
         */
        function ($event) {
            if (navigator && navigator.userAgent) {
                this.analytics = window['analytics'];
                if (this.event.action) {
                    this.analytics.track(this.event.action, {
                        category: this.event.category,
                        label: this.event.label,
                        value: this.event.value
                    });
                }
            }
        };
        EventTrackerDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[VerdocsEventTracker]'
                    },] }
        ];
        /** @nocollapse */
        EventTrackerDirective.ctorParameters = function () { return []; };
        EventTrackerDirective.propDecorators = {
            event: [{ type: core.Input, args: ['VerdocsEventTracker',] }],
            onClick: [{ type: core.HostListener, args: ['click', ['$event'],] }]
        };
        return EventTrackerDirective;
    }());
    if (false) {
        /**
         * @type {?}
         * @private
         */
        EventTrackerDirective.prototype.analytics;
        /** @type {?} */
        EventTrackerDirective.prototype.event;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var EventTrackerModule = /** @class */ (function () {
        function EventTrackerModule() {
        }
        EventTrackerModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [],
                        declarations: [
                            EventTrackerDirective
                        ],
                        providers: [
                            EventTrackerService
                        ],
                        exports: [
                            EventTrackerDirective
                        ]
                    },] }
        ];
        return EventTrackerModule;
    }());

    exports.EventTrackerDirective = EventTrackerDirective;
    exports.EventTrackerModule = EventTrackerModule;
    exports.EventTrackerService = EventTrackerService;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
