// @verdocs/essentials/esm5/lib/dialogs/plan-dialogs/plans-dialog.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { Router } from '@angular/router';
import { MatDialogRef } from '@angular/material/dialog';
var PlansDialog = /** @class */ (function () {
    function PlansDialog(router, dialog) {
        this.router = router;
        this.dialog = dialog;
        this.redirectUrl = null;
        this.type = null;
        this.dialogTitle = null;
        this.dialogMessage = null;
    }
    /**
     * @return {?}
     */
    PlansDialog.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        switch (this.type) {
            case 'free':
                this.dialogTitle = 'Upgrade to create another document';
                this.dialogMessage = 'We are glad to see you are drawing value from Verdocs. You have reached your limit of three documents per month with your free account. Please upgrade to an Essentials or Pro subscription to create more documents.';
                break;
            case 'essential':
                this.dialogTitle = 'Upgrade to access this feature';
                this.dialogMessage = 'This feature requires an essentials subscription or above. Please upgrade to access this feature.';
                break;
            case 'pro':
                this.dialogTitle = 'Upgrade to access this feature';
                this.dialogMessage = 'This feature requires a pro subscription. Please upgrade to access this feature.';
                break;
            default:
                this.dialog.close();
                break;
        }
    };
    /**
     * @return {?}
     */
    PlansDialog.prototype.goToBilling = /**
     * @return {?}
     */
    function () {
        if (window && !!this.redirectUrl && !!this.type) {
            window.location.href = (this.redirectUrl);
        }
        else {
            this.dialog.close();
        }
    };
    /**
     * @return {?}
     */
    PlansDialog.prototype.close = /**
     * @return {?}
     */
    function () {
        this.dialog.close();
    };
    PlansDialog.decorators = [
        { type: Component, args: [{
                    selector: 'plans-dialog',
                    template: "<div class=\"plans-dialog__container\">\n  <div class=\"plans-dialog__title\">{{dialogTitle}}</div>\n  <div class=\"plans-dialog__message\">\n      {{dialogMessage}}\n  </div>\n  <div class=\"plans-dialog__action\">\n    <button class=\"plans-dialog__button\" mat-button (click)=\"goToBilling()\">let's go</button>\n    <button class=\"plans-dialog__button\" mat-button (click)=\"close()\">not now</button>\n  </div>\n</div>",
                    styles: [".plans-dialog__title{padding:16px;font-size:24px;font-weight:500;color:rgba(0,0,0,.87)}.plans-dialog__message{padding:0 16px 16px;font-size:16px;line-height:24px;color:rgba(0,0,0,.54)}.plans-dialog__action{padding-top:8px;text-align:right}.plans-dialog__button{color:#37474f;text-transform:uppercase}"]
                }] }
    ];
    /** @nocollapse */
    PlansDialog.ctorParameters = function () { return [
        { type: Router },
        { type: MatDialogRef }
    ]; };
    return PlansDialog;
}());
export { PlansDialog };
if (false) {
    /** @type {?} */
    PlansDialog.prototype.redirectUrl;
    /** @type {?} */
    PlansDialog.prototype.type;
    /** @type {?} */
    PlansDialog.prototype.dialogTitle;
    /** @type {?} */
    PlansDialog.prototype.dialogMessage;
    /**
     * @type {?}
     * @private
     */
    PlansDialog.prototype.router;
    /**
     * @type {?}
     * @private
     */
    PlansDialog.prototype.dialog;
}
