// @verdocs/header/esm2015/lib/models/verdocs.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function ITemplate() { }
if (false) {
    /** @type {?|undefined} */
    ITemplate.prototype.template_documents;
    /** @type {?|undefined} */
    ITemplate.prototype.pages;
    /** @type {?|undefined} */
    ITemplate.prototype.roles;
    /** @type {?|undefined} */
    ITemplate.prototype.counter;
    /** @type {?} */
    ITemplate.prototype.name;
    /** @type {?|undefined} */
    ITemplate.prototype.id;
    /** @type {?|undefined} */
    ITemplate.prototype.profile_id;
    /** @type {?|undefined} */
    ITemplate.prototype.created_at;
    /** @type {?|undefined} */
    ITemplate.prototype.updated_at;
    /** @type {?|undefined} */
    ITemplate.prototype.token;
    /** @type {?|undefined} */
    ITemplate.prototype.tags;
    /** @type {?|undefined} */
    ITemplate.prototype.description;
    /** @type {?|undefined} */
    ITemplate.prototype.reminder_id;
    /** @type {?|undefined} */
    ITemplate.prototype.reminder;
    /** @type {?|undefined} */
    ITemplate.prototype.template_stars;
    /** @type {?|undefined} */
    ITemplate.prototype.organization_id;
    /** @type {?|undefined} */
    ITemplate.prototype.is_personal;
    /** @type {?|undefined} */
    ITemplate.prototype.is_public;
    /** @type {?|undefined} */
    ITemplate.prototype.sender;
    /** @type {?|undefined} */
    ITemplate.prototype.last_used;
}
/**
 * @record
 */
export function Envelope() { }
if (false) {
    /** @type {?|undefined} */
    Envelope.prototype.id;
    /** @type {?} */
    Envelope.prototype.template_id;
    /** @type {?|undefined} */
    Envelope.prototype.status;
    /** @type {?} */
    Envelope.prototype.user_id;
    /** @type {?} */
    Envelope.prototype.envelope_document_id;
    /** @type {?|undefined} */
    Envelope.prototype.created_at;
    /** @type {?|undefined} */
    Envelope.prototype.updated_at;
    /** @type {?|undefined} */
    Envelope.prototype.canceled_at;
    /** @type {?|undefined} */
    Envelope.prototype.recipients;
    /** @type {?|undefined} */
    Envelope.prototype.document;
    /** @type {?|undefined} */
    Envelope.prototype.component;
    /** @type {?|undefined} */
    Envelope.prototype.histories;
}
/** @enum {string} */
const TemplateSenderTypes = {
    CREATOR: 'creator',
    ORGANIZATION_MEMBER: 'organization_member',
    ORGANIZATION_MEMBER_AS_CREATOR: 'organization_member_as_creator',
    EVERYONE: 'everyone',
    EVERYONE_AS_CREATOR: 'everyone_as_creator' // Creator would be sender of envelope no matter who creates the envelope
    ,
};
export { TemplateSenderTypes };
/** @enum {string} */
const TemplatePermissions = {
    TEMPLATE_CREATOR_CREATE_PUBLIC: 'template:creator:create:public',
    TEMPLATE_CREATOR_CREATE_ORG: 'template:creator:create:org',
    TEMPLATE_CREATOR_CREATE_PERSONAL: 'template:creator:create:personal',
    TEMPLATE_CREATOR_DELETE: 'template:creator:delete',
    TEMPLATE_CREATOR_VISIBILITY: 'template:creator:visibility',
    TEMPLATE_MEMBER_READ: 'template:member:read',
    TEMPLATE_MEMBER_WRITE: 'template:member:write',
    TEMPLATE_MEMBER_DELETE: 'template:member:delete',
    TEMPLATE_MEMBER_VISIBILITY: 'template:member:visibility',
};
export { TemplatePermissions };
;
/** @enum {string} */
const TemplateActions = {
    CREATE_PERSONAL: 'create_personal',
    CREATE_ORG: 'create_org',
    CREATE_PUBLIC: 'create_public',
    READ: 'read',
    WRITE: 'write',
    DELETE: 'delete',
    CHANGE_VISIBILITY_PERSONAL: 'change_visibility_personal',
    CHANGE_VISIBILITY_ORG: 'change_visibility_org',
    CHANGE_VISIBILITY_PUBLIC: 'change_visibility_public',
};
export { TemplateActions };
/**
 * @record
 */
export function IRole() { }
if (false) {
    /** @type {?|undefined} */
    IRole.prototype.id;
    /** @type {?} */
    IRole.prototype.template_id;
    /** @type {?} */
    IRole.prototype.name;
    /** @type {?|undefined} */
    IRole.prototype.full_name;
    /** @type {?|undefined} */
    IRole.prototype.old_name;
    /** @type {?|undefined} */
    IRole.prototype.email;
    /** @type {?} */
    IRole.prototype.type;
    /** @type {?} */
    IRole.prototype.sequence;
    /** @type {?|undefined} */
    IRole.prototype.fields;
    /** @type {?|undefined} */
    IRole.prototype.message;
    /** @type {?|undefined} */
    IRole.prototype.rgba;
    /** @type {?} */
    IRole.prototype.delegator;
}
/**
 * @record
 */
export function IPage() { }
if (false) {
    /** @type {?} */
    IPage.prototype.template_id;
    /** @type {?} */
    IPage.prototype.document_id;
    /** @type {?|undefined} */
    IPage.prototype.document;
    /** @type {?} */
    IPage.prototype.sequence;
    /** @type {?} */
    IPage.prototype.page_number;
    /** @type {?|undefined} */
    IPage.prototype.thumbnail_url;
    /** @type {?|undefined} */
    IPage.prototype.thumbnail_safe;
    /** @type {?|undefined} */
    IPage.prototype.fields;
}
/**
 * @record
 */
export function IField() { }
if (false) {
    /** @type {?} */
    IField.prototype.label;
    /** @type {?} */
    IField.prototype.name;
    /** @type {?} */
    IField.prototype.role_name;
    /** @type {?} */
    IField.prototype.template_id;
    /** @type {?} */
    IField.prototype.type;
    /** @type {?} */
    IField.prototype.required;
    /** @type {?} */
    IField.prototype.setting;
    /** @type {?} */
    IField.prototype.page_sequence;
    /** @type {?|undefined} */
    IField.prototype.validator;
}
/**
 * @record
 */
export function ITemplateDocument() { }
if (false) {
    /** @type {?} */
    ITemplateDocument.prototype.url;
    /** @type {?} */
    ITemplateDocument.prototype.name;
    /** @type {?} */
    ITemplateDocument.prototype.page_numbers;
    /** @type {?|undefined} */
    ITemplateDocument.prototype.id;
    /** @type {?|undefined} */
    ITemplateDocument.prototype.updated_at;
    /** @type {?|undefined} */
    ITemplateDocument.prototype.created_at;
    /** @type {?} */
    ITemplateDocument.prototype.template_id;
    /** @type {?} */
    ITemplateDocument.prototype.thumbnail_url;
    /** @type {?|undefined} */
    ITemplateDocument.prototype.thumbnail_safe;
    /** @type {?} */
    ITemplateDocument.prototype.mime;
}
/**
 * @record
 */
export function IEnvelopeDocument() { }
if (false) {
    /** @type {?|undefined} */
    IEnvelopeDocument.prototype.id;
    /** @type {?} */
    IEnvelopeDocument.prototype.url;
    /** @type {?} */
    IEnvelopeDocument.prototype.name;
    /** @type {?|undefined} */
    IEnvelopeDocument.prototype.created_at;
    /** @type {?|undefined} */
    IEnvelopeDocument.prototype.updated_at;
    /** @type {?} */
    IEnvelopeDocument.prototype.page_numbers;
    /** @type {?} */
    IEnvelopeDocument.prototype.mime;
}
/**
 * @record
 */
export function IEnvelopeHistory() { }
if (false) {
    /** @type {?|undefined} */
    IEnvelopeHistory.prototype.id;
    /** @type {?|undefined} */
    IEnvelopeHistory.prototype.envelope_id;
    /** @type {?|undefined} */
    IEnvelopeHistory.prototype.role_name;
    /** @type {?} */
    IEnvelopeHistory.prototype.event;
    /** @type {?|undefined} */
    IEnvelopeHistory.prototype.event_detail;
    /** @type {?} */
    IEnvelopeHistory.prototype.created_at;
}
/**
 * @record
 */
export function IActivity() { }
if (false) {
    /** @type {?} */
    IActivity.prototype.date;
    /** @type {?} */
    IActivity.prototype.log;
    /** @type {?} */
    IActivity.prototype.icon;
}
/**
 * @record
 */
export function Recipient() { }
if (false) {
    /** @type {?} */
    Recipient.prototype.role_name;
    /** @type {?} */
    Recipient.prototype.envelope_id;
    /** @type {?} */
    Recipient.prototype.status;
    /** @type {?} */
    Recipient.prototype.email;
    /** @type {?} */
    Recipient.prototype.full_name;
    /** @type {?} */
    Recipient.prototype.sequence;
    /** @type {?} */
    Recipient.prototype.type;
    /** @type {?} */
    Recipient.prototype.user_id;
    /** @type {?} */
    Recipient.prototype.created_at;
    /** @type {?} */
    Recipient.prototype.updated_at;
    /** @type {?} */
    Recipient.prototype.token;
    /** @type {?} */
    Recipient.prototype.message;
    /** @type {?|undefined} */
    Recipient.prototype.signatureUrl;
    /** @type {?} */
    Recipient.prototype.signatures;
    /** @type {?|undefined} */
    Recipient.prototype.rgba;
}
/**
 * @record
 */
export function IEnvelopeSignature() { }
if (false) {
    /** @type {?} */
    IEnvelopeSignature.prototype.envelope_id;
    /** @type {?} */
    IEnvelopeSignature.prototype.signature_id;
    /** @type {?} */
    IEnvelopeSignature.prototype.role_name;
    /** @type {?|undefined} */
    IEnvelopeSignature.prototype.signed_at;
    /** @type {?|undefined} */
    IEnvelopeSignature.prototype.ip_address;
    /** @type {?|undefined} */
    IEnvelopeSignature.prototype.hash;
}
/**
 * @record
 */
export function EnvelopeComponent() { }
if (false) {
    /** @type {?|undefined} */
    EnvelopeComponent.prototype.progress;
    /** @type {?|undefined} */
    EnvelopeComponent.prototype.status_color;
    /** @type {?|undefined} */
    EnvelopeComponent.prototype.selected_format_time;
    /** @type {?|undefined} */
    EnvelopeComponent.prototype.format_time;
    /** @type {?|undefined} */
    EnvelopeComponent.prototype.selected;
    /** @type {?|undefined} */
    EnvelopeComponent.prototype.envelope__list__class;
}
/**
 * @record
 */
export function IEnvelopeSearchParams() { }
if (false) {
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.envelope_status;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.recipient_status;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.envelope_name;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.recipient_name;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.recipient_email;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.updated_at;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.canceled_at;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.created_at;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.text_field_value;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.is_owner;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.is_recipient;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.sort_by;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.ascending;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.row;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.page;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.template_id;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.recipient_claimed;
}
/**
 * @record
 */
export function ITimePeriod() { }
if (false) {
    /** @type {?} */
    ITimePeriod.prototype.start_time;
    /** @type {?} */
    ITimePeriod.prototype.end_time;
}
/** @enum {string} */
const SortOptions = {
    created_at: 'created_at',
    updated_at: 'updated_at',
    envelope_name: 'envelope_name',
    canceled_at: 'canceled_at',
    envelope_status: 'envelope_status',
};
export { SortOptions };
/**
 * @record
 */
export function ITemplateSearchParams() { }
if (false) {
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.name;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.sender;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.description;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.profile_id;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.organization_id;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.updated_at;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.created_at;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.last_used_at;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.is_personal;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.is_public;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.tags;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.document_name;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.sort_by;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.ascending;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.row;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.page;
}
/** @enum {string} */
const TemplateSortOptions = {
    CREATED_AT: 'created_at',
    UPDATED_AT: 'updated_at',
    NAME: 'name',
    LAST_USED_AT: 'last_used_at',
    COUNTER: 'counter',
    STAR_COUNTER: 'star_counter',
};
export { TemplateSortOptions };
