// @verdocs/essentials/esm5/lib/fields/conversion.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { v4 as uuid } from 'uuid';
import { getDefaultHeight, getDefaultWidth, getUIFieldStyling } from '../functions/conversion';
var ConversionService = /** @class */ (function () {
    function ConversionService() {
        this.ITextWidth = 612;
        this.ITextHeight = 792;
        this.XRatio = 1;
        this.YRatio = 1;
        this.a3Portrait = {
            x: 842,
            y: 1191
        };
        this.a3Landscape = {
            x: this.a3Portrait.y,
            y: this.a3Portrait.x
        };
        this.a4Portrait = {
            x: 595,
            y: 842
        };
        this.a4Landscape = {
            x: this.a4Portrait.y,
            y: this.a4Portrait.x
        };
        this.a5Portrait = {
            x: 420,
            y: 595
        };
        this.a5Landscape = {
            x: this.a5Portrait.y,
            y: this.a5Portrait.x
        };
        this.letterPortrait = {
            x: 612,
            y: 792
        };
        this.letterLandscape = {
            x: this.letterPortrait.y,
            y: this.letterPortrait.x
        };
        this.legalPortrait = {
            x: 612,
            y: 1008
        };
        this.legalLandscape = {
            x: this.legalPortrait.y,
            y: this.legalPortrait.x
        };
        this.portraitSizes = [
            this.a3Portrait, this.a4Portrait, this.a5Portrait, this.letterPortrait, this.legalPortrait
        ];
        this.landscapeSizes = [
            this.a3Landscape, this.a4Landscape, this.a5Landscape, this.letterLandscape, this.legalLandscape
        ];
        this.rFormWidth = 0;
        this.rFormHeight = 0;
        this.pageCache = [];
    }
    /**
     * @param {?} width
     * @param {?} height
     * @return {?}
     */
    ConversionService.prototype.setRValue = /**
     * @param {?} width
     * @param {?} height
     * @return {?}
     */
    function (width, height) {
        this.rFormWidth = width;
        this.rFormHeight = height;
    };
    /**
     * @param {?} pageCache
     * @return {?}
     */
    ConversionService.prototype.setPageCache = /**
     * @param {?} pageCache
     * @return {?}
     */
    function (pageCache) {
        this.pageCache = pageCache;
    };
    Object.defineProperty(ConversionService.prototype, "xRatio", {
        get: /**
         * @return {?}
         */
        function () {
            return this.XRatio;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} value
     * @return {?}
     */
    ConversionService.prototype.setXRatio = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.XRatio = value;
    };
    Object.defineProperty(ConversionService.prototype, "yRatio", {
        get: /**
         * @return {?}
         */
        function () {
            return this.YRatio;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} value
     * @return {?}
     */
    ConversionService.prototype.setYRatio = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.YRatio = value;
    };
    /**
     * @param {?} y
     * @param {?} fieldHeight
     * @param {?=} iTextHeight
     * @param {?=} yRatio
     * @return {?}
     */
    ConversionService.prototype.getRTop = /**
     * @param {?} y
     * @param {?} fieldHeight
     * @param {?=} iTextHeight
     * @param {?=} yRatio
     * @return {?}
     */
    function (y, fieldHeight, iTextHeight, yRatio) {
        if (!iTextHeight) {
            iTextHeight = this.ITextHeight;
        }
        if (!yRatio) {
            yRatio = this.yRatio;
        }
        return iTextHeight - ((y + fieldHeight) * yRatio);
    };
    /**
     * @param {?} y
     * @param {?=} ratio
     * @return {?}
     */
    ConversionService.prototype.getRValue = /**
     * @param {?} y
     * @param {?=} ratio
     * @return {?}
     */
    function (y, ratio) {
        if (!ratio) {
            ratio = this.yRatio;
        }
        return y * ratio;
    };
    /**
     * @param {?} x
     * @param {?=} ratio
     * @return {?}
     */
    ConversionService.prototype.getRLeft = /**
     * @param {?} x
     * @param {?=} ratio
     * @return {?}
     */
    function (x, ratio) {
        if (!ratio) {
            ratio = this.xRatio;
        }
        return x * ratio;
    };
    /**
     * @param {?} x
     * @param {?} xRatio
     * @return {?}
     */
    ConversionService.prototype.getITextLeft = /**
     * @param {?} x
     * @param {?} xRatio
     * @return {?}
     */
    function (x, xRatio) {
        if (!xRatio) {
            xRatio = this.xRatio;
        }
        return x / xRatio;
    };
    /**
     * @param {?} y
     * @param {?} fieldHeight
     * @param {?=} rFormHeight
     * @param {?=} yRatio
     * @return {?}
     */
    ConversionService.prototype.getITextBottomFromRTop = /**
     * @param {?} y
     * @param {?} fieldHeight
     * @param {?=} rFormHeight
     * @param {?=} yRatio
     * @return {?}
     */
    function (y, fieldHeight, rFormHeight, yRatio) {
        if (!rFormHeight) {
            rFormHeight = this.rFormHeight;
        }
        /** @type {?} */
        var uiHeight = fieldHeight;
        return (rFormHeight - y - uiHeight);
    };
    /**
     * @param {?} y
     * @return {?}
     */
    ConversionService.prototype.getITextBottom = /**
     * @param {?} y
     * @return {?}
     */
    function (y) {
        return y;
    };
    /**
     * @param {?} field
     * @return {?}
     */
    ConversionService.prototype.getDefaultHeight = /**
     * @param {?} field
     * @return {?}
     */
    function (field) {
        return getDefaultHeight(field);
    };
    /**
     * @param {?} field
     * @return {?}
     */
    ConversionService.prototype.getDefaultWidth = /**
     * @param {?} field
     * @return {?}
     */
    function (field) {
        return getDefaultWidth(field);
    };
    /**
     * @param {?} field
     * @param {?} pageHeight
     * @param {?=} xRatio
     * @param {?=} yRatio
     * @return {?}
     */
    ConversionService.prototype.getUIFieldStyling = /**
     * @param {?} field
     * @param {?} pageHeight
     * @param {?=} xRatio
     * @param {?=} yRatio
     * @return {?}
     */
    function (field, pageHeight, xRatio, yRatio) {
        if (!xRatio) {
            xRatio = this.xRatio;
        }
        if (!yRatio) {
            yRatio = this.yRatio;
        }
        return getUIFieldStyling(field);
    };
    /**
     * @param {?} field
     * @param {?} x
     * @param {?} y
     * @param {?} pageHeight
     * @return {?}
     */
    ConversionService.prototype.getUIFieldSetting = /**
     * @param {?} field
     * @param {?} x
     * @param {?} y
     * @param {?} pageHeight
     * @return {?}
     */
    function (field, x, y, pageHeight) {
        /** @type {?} */
        var setting;
        if (field.type === 'textbox' || field.type === 'date' || field.type === 'placeholder') {
            setting = {
                width: field.setting.width ? field.setting.width : 150,
                height: field.setting.height ? field.setting.height : 15,
                x: field.setting.x ? field.setting.x : x,
                y: this.getITextBottomFromRTop(y, field.setting.height ? field.setting.height : 15, pageHeight)
            };
            if (field.type === 'date') {
                setting.width = 64;
            }
        }
        else if (field.type === 'timestamp') {
            setting = {
                width: field.setting.width ? field.setting.width : 64,
                height: field.setting.height ? field.setting.height : 15,
                x: field.setting.x ? field.setting.x : x,
                y: this.getITextBottomFromRTop(y, field.setting.height ? field.setting.height : 15, pageHeight)
            };
        }
        else if (field.type === 'signature' || field.type === 'initial') {
            setting = {
                x: field.setting.x ? field.setting.x : x,
                y: this.getITextBottomFromRTop(y, 36, pageHeight)
            };
        }
        else if (field.type === 'checkbox_group') {
            setting = {
                options: [{
                        id: uuid(),
                        x: field.setting.x ? field.setting.x : x,
                        y: this.getITextBottomFromRTop(y, 13.5, pageHeight),
                        checked: false,
                        value: 'Option 1'
                    }]
            };
        }
        else if (field.type === 'radio_button_group') {
            setting = {
                options: [{
                        id: uuid(),
                        x: field.setting.x ? field.setting.x : x,
                        y: this.getITextBottomFromRTop(y, 13.5, pageHeight),
                        selected: false,
                        value: 'Option 1'
                    }]
            };
        }
        else if (field.type === 'dropdown') {
            setting = {
                x: field.setting.x ? field.setting.x : x,
                y: this.getITextBottomFromRTop(y, 15, pageHeight),
                width: field.setting.width ? field.setting.width : 64,
                height: field.setting.height ? field.setting.height : 15,
                value: null,
                placeholder: 'Choose',
                options: [{
                        id: uuid(),
                        value: 'Option 1'
                    }]
            };
        }
        else if (field.type === 'attachment') {
            setting = {
                x: field.setting.x ? field.setting.x : x,
                y: this.getITextBottomFromRTop(y, 24, pageHeight),
                label: ''
            };
        }
        else if (field.type === 'payment') {
            setting = {
                x: field.setting.x ? field.setting.x : x,
                y: this.getITextBottomFromRTop(y, 24, pageHeight),
                account_id: '',
                currency: 'usd',
                payment_fee_type: 'static',
                amount: 1,
                if_transfer_fee: false
            };
        }
        return setting;
    };
    /**
     * @param {?} newHeight
     * @param {?} newWidth
     * @return {?}
     */
    ConversionService.prototype.updateITextValues = /**
     * @param {?} newHeight
     * @param {?} newWidth
     * @return {?}
     */
    function (newHeight, newWidth) {
        this.ITextHeight = newHeight;
        this.ITextWidth = newWidth;
    };
    ConversionService.decorators = [
        { type: Injectable }
    ];
    return ConversionService;
}());
export { ConversionService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.ITextWidth;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.ITextHeight;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.XRatio;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.YRatio;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.a3Portrait;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.a3Landscape;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.a4Portrait;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.a4Landscape;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.a5Portrait;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.a5Landscape;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.letterPortrait;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.letterLandscape;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.legalPortrait;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.legalLandscape;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.portraitSizes;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.landscapeSizes;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.rFormWidth;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.rFormHeight;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.pageCache;
}
