// @verdocs/essentials/esm5/lib/fields/envelope-view.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ReplaySubject } from 'rxjs';
import { SignatureService } from './envelope-signature.service';
var EnvelopeViewService = /** @class */ (function () {
    function EnvelopeViewService(signatureService) {
        var _this = this;
        this.signatureService = signatureService;
        this.recipientSubject = new ReplaySubject();
        this.jumpCoordinateSubject = new ReplaySubject();
        this.fieldTypeSubject = new ReplaySubject();
        this.pdfUrlSubject = new ReplaySubject();
        this.pdfBlobSubject = new ReplaySubject();
        this.toggleNextSubject = new ReplaySubject();
        this.viewModeSubject = new ReplaySubject();
        this.agreedSubject = new ReplaySubject(1, 500);
        this.attachmentsSubject = new ReplaySubject();
        this.formValiditySubject = new ReplaySubject();
        this.requestFormValiditySubject = new ReplaySubject();
        this.signatureService._rName.subscribe((/**
         * @param {?} role
         * @return {?}
         */
        function (role) {
            _this.rName = role;
        }));
    }
    /**
     * @param {?} mode
     * @return {?}
     */
    EnvelopeViewService.prototype.setMode = /**
     * @param {?} mode
     * @return {?}
     */
    function (mode) {
        this.mode = mode;
        this.viewModeSubject.next(this.mode);
    };
    /**
     * @param {?} signatureService
     * @return {?}
     */
    EnvelopeViewService.prototype.setSignatureService = /**
     * @param {?} signatureService
     * @return {?}
     */
    function (signatureService) {
        this.signatureService = signatureService;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    EnvelopeViewService.prototype.requestFormValidity = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.requestFormValiditySubject.next(value);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    EnvelopeViewService.prototype.setFormValidityData = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.formValiditySubject.next(value);
    };
    /**
     * @param {?} type
     * @return {?}
     */
    EnvelopeViewService.prototype.ifBelongsToPreparer = /**
     * @param {?} type
     * @return {?}
     */
    function (type) {
        return type === 'textbox' || type === 'date' || type === 'checkbox';
    };
    /**
     * @param {?} field
     * @return {?}
     */
    EnvelopeViewService.prototype.ifBelongsToCurrentRecipient = /**
     * @param {?} field
     * @return {?}
     */
    function (field) {
        if (this.mode === 'prepareview') {
            return this.ifBelongsToPreparer(field.controlType);
        }
        else {
            return field.recipientRole === this.rName;
        }
    };
    EnvelopeViewService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    EnvelopeViewService.ctorParameters = function () { return [
        { type: SignatureService }
    ]; };
    return EnvelopeViewService;
}());
export { EnvelopeViewService };
if (false) {
    /** @type {?} */
    EnvelopeViewService.prototype.recipientSubject;
    /** @type {?} */
    EnvelopeViewService.prototype.jumpCoordinateSubject;
    /** @type {?} */
    EnvelopeViewService.prototype.fieldTypeSubject;
    /** @type {?} */
    EnvelopeViewService.prototype.pdfUrlSubject;
    /** @type {?} */
    EnvelopeViewService.prototype.pdfBlobSubject;
    /** @type {?} */
    EnvelopeViewService.prototype.toggleNextSubject;
    /** @type {?} */
    EnvelopeViewService.prototype.viewModeSubject;
    /** @type {?} */
    EnvelopeViewService.prototype.agreedSubject;
    /** @type {?} */
    EnvelopeViewService.prototype.attachmentsSubject;
    /** @type {?} */
    EnvelopeViewService.prototype.formValiditySubject;
    /** @type {?} */
    EnvelopeViewService.prototype.requestFormValiditySubject;
    /** @type {?} */
    EnvelopeViewService.prototype.mode;
    /**
     * @type {?}
     * @private
     */
    EnvelopeViewService.prototype.rName;
    /**
     * @type {?}
     * @private
     */
    EnvelopeViewService.prototype.signatureService;
}
