// @verdocs/sdk/esm2015/public-api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * Public API Surface of angular-sdk
 */
export {} from './lib/definitions/componentData';
export { TemplateSenderTypes, TemplatePermissions, TemplateActions } from './lib/definitions/template.enums';
export {} from './lib/models/document.model';
export {} from './lib/models/document_fields.model';
export {} from './lib/models/document_histories.model';
export { DocumentSortOptions } from './lib/models/document_search.model';
export {} from './lib/models/envelope_document.model';
export { FieldData } from './lib/models/field-data.model';
export {} from './lib/models/field-role.model';
export {} from './lib/models/field.model';
export {} from './lib/models/options.model';
export {} from './lib/models/page.model';
export {} from './lib/models/recipient.model';
export {} from './lib/models/reminder.model';
export {} from './lib/models/role.model';
export {} from './lib/models/template.model';
export {} from './lib/models/template_document.model';
export { SortOptions } from './lib/models/template_search.model';
export {} from './lib/models/time_period.model';
export { regParse } from './lib/functions/regex';
export { getRGB, getRGBA, nameToRGBA } from './lib/functions/rgb';
export { getCountryByCode, isFrenchGuiana, isGuadeloupe, isMartinique, isMayotte, getPlusOneCountry, isCanada, isAmericanSamoa, isDominicanRepublic, isPuertoRico, getMatchingCountry, simpleE164Validator, countries } from './lib/functions/country-code';
export { ValidatorService } from './lib/validators/validator.service';
export { PhoneValidator } from './lib/validators/phone.validator';
export { roleNameValidator } from './lib/validators/role-name.validator';
export { TagDuplicateValidator, TagValidator, TagErrorStateMatcher } from './lib/validators/tag.validator';
export { AuthGuardService } from './lib/guards/auth-guard.service';
export { DocumentsService } from './lib/documents/document.service';
export { EnvelopRemindersService } from './lib/documents/envelope-reminders.service';
export { RecipientService } from './lib/recipients/recipients.service';
export { TemplatesGuardService } from './lib/templates/templates.guard';
export { TemplatesService } from './lib/templates/templates.service';
export { TemplatesRemindersService } from './lib/templates/template-reminders.service';
export { BuilderService } from './lib/templates/builder.service';
export { VSDKConfigToken, AngularApiModule } from './lib/angular-api.module';
