// @verdocs/sdk/esm2015/lib/models/document_search.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function IEnvelopeSearchParams() { }
if (false) {
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.envelope_status;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.recipient_status;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.envelope_name;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.recipient_name;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.recipient_email;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.updated_at;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.canceled_at;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.created_at;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.text_field_value;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.is_owner;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.is_recipient;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.sort_by;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.ascending;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.row;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.page;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.template_id;
    /** @type {?|undefined} */
    IEnvelopeSearchParams.prototype.recipient_claimed;
}
/**
 * @record
 */
export function IEnvelopeSearchResult() { }
if (false) {
    /** @type {?} */
    IEnvelopeSearchResult.prototype.page;
    /** @type {?} */
    IEnvelopeSearchResult.prototype.row;
    /** @type {?} */
    IEnvelopeSearchResult.prototype.result;
    /** @type {?} */
    IEnvelopeSearchResult.prototype.total;
}
/**
 * @record
 */
export function IEnvelopeFilterParams() { }
if (false) {
    /** @type {?} */
    IEnvelopeFilterParams.prototype.type;
    /** @type {?|undefined} */
    IEnvelopeFilterParams.prototype.ascending;
    /** @type {?|undefined} */
    IEnvelopeFilterParams.prototype.row;
    /** @type {?|undefined} */
    IEnvelopeFilterParams.prototype.page;
    /** @type {?|undefined} */
    IEnvelopeFilterParams.prototype.filterTime;
    /** @type {?|undefined} */
    IEnvelopeFilterParams.prototype.sortBy;
    /** @type {?|undefined} */
    IEnvelopeFilterParams.prototype.customerFilters;
}
/** @enum {string} */
const DocumentSortOptions = {
    created_at: 'created_at',
    updated_at: 'updated_at',
    envelope_name: 'envelope_name',
    canceled_at: 'canceled_at',
    envelope_status: 'envelope_status',
};
export { DocumentSortOptions };
