// @verdocs/sdk/esm2015/lib/models/field-data.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @template T
 */
export class FieldData {
    /**
     * @param {?=} options
     */
    constructor(options = {}) {
        this.value = options.value;
        this.pageNum = options.pageNum;
        this.id = options.id;
        this.fName = options.fName;
        this.required = options.required;
        this.order = options.order;
        this.vName = options.vName;
        this.error = options.error || '';
        this.type = options.type;
    }
}
if (false) {
    /** @type {?} */
    FieldData.prototype.value;
    /** @type {?} */
    FieldData.prototype.pageNum;
    /** @type {?} */
    FieldData.prototype.id;
    /** @type {?} */
    FieldData.prototype.fName;
    /** @type {?} */
    FieldData.prototype.required;
    /** @type {?} */
    FieldData.prototype.order;
    /** @type {?} */
    FieldData.prototype.vName;
    /** @type {?} */
    FieldData.prototype.error;
    /** @type {?} */
    FieldData.prototype.type;
}
