// @verdocs/essentials/esm2015/public-api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * Public API Surface of angular-sdk
 */
export { DynamicField } from './lib/models/dynamic-field.model';
export { setRatio, constructPages, constructGroupFields, generateInitialStyle, generateStyle, generateGroupLabel, getRoleColor, convertToUIRatio } from './lib/functions/builder-fields';
export { isInView, topOfPageIsInView, middleOfPageIsInView, bottomOfPageIsInView } from './lib/functions/builder-page';
export { getDefaultHeight, getDefaultWidth, getRTop, getRLeft, getRValue, getUIFieldStyling, blobToBase64 } from './lib/functions/conversion';
export { timePeriod } from './lib/functions/filter';
export { regParse, dataURLtoBlob, printPdfUrl } from './lib/functions/utils';
export { buildFields, getFieldsMap, updateElementStyles, getInputStyle, generateCheckboxLabel, getLetterSpacing, rescale, prepareFieldsForSigner } from './lib/functions/viewer-fields';
export { SnackbarService } from './lib/snackbar/snackbar.service';
export { AccountService } from './lib/account/account.service';
export { BrowserToiTextService } from './lib/builder/browser-to-iText.service';
export { BuilderDataService } from './lib/builder/builder-data.service';
export { BuilderService } from './lib/builder/builder-settings.service';
export { ConversionService } from './lib/fields/conversion.service';
export { SignatureService } from './lib/fields/envelope-signature.service';
export { EnvelopeViewService } from './lib/fields/envelope-view.service';
export { PageService } from './lib/fields/page.service';
export { FourOhOneDialog } from './lib/dialogs/error-dialogs/four-oh-one.dialog';
export { PlansDialog } from './lib/dialogs/plan-dialogs/plans-dialog.component';
export { ErrorDialogService } from './lib/error-dialogs/error-dialog.service';
export { GuardService } from './lib/guards/guard.service';
export { ModalService } from './lib/modal/modal.service';
export { EssentialsConfigToken, EssentialsModule } from './lib/essentials.module';
