// @verdocs/sdk/esm5/lib/guards/auth-guard.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Inject } from '@angular/core';
import { DOCUMENT } from "@angular/common";
import { VerdocsAuthService, VerdocsAuthGuardService, VerdocsStateService } from '@verdocs/tokens';
var AuthGuardService = /** @class */ (function () {
    function AuthGuardService(vTokenAuth, vTokenAuthGuard, vTokenStateService, document) {
        this.vTokenAuth = vTokenAuth;
        this.vTokenAuthGuard = vTokenAuthGuard;
        this.vTokenStateService = vTokenStateService;
        this.document = document;
    }
    /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    AuthGuardService.prototype.canActivate = /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    function (route, state) {
        /** @type {?} */
        var signerToken = this.vTokenStateService.getOtherCookieObject('signer_token');
        /** @type {?} */
        var currentPath = route && route.routeConfig && route.routeConfig.path ? route.routeConfig.path : route.path;
        /** @type {?} */
        var matchesPath = currentPath ? currentPath.match(/(:id\/role\/:role)/g) || currentPath.includes('envelope') : false;
        if (route.routeConfig && !currentPath) {
            return true;
        }
        else if (signerToken && !!matchesPath) {
            return true;
        }
        else {
            return this.checkLogin(route, state.url);
        }
    };
    /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    AuthGuardService.prototype.canActivateChild = /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    function (route, state) {
        return this.canActivate(route, state);
    };
    /**
     * @param {?} route
     * @param {?} url
     * @return {?}
     */
    AuthGuardService.prototype.checkLogin = /**
     * @param {?} route
     * @param {?} url
     * @return {?}
     */
    function (route, url) {
        if (this.vTokenAuth.isAuthenticated() || this.isGuestLink(route)) {
            return true;
        }
        this.vTokenAuthGuard.signOut();
        return false;
    };
    /**
     * @param {?} route
     * @return {?}
     */
    AuthGuardService.prototype.canLoad = /**
     * @param {?} route
     * @return {?}
     */
    function (route) {
        /** @type {?} */
        var url = "/" + route.path;
        return this.checkLogin(route, url);
    };
    /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    AuthGuardService.prototype.canLoadBuilder = /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    function (route, state) {
    };
    /**
     * @param {?} route
     * @return {?}
     */
    AuthGuardService.prototype.isGuestLink = /**
     * @param {?} route
     * @return {?}
     */
    function (route) {
        /** @type {?} */
        var signer_token = this.vTokenStateService.getOtherCookie('signer_token');
        /** @type {?} */
        var currentPath = route && route.routeConfig && route.routeConfig.path ? route.routeConfig.path : route.path;
        /** @type {?} */
        var matchesPath = currentPath ? currentPath.match(/(:id\/role\/:role)/g) || currentPath.includes('envelope') : false;
        if (route && !!matchesPath && !!signer_token) {
            if (signer_token) {
                return true;
            }
            this.vTokenStateService.setTimer(true);
        }
        return route._routeConfig && route._routeConfig.path.match(/(:id\/role\/:role)/g) && !!signer_token;
    };
    AuthGuardService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AuthGuardService.ctorParameters = function () { return [
        { type: VerdocsAuthService },
        { type: VerdocsAuthGuardService },
        { type: VerdocsStateService },
        { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
    ]; };
    return AuthGuardService;
}());
export { AuthGuardService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AuthGuardService.prototype.vTokenAuth;
    /**
     * @type {?}
     * @private
     */
    AuthGuardService.prototype.vTokenAuthGuard;
    /**
     * @type {?}
     * @private
     */
    AuthGuardService.prototype.vTokenStateService;
    /**
     * @type {?}
     * @private
     */
    AuthGuardService.prototype.document;
}
