// @verdocs/header/esm5/public-api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * Public API Surface of header
 */
export {} from './lib/models/notification.interface';
export { TemplateSenderTypes, TemplatePermissions, TemplateActions, SortOptions, TemplateSortOptions } from './lib/models/verdocs.model';
export { connectionConfigToken, HeaderModule } from './lib/header.module';
export { VerdocsHeaderService } from './lib/services/header.service';
export { TemplatesGuardService } from './lib/services/templates.guard';
export { HeaderToolsComponent } from './lib/components/header-tools.component';
export { HeaderComponent } from './lib/header.component';
export { RecipientEditorComponent } from './lib/template/recipients/recipients-editor.component';
export { TemplateSourceComponent } from './lib/template/source-file/source-file.component';
export { VisibilityEditorComponent } from './lib/template/visibility/visibility-editor.component';
export { FourOhOneDialog } from './lib/error-dialogs/four-oh-one.dialog';
