// @verdocs/header/esm5/lib/header.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ViewChild, ElementRef, Compiler, Input, ChangeDetectorRef, Injector, Inject, PLATFORM_ID } from '@angular/core';
import { Router, NavigationEnd, NavigationStart } from '@angular/router';
import { isPlatformBrowser } from '@angular/common';
import { HttpResponse } from '@angular/common/http';
import { VerdocsStateService, VerdocsAuthService, VerdocsTokenObjectService } from '@verdocs/tokens';
import { Subscription, fromEvent } from 'rxjs';
import { debounceTime, distinctUntilChanged } from 'rxjs/operators';
import { findIndex } from 'lodash';
import { VerdocsHeaderService } from './services/header.service';
import { TemplatesGuardService } from './services/templates.guard';
import { TemplateActions } from './models/verdocs.model';
import { connectionConfigToken } from './header.module';
var HeaderComponent = /** @class */ (function () {
    function HeaderComponent(verdocsHeaderService, vTokenStateService, vTokenObjectService, vAuthService, templateGuard, router, cdr, compiler, injector, platform) {
        this.verdocsHeaderService = verdocsHeaderService;
        this.vTokenStateService = vTokenStateService;
        this.vTokenObjectService = vTokenObjectService;
        this.vAuthService = vAuthService;
        this.templateGuard = templateGuard;
        this.router = router;
        this.cdr = cdr;
        this.compiler = compiler;
        this.injector = injector;
        this.platform = platform;
        this.pageLoading = true;
        this.isAuthenticated = false;
        this.headerTitle = 'Envelope';
        this.mode = '';
        this.searchActive = false;
        this.templateName = '';
        this.newTemplateName = null;
        this.saveStatus = '';
        this.notifications = [];
        this.loading = false;
        this.csvBlob = null;
        this.templateDetailsExpanded = true;
        this.template = null;
        this.isSearching = false;
        this.currentProfile = null;
        this.inputFocused = false;
        this.isLiveView = false;
        this.app_version = '';
        this.currentAt = '';
        this._canStar = null;
        this._canCreate = null;
        this._canEdit = null;
        this._userStarred = null;
        this.metaDataSubscription = new Subscription();
        this.titleSubscription = new Subscription();
        this.titleInputSubscription = new Subscription();
        this.titleKeyDownSubscription = new Subscription();
        this.templateNameSubscription = new Subscription();
        this.templateSubscription = new Subscription();
        this.saveStatusSubscription = new Subscription();
        this.profileLoadingSubscription = new Subscription();
        this.userDataSubscription = new Subscription();
        this.modeSubscription = new Subscription();
        this.previewInputSubscription = new Subscription();
        this.generalInputSubscription = new Subscription();
        this.navigationEventSubscription = new Subscription();
        this.csvBlobSubscription = new Subscription();
        this.connections = this.injector.get(connectionConfigToken);
        this.myAccountUrl = this.connections.rAccount_frontend_url + '/rAccount';
        this.termsUrl = this.connections.rSecure_frontend_url + '/terms';
        this.privacyUrl = this.connections.rSecure_frontend_url + '/privacy';
        this.organizationImgSrc = this.connections.rAccount_frontend_url + "/assets/organization/select-profile.png";
        this.verdocs_url = this.connections.rForm_frontend_url;
        this.manageLink = this.verdocs_url + "/manage";
        this.templatesLink = this.verdocs_url + "/templates";
        this.rSecureFrontEnd = this.connections.rSecure_frontend_url;
        this.rFormFrontEnd = this.connections.rForm_frontend_url;
        this.currentOrigin = this.connections.origin;
        this.currentAt = this.currentOrigin;
    }
    /**
     * @return {?}
     */
    HeaderComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.templateSubscription = this.verdocsHeaderService.templateSubject.subscribe((/**
         * @param {?} template
         * @return {?}
         */
        function (template) {
            _this.template = template;
            _this.updateStar();
        }));
        if (isPlatformBrowser(this.platform)) {
            if (window && window.location || this.connections.app_version) {
                this.origin = window.location.protocol + "//" + window.location.hostname + (window.location.port ? ':' + window.location.port : '');
            }
            if (this.origin && this.origin.includes('/stage.') || this.origin.includes('/localhost')) {
                this.app_version = this.connections.app_version;
            }
            else {
                this.app_version = '';
            }
        }
        this.isAuthenticated = this.vAuthService.isAuthenticated();
        this.emailVerified = this.vTokenStateService.getEmailVerification();
        this.navigationEventSubscription = this.router.events.subscribe((/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            if (event instanceof NavigationEnd) {
                _this.isAuthenticated = _this.vAuthService.isAuthenticated();
                if (event && event.url && ((event.url.includes('document/') && !event.url.includes('envelopes/document')) || event.url.includes('view/live/'))) {
                    _this.isAtOriginalOrigin = _this.verdocsHeaderService.getAtOrigin(_this.currentOrigin);
                    _this.currentAt = _this.currentOrigin;
                    _this.mode = 'preview';
                    if (event.url.includes('view/live/')) {
                        _this.isLiveView = true;
                        _this.templateDetailsExpanded = false;
                    }
                    else {
                        _this.isLiveView = false;
                    }
                }
                else {
                    _this.mode = 'not-preview';
                    _this.isLiveView = false;
                }
                if (_this.mode !== 'preview') {
                    _this.image_url = _this.connections.rForm_frontend_url + "/assets/white-logo.svg";
                }
                else {
                    _this.image_url = _this.connections.rForm_frontend_url + "/assets/break-black-logo.svg";
                }
                _this.getProfile();
                _this.updateStar();
                _this.checkCanStar();
                _this._canEdit = undefined;
                _this._canCreate = undefined;
                _this.pageLoading = false;
            }
            else if (event instanceof NavigationStart) {
                _this.pageLoading = true;
            }
        }));
        this.titleSubscription = this.verdocsHeaderService.titleSubject.subscribe((/**
         * @param {?} title
         * @return {?}
         */
        function (title) {
            if (title) {
                _this.title = title;
            }
            else {
                _this.title = null;
            }
        }));
        this.csvBlobSubscription = this.verdocsHeaderService.csvBlobSubject.subscribe((/**
         * @param {?} blob
         * @return {?}
         */
        function (blob) {
            if (blob) {
                _this.csvBlob = blob;
            }
            else {
                _this.csvBlob = null;
            }
        }));
        this.saveStatusSubscription = this.verdocsHeaderService.saveStatusSubject.subscribe((/**
         * @param {?} status
         * @return {?}
         */
        function (status) {
            _this.saveStatus = status;
        }));
        this.profileLoadingSubscription = this.verdocsHeaderService.profileLoadingSubject.subscribe((/**
         * @param {?} status
         * @return {?}
         */
        function (status) {
            _this.loading = status;
        }));
    };
    /**
     * @return {?}
     */
    HeaderComponent.prototype.ngAfterViewChecked = /**
     * @return {?}
     */
    function () {
        this.cdr.detectChanges();
    };
    /**
     * @return {?}
     */
    HeaderComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.metaDataSubscription.unsubscribe();
        this.titleSubscription.unsubscribe();
        this.titleInputSubscription.unsubscribe();
        this.titleKeyDownSubscription.unsubscribe();
        this.templateNameSubscription.unsubscribe();
        this.templateSubscription.unsubscribe();
        this.saveStatusSubscription.unsubscribe();
        this.profileLoadingSubscription.unsubscribe();
        this.userDataSubscription.unsubscribe();
        this.modeSubscription.unsubscribe();
        if (!!this.generalInputSubscription) {
            this.generalInputSubscription.unsubscribe();
        }
        if (!!this.previewInputSubscription) {
            this.previewInputSubscription.unsubscribe();
        }
        this.navigationEventSubscription.unsubscribe();
        this.csvBlobSubscription.unsubscribe();
        this.compiler.clearCache();
    };
    /**
     * @return {?}
     */
    HeaderComponent.prototype.init = /**
     * @return {?}
     */
    function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(this.vAuthService.isAuthenticated() && this.mode !== 'view')) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.verdocsHeaderService.initNotificationService()];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2: return [2 /*return*/];
                }
            });
        });
    };
    Object.defineProperty(HeaderComponent.prototype, "home_url", {
        get: /**
         * @return {?}
         */
        function () {
            return this.vTokenStateService.getIDToken() ? this.connections.rForm_frontend_url + '/dashboard' : this.connections.rForm_frontend_url;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} event
     * @return {?}
     */
    HeaderComponent.prototype.previewInputEvent = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var _this = this;
        clearTimeout(this.inputTimer);
        this.inputTimer = setTimeout((/**
         * @return {?}
         */
        function () {
            _this.getSearchResults(_this.previewInput.nativeElement.value);
            clearTimeout(_this.inputTimer);
        }), 500);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    HeaderComponent.prototype.generalInputEvent = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var _this = this;
        clearTimeout(this.inputTimer);
        this.inputTimer = setTimeout((/**
         * @return {?}
         */
        function () {
            _this.getSearchResults(_this.generalInput.nativeElement.value);
            clearTimeout(_this.inputTimer);
        }), 500);
    };
    /**
     * @return {?}
     */
    HeaderComponent.prototype.prepareSearchInput = /**
     * @return {?}
     */
    function () {
        if (this.mode === 'preview') {
            this.listenForPreviewInput();
        }
        else {
            this.listenForGeneralInput();
        }
    };
    /**
     * @return {?}
     */
    HeaderComponent.prototype.listenForPreviewInput = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.generalInputSubscription) {
            this.generalInputSubscription.unsubscribe();
        }
        if (this.previewInput) {
            if (!this.previewInputSubscription) {
                this.previewInputSubscription = fromEvent(this.previewInput.nativeElement, 'keyup')
                    .pipe(debounceTime(500), distinctUntilChanged()).subscribe((/**
                 * @param {?} event
                 * @return {?}
                 */
                function (event) {
                    /** @type {?} */
                    var keyName = event['code'] || event['key'];
                    if (keyName) {
                        if (keyName !== 'Enter' && keyName !== 'ArrowDown' && keyName !== 'ArrowUp') {
                            _this.getSearchResults(_this.previewInput.nativeElement.value);
                        }
                    }
                    else {
                        _this.getSearchResults(_this.previewInput.nativeElement.value);
                    }
                }));
            }
        }
    };
    /**
     * @return {?}
     */
    HeaderComponent.prototype.listenForGeneralInput = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.previewInputSubscription) {
            this.previewInputSubscription.unsubscribe();
        }
        if (this.generalInput) {
            if (!this.generalInputSubscription) {
                this.generalInputSubscription = fromEvent(this.generalInput.nativeElement, 'keyup')
                    .pipe(debounceTime(500), distinctUntilChanged()).subscribe((/**
                 * @param {?} event
                 * @return {?}
                 */
                function (event) {
                    /** @type {?} */
                    var keyName = event['code'] || event['key'];
                    if (keyName) {
                        if (keyName !== 'Enter' && keyName !== 'ArrowDown' && keyName !== 'ArrowUp') {
                            _this.getSearchResults(_this.generalInput.nativeElement.value);
                        }
                    }
                    else {
                        _this.getSearchResults(_this.generalInput.nativeElement.value);
                    }
                }));
            }
        }
    };
    Object.defineProperty(HeaderComponent.prototype, "canStar", {
        get: /**
         * @return {?}
         */
        function () {
            return this._canStar;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HeaderComponent.prototype, "userStarred", {
        get: /**
         * @return {?}
         */
        function () {
            return this._userStarred;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HeaderComponent.prototype, "canEdit", {
        get: /**
         * @return {?}
         */
        function () {
            if (typeof (this._canEdit) !== 'boolean' && this.templateGuard && this.isAuthenticated && this.template) {
                /** @type {?} */
                var response = this.templateGuard.canPerformAction(TemplateActions.WRITE, this.template);
                this._canEdit = response['canPerform'];
            }
            if (!this.isAuthenticated && this.template) {
                this._canEdit = false;
            }
            return this._canEdit;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HeaderComponent.prototype, "canCreate", {
        get: /**
         * @return {?}
         */
        function () {
            var _this = this;
            if (this.isAuthenticated) {
                if (typeof (this._canCreate) !== 'boolean') {
                    /** @type {?} */
                    var permissions = this.vTokenObjectService.getVerdocsPermissions();
                    this._canCreate = false;
                    if (permissions && permissions.length > 0) {
                        permissions.forEach((/**
                         * @param {?} permission
                         * @return {?}
                         */
                        function (permission) {
                            if (permission.includes('creator:create')) {
                                return _this._canCreate = true;
                            }
                        }));
                    }
                }
                return this._canCreate;
            }
            else {
                return false;
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    HeaderComponent.prototype.getProfile = /**
     * @return {?}
     */
    function () {
        if (this.isAuthenticated) {
            this.currentProfile = this.vTokenObjectService.getProfile();
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    HeaderComponent.prototype.selected = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if ((this.generalInput && this.generalInput.nativeElement.value) || (this.previewInput && this.previewInput.nativeElement.value)) {
            event.option._element.nativeElement.click();
        }
    };
    /**
     * @return {?}
     */
    HeaderComponent.prototype.checkCanStar = /**
     * @return {?}
     */
    function () {
        this._canStar = this.isAuthenticated;
    };
    /**
     * @param {?} search_string
     * @return {?}
     */
    HeaderComponent.prototype.getSearchResults = /**
     * @param {?} search_string
     * @return {?}
     */
    function (search_string) {
        var _this = this;
        this.isSearching = true;
        if (this.isAuthenticated) {
            this.verdocsHeaderService.searchEnvelopes({ envelope_name: search_string, row: 3 }).subscribe((/**
             * @param {?} result
             * @return {?}
             */
            function (result) {
                if (result instanceof HttpResponse) {
                    _this.envelopeResult = (/** @type {?} */ (result.body['result']));
                    _this.verdocsHeaderService.searchTemplates({ name: search_string, row: 3 }).subscribe((/**
                     * @param {?} result
                     * @return {?}
                     */
                    function (result) {
                        if (result instanceof HttpResponse) {
                            _this.templateResult = (/** @type {?} */ (result.body['result']));
                            _this.isSearching = false;
                        }
                    }));
                }
            }));
        }
        else {
            this.verdocsHeaderService.searchTemplates({ name: search_string, row: 3 }).subscribe((/**
             * @param {?} result
             * @return {?}
             */
            function (result) {
                if (result instanceof HttpResponse) {
                    _this.templateResult = (/** @type {?} */ (result.body['result']));
                    _this.isSearching = false;
                }
            }));
        }
    };
    /**
     * @param {?} index
     * @param {?} item
     * @return {?}
     */
    HeaderComponent.prototype.recipientTracking = /**
     * @param {?} index
     * @param {?} item
     * @return {?}
     */
    function (index, item) {
        return index;
    };
    /**
     * @return {?}
     */
    HeaderComponent.prototype.toggleMenu = /**
     * @return {?}
     */
    function () {
        this.verdocsHeaderService.toggleSidenav(!this.sidenavOpen);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    HeaderComponent.prototype.setInputFocused = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.inputFocused = value;
    };
    /**
     * @param {?} obj
     * @param {?} key
     * @return {?}
     */
    HeaderComponent.prototype.hasProp = /**
     * @param {?} obj
     * @param {?} key
     * @return {?}
     */
    function (obj, key) {
        return obj.hasOwnProperty(key);
    };
    /**
     * @param {?} name
     * @return {?}
     */
    HeaderComponent.prototype.roleColor = /**
     * @param {?} name
     * @return {?}
     */
    function (name) {
        return this.verdocsHeaderService.getRoleColor(name);
    };
    /**
     * @return {?}
     */
    HeaderComponent.prototype.toggleSidenav = /**
     * @return {?}
     */
    function () {
        this.verdocsHeaderService.toggleBuilderSideNav();
    };
    /**
     * @return {?}
     */
    HeaderComponent.prototype.toggleTemplateDetails = /**
     * @return {?}
     */
    function () {
        this.templateDetailsExpanded = !this.templateDetailsExpanded;
        if (this.templateDetailsExpanded) {
            this.verdocsHeaderService.templateDetailToggleSubject.next(true);
        }
    };
    /**
     * @return {?}
     */
    HeaderComponent.prototype.reset = /**
     * @return {?}
     */
    function () {
        this.templateName = 'Unititled';
        // this.verdocsHeaderService.templateNameSubject.next('Untitled');
    };
    /**
     * @param {?} roles
     * @return {?}
     */
    HeaderComponent.prototype.getRoleNames = /**
     * @param {?} roles
     * @return {?}
     */
    function (roles) {
        if (roles && roles.length > 0) {
            /** @type {?} */
            var last_index = roles.length - 1;
            /** @type {?} */
            var names_string = '';
            for (var x = 0; x < roles.length; x++) {
                if (x < last_index) {
                    names_string += (/** @type {?} */ (roles[x]['full_name'] + ', '));
                }
                else {
                    names_string += (/** @type {?} */ (roles[x]['full_name']));
                }
            }
            return names_string;
        }
    };
    /**
     * @param {?} template
     * @return {?}
     */
    HeaderComponent.prototype.getTemplateStats = /**
     * @param {?} template
     * @return {?}
     */
    function (template) {
        return this.getTemplatePages(template) + "  " + this.getTemplateSequence(template) + "  " + this.getTemplateRecipients(template);
    };
    /**
     * @param {?} template
     * @return {?}
     */
    HeaderComponent.prototype.getTemplatePages = /**
     * @param {?} template
     * @return {?}
     */
    function (template) {
        if (template && template.pages) {
            return "" + (template.pages.length >= 2 ? template.pages.length + ' Pages' : '1 Page');
        }
        else {
            return '0 Page';
        }
    };
    /**
     * @param {?} template
     * @return {?}
     */
    HeaderComponent.prototype.getTemplateSequence = /**
     * @param {?} template
     * @return {?}
     */
    function (template) {
        if (template && template.roles) {
            /** @type {?} */
            var hasPreparer = false;
            /** @type {?} */
            var maxSequence = -1;
            for (var x = 0; x < template.roles.length; x++) {
                if (template.roles[x].sequence > maxSequence) {
                    maxSequence = template.roles[x].sequence;
                }
                if (template.roles[x].sequence === 0) {
                    hasPreparer = true;
                }
            }
            /** @type {?} */
            var totalSequence = hasPreparer ? maxSequence + 1 : maxSequence;
            return totalSequence > 1 ? totalSequence + ' Sequences' : '1 Sequence';
        }
        else {
            return '0 Sequence';
        }
    };
    /**
     * @param {?} template
     * @return {?}
     */
    HeaderComponent.prototype.getTemplateRecipients = /**
     * @param {?} template
     * @return {?}
     */
    function (template) {
        if (template && template.roles) {
            /** @type {?} */
            var totalRecipients = template.roles.length;
            return totalRecipients > 1 ? totalRecipients + ' Recipients' : '1 Recipient';
        }
        else {
            return '0 Recipient';
        }
    };
    /**
     * @param {?} envelope_id
     * @return {?}
     */
    HeaderComponent.prototype.goToEnvelope = /**
     * @param {?} envelope_id
     * @return {?}
     */
    function (envelope_id) {
        if (this.currentOrigin === 'rForm' && this.isAtOriginalOrigin) {
            this.router.navigate(["/envelope/" + envelope_id]);
        }
        else {
            location.href = this.connections.rForm_frontend_url + "/envelope/" + envelope_id;
        }
        this.clearSearchInputs();
    };
    /**
     * @param {?} template
     * @return {?}
     */
    HeaderComponent.prototype.goToTemplate = /**
     * @param {?} template
     * @return {?}
     */
    function (template) {
        this.verdocsHeaderService.openTemplate(template);
        this.clearSearchInputs();
    };
    /**
     * @param {?} keyword
     * @return {?}
     */
    HeaderComponent.prototype.goToTemplates = /**
     * @param {?} keyword
     * @return {?}
     */
    function (keyword) {
        this.verdocsHeaderService.openTemplates(keyword);
    };
    /**
     * @return {?}
     */
    HeaderComponent.prototype.searchEnvelopes = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var keyword = this.getSearchKeyword();
        if (keyword !== '') {
            this.router.navigate(["/manage"], { queryParams: { envelope_name: keyword } });
            this.clearSearchInputs();
        }
    };
    /**
     * @return {?}
     */
    HeaderComponent.prototype.searchRecipients = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var keyword = this.getSearchKeyword();
        if (keyword !== '') {
            this.router.navigate(["/manage"], { queryParams: { recipient_name: keyword } });
            this.clearSearchInputs();
        }
    };
    /**
     * @return {?}
     */
    HeaderComponent.prototype.searchTemplates = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var keyword = this.getSearchKeyword();
        if (keyword !== '') {
            this.router.navigate(["/templates"], { queryParams: { name: keyword } });
            this.clearSearchInputs();
        }
    };
    /**
     * @return {?}
     */
    HeaderComponent.prototype.getSearchKeyword = /**
     * @return {?}
     */
    function () {
        if (this.previewInput && this.previewInput.nativeElement.value) {
            return this.previewInput.nativeElement.value;
        }
        else if (this.generalInput && this.generalInput.nativeElement.value) {
            return this.generalInput.nativeElement.value;
        }
        return '';
    };
    /**
     * @return {?}
     */
    HeaderComponent.prototype.clearSearchInputs = /**
     * @return {?}
     */
    function () {
        if (this.previewInput) {
            this.previewInput.nativeElement.value = '';
        }
        if (this.generalInput) {
            this.generalInput.nativeElement.value = '';
        }
        this.envelopeResult = [];
        this.templateResult = [];
    };
    /**
     * @return {?}
     */
    HeaderComponent.prototype.updateStar = /**
     * @return {?}
     */
    function () {
        if (this.template && this.currentProfile && this.template.template_stars) {
            /** @type {?} */
            var existingStarIndex = findIndex(this.template.template_stars, { profile_id: this.currentProfile.id, template_id: this.template.id });
            this._userStarred = existingStarIndex > -1;
        }
        else {
            return false;
        }
    };
    /**
     * @return {?}
     */
    HeaderComponent.prototype.starTemplate = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.canStar) {
            this.verdocsHeaderService.starTemplate(this.template.id).subscribe((/**
             * @param {?} res
             * @return {?}
             */
            function (res) {
                _this.template.template_stars.push(res);
                _this.updateStar();
            }));
        }
        else if (!this.isAuthenticated) {
            this.createTemplate();
        }
    };
    /**
     * @return {?}
     */
    HeaderComponent.prototype.unstarTemplate = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.canStar) {
            this.verdocsHeaderService.unstarTemplate(this.template.id).subscribe((/**
             * @param {?} res
             * @return {?}
             */
            function (res) {
                /** @type {?} */
                var removedStarIndex = findIndex(_this.template.template_stars, { profile_id: _this.currentProfile.id, template_id: _this.template.id });
                if (removedStarIndex > -1) {
                    _this.template.template_stars.splice(removedStarIndex, 1);
                    _this.updateStar();
                }
            }));
        }
        else if (!this.isAuthenticated) {
            this.createTemplate();
        }
    };
    /**
     * @return {?}
     */
    HeaderComponent.prototype.editTemplate = /**
     * @return {?}
     */
    function () {
        if (this.canEdit) {
            this.router.navigate(["/builder/" + this.template.id + "/fields"]);
        }
        else if (!this.isAuthenticated) {
            this.createTemplate();
        }
    };
    /**
     * @return {?}
     */
    HeaderComponent.prototype.createTemplate = /**
     * @return {?}
     */
    function () {
        this.verdocsHeaderService.toggleCreate();
    };
    HeaderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'verdocs-header',
                    template: "<div class=\"header__placeholder\" *ngIf=\"pageLoading\"></div>\n<div *ngIf=\"!pageLoading\">\n  <header #headerContainer>\n    <div class=\"header-row\" fxLayout=\"row\" fxLayoutAlign=\"space-between center\"\n      [ngClass]=\"{authenticated: isAuthenticated, preview: mode === 'preview'}\">\n      <div class=\"header-util-wrapper collapsed\" fxLayout=\"row\">\n        <button mat-icon-button (click)=\"toggleMenu()\">\n          <mat-icon>menu</mat-icon>\n        </button>\n      </div>\n      <div class=\"header__dynamic-section\"\n        [ngClass]=\"{preview: mode === 'preview', expanded: templateDetailsExpanded, 'live-view': isLiveView}\">\n        <div class=\"header__logo-wrapper preview\" *ngIf=\"mode === 'preview'\"\n          [ngClass]=\"{authenticated: isAuthenticated, expanded: templateDetailsExpanded}\" matAutocompleteOrigin\n          #previewOrigin=\"matAutocompleteOrigin\">\n          <div class=\"header__dynamic-mobile-menu collapsed\">\n            <button mat-icon-button (click)=\"toggleMenu()\">\n              <mat-icon>menu</mat-icon>\n            </button>\n          </div>\n          <a class=\"header__title preview\" href=\"{{home_url}}\" [ngClass]=\"{authenticated: isAuthenticated}\"\n            *ngIf=\"currentOrigin !== 'rForm'\">\n            <img class=\"header__logo\" [src]=\"image_url\"\n              [ngClass]=\"{preview: mode === 'preview', authenticated: isAuthenticated}\" />\n              <div class=\"header__app-version\" [ngClass]=\"{preview: mode === 'preview', authenticated: isAuthenticated}\">{{app_version}}</div>\n          </a>\n          <a class=\"header__title preview\" [routerLink]=\"(isAuthenticated?'/dashboard': '/')\" [ngClass]=\"{authenticated: isAuthenticated}\"\n            *ngIf=\"currentOrigin === 'rForm'\">\n            <img class=\"header__logo\" [src]=\"image_url\"\n              [ngClass]=\"{preview: mode === 'preview', authenticated: isAuthenticated}\" />\n              <div class=\"header__app-version\" [ngClass]=\"{preview: mode === 'preview', authenticated: isAuthenticated}\">{{app_version}}</div>\n          </a>\n          <div class=\"header__general-search preview\">\n            <input class=\"header__general-search-input\" #previewInput type=\"text\"\n              [ngClass]=\"{preview: mode === 'preview', stretch: previewInput.value !== ''}\"\n              placeholder=\"Search or jump to...\" [matAutocompleteConnectedTo]=\"previewOrigin\"\n              [matAutocomplete]=\"autocomplete\" (focus)=\"setInputFocused(true)\" (blur)=\"setInputFocused(false)\"\n              (input)=\"previewInputEvent($event)\">\n            <button class=\"header__search-button preview\" mat-icon-button *ngIf=\"!isSearching\"\n              (click)=\"previewInput.focus()\">\n              <mat-icon>search</mat-icon>\n            </button>\n            <div class=\"header__general-search-spinner\" *ngIf=\"isSearching\">\n              <mat-spinner [diameter]=\"24\"></mat-spinner>\n            </div>\n          </div>\n          <mat-autocomplete class=\"header__search-autocomplete\" #autocomplete=\"matAutocomplete\"\n            (optionSelected)=\"selected($event)\">\n            <mat-option class=\"header__search-autocomplete-option\" *ngFor=\"let envelope of envelopeResult\"\n              [value]=\"envelope['name']\" (keydown.enter)=\"goToEnvelope(envelope.id)\" (click)=\"goToEnvelope(envelope.id)\">\n              <div class=\"header__search-autocomplete-wrapper\">\n                <div class=\"header__search-autocomplete-left\">\n                  <mat-icon>mail</mat-icon>\n                </div>\n                <div class=\"header__search-autocomplete-middle\">\n                  <div class=\"header__search-autocomplete-header\">\n                    {{envelope['name']}}\n                  </div>\n                  <div class=\"header__search-autocomplete-subheader\">\n                    {{getRoleNames(envelope.recipients)}}\n                  </div>\n                </div>\n                <div class=\"header__search-autocomplete-right\"></div>\n              </div>\n            </mat-option>\n            <mat-option class=\"header__search-autocomplete-option\" *ngIf=\"envelopeResult && envelopeResult.length > 0\"\n              [value]=\"previewInput.value\" (keydown.enter)=\"searchEnvelopes()\" (click)=\"searchEnvelopes()\">\n              <div class=\"header__search-autocomplete-wrapper\">\n                <div class=\"header__search-autocomplete-left\">\n                  <img class=\"header__search-verdocs-image\" src=\"{{verdocs_url}}/assets/v-blue-logo.svg\" />\n                </div>\n                <div class=\"header__search-autocomplete-middle\">\n                  <div class=\"header__search-autocomplete-header\">\n                    Show more Verdocs\n                  </div>\n                </div>\n                <div class=\"header__search-autocomplete-right\"></div>\n              </div>\n            </mat-option>\n            <hr *ngIf=\"envelopeResult && envelopeResult.length > 0\">\n            <mat-option class=\"header__search-autocomplete-option\" *ngFor=\"let template of templateResult\"\n              [value]=\"template.name\" (keydown.enter)=\"goToTemplate(template)\" (click)=\"goToTemplate(template)\">\n              <div class=\"header__search-autocomplete-wrapper\">\n                <div class=\"header__search-autocomplete-left\">\n                  <mat-icon>description</mat-icon>\n                </div>\n                <div class=\"header__search-autocomplete-middle\">\n                  <div class=\"header__search-autocomplete-header\">\n                    <div class=\"header__search-autocomplete-header\">\n                      {{template.name}}\n                    </div>\n                    <div class=\"header__search-autocomplete-subheader\">\n                      {{getTemplateStats(template)}}\n                    </div>\n                  </div>\n                </div>\n                <div class=\"header__search-autocomplete-right\"></div>\n              </div>\n            </mat-option>\n            <mat-option class=\"header__search-autocomplete-option\" *ngIf=\"templateResult && templateResult.length > 0\"\n              [value]=\"previewInput.value\" (keydown.enter)=\"goToTemplates(previewInput.value)\" (click)=\"goToTemplates(previewInput.value)\">\n              <div class=\"header__search-autocomplete-wrapper\">\n                <div class=\"header__search-autocomplete-left\">\n                  <img class=\"header__search-verdocs-image\" src=\"{{verdocs_url}}/assets/v-blue-logo.svg\" />\n                </div>\n                <div class=\"header__search-autocomplete-middle\">\n                  <div class=\"header__search-autocomplete-header\">\n                    Show more Documents\n                  </div>\n                </div>\n                <div class=\"header__search-autocomplete-right\"></div>\n              </div>\n            </mat-option>\n            <hr *ngIf=\"templateResult && templateResult.length > 0 && isAuthenticated\">\n            <mat-option class=\"header__search-autocomplete-option\" [value]=\"previewInput.value\"\n              (keydown.enter)=\"searchRecipients()\" (click)=\"searchRecipients()\"\n              *ngIf=\"isAuthenticated && previewInput.value !== ''\">\n              <div class=\"header__search-autocomplete-wrapper\">\n                <div class=\"header__search-autocomplete-left\">\n                  <mat-icon>search</mat-icon>\n                </div>\n                <div class=\"header__search-autocomplete-middle\">\n                  <div class=\"header__search-autocomplete-header\">\n                    Search Recipients with \"{{previewInput.value}}\"\n                  </div>\n                </div>\n                <div class=\"header__search-autocomplete-right\"></div>\n              </div>\n            </mat-option>\n          </mat-autocomplete>\n        </div>\n        <div class=\"header__logo-wrapper\" *ngIf=\"mode !== 'preview'\"\n          [ngClass]=\"{preview: mode === 'preview', authenticated: isAuthenticated, expanded: templateDetailsExpanded}\">\n          <a class=\"header__title\" href=\"{{home_url}}\" [ngClass]=\"{preview: mode === 'preview', authenticated: isAuthenticated}\"\n            *ngIf=\"currentOrigin !== 'rForm'\">\n            <img class=\"header__logo\" [src]=\"image_url\"\n              [ngClass]=\"{preview: mode === 'preview', authenticated: isAuthenticated}\" />\n              <div class=\"header__app-version\" [ngClass]=\"{preview: mode === 'preview', authenticated: isAuthenticated}\">{{app_version}}</div>\n          </a>\n          <a class=\"header__title\" [routerLink]=\"(isAuthenticated?'/dashboard': '/')\" [ngClass]=\"{preview: mode === 'preview', authenticated: isAuthenticated}\"\n            *ngIf=\"currentOrigin === 'rForm'\">\n            <img class=\"header__logo\" [src]=\"image_url\"\n              [ngClass]=\"{preview: mode === 'preview', authenticated: isAuthenticated}\" />\n              <div class=\"header__app-version\" [ngClass]=\"{preview: mode === 'preview', authenticated: isAuthenticated}\">{{app_version}}</div>\n          </a>\n          <div class=\"header__general-search\"\n            [ngClass]=\"{preview: mode === 'preview', extend: inputFocused || generalInput.value != '', unauthenticated: !isAuthenticated, landing: currentAt === 'rSecure'}\"\n            matAutocompleteOrigin #generalOrigin=\"matAutocompleteOrigin\">\n            <input class=\"header__general-search-input\" #generalInput type=\"text\"\n              [ngClass]=\"{preview: mode === 'preview', stretch: generalInput.value !== '', unauthenticated: !isAuthenticated}\"\n              placeholder=\"Search or jump to...\" [matAutocompleteConnectedTo]=\"generalOrigin\"\n              [matAutocomplete]=\"autocomplete\" (focus)=\"setInputFocused(true)\" (blur)=\"setInputFocused(false)\"\n              (input)=\"generalInputEvent($event)\">\n            <button class=\"header__search-button\" mat-icon-button [ngClass]=\"{preview: mode === 'preview'}\"\n              *ngIf=\"!isSearching\" (click)=\"generalInput.focus()\">\n              <mat-icon>search</mat-icon>\n            </button>\n            <div class=\"header__general-search-spinner\" *ngIf=\"isSearching\">\n              <mat-spinner [diameter]=\"24\"></mat-spinner>\n            </div>\n          </div>\n          <mat-autocomplete class=\"header__search-autocomplete\" #autocomplete=\"matAutocomplete\"\n            (optionSelected)=\"selected($event)\">\n            <mat-option class=\"header__search-autocomplete-option\" *ngFor=\"let envelope of envelopeResult\"\n              [value]=\"envelope.name\" (keydown.enter)=\"goToEnvelope(envelope.id)\" (click)=\"goToEnvelope(envelope.id)\">\n              <div class=\"header__search-autocomplete-wrapper\">\n                <div class=\"header__search-autocomplete-left\">\n                  <mat-icon>mail</mat-icon>\n                </div>\n                <div class=\"header__search-autocomplete-middle\">\n                  <div class=\"header__search-autocomplete-header\">\n                    {{envelope.name}}\n                  </div>\n                  <div class=\"header__search-autocomplete-subheader\">\n                    {{getRoleNames(envelope.recipients)}}\n                  </div>\n                </div>\n                <div class=\"header__search-autocomplete-right\"></div>\n              </div>\n            </mat-option>\n            <mat-option class=\"header__search-autocomplete-option\" *ngIf=\"envelopeResult && envelopeResult.length > 0\"\n              [value]=\"generalInput.value\" (keydown.enter)=\"searchEnvelopes()\" (click)=\"searchEnvelopes()\">\n              <div class=\"header__search-autocomplete-wrapper\">\n                <div class=\"header__search-autocomplete-left\">\n                  <img class=\"header__search-verdocs-image\" src=\"{{verdocs_url}}/assets/v-blue-logo.svg\" />\n                </div>\n                <div class=\"header__search-autocomplete-middle\">\n                  <div class=\"header__search-autocomplete-header\">\n                    Show more Verdocs\n                  </div>\n                </div>\n                <div class=\"header__search-autocomplete-right\"></div>\n              </div>\n            </mat-option>\n            <hr *ngIf=\"envelopeResult && envelopeResult.length > 0\">\n            <mat-option class=\"header__search-autocomplete-option\" *ngFor=\"let template of templateResult\"\n              (keydown.enter)=\"goToTemplate(template)\" (click)=\"goToTemplate(template)\">\n              <div class=\"header__search-autocomplete-wrapper\">\n                <div class=\"header__search-autocomplete-left\">\n                  <mat-icon>description</mat-icon>\n                </div>\n                <div class=\"header__search-autocomplete-middle\">\n                  <div class=\"header__search-autocomplete-header\">\n                    {{template?.name}}\n                  </div>\n                  <div class=\"header__search-autocomplete-subheader\">\n                    {{getTemplateStats(template)}}\n                  </div>\n                </div>\n                <div class=\"header__search-autocomplete-right\"></div>\n              </div>\n            </mat-option>\n            <mat-option class=\"header__search-autocomplete-option\" *ngIf=\"templateResult && templateResult.length > 0\"\n              [value]=\"generalInput.value\" (keydown.enter)=\"goToTemplates(generalInput.value)\" (click)=\"goToTemplates(generalInput.value)\">\n              <div class=\"header__search-autocomplete-wrapper\">\n                <div class=\"header__search-autocomplete-left\">\n                  <img class=\"header__search-verdocs-image\" src=\"{{verdocs_url}}/assets/v-blue-logo.svg\" />\n                </div>\n                <div class=\"header__search-autocomplete-middle\">\n                  <div class=\"header__search-autocomplete-header\">\n                    Show more Documents\n                  </div>\n                </div>\n                <div class=\"header__search-autocomplete-right\"></div>\n              </div>\n            </mat-option>\n            <hr *ngIf=\"templateResult && templateResult.length > 0 && isAuthenticated\">\n            <mat-option class=\"header__search-autocomplete-option\" [value]=\"generalInput.value\"\n              (keydown.enter)=\"searchRecipients()\" (click)=\"searchRecipients()\"\n              *ngIf=\"isAuthenticated && generalInput.value !== ''\">\n              <div class=\"header__search-autocomplete-wrapper\">\n                <div class=\"header__search-autocomplete-left\">\n                  <mat-icon>search</mat-icon>\n                </div>\n                <div class=\"header__search-autocomplete-middle\">\n                  <div class=\"header__search-autocomplete-header\">\n                    Search Recipients with \"{{generalInput.value}}\"\n                  </div>\n                </div>\n                <div class=\"header__search-autocomplete-right\"></div>\n              </div>\n            </mat-option>\n          </mat-autocomplete>\n        </div>\n        <div class=\"header__template-details\"\n          [ngClass]=\"{show: mode === 'preview' && !isLiveView, expanded: templateDetailsExpanded}\">\n          <div class=\"header__template-detail-container\" [ngClass]=\"{show: templateDetailsExpanded}\">\n            <div class=\"header__template-detail-wrapper\">\n              <div class=\"header__template-detail-header no-top-border\">\n                <div class=\"header__template-detail-title bold\">\n                  <h1>{{template?.name}}</h1>\n                </div>\n                <div class=\"header__template-detail-button-container\"></div>\n              </div>\n              <div class=\"header__template-detail-body\" *ngIf=\"template && template.description\">\n                <div class=\"header__template-detail-body-text\">\n                  {{template?.description}}\n                </div>\n              </div>\n            </div>\n            <div class=\"header__template-detail-wrapper\" *ngIf=\"template\">\n              <div class=\"header__template-detail-header\">\n                <div class=\"header__template-detail-title\">Tags</div>\n                <div class=\"header__template-detail-button-container\"></div>\n              </div>\n              <div class=\"header__template-detail-body\">\n                <div class=\"header__template-detail-tags\">\n                  <div class=\"header__template-detail-tag\" *ngFor=\"let tag of template.tags\">\n                    {{tag.tag_name}}\n                  </div>\n                </div>\n              </div>\n            </div>\n            <div class=\"header__template-detail-wrapper\" *ngIf=\"template\">\n              <div class=\"header__template-detail-header\">\n                <div class=\"header__template-detail-title\">Visibility</div>\n              </div>\n              <div class=\"header__template-detail-button-container\">\n                <app-preview-visibility [template]=\"template\"\n                  [visibility]=\"{is_personal: template?.is_personal, is_public: template?.is_public}\">\n                </app-preview-visibility>\n              </div>\n              <div class=\"header__template-detail-body\"></div>\n            </div>\n            <div class=\"header__template-detail-wrapper\" *ngIf=\"template\">\n              <div class=\"header__template-detail-header\">\n                <div class=\"header__template-detail-title\">Source documents</div>\n              </div>\n              <div class=\"header__template-detail-body\">\n                <app-template-source [template]=\"template\"></app-template-source>\n              </div>\n            </div>\n            <div class=\"header__template-detail-wrapper\" *ngIf=\"template\">\n              <div class=\"header__template-detail-header\">\n                <div class=\"header__template-detail-title\">Workflow</div>\n              </div>\n              <div class=\"header__template-detail-body\">\n                <app-recipient-editor [template]=\"template\"></app-recipient-editor>\n              </div>\n            </div>\n            <div class=\"header__template-detail-wrapper\" *ngIf=\"template\">\n              <div class=\"header__template-detail-header\">\n                <div class=\"header__template-detail-title\">Metrics</div>\n              </div>\n              <div class=\"header__template-detail-body\">\n                <div class=\"header__template-detail-stats\">\n                  <div class=\"header__counter-stars\">\n                    <mat-icon class=\"header__counter-icon\">star</mat-icon>\n                    <span class=\"header__counter-number\">\n                      {{template.template_stars ? template.template_stars.length : 0}}\n                    </span>\n                  </div>\n                  <div class=\"header__counter-envelopes\">\n                    <mat-icon class=\"header__counter-icon\">mail</mat-icon>\n                    <span class=\"header__counter-number\">\n                      {{template.counter}}\n                    </span>\n                  </div>\n                </div>\n              </div>\n            </div>\n          </div>\n          <div class=\"header__template-detail-options\" [ngClass]=\"{show: templateDetailsExpanded}\">\n            <div class=\"header__template-detail-option-button-wrapper\">\n              <button mat-raised-button class=\"header__template-detail-option-button\" *ngIf=\"!userStarred\"\n                (click)=\"starTemplate()\">\n                <mat-icon class=\"header__template-detail-option-icon\">star</mat-icon>\n                <span class=\"header__template-detail-option-text\">Star</span>\n              </button>\n              <button mat-raised-button class=\"header__template-detail-option-button\" *ngIf=\"userStarred\"\n                (click)=\"unstarTemplate()\">\n                <mat-icon class=\"header__template-detail-option-icon\">star_outline</mat-icon>\n                <span class=\"header__template-detail-option-text\">Unstar</span>\n              </button>\n            </div>\n            <div class=\"header__template-detail-option-button-wrapper\">\n              <button mat-raised-button class=\"header__template-detail-option-button\" (click)=\"createTemplate()\">\n                <mat-icon class=\"header__template-detail-option-icon\">drafts</mat-icon>\n                <span class=\"header__template-detail-option-text\">Create</span>\n              </button>\n            </div>\n            <div class=\"header__template-detail-option-button-wrapper\" matTooltip=\"Edit privileges required.\">\n              <button mat-raised-button class=\"header__template-detail-option-button\" (click)=\"editTemplate()\" [disabled]=\"!canEdit && isAuthenticated\">\n                <mat-icon class=\"header__template-detail-option-icon\">build</mat-icon>\n                <span class=\"header__template-detail-option-text\">Edit</span>\n              </button>\n            </div>\n          </div>\n          <div class=\"header__template-details-expansion\" (click)=\"toggleTemplateDetails()\"\n            [ngClass]=\"{expanded: templateDetailsExpanded}\">\n            <button class=\"header__template-details-button\" mat-icon-button>\n              <mat-icon class=\"header__template-details-arrow\" [ngClass]=\"{expanded: templateDetailsExpanded}\">\n                keyboard_arrow_down</mat-icon>\n            </button>\n            <div class=\"header__template-details-text\" *ngIf=\"!templateDetailsExpanded\">Show details</div>\n            <div class=\"header__template-details-text\" *ngIf=\"templateDetailsExpanded\">Hide</div>\n          </div>\n        </div>\n      </div>\n      <div class=\"header__nav\" fxFlex [ngClass]=\"{preview: mode === 'preview', ua: !isAuthenticated}\">\n        <div class=\"header__nav-authenticated\" *ngIf=\"isAuthenticated && currentOrigin === 'rForm'\">\n          <a class=\"header__link\" [routerLink]=\"'/manage'\">Manage</a>\n          <a class=\"header__link\" [routerLink]=\"'/documents'\">Documents</a>\n        </div>\n        <div class=\"header__nav-authenticated\" *ngIf=\"isAuthenticated && currentOrigin !== 'rForm' && currentOrigin !== 'rSecure'\">\n          <a class=\"header__link\" [href]=\"rFormFrontEnd + '/manage'\">Manage</a>\n          <a class=\"header__link\" [href]=\"rFormFrontEnd + '/documents'\">Documents</a>\n        </div>\n        <div class=\"header__nav-unauthenticated\" *ngIf=\"(!isAuthenticated && currentOrigin !== 'rSecure') || currentOrigin === 'rSecure'\">\n          <div class=\"header__link-wrapper realster\">\n            <div class=\"header__link realster\">Why Verdocs <mat-icon>expand_more</mat-icon>\n            </div>\n            <div class=\"header__sub-link realster\" *ngIf=\"currentOrigin !== 'rSecure'\">\n              <a [href]=\"rSecureFrontEnd + '/features'\" class=\"header__sub-link-nav main\">Features</a>\n              <a [href]=\"rSecureFrontEnd + '/features/builder'\" class=\"header__sub-link-nav\">Builder</a>\n              <a [href]=\"rSecureFrontEnd + '/features/documents'\" class=\"header__sub-link-nav\">Documents</a>\n              <a [href]=\"rSecureFrontEnd + '/features/attachments'\" class=\"header__sub-link-nav\">Collect Attachments</a>\n              <a [href]=\"rSecureFrontEnd + '/features/payments'\" class=\"header__sub-link-nav\">Collect Payments</a>\n              <a [href]=\"rSecureFrontEnd + '/features/export'\" class=\"header__sub-link-nav\">Export</a>\n              <a [href]=\"rSecureFrontEnd + '/features/audit-trail'\" class=\"header__sub-link-nav\">Tamper-Proof Audit\n                Trail</a>\n              <a [href]=\"rSecureFrontEnd + '/features/reminders'\" class=\"header__sub-link-nav\">Reminders</a>\n              <a [href]=\"rSecureFrontEnd + '/features/flexible-workflows'\" class=\"header__sub-link-nav\">Flexible\n                Workflows</a>\n              <a [href]=\"rSecureFrontEnd + '/features/team-management'\" class=\"header__sub-link-nav\">Team Management</a>\n              <hr class=\"header__navigation-divider realster\">\n              <a [href]=\"rSecureFrontEnd + '/security'\" class=\"header__sub-link-nav main\">Security</a>\n              <a href=\"https://verdocs.com/resources/category/customer-stories\"\n                class=\"header__sub-link-nav main\">Customer\n                Stories</a>\n            </div>\n            <div class=\"header__sub-link realster local\" *ngIf=\"currentOrigin === 'rSecure'\">\n              <a [routerLink]=\"'/features'\" class=\"header__sub-link-nav main\">Features</a>\n              <a [routerLink]=\"'/features/builder'\" class=\"header__sub-link-nav\">Builder</a>\n              <a [routerLink]=\"'/features/documents'\" class=\"header__sub-link-nav\">Documents</a>\n              <a [routerLink]=\"'/features/attachments'\" class=\"header__sub-link-nav\">Collect Attachments</a>\n              <a [routerLink]=\"'/features/payments'\" class=\"header__sub-link-nav\">Collect Payments</a>\n              <a [routerLink]=\"'/features/export'\" class=\"header__sub-link-nav\">Export</a>\n              <a [routerLink]=\"'/features/audit-trail'\" class=\"header__sub-link-nav\">Tamper-Proof Audit Trail</a>\n              <a [routerLink]=\"'/features/reminders'\" class=\"header__sub-link-nav\">Reminders</a>\n              <a [routerLink]=\"'/features/flexible-workflows'\" class=\"header__sub-link-nav\">Flexible Workflows</a>\n              <a [routerLink]=\"'/features/team-management'\" class=\"header__sub-link-nav\">Team Management</a>\n              <hr class=\"header__navigation-divider realster\">\n              <a [routerLink]=\"'/security'\" class=\"header__sub-link-nav main\">Security</a>\n              <a href=\"https://verdocs.com/resources/category/customer-stories\"\n                class=\"header__sub-link-nav main\">Customer\n                Stories</a>\n            </div>\n          </div>\n          <div class=\"header__link-wrapper pricing\">\n            <div class=\"header__link pricing\">Pricing <mat-icon>expand_more</mat-icon>\n            </div>\n            <div class=\"header__sub-link pricing\" *ngIf=\"currentOrigin !== 'rSecure'\">\n              <a [href]=\"rSecureFrontEnd + '/pricing'\" class=\"header__sub-link-nav main\">Plans</a>\n              <a [href]=\"rSecureFrontEnd + '/pricing/contact-sales'\" class=\"header__sub-link-nav\">Contact Sales</a>\n              <hr class=\"nav-header__navigation-divider pricing\">\n              <!-- <a [href]=\"rSecureFrontEnd + '/pricing/real-estate'\" class=\"nav-header__navigation-link\">Real Estate</a> -->\n              <a [href]=\"rSecureFrontEnd + '/pricing/nonprofit'\" class=\"header__sub-link-nav\">Nonprofit</a>\n            </div>\n            <div class=\"header__sub-link pricing local\" *ngIf=\"currentOrigin === 'rSecure'\">\n              <a [routerLink]=\"'/pricing'\" class=\"header__sub-link-nav main\">Plans</a>\n              <a [routerLink]=\"'/pricing/contact-sales'\" class=\"header__sub-link-nav\">Contact Sales</a>\n              <hr class=\"nav-header__navigation-divider pricing\">\n              <!-- <a [routerLink]=\"'/pricing/real-estate'\" class=\"nav-header__navigation-link\">Real Estate</a> -->\n              <a [routerLink]=\"'/pricing/nonprofit'\" class=\"header__sub-link-nav\">Nonprofit</a>\n            </div>\n          </div>\n          <a class=\"header__link\" href=\"https://verdocs.com/resources\">Resources</a>\n        </div>\n      </div>\n      <div class=\"header-util-mg\" fxFlex>\n        <verdocs-header-tools [mode]=\"mode\"></verdocs-header-tools>\n      </div>\n    </div>\n    <svg class=\"ghost\" version=\"1.1\" xmlns=\"http://www.w3.org/2000/svg\">\n      <defs>\n        <linearGradient id=\"linear\" x1=\"0%\" y1=\"0%\" x2=\"100%\" y2=\"0%\">\n          <stop offset=\"0%\" stop-color=\"#50BD80\" />\n          <stop offset=\"100%\" stop-color=\"#4856CA\" />\n        </linearGradient>\n      </defs>\n    </svg>\n  </header>\n</div>",
                    styles: ["app-header{position:fixed;top:0;left:0;width:100%}header{position:relative;z-index:99}.header__placeholder{background-color:#f6f8fa;height:56px}.header__nav{margin-left:8px;display:flex;align-items:center;height:64px;flex:1 1 0%;box-sizing:border-box}.header__nav .header__link,.header__nav .header__link:visited{display:flex;align-items:center;padding-left:8px;padding-right:8px;color:#fff;text-decoration:none;height:64px}.header__nav.preview .header__link,.header__nav.preview .header__link:visited{color:rgba(0,0,0,.87)}.header__nav.preview{position:absolute;top:0;left:378px}@media only screen and (max-width:854px){.header__nav{display:none}.header__app-version.preview{padding-left:6px}}@media only screen and (max-width:600px){.header__nav{height:56px}}.header__nav-authenticated,.header__nav-unauthenticated{position:relative;display:flex;align-items:center}.header__app-version{position:absolute;top:20px;font-size:10px;padding-left:12px}.header__app-version.preview{color:rgba(0,0,0,.54);padding-left:6px}@media only screen and (max-width:768px){.header__app-version.preview{padding-left:6px}.header__app-version{padding-left:0}}.header__link-wrapper{position:relative;overflow:visible}.header__link-wrapper:hover{cursor:pointer}.header__link-wrapper.realster{width:143.64px;max-width:143.64px}.header__link-wrapper.realster:hover .header__sub-link.realster{visibility:visible;opacity:1;height:466px;width:240px;background-color:#fff}.header__link-wrapper.realster:hover .header__sub-link.realster:after,.header__link-wrapper.realster:hover .header__sub-link.realster:before{content:'';position:absolute;box-shadow:0 1px 8px 0 rgba(0,0,0,.2),0 3px 3px -2px rgba(0,0,0,.12),0 3px 4px 0 rgba(0,0,0,.14);z-index:-1}.header__link-wrapper.realster:hover .header__sub-link.realster:before{height:100%;width:100%;left:0;top:0}.header__link-wrapper.realster:hover .header__sub-link.realster:after{content:'';width:143.64px;height:54px;top:-52px;left:0;background-color:#fff;box-shadow:0 1px 8px 0 rgba(0,0,0,.2),0 3px 3px -2px rgba(0,0,0,.12),0 3px 4px 0 rgba(0,0,0,.14)}.header__link-wrapper.realster:hover .header__sub-link.realster .header__sub-link-nav{visibility:visible;opacity:1;height:38px}.header__link-wrapper.pricing{width:92.52px;max-width:92.52px}.header__link-wrapper.pricing:hover .header__sub-link.pricing{visibility:visible;opacity:1;height:120px;width:160px;background-color:#fff}.header__link-wrapper.pricing:hover .header__sub-link.pricing:after,.header__link-wrapper.pricing:hover .header__sub-link.pricing:before{content:'';position:absolute;box-shadow:0 1px 8px 0 rgba(0,0,0,.2),0 3px 3px -2px rgba(0,0,0,.12),0 3px 4px 0 rgba(0,0,0,.14);z-index:-1}.header__link-wrapper.pricing:hover .header__sub-link.pricing:before{height:100%;width:100%;left:0;top:0}.header__link-wrapper.pricing:hover .header__sub-link.pricing:after{content:'';width:92.52px;height:54px;top:-52px;left:0;background-color:#fff;box-shadow:0 1px 8px 0 rgba(0,0,0,.2),0 3px 3px -2px rgba(0,0,0,.12),0 3px 4px 0 rgba(0,0,0,.14)}.header__link-wrapper.pricing:hover .header__sub-link.pricing .header__sub-link-nav{visibility:visible;opacity:1;height:38px}.header__sub-link{position:absolute;padding-top:8px;top:60px;left:0;visibility:hidden;opacity:0;height:0;transition:height .3s cubic-bezier(.17,.04,.03,.94);border-bottom-left-radius:2px;border-bottom-right-radius:2px}.header__sub-link:hover{visibility:visible;opacity:1;height:466px;width:240px;background-color:#fff}.header__sub-link:hover:after,.header__sub-link:hover:before{content:'';position:absolute;box-shadow:0 1px 8px 0 rgba(0,0,0,.2),0 3px 3px -2px rgba(0,0,0,.12),0 3px 4px 0 rgba(0,0,0,.14);z-index:-1}.header__sub-link:hover:before{height:100%;width:100%;left:0;top:0}.header__sub-link:hover:after{content:'';width:143.64px;height:54px;top:-52px;left:0;background-color:#fff;box-shadow:0 1px 8px 0 rgba(0,0,0,.2),0 3px 3px -2px rgba(0,0,0,.12),0 3px 4px 0 rgba(0,0,0,.14)}@media only screen and (max-width:600px){.header__sub-link{top:56px}}.header__sub-link-nav{padding:8px 16px;display:block;text-decoration:none;color:rgba(0,0,0,.87);visibility:none;opacity:0;transition:250ms cubic-bezier(.17,.04,.03,.94);transition-delay:50ms;height:0}.header__sub-link-nav.main{font-weight:600}.header__sub-link-nav:visited{color:rgba(0,0,0,.87)}.header-row{position:relative;height:56px;display:flex;align-items:center;place-content:center space-between;background-color:#33354c;color:#fff}.header-row.authenticated{background-color:#33354c;color:#fff}.header-row.preview,.header-row.preview.authenticated{background-color:transparent;color:rgba(0,0,0,.87)}@media only screen and (min-width:601px){.header-row{height:64px}}.header-util-wrapper{margin-left:12px;visibility:visible;opacity:1;transform:translateX(0);transition:.2s ease-out}.header-util-wrapper button:hover{background-color:rgba(0,0,0,.12)}@media only screen and (max-width:415px){.header-util-wrapper.search-active{visibility:hidden;opacity:0;width:0;transform:translateX(-200%)}}@media only screen and (min-width:855px){.header-util-wrapper{visibility:0;opacity:0;display:none!important}}.header__dynamic-mobile-menu{display:none}.header__dynamic-section.preview{position:absolute;top:8px;left:0;height:96px;max-width:370.5px;margin-left:8px;background-color:#f3f3f3;border-radius:2px;box-shadow:0 3px 14px 2px rgba(0,0,0,.12),0 8px 10px 1px rgba(0,0,0,.14),0 5px 5px -3px rgba(0,0,0,.2);transition:height .3s cubic-bezier(.17,.04,.03,.94)}.header__dynamic-section.preview.live-view{height:0}.header__dynamic-section.preview.expanded{max-width:386.5px;margin-left:0;top:0;height:100vh;height:calc(var(--vh,1vh) * 100)}.header__general-search{display:flex;align-items:center;height:48px;background-color:rgba(0,0,0,.15);border-radius:2px}.header__general-search.preview{background-color:transparent}@media only screen and (max-width:854px){.header__dynamic-mobile-menu{display:flex;padding-left:4px}.header__dynamic-section.preview{width:calc(100% - 16px);max-width:calc(100% - 16px)}.header__dynamic-section.preview.expanded{width:100%;max-width:100%}.header__general-search.extend{background-color:rgba(0,0,0,.15);width:calc(100vw - 16px);margin-left:-150px}.header__general-search{background-color:transparent}}.header__general-search-field{height:48px}.header__general-search-input{padding-left:8px;border:none;height:48px;font-size:16px;line-height:48px;outline:0;border-radius:2px;background-color:transparent;color:#fff;width:203px;transition:.3s cubic-bezier(.17,.04,.03,.94)}.header__general-search-input.stretch,.header__general-search-input:focus{width:340px}.header__general-search-input.landing{width:130px}.header__general-search-input.landing.stretch{width:340px}.header__general-search-input.preview,.header__general-search-input.preview.authenticated{background-color:transparent;color:rgba(0,0,0,.87)}.header__general-search-input.preview.authenticated::-webkit-input-placeholder,.header__general-search-input.preview::-webkit-input-placeholder{color:rgba(0,0,0,.54)}.header__general-search-input.preview.authenticated::-moz-placeholder,.header__general-search-input.preview::-moz-placeholder{color:rgba(0,0,0,.54)}.header__general-search-input.preview.authenticated::-ms-input-placeholder,.header__general-search-input.preview::-ms-input-placeholder{color:rgba(0,0,0,.54)}.header__general-search-input.preview.authenticated::placeholder,.header__general-search-input.preview::placeholder{color:rgba(0,0,0,.54)}.header__general-search-input.preview.authenticated.stretch,.header__general-search-input.preview.authenticated:focus,.header__general-search-input.preview.stretch,.header__general-search-input.preview:focus{width:203px}.header__general-search-input::-webkit-input-placeholder{color:rgba(255,255,255,.54)}.header__general-search-input::-moz-placeholder{color:rgba(255,255,255,.54)}.header__general-search-input::-ms-input-placeholder{color:rgba(255,255,255,.54)}.header__general-search-input::placeholder{color:rgba(255,255,255,.54)}@media only screen and (max-width:854px){.header__general-search-input{width:0;border:none}.header__general-search-input.stretch,.header__general-search-input:focus{width:calc(100vw - 72px)!important;background-color:#292d3c;z-index:100}.header__general-search-input.stretch.preview,.header__general-search-input:focus.preview{margin-left:-152px;background-color:#fff}}.header__general-search-spinner{margin-left:8px;display:flex;align-items:center;justify-content:center;height:40px;width:40px}.header__template-details{position:absolute;top:-120px;height:0;visibility:hidden;opacity:0;transition:.3s cubic-bezier(.17,.04,.03,.94)}.header__template-details.show{position:relative;top:0;visibility:visible;opacity:1;height:48px}.header__template-details.show.expanded{height:calc(100% - 64px)}.header__template-details-expansion{position:absolute;bottom:0;left:0;padding-left:8px;padding-right:8px;display:flex;height:48px;align-items:center;width:100%}.header__template-details-expansion.expanded{padding-left:16px;border-top:2px solid rgba(0,0,0,.12)}.header__template-details-expansion:hover{cursor:pointer}.header__template-details-arrow{transition:.3s cubic-bezier(.17,.04,.03,.94)}.header__template-details-arrow.expanded{transform:rotate(-180deg)}.header__template-details-text{font-size:12px;color:rgba(0,0,0,.54)}.header__template-detail-container{background-color:#fff;box-shadow:0 1px 3px 0 rgba(0,0,0,.2),0 2px 2px 0 rgba(0,0,0,.12),0 0 2px 0 rgba(0,0,0,.14);visibility:hidden;opacity:0;height:0;transition:.3s cubic-bezier(.17,.04,.03,.94);transition-delay:25ms}.header__template-detail-container.show{margin:8px 8px 0;padding:8px;visibility:visible;opacity:1;height:auto;max-height:calc(100% - 128px);overflow-y:auto}.header__template-detail-options{visibility:hidden;opacity:0;height:0;transition:.3s cubic-bezier(.17,.04,.03,.94);transition-delay:25ms}.header__template-detail-options.show{display:flex;padding:8px;visibility:visible;opacity:1;width:100%;height:auto}.header__template-detail-option-button{width:100%}.header__template-detail-option-button-wrapper{width:calc(41% - 32px);margin-right:8px}.header__template-detail-option-button-wrapper:last-child{margin-right:0}.header__template-detail-option-icon{margin-right:4px;color:rgba(0,0,0,.54)}.header__template-detail-option-text{text-transform:uppercase;color:rgba(0,0,0,.87)}.header__template-detail-wrapper{padding-left:8px;padding-right:8px}.header__template-detail-header{padding-top:8px;padding-bottom:8px;border-top:1px solid #d1d5da}.header__template-detail-header.no-top-border{border-top:none}.header__template-detail-title{font-size:12px;color:rgba(0,0,0,.87)}.header__template-detail-title.bold{padding-bottom:8px;font-weight:800;font-size:20px}.header__template-detail-title.bold h1{color:rgba(0,0,0,.87);padding-bottom:8px;font-weight:800;font-size:20px;margin:0}.header__template-detail-body{padding-bottom:8px}.header__template-detail-tags{display:flex;padding-top:2px;align-items:center}.header__template-detail-tag{margin-left:8px;padding-left:8px;padding-right:8px;height:20px;line-height:20px;background-color:rgba(55,71,79,.34);border-radius:2px;font-size:10px;color:#000}.header__template-detail-tag:first-child{margin-left:0}.header__template-detail-stats{display:flex;align-items:center;padding-top:8px;line-height:24px}.header__logo{display:inline-block;height:20px;width:77.5px;margin-right:16px;background-size:contain;background-position:center;background-repeat:no-repeat}@media only screen and (max-width:854px){.header__logo{margin:0}}.header__logo-wrapper{display:flex;height:48px;align-items:center}.header__logo-wrapper.preview,.header__logo-wrapper.preview.authenticated{border-radius:2px;background-color:#fff;box-shadow:0 1px 3px 0 rgba(0,0,0,.2),0 2px 1px -1px rgba(0,0,0,.12),0 1px 1px 0 rgba(0,0,0,.14);max-width:370.5px}.header__logo-wrapper.preview.authenticated.expanded,.header__logo-wrapper.preview.expanded{margin-top:8px;margin-left:8px;margin-right:8px;max-width:370.5px}.header__logo-wrapper.preview .header__logo,.header__logo-wrapper.preview.authenticated .header__logo{margin-left:6px}@media screen and (max-width:854px){.header__logo-wrapper.preview,.header__logo-wrapper.preview.authenticated{width:100%;max-width:100%}.header__logo-wrapper.preview.authenticated.expanded,.header__logo-wrapper.preview.expanded{width:calc(100% - 16px);max-width:calc(100% - 16px)}.header__logo-wrapper{justify-content:space-between;width:calc(100vw - 64px)}}@media screen and (min-width:855px){.header__logo{margin-left:12px}}.header__search-button{font-size:24px;margin-left:8px}.header__search-button.preview{color:rgba(0,0,0,.54)}.header__search-button.preview:hover{color:rgba(0,0,0,.87)}.header__search-verdocs-image{margin-left:2px;width:20px}.header__search-autocomplete{position:relative;padding-top:8px;max-height:515px}.header__search-autocomplete-option{display:flex;align-items:center;height:60px;line-height:60px}.header__search-autocomplete-wrapper{display:flex;align-items:center}.header__search-autocomplete-right{display:flex;align-items:center;width:40px;color:rgba(0,0,0,.54)}.header__search-autocomplete-middle{max-width:calc(100% - 80px)}.header__search-autocomplete-header{font-size:16px;color:rgba(0,0,0,.87);height:20px;line-height:20px}.header__search-autocomplete-subheader{font-size:14px;color:rgba(0,0,0,.54);height:20px;line-height:20px}.header__search-autocomplete-left{display:flex;align-items:center;width:40px;color:rgba(0,0,0,.54)}.header__title{position:relative;visibility:visible;opacity:1;padding-left:12px;margin-right:8px;width:calc(100% - 120px);text-align:left;font-size:20px;font-weight:500;text-overflow:ellipsis;text-decoration:none;white-space:nowrap;overflow:visible;color:#fff;transition:.2s ease-out}@media only screen and (max-width:415px){.header__title.search-active{visibility:hidden;width:0;opacity:0;transform:translateX(-200%)}}@media only screen and (max-width:854px){.header__title{margin:0 auto;width:auto;padding-left:0;padding-right:0;transform:translateX(8px)}}.header__title-input{border:none;outline:0;max-width:500px;width:100%;font-size:20px;font-weight:500;background:0 0;color:#fff}.header__title-input:not(:focus){text-overflow:ellipsis}.header__counter-envelopes,.header__counter-stars{margin-right:8px;display:flex;align-items:center}.header__counter-icon{margin-right:4px;color:rgba(0,0,0,.54)}.header__counter-number{margin-right:6px;font-size:16px;line-height:24px;height:24px;color:rgba(0,0,0,.87)}.header-util-mg{margin-right:12px;font-size:1.125rem;font-weight:300;height:64px;flex:1 1 0%}.header-util-mg button{min-width:48px;min-height:48px;background-color:transparent}@media screen and (max-width:768px){.header-util-mg{flex:none!important}}@media only screen and (max-width:600px){.header-util-mg{margin-right:0}.header-user__button{display:none}}.header__app{float:left;width:48px;height:48px;padding-top:5px}.header__notification{float:left;width:48px;height:48px;margin-top:-8px;margin-right:8px}.header-user__button{float:left;height:48px;width:48px;border:none;padding-top:5px;border-radius:50%}.header-user__button:hover{cursor:pointer}.header__nav-back,.header__nav-settings{line-height:40px!important;vertical-align:middle}.header-nav-route{padding-top:10px;padding-bottom:10px;color:#fff;text-align:center;text-transform:uppercase;font-size:1.125em;font-weight:400}.header-nav-filter{margin-right:30px}.header-nav-menu{margin-left:30px}.header__recipients{text-align:right}.header__template-status{position:absolute;bottom:4px;left:56px;font-size:12px;margin-left:8px;color:#fff}@media only screen and (max-width:675px){.header__title{padding-left:0}.header__recipients{max-width:195px!important;min-width:150px!important}.header__template-status{left:45px}}@media only screen and (max-width:360px){.header__template-status{display:none}}"]
                }] }
    ];
    /** @nocollapse */
    HeaderComponent.ctorParameters = function () { return [
        { type: VerdocsHeaderService },
        { type: VerdocsStateService },
        { type: VerdocsTokenObjectService },
        { type: VerdocsAuthService },
        { type: TemplatesGuardService },
        { type: Router },
        { type: ChangeDetectorRef },
        { type: Compiler },
        { type: Injector },
        { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
    ]; };
    HeaderComponent.propDecorators = {
        currentOrigin: [{ type: Input }],
        header: [{ type: ViewChild, args: ['headerContainer', { static: false },] }],
        searchBar: [{ type: ViewChild, args: ['search', { static: false },] }],
        templateTitleInput: [{ type: ViewChild, args: ['templateTitle', { static: false },] }],
        previewInput: [{ type: ViewChild, args: ['previewInput', { static: false },] }],
        generalInput: [{ type: ViewChild, args: ['generalInput', { static: false },] }]
    };
    return HeaderComponent;
}());
export { HeaderComponent };
if (false) {
    /** @type {?} */
    HeaderComponent.prototype.pageLoading;
    /** @type {?} */
    HeaderComponent.prototype.connections;
    /** @type {?} */
    HeaderComponent.prototype.title;
    /** @type {?} */
    HeaderComponent.prototype.isAuthenticated;
    /** @type {?} */
    HeaderComponent.prototype.headerTitle;
    /** @type {?} */
    HeaderComponent.prototype.mode;
    /** @type {?} */
    HeaderComponent.prototype.searchActive;
    /** @type {?} */
    HeaderComponent.prototype.templateName;
    /** @type {?} */
    HeaderComponent.prototype.newTemplateName;
    /** @type {?} */
    HeaderComponent.prototype.saveStatus;
    /** @type {?} */
    HeaderComponent.prototype.notifications;
    /** @type {?} */
    HeaderComponent.prototype.loading;
    /** @type {?} */
    HeaderComponent.prototype.myAccountUrl;
    /** @type {?} */
    HeaderComponent.prototype.termsUrl;
    /** @type {?} */
    HeaderComponent.prototype.privacyUrl;
    /** @type {?} */
    HeaderComponent.prototype.newProfile;
    /** @type {?} */
    HeaderComponent.prototype.organization_id;
    /** @type {?} */
    HeaderComponent.prototype.emailVerified;
    /** @type {?} */
    HeaderComponent.prototype.organizationImgSrc;
    /** @type {?} */
    HeaderComponent.prototype.csvBlob;
    /** @type {?} */
    HeaderComponent.prototype.templateDetailsExpanded;
    /** @type {?} */
    HeaderComponent.prototype.template;
    /** @type {?} */
    HeaderComponent.prototype.isSearching;
    /** @type {?} */
    HeaderComponent.prototype.envelopeResult;
    /** @type {?} */
    HeaderComponent.prototype.templateResult;
    /** @type {?} */
    HeaderComponent.prototype.currentProfile;
    /** @type {?} */
    HeaderComponent.prototype.inputFocused;
    /** @type {?} */
    HeaderComponent.prototype.manageLink;
    /** @type {?} */
    HeaderComponent.prototype.templatesLink;
    /** @type {?} */
    HeaderComponent.prototype.rSecureFrontEnd;
    /** @type {?} */
    HeaderComponent.prototype.rFormFrontEnd;
    /** @type {?} */
    HeaderComponent.prototype.isLiveView;
    /** @type {?} */
    HeaderComponent.prototype.verdocs_url;
    /** @type {?} */
    HeaderComponent.prototype.image_url;
    /** @type {?} */
    HeaderComponent.prototype.isAtOriginalOrigin;
    /** @type {?} */
    HeaderComponent.prototype.app_version;
    /** @type {?} */
    HeaderComponent.prototype.currentAt;
    /** @type {?} */
    HeaderComponent.prototype.currentOrigin;
    /** @type {?} */
    HeaderComponent.prototype.header;
    /** @type {?} */
    HeaderComponent.prototype.searchBar;
    /** @type {?} */
    HeaderComponent.prototype.templateTitleInput;
    /** @type {?} */
    HeaderComponent.prototype.previewInput;
    /** @type {?} */
    HeaderComponent.prototype.generalInput;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.sidenavOpen;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype._canStar;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype._canCreate;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype._canEdit;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype._userStarred;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.metaDataSubscription;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.titleSubscription;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.titleInputSubscription;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.titleKeyDownSubscription;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.templateNameSubscription;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.templateSubscription;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.saveStatusSubscription;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.profileLoadingSubscription;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.userDataSubscription;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.modeSubscription;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.previewInputSubscription;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.generalInputSubscription;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.navigationEventSubscription;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.csvBlobSubscription;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.inputTimer;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.origin;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.verdocsHeaderService;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.vTokenStateService;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.vTokenObjectService;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.vAuthService;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.templateGuard;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.router;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.cdr;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.compiler;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    HeaderComponent.prototype.platform;
}
