// @verdocs/tokens/esm5/lib/directives/realster-guard.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Output, Input, EventEmitter, Inject, forwardRef, PLATFORM_ID } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
import { MatDialog } from '@angular/material';
import { Subscription } from 'rxjs';
import { VerifyEmailDialogComponent } from '../dialogs/verify-email-dialog.component';
import { VerdocsAuthGuardService } from '../services/authguard.service';
import { VerdocsStateService } from '../services/state.service';
import { VerdocsAuthService } from '../services/auth.service';
var VerdocsTokenGuardDirective = /** @class */ (function () {
    function VerdocsTokenGuardDirective(dialog, verdocsAuthGuard, verdocsStateService, verdocsAuthService, platform) {
        this.dialog = dialog;
        this.verdocsAuthGuard = verdocsAuthGuard;
        this.verdocsStateService = verdocsStateService;
        this.verdocsAuthService = verdocsAuthService;
        this.platform = platform;
        this.verifyEmailDialogIsOpen = false;
        this.idTokenSubscription = new Subscription();
        this.resendEmailVerificationEvent = new EventEmitter();
    }
    /**
     * @return {?}
     */
    VerdocsTokenGuardDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.idTokenSubscription = this.verdocsStateService.decodedIdTokenSubject.subscribe((/**
         * @param {?} decodedIdToken
         * @return {?}
         */
        function (decodedIdToken) {
            _this.emailAddress = decodedIdToken.email;
        }));
        setInterval((/**
         * @return {?}
         */
        function () {
            if (_this.enforceAuthentication) {
                _this.verdocsAuthGuard.checkAuthentication();
            }
            if (_this.verdocsAuthService.isAuthenticated()) {
                _this.checkEmailVerification();
            }
        }), 5000);
    };
    /**
     * @return {?}
     */
    VerdocsTokenGuardDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.idTokenSubscription.unsubscribe();
    };
    /**
     * @private
     * @return {?}
     */
    VerdocsTokenGuardDirective.prototype.checkEmailVerification = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var emailVerified = this.verdocsStateService.getEmailVerification();
        if (!emailVerified && !this.verifyEmailDialogIsOpen && isPlatformBrowser(this.platform)) {
            this.verifyEmailDialogIsOpen = true;
            /** @type {?} */
            var email_data = {
                resendEmailVerification: (/**
                 * @return {?}
                 */
                function () { return _this.resendEmailVerification(); }),
                emailAddress: this.emailAddress
            };
            /** @type {?} */
            var verifyEmailDialog = this.dialog.open(VerifyEmailDialogComponent, {
                data: email_data,
                disableClose: true
            });
            verifyEmailDialog.afterClosed().subscribe((/**
             * @return {?}
             */
            function () {
                _this.verifyEmailDialogIsOpen = false;
                location.reload();
            }));
        }
    };
    /**
     * @return {?}
     */
    VerdocsTokenGuardDirective.prototype.resendEmailVerification = /**
     * @return {?}
     */
    function () {
        this.resendEmailVerificationEvent.emit();
    };
    VerdocsTokenGuardDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[verdocs-guard]'
                },] }
    ];
    /** @nocollapse */
    VerdocsTokenGuardDirective.ctorParameters = function () { return [
        { type: MatDialog },
        { type: VerdocsAuthGuardService, decorators: [{ type: Inject, args: [forwardRef((/**
                         * @return {?}
                         */
                        function () { return VerdocsAuthGuardService; })),] }] },
        { type: VerdocsStateService, decorators: [{ type: Inject, args: [forwardRef((/**
                         * @return {?}
                         */
                        function () { return VerdocsStateService; })),] }] },
        { type: VerdocsAuthService, decorators: [{ type: Inject, args: [forwardRef((/**
                         * @return {?}
                         */
                        function () { return VerdocsAuthService; })),] }] },
        { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
    ]; };
    VerdocsTokenGuardDirective.propDecorators = {
        resendEmailVerificationEvent: [{ type: Output }],
        enforceAuthentication: [{ type: Input }]
    };
    return VerdocsTokenGuardDirective;
}());
export { VerdocsTokenGuardDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    VerdocsTokenGuardDirective.prototype.verifyEmailDialogIsOpen;
    /**
     * @type {?}
     * @private
     */
    VerdocsTokenGuardDirective.prototype.emailAddress;
    /**
     * @type {?}
     * @private
     */
    VerdocsTokenGuardDirective.prototype.idTokenSubscription;
    /**
     * @type {?}
     * @private
     */
    VerdocsTokenGuardDirective.prototype.resendEmailVerificationEvent;
    /**
     * @type {?}
     * @private
     */
    VerdocsTokenGuardDirective.prototype.enforceAuthentication;
    /**
     * @type {?}
     * @private
     */
    VerdocsTokenGuardDirective.prototype.dialog;
    /**
     * @type {?}
     * @private
     */
    VerdocsTokenGuardDirective.prototype.verdocsAuthGuard;
    /**
     * @type {?}
     * @private
     */
    VerdocsTokenGuardDirective.prototype.verdocsStateService;
    /**
     * @type {?}
     * @private
     */
    VerdocsTokenGuardDirective.prototype.verdocsAuthService;
    /**
     * @type {?}
     * @private
     */
    VerdocsTokenGuardDirective.prototype.platform;
}
