// @verdocs/sdk/esm5/lib/models/template_search.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function ITemplateSearchParams() { }
if (false) {
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.name;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.sender;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.description;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.profile_id;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.organization_id;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.updated_at;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.created_at;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.last_used_at;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.is_personal;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.is_public;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.tags;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.document_name;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.sort_by;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.ascending;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.row;
    /** @type {?|undefined} */
    ITemplateSearchParams.prototype.page;
}
/**
 * @record
 */
export function ITemplateSearchResult() { }
if (false) {
    /** @type {?} */
    ITemplateSearchResult.prototype.page;
    /** @type {?} */
    ITemplateSearchResult.prototype.row;
    /** @type {?} */
    ITemplateSearchResult.prototype.total;
    /** @type {?} */
    ITemplateSearchResult.prototype.result;
}
/** @enum {string} */
var SortOptions = {
    CREATED_AT: 'created_at',
    UPDATED_AT: 'updated_at',
    NAME: 'name',
    LAST_USED_AT: 'last_used_at',
    COUNTER: 'counter',
    STAR_COUNTER: 'star_counter',
};
export { SortOptions };
