// @verdocs/essentials/esm5/lib/essentials.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule, InjectionToken } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MatDialogModule } from '@angular/material/dialog';
import { AngularApiModule } from '@verdocs/sdk';
import { SnackbarService } from './snackbar/snackbar.service';
import { AccountService } from './account/account.service';
import { BrowserToiTextService } from './builder/browser-to-iText.service';
import { BuilderDataService } from './builder/builder-data.service';
import { BuilderService } from './builder/builder-settings.service';
import { ConversionService } from './fields/conversion.service';
import { SignatureService } from './fields/envelope-signature.service';
import { EnvelopeViewService } from './fields/envelope-view.service';
import { PageService } from './fields/page.service';
import { ModalService } from './modal/modal.service';
import { ErrorDialogService } from './error-dialogs/error-dialog.service';
import { GuardService } from './guards/guard.service';
import { FourOhOneDialog } from './dialogs/error-dialogs/four-oh-one.dialog';
import { PlansDialog } from './dialogs/plan-dialogs/plans-dialog.component';
/**
 * @record
 */
export function EssentialsConfig() { }
if (false) {
    /** @type {?} */
    EssentialsConfig.prototype.production;
    /** @type {?} */
    EssentialsConfig.prototype.domain;
    /** @type {?} */
    EssentialsConfig.prototype.rForm_backend_url;
    /** @type {?} */
    EssentialsConfig.prototype.rForm_frontend_url;
    /** @type {?} */
    EssentialsConfig.prototype.rSecure_frontend_url;
    /** @type {?} */
    EssentialsConfig.prototype.rSecure_backend_url;
    /** @type {?} */
    EssentialsConfig.prototype.rAccount_frontend_url;
    /** @type {?} */
    EssentialsConfig.prototype.rAccount_backend_url;
    /** @type {?} */
    EssentialsConfig.prototype.rNotification_backend_url;
    /** @type {?} */
    EssentialsConfig.prototype.rForm_cookie_name;
    /** @type {?} */
    EssentialsConfig.prototype.loginURL;
    /** @type {?} */
    EssentialsConfig.prototype.origin;
    /** @type {?} */
    EssentialsConfig.prototype.plans;
}
/** @type {?} */
export var EssentialsConfigToken = new InjectionToken('EssentialsConfig');
var EssentialsModule = /** @class */ (function () {
    function EssentialsModule() {
    }
    /**
     * @param {?} config
     * @return {?}
     */
    EssentialsModule.initializeEssentials = /**
     * @param {?} config
     * @return {?}
     */
    function (config) {
        return {
            ngModule: EssentialsModule,
            import: [
                CommonModule,
                AngularApiModule.initSDK({
                    production: config.production,
                    domain: config.domain,
                    origin: config.origin,
                    loginURL: config.loginURL,
                    rForm_backend_url: config.rForm_backend_url,
                    rForm_frontend_url: config.rForm_frontend_url,
                    rSecure_backend_url: config.rSecure_backend_url,
                    rSecure_frontend_url: config.rSecure_frontend_url,
                    rAccount_backend_url: config.rAccount_backend_url,
                    rAccount_frontend_url: config.rAccount_frontend_url,
                    rNotification_backend_url: config.rNotification_backend_url,
                    plans: config.plans
                }),
                MatDialogModule
            ],
            providers: [
                {
                    provide: EssentialsConfigToken,
                    useValue: config
                },
                SnackbarService,
                AccountService,
                BrowserToiTextService,
                BuilderDataService,
                BuilderService,
                ConversionService,
                SignatureService,
                EnvelopeViewService,
                PageService,
                ModalService,
                ErrorDialogService,
                GuardService
            ],
            declarations: [
                FourOhOneDialog,
                PlansDialog
            ],
            EntryComponents: [
                FourOhOneDialog,
                PlansDialog
            ],
            exports: [
                FourOhOneDialog,
                PlansDialog
            ]
        };
    };
    EssentialsModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule
                    ],
                    providers: [],
                    declarations: [
                        FourOhOneDialog,
                        PlansDialog
                    ]
                },] }
    ];
    return EssentialsModule;
}());
export { EssentialsModule };
