// @verdocs/event-tracker/esm2015/lib/event-tracker.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostListener, Input } from '@angular/core';
export class EventTrackerDirective {
    constructor() { }
    /**
     * @param {?} $event
     * @return {?}
     */
    onClick($event) {
        if (navigator && navigator.userAgent) {
            this.analytics = window['analytics'];
            if (this.event.action) {
                this.analytics.track(this.event.action, {
                    category: this.event.category,
                    label: this.event.label,
                    value: this.event.value
                });
            }
        }
    }
}
EventTrackerDirective.decorators = [
    { type: Directive, args: [{
                selector: '[VerdocsEventTracker]'
            },] }
];
/** @nocollapse */
EventTrackerDirective.ctorParameters = () => [];
EventTrackerDirective.propDecorators = {
    event: [{ type: Input, args: ['VerdocsEventTracker',] }],
    onClick: [{ type: HostListener, args: ['click', ['$event'],] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    EventTrackerDirective.prototype.analytics;
    /** @type {?} */
    EventTrackerDirective.prototype.event;
}
