// @verdocs/sdk/esm2015/lib/templates/templates.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Inject, Injector, PLATFORM_ID } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
import { HttpClient, HttpResponse, HttpRequest } from '@angular/common/http';
// import { MatDialog } from '@angular/material/dialog'
import { catchError, map } from 'rxjs/operators';
import { Subject, Observable, EMPTY, of } from 'rxjs';
import { VerdocsTokenObjectService } from '@verdocs/tokens';
import { filter } from 'lodash';
// import { environment } from '../../../environments/environment';
import { TemplatesGuardService } from './templates.guard';
import { TemplateActions } from '../definitions/template.enums';
import { VSDKConfigToken } from '../angular-api.module';
export class TemplatesService {
    /**
     * @param {?} http
     * @param {?} vTokenObjectService
     * @param {?} templateGuard
     * @param {?} injector
     * @param {?} platform
     */
    constructor(http, vTokenObjectService, templateGuard, injector, platform) {
        this.http = http;
        this.vTokenObjectService = vTokenObjectService;
        this.templateGuard = templateGuard;
        this.injector = injector;
        this.platform = platform;
        this.templates = new Subject();
        this.template = new Subject();
        this.loadingStatus = new Subject();
        this._config_token = this.injector.get(VSDKConfigToken);
        this.rForm_backend_url = this._config_token.rForm_backend_url;
    }
    /**
     * @param {?} err
     * @return {?}
     */
    errorMessageNotVerified(err) {
        return err.error.error === 'email is not verified' && err.error.code === 401;
    }
    /**
     * @param {?=} filter
     * @return {?}
     */
    getTemplates(filter) {
        this.loadingStatus.next(true);
        this.templates.next([]);
        /** @type {?} */
        let query = null;
        if (filter) {
            switch (filter) {
                case 'creator':
                    query = '?is_creator=true';
                    break;
                case 'organization':
                    query = '?is_organization=true';
                    break;
                case 'starred':
                    query = '?is_starred=true';
                    break;
                default:
                    query = '';
                    break;
            }
        }
        /** @type {?} */
        const requestUrl = this.rForm_backend_url + '/templates' + (query ? query : '');
        return this.http.request(new HttpRequest('GET', requestUrl, {}, { reportProgress: true })).pipe(map((/**
         * @param {?} event
         * @return {?}
         */
        (event) => {
            if (event instanceof HttpResponse) {
                /** @type {?} */
                const template = (/** @type {?} */ (event.body));
                this.templates.next(template);
                this.loadingStatus.next(false);
                return (/** @type {?} */ (event));
            }
            return event;
        })), catchError((/**
         * @param {?} err
         * @return {?}
         */
        (err) => {
            if (err && err.status === 401 && !this.errorMessageNotVerified(err)) {
                // const errorDialog = this.dialog.open(FourOhOneDialog, {
                //   panelClass: 'error__dialog',
                //   disableClose: true
                // })
                // errorDialog.componentInstance.error = err;
            }
            this.loadingStatus.next(false);
            return Observable.throw(err);
        })));
    }
    // Move this to another service
    /**
     * @param {?} id
     * @param {?=} thumbnail
     * @return {?}
     */
    getTemplateObservable(id, thumbnail) {
        this.loadingStatus.next(true);
        /** @type {?} */
        let templateUrl = this.rForm_backend_url + '/templates/' + id;
        if (thumbnail === true) {
            templateUrl += '?thumbnail=true';
        }
        return this.http.get(templateUrl)
            .pipe(map((/**
         * @param {?} template
         * @return {?}
         */
        template => {
            this.loadingStatus.next(false);
            return template;
        })), catchError((/**
         * @param {?} err
         * @return {?}
         */
        err => {
            this.loadingStatus.next(false);
            if (err && err.status === 401 && !this.errorMessageNotVerified(err)) {
                // const errorDialog = this.dialog.open(FourOhOneDialog, {
                //   panelClass: 'error__dialog',
                //   disableClose: true
                // })
                // errorDialog.componentInstance.error = err;
                return of(err);
            }
            return EMPTY;
        })));
    }
    /**
     * @param {?} id
     * @return {?}
     */
    starTemplate(id) {
        return this.http.post(this.rForm_backend_url + `/templates/${id}/stars`, {}).pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        (res) => res)), catchError((/**
         * @param {?} err
         * @return {?}
         */
        (err) => {
            if (err && err.status === 401 && this.errorMessageNotVerified(err)) {
                // const errorDialog = this.dialog.open(FourOhOneDialog, {
                //   panelClass: 'error__dialog',
                //   disableClose: true
                // })
                // errorDialog.componentInstance.error = err;
                return of(err);
            }
            return Observable.throw(err);
        })));
    }
    /**
     * @param {?} id
     * @return {?}
     */
    unstarTemplate(id) {
        return this.http.delete(this.rForm_backend_url + `/templates/${id}/stars`).pipe(catchError((/**
         * @param {?} err
         * @return {?}
         */
        err => {
            if (err && err.status === 401 && this.errorMessageNotVerified(err)) {
                // const errorDialog = this.dialog.open(FourOhOneDialog, {
                //   panelClass: 'error__dialog',
                //   disableClose: true
                // })
                // errorDialog.componentInstance.error = err;
                return of(err);
            }
            return Observable.throw((/** @type {?} */ (err)));
        })));
    }
    /**
     * @param {?} id
     * @return {?}
     */
    getTemplate(id) {
        return this.getTemplateObservable(id).pipe(map((/**
         * @param {?} template
         * @return {?}
         */
        (template) => {
            this.template.next(template);
            return template;
        })));
    }
    /**
     * @return {?}
     */
    getCreatorTemplates() {
        return this.getTemplates('creator');
    }
    /**
     * @return {?}
     */
    getOrganizationTemplates() {
        return this.getTemplates('organization');
    }
    /**
     * @return {?}
     */
    getStarredTemplates() {
        return this.getTemplates('starred');
    }
    /**
     * @return {?}
     */
    getPermissions() {
        return this.vTokenObjectService.getPermissions();
    }
    /**
     * @param {?} id
     * @return {?}
     */
    getTemplateOwnerInfo(id) {
        return this.http.get(`${this.rForm_backend_url}/templates/${id}?owner_info=true`)
            .toPromise().then((/**
         * @param {?} res
         * @return {?}
         */
        res => {
            return res;
        }));
    }
    /**
     * @param {?} template_id
     * @param {?} template_document
     * @return {?}
     */
    getTemplateDocumentPDFObservable(template_id, template_document) {
        return this.http.get(this.rForm_backend_url + '/templates/' + template_id + '/documents/' +
            template_document.id + '?file=true', { responseType: 'blob' });
    }
    /**
     * @param {?} templateId
     * @param {?} templateDocument
     * @return {?}
     */
    getTemplateDocument(templateId, templateDocument) {
        return this.getTemplateDocumentPDFObservable(templateId, templateDocument).toPromise();
    }
    /**
     * @param {?} templateId
     * @param {?} templateDocumentId
     * @return {?}
     */
    getTemplateThumbnail(templateId, templateDocumentId) {
        return this.http.get(this.rForm_backend_url + '/templates/' + templateId + '/documents/' +
            templateDocumentId + '?thumbnail=true', { responseType: 'blob' }).toPromise();
    }
    /**
     * @param {?} templateId
     * @return {?}
     */
    getAllTemplateDocumentsObservable(templateId) {
        return this.http.get(this.rForm_backend_url + '/templates/' + templateId + '/documents');
    }
    /**
     * @param {?} templateId
     * @return {?}
     */
    getAllTemplateDocuments(templateId) {
        return this.http.get(this.rForm_backend_url + '/templates/' + templateId + '/documents')
            .toPromise();
    }
    /**
     * @param {?} template
     * @return {?}
     */
    downloadTemplateDocument(template) {
        if ((isPlatformBrowser(this.platform))) {
            this.getTemplateDocument(template.id, template.template_documents[0]).then((/**
             * @param {?} template_file
             * @return {?}
             */
            template_file => {
                this.prepareSave(template, template_file);
            }));
        }
    }
    /**
     * @param {?} template
     * @param {?} template_file
     * @return {?}
     */
    prepareSave(template, template_file) {
        if (window && window.URL) {
            /** @type {?} */
            const fileUrl = URL.createObjectURL(template_file);
            // saveAs(fileUrl, template.template_documents[0].name + '.pdf');
        }
    }
    /**
     * @param {?} searchParams
     * @return {?}
     */
    searchTemplates(searchParams) {
        for (const key in searchParams) {
            if (searchParams.hasOwnProperty(key)) {
                if (searchParams[key] === null) {
                    delete searchParams[key];
                }
            }
        }
        return this.http.request(new HttpRequest('POST', this.rForm_backend_url + '/templates/search', searchParams, {
            reportProgress: true,
            responseType: 'json'
        })).pipe(map((/**
         * @param {?} event
         * @return {?}
         */
        event => {
            return event;
        })));
    }
    /**
     * @param {?} templateId
     * @param {?} body
     * @return {?}
     */
    updateTemplateObservable(templateId, body) {
        return this.http.put(this.rForm_backend_url + '/templates/' + templateId, body);
    }
    /**
     * @param {?} templateId
     * @param {?} body
     * @return {?}
     */
    updateTemplate(templateId, body) {
        return this.updateTemplateObservable(templateId, body)
            .toPromise().then((/**
         * @param {?} template
         * @return {?}
         */
        template => template));
    }
    /**
     * @param {?} templateId
     * @return {?}
     */
    deleteTemplate(templateId) {
        return this.http.delete(this.rForm_backend_url + '/templates/' + templateId);
    }
    /**
     * @param {?} templateId
     * @param {?} sequence_number
     * @return {?}
     */
    deleteSequence(templateId, sequence_number) {
        return this.http.delete(this.rForm_backend_url + '/templates/' + templateId + '/roles?sequence=' + sequence_number).toPromise();
    }
    /**
     * @param {?} templates
     * @return {?}
     */
    updateTemplates(templates) {
        this.templates.next(templates);
    }
    /**
     * @param {?} template
     * @return {?}
     */
    canSendEnvelope(template) {
        if (template) {
            /** @type {?} */
            const signers = filter(template.roles, { type: 'signer' });
            /** @type {?} */
            const hasSigner = signers.length > 0;
            /** @type {?} */
            let signersHaveFields = false;
            if (hasSigner) {
                for (const signer of signers) {
                    signersHaveFields = signer['fields'] && signer['fields'].length > 0;
                }
            }
            /** @type {?} */
            const hasAccessToTemplate = this.templateGuard.canPerformAction(TemplateActions.READ, template).canPerform;
            return hasAccessToTemplate && hasSigner && signersHaveFields && this.templateGuard.canBeSender(template);
        }
        else {
            return false;
        }
    }
}
TemplatesService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
TemplatesService.ctorParameters = () => [
    { type: HttpClient },
    { type: VerdocsTokenObjectService },
    { type: TemplatesGuardService },
    { type: Injector },
    { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
];
if (false) {
    /** @type {?} */
    TemplatesService.prototype.templates;
    /** @type {?} */
    TemplatesService.prototype.template;
    /** @type {?} */
    TemplatesService.prototype.loadingStatus;
    /**
     * @type {?}
     * @private
     */
    TemplatesService.prototype._config_token;
    /**
     * @type {?}
     * @private
     */
    TemplatesService.prototype.rForm_backend_url;
    /**
     * @type {?}
     * @private
     */
    TemplatesService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    TemplatesService.prototype.vTokenObjectService;
    /**
     * @type {?}
     * @private
     */
    TemplatesService.prototype.templateGuard;
    /**
     * @type {?}
     * @private
     */
    TemplatesService.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    TemplatesService.prototype.platform;
}
