// @verdocs/sdk/esm2015/lib/templates/template-reminders.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { VSDKConfigToken } from '../angular-api.module';
export class TemplatesRemindersService {
    /**
     * @param {?} http
     * @param {?} injector
     */
    constructor(http, injector) {
        this.http = http;
        this.injector = injector;
        this._config_token = this.injector.get(VSDKConfigToken);
        this._rForm_backend_url = this._config_token.rForm_backend_url;
    }
    /**
     * @param {?} templateId
     * @param {?} reminderId
     * @return {?}
     */
    getReminder(templateId, reminderId) {
        return this.http.get(`${this._rForm_backend_url}/templates/${templateId}/reminder/${reminderId}`).toPromise()
            .then((/**
         * @param {?} reminder
         * @return {?}
         */
        (reminder) => {
            return reminder;
        }));
    }
    /**
     * @param {?} templateId
     * @param {?} reminderId
     * @param {?} body
     * @return {?}
     */
    updateReminder(templateId, reminderId, body) {
        return this.http
            .put(`${this._rForm_backend_url}/templates/${templateId}/reminder/${reminderId}`, body)
            .toPromise();
    }
    /**
     * @param {?} templateId
     * @param {?} reminderId
     * @return {?}
     */
    deleteReminder(templateId, reminderId) {
        return this.http.delete(`${this._rForm_backend_url}/templates/${templateId}/reminder/${reminderId}`).toPromise();
    }
    /**
     * @param {?} templateId
     * @param {?} body
     * @return {?}
     */
    createReminder(templateId, body) {
        return this.http.post(`${this._rForm_backend_url}/templates/${templateId}/reminder`, body).toPromise()
            .then((/**
         * @param {?} reminder
         * @return {?}
         */
        (reminder) => {
            return reminder;
        }));
    }
}
TemplatesRemindersService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
TemplatesRemindersService.ctorParameters = () => [
    { type: HttpClient },
    { type: Injector }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    TemplatesRemindersService.prototype._config_token;
    /**
     * @type {?}
     * @private
     */
    TemplatesRemindersService.prototype._rForm_backend_url;
    /**
     * @type {?}
     * @private
     */
    TemplatesRemindersService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    TemplatesRemindersService.prototype.injector;
}
