// @verdocs/essentials/esm2015/lib/modal/modal.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { find } from 'lodash';
export class ModalService {
    constructor() {
        this.modals = [];
        this.subject = new Subject();
    }
    /**
     * @param {?} modal
     * @return {?}
     */
    add(modal) {
        this.modals.push(modal);
    }
    /**
     * @param {?} id
     * @return {?}
     */
    open(id) {
        /** @type {?} */
        const modal = find(this.modals, { id: id });
    }
    /**
     * @param {?} id
     * @return {?}
     */
    close(id) {
        /** @type {?} */
        const modal = find(this.modals, { id: id });
        modal.close();
    }
    /**
     * @param {?} componentData
     * @return {?}
     */
    setComponentData(componentData) {
        this.componentData = componentData;
        this.subject.next(componentData);
    }
    /**
     * @return {?}
     */
    getComponentData() {
        return this.subject.asObservable();
    }
}
ModalService.decorators = [
    { type: Injectable }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    ModalService.prototype.modals;
    /**
     * @type {?}
     * @private
     */
    ModalService.prototype.componentData;
    /**
     * @type {?}
     * @private
     */
    ModalService.prototype.subject;
}
