// @verdocs/sdk/esm5/lib/documents/envelope-reminders.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { VSDKConfigToken } from '../angular-api.module';
var EnvelopRemindersService = /** @class */ (function () {
    function EnvelopRemindersService(http, injector) {
        this.http = http;
        this.injector = injector;
        this._config_token = this.injector.get(VSDKConfigToken);
        this._rForm_backend_url = this._config_token.rForm_backend_url;
    }
    /**
     * @param {?} envelopeId
     * @param {?} reminderId
     * @return {?}
     */
    EnvelopRemindersService.prototype.getReminder = /**
     * @param {?} envelopeId
     * @param {?} reminderId
     * @return {?}
     */
    function (envelopeId, reminderId) {
        return this.http.get(this._rForm_backend_url + "/envelopes/" + envelopeId + "/reminder/" + reminderId).toPromise()
            .then((/**
         * @param {?} reminder
         * @return {?}
         */
        function (reminder) {
            return reminder;
        }));
    };
    /**
     * @param {?} envelopeId
     * @param {?} reminderId
     * @param {?} body
     * @return {?}
     */
    EnvelopRemindersService.prototype.updateReminder = /**
     * @param {?} envelopeId
     * @param {?} reminderId
     * @param {?} body
     * @return {?}
     */
    function (envelopeId, reminderId, body) {
        return this.http
            .put(this._rForm_backend_url + "/envelopes/" + envelopeId + "/reminder/" + reminderId, body)
            .toPromise();
    };
    /**
     * @param {?} envelopeId
     * @param {?} reminderId
     * @return {?}
     */
    EnvelopRemindersService.prototype.deleteReminder = /**
     * @param {?} envelopeId
     * @param {?} reminderId
     * @return {?}
     */
    function (envelopeId, reminderId) {
        return this.http.delete(this._rForm_backend_url + "/envelopes/" + envelopeId + "/reminder/" + reminderId).toPromise();
    };
    /**
     * @param {?} envelopeId
     * @param {?} body
     * @return {?}
     */
    EnvelopRemindersService.prototype.createReminder = /**
     * @param {?} envelopeId
     * @param {?} body
     * @return {?}
     */
    function (envelopeId, body) {
        return this.http.post(this._rForm_backend_url + "/envelopes/" + envelopeId + "/reminder", body).toPromise()
            .then((/**
         * @param {?} reminder
         * @return {?}
         */
        function (reminder) {
            return reminder;
        }));
    };
    EnvelopRemindersService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    EnvelopRemindersService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: Injector }
    ]; };
    return EnvelopRemindersService;
}());
export { EnvelopRemindersService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    EnvelopRemindersService.prototype._config_token;
    /**
     * @type {?}
     * @private
     */
    EnvelopRemindersService.prototype._rForm_backend_url;
    /**
     * @type {?}
     * @private
     */
    EnvelopRemindersService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    EnvelopRemindersService.prototype.injector;
}
