// @verdocs/header/esm2015/lib/header.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule, InjectionToken } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { MatAutocompleteModule } from '@angular/material/autocomplete';
import { MatButtonModule } from '@angular/material/button';
import { MatChipsModule } from '@angular/material/chips';
import { MatDialogModule } from '@angular/material/dialog';
import { MatIconModule } from '@angular/material/icon';
import { MatRadioModule } from '@angular/material/radio';
import { MatSelectModule } from '@angular/material/select';
import { MatSnackBarModule } from '@angular/material/snack-bar';
import { MatProgressSpinnerModule } from '@angular/material/progress-spinner';
import { MatTooltipModule } from '@angular/material/tooltip';
import { NotificationsModule } from '@verdocs/notifications';
import { TokensModule } from '@verdocs/tokens';
import { ProfilesModule } from '@verdocs/profiles';
import { RecipientEditorComponent } from './template/recipients/recipients-editor.component';
import { TemplateSourceComponent } from './template/source-file/source-file.component';
import { VisibilityEditorComponent } from './template/visibility/visibility-editor.component';
import { HeaderComponent } from './header.component';
import { VerdocsHeaderService } from './services/header.service';
import { TemplatesGuardService } from './services/templates.guard';
import { HeaderToolsComponent } from './components/header-tools.component';
import { FourOhOneDialog } from './error-dialogs/four-oh-one.dialog';
/**
 * @record
 */
export function connectionConfig() { }
if (false) {
    /** @type {?} */
    connectionConfig.prototype.rForm_backend_url;
    /** @type {?} */
    connectionConfig.prototype.rForm_frontend_url;
    /** @type {?} */
    connectionConfig.prototype.rSecure_frontend_url;
    /** @type {?} */
    connectionConfig.prototype.rSecure_backend_url;
    /** @type {?} */
    connectionConfig.prototype.rAccount_frontend_url;
    /** @type {?} */
    connectionConfig.prototype.rAccount_backend_url;
    /** @type {?} */
    connectionConfig.prototype.rNotification_backend_url;
    /** @type {?|undefined} */
    connectionConfig.prototype.origin;
    /** @type {?|undefined} */
    connectionConfig.prototype.app_version;
}
/** @type {?} */
export const connectionConfigToken = new InjectionToken('connectionConfig');
export class HeaderModule {
    /**
     * @param {?} config
     * @return {?}
     */
    static initVerdocsHeader(config) {
        return {
            ngModule: HeaderModule,
            providers: [
                {
                    provide: connectionConfigToken,
                    useValue: config
                }
            ]
        };
    }
}
HeaderModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    HeaderComponent,
                    RecipientEditorComponent,
                    TemplateSourceComponent,
                    VisibilityEditorComponent,
                    HeaderToolsComponent,
                    FourOhOneDialog
                ],
                entryComponents: [
                    FourOhOneDialog
                ],
                imports: [
                    CommonModule,
                    FormsModule,
                    HttpClientModule,
                    RouterModule,
                    ReactiveFormsModule,
                    MatSnackBarModule,
                    MatIconModule,
                    MatAutocompleteModule,
                    MatButtonModule,
                    MatSelectModule,
                    MatDialogModule,
                    MatChipsModule,
                    MatProgressSpinnerModule,
                    MatRadioModule,
                    MatTooltipModule,
                    NotificationsModule,
                    ProfilesModule,
                    TokensModule
                ],
                providers: [
                    TemplatesGuardService,
                    VerdocsHeaderService
                ],
                exports: [
                    HeaderComponent,
                    RecipientEditorComponent,
                    TemplateSourceComponent,
                    VisibilityEditorComponent,
                    HeaderToolsComponent,
                    FourOhOneDialog
                ]
            },] }
];
