// @verdocs/essentials/esm5/lib/account/account.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable, Injector } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Router } from '@angular/router';
import { ReplaySubject } from 'rxjs';
import { VerdocsStateService } from '@verdocs/tokens';
import { ProfileModel, ProfileCollection } from '@verdocs/profiles';
import { EssentialsConfigToken } from '../essentials.module';
import { findIndex } from 'lodash';
var AccountService = /** @class */ (function () {
    function AccountService(http, router, vTokenStateService, injector) {
        this.http = http;
        this.router = router;
        this.vTokenStateService = vTokenStateService;
        this.injector = injector;
        this.userDataSubject = new ReplaySubject();
        this.userProfileSubject = new ReplaySubject();
        this.profileLoadingSubject = new ReplaySubject();
        this.currentProfileSubject = new ReplaySubject();
        this.personalAccountSubject = new ReplaySubject();
        this.profilesSubject = new ReplaySubject();
        this.userFullname = '';
        this.userInitial = '';
        this.userEmail = '';
        this.userPhone = '';
        this.userData = {
            name: this.userFullname,
            initial: null,
            email: this.userEmail,
            phone: this.userPhone
        };
        this.currentOrganizationId = null;
        this._config_token = this.injector.get(EssentialsConfigToken);
        this.backend_url = this._config_token.rAccount_backend_url;
        this.rForm_backend_url = this._config_token.rForm_backend_url;
        this.rSecure_backend_url = this._config_token.rSecure_backend_url;
    }
    /**
     * @return {?}
     */
    AccountService.prototype.getAccountData = /**
     * @return {?}
     */
    function () {
        /***********
        * Refactor *
        ***********/
        // const headerService = this.injector.get(HeaderService);
        // headerService.noHeader.subscribe(status => {
        // if (status === false) {
        this.getProfiles();
        // }
        // });
    };
    /**
     * @return {?}
     */
    AccountService.prototype.getProfiles = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return this.http.get(this.backend_url + '/profiles').toPromise().then((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            /** @type {?} */
            var personal = res.splice(findIndex(res, (/**
             * @param {?} p
             * @return {?}
             */
            function (p) { return p.organization.name === 'realster'; })), 1);
            _this.personalAccountSubject.next(personal[0]);
            _this.profileData = res.sort((/**
             * @param {?} a
             * @param {?} b
             * @return {?}
             */
            function (a, b) {
                if (a.organization.name < b.organization.name) {
                    return -1;
                }
                if (a.organization.name > b.organization.name) {
                    return 1;
                }
                return 0;
            }));
            _this.profileData.unshift(personal[0]);
            /** @type {?} */
            var currentIndex = findIndex(_this.profileData, { current: true });
            if (_this.profileData[currentIndex]['first_name']) {
                _this.userFullname = _this.profileData[currentIndex]['first_name'] + ' ' + _this.profileData[currentIndex]['last_name'];
                _this.userInitial = _this.profileData[currentIndex]['first_name'].charAt(0) +
                    _this.profileData[currentIndex]['last_name'].charAt(0);
            }
            else {
                _this.userFullname = '';
            }
            if (_this.profileData[currentIndex]['phone']) {
                _this.userPhone = _this.profileData[currentIndex]['phone'];
            }
            else {
                _this.userPhone = '';
            }
            if (_this.profileData[currentIndex]['email']) {
                _this.userEmail = _this.profileData[currentIndex]['email'];
            }
            _this.userData = {
                name: _this.userFullname,
                initial: _this.userInitial,
                email: _this.userEmail,
                phone: _this.userPhone
            };
            _this.userProfileSubject.next(_this.profileData);
            /** @type {?} */
            var currentProfile = _this.profileData[currentIndex];
            _this.currentProfileSubject.next(currentProfile);
            _this.currentOrganizationId = currentProfile.organization_id;
            _this.userDataSubject.next(_this.userData);
            return _this.userData;
        }));
    };
    /**
     * @param {?} profiles
     * @return {?}
     */
    AccountService.prototype.getProfileCollection = /**
     * @param {?} profiles
     * @return {?}
     */
    function (profiles) {
        /** @type {?} */
        var profileModels = [];
        profiles.forEach((/**
         * @param {?} profile
         * @return {?}
         */
        function (profile) {
            profileModels.push(new ProfileModel(profile));
        }));
        /** @type {?} */
        var profileCollection = new ProfileCollection(profileModels);
        return profileCollection;
    };
    /**
     * @param {?} id
     * @return {?}
     */
    AccountService.prototype.removeFromProfileData = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        /** @type {?} */
        var removeIndex = findIndex(this.profileData, { organization_id: name });
        if (removeIndex >= 0) {
            this.profileData.splice(removeIndex, 1);
            this.userProfileSubject.next(this.profileData);
        }
    };
    /**
     * @return {?}
     */
    AccountService.prototype.getAllProfiles = /**
     * @return {?}
     */
    function () {
        return this.http.get(this.backend_url + '/profiles/').toPromise();
    };
    /**
     * @param {?} profile_id
     * @return {?}
     */
    AccountService.prototype.switchProfile = /**
     * @param {?} profile_id
     * @return {?}
     */
    function (profile_id) {
        var _this = this;
        this.profileLoadingSubject.next(true);
        return this.http.post(this.backend_url + '/profiles/' + profile_id + '/switch', null).toPromise().then((/**
         * @param {?} res
         * @return {?}
         */
        function (res) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                this.vTokenStateService.setTokens(res.tokens);
                this.profileLoadingSubject.next(false);
                return [2 /*return*/, Promise.resolve(res)];
            });
        }); }));
    };
    /**
     * @param {?=} sendMessage
     * @return {?}
     */
    AccountService.prototype.updateTokens = /**
     * @param {?=} sendMessage
     * @return {?}
     */
    function (sendMessage) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                return [2 /*return*/, new Promise((/**
                     * @param {?} resolve
                     * @param {?} reject
                     * @return {?}
                     */
                    function (resolve, reject) {
                        _this.vTokenStateService.updateTokens(_this.rSecure_backend_url).subscribe((/**
                         * @param {?} tokens
                         * @return {?}
                         */
                        function (tokens) {
                            if (tokens && tokens.length > 0) {
                                if (sendMessage) {
                                    // this.snackbar.dismiss();
                                    _this.createSnackbar('Status updated');
                                }
                            }
                            _this.getAccountData();
                            return resolve();
                        }));
                    }))];
            });
        });
    };
    /**
     * @return {?}
     */
    AccountService.prototype.resendEmailVerification = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.http.post(this.rSecure_backend_url + '/user/email_verification', null).toPromise().then((/**
         * @return {?}
         */
        function () {
            _this.createSnackbar('Verification email sent');
        }));
    };
    /********************************
    * replace with snackbar service *
    ********************************/
    /**
     * *****************************
     * replace with snackbar service *
     * ******************************
     * @param {?} message
     * @return {?}
     */
    AccountService.prototype.createSnackbar = /**
     * *****************************
     * replace with snackbar service *
     * ******************************
     * @param {?} message
     * @return {?}
     */
    function (message) {
        // let snackbarConfig: MatSnackBarConfig
        // if (window.innerWidth >= 920) {
        //   snackbarConfig = {
        //     verticalPosition: 'bottom',
        //     horizontalPosition: 'left',
        //     duration: 5000
        //   }
        // } else {
        //   snackbarConfig = {
        //     verticalPosition: 'top',
        //     duration: 5000
        //   }
        // }
        // this.snackbar.open(message, null, snackbarConfig);
    };
    /**
     * @param {?} userData
     * @return {?}
     */
    AccountService.prototype.putAccountData = /**
     * @param {?} userData
     * @return {?}
     */
    function (userData) {
        var _this = this;
        if (userData.email) {
            this.userData.email = userData.email;
        }
        if (userData.first_name && userData.last_name) {
            this.userData.name = userData.first_name + ' ' + userData.last_name;
            this.userData.initial = userData.first_name.charAt(0) + userData.last_name.charAt(0);
        }
        if (userData.phone) {
            this.userData.phone = userData.phone || '';
        }
        this.http.put(this.backend_url + '/profiles/' + userData.id, userData).toPromise().then((/**
         * @return {?}
         */
        function () {
            _this.userDataSubject.next(_this.userData);
        }));
    };
    /**
     * @return {?}
     */
    AccountService.prototype.deleteAccount = /**
     * @return {?}
     */
    function () {
        return this.http.delete(this.backend_url).toPromise();
    };
    /**
     * @param {?} email
     * @param {?} oldPassword
     * @param {?} newPassword
     * @return {?}
     */
    AccountService.prototype.changePassword = /**
     * @param {?} email
     * @param {?} oldPassword
     * @param {?} newPassword
     * @return {?}
     */
    function (email, oldPassword, newPassword) {
        return this.http.put(this.backend_url + 'updatepassword', {
            email: email,
            oldPassword: oldPassword,
            newPassword: newPassword
        }).toPromise();
    };
    /**
     * @param {?} name
     * @return {?}
     */
    AccountService.prototype.getNameValidity = /**
     * @param {?} name
     * @return {?}
     */
    function (name) {
        return this.http.get(this.backend_url + '/organizations/is_valid?name=' + name, this.getOptions())
            .toPromise()
            .then((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            return Promise.resolve(res);
        }));
    };
    /**
     * @param {?} body
     * @return {?}
     */
    AccountService.prototype.createOrganization = /**
     * @param {?} body
     * @return {?}
     */
    function (body) {
        return this.http.post(this.backend_url + '/organizations', body, this.getOptions())
            .toPromise()
            .then((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            return Promise.resolve(res);
        }));
    };
    /**
     * @private
     * @return {?}
     */
    AccountService.prototype.getOptions = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var rTokenCookie = this.vTokenStateService.getOtherCookieObject('rAccount_Verdocs');
        /** @type {?} */
        var authValue = 'Bearer ' + rTokenCookie['accessToken'];
        /** @type {?} */
        var headers = new Headers({
            'Cache-Control': 'no-cache',
            'Pragma': 'no-cache',
            'Expires': 'Sat, 01 Jan 2000 00:00:00 GMT',
            'If-Modified-Since': '0',
            'Content-Type': 'application/json',
            'Authorization': authValue,
            'Accept': 'application/json'
        });
        return { 'headers': headers };
    };
    /**
     * @param {?} id
     * @param {?} roleName
     * @param {?} invite
     * @param {?=} redirectReq
     * @return {?}
     */
    AccountService.prototype.fetchToken = /**
     * @param {?} id
     * @param {?} roleName
     * @param {?} invite
     * @param {?=} redirectReq
     * @return {?}
     */
    function (id, roleName, invite, redirectReq) {
        var _this = this;
        /** @type {?} */
        var req = this.rForm_backend_url + ("/envelopes/" + id + "/recipients/" + roleName + "/invitation/" + invite);
        return this.http.get(req, { observe: 'response' }).subscribe((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            _this.vTokenStateService.storeOtherCookie('signer_token', res['headers'].get('signer_token'));
            if (redirectReq) {
                _this.router.navigate(["/view/sign/" + id + "/role/" + roleName], { queryParams: { redirectReq: redirectReq } });
            }
            else {
                _this.router.navigate(["/view/sign/" + id + "/role/" + roleName]);
            }
        }), (/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            _this.router.navigate(["/view/sign/" + id + "/role/" + roleName, { error: err }]);
        }));
    };
    AccountService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AccountService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: Router },
        { type: VerdocsStateService },
        { type: Injector }
    ]; };
    return AccountService;
}());
export { AccountService };
if (false) {
    /** @type {?} */
    AccountService.prototype.userDataSubject;
    /** @type {?} */
    AccountService.prototype.userProfileSubject;
    /** @type {?} */
    AccountService.prototype.profileLoadingSubject;
    /** @type {?} */
    AccountService.prototype.currentProfileSubject;
    /** @type {?} */
    AccountService.prototype.personalAccountSubject;
    /** @type {?} */
    AccountService.prototype.profilesSubject;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype._config_token;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.userFullname;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.userInitial;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.userEmail;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.userPhone;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.userData;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.profileData;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.backend_url;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.rSecure_backend_url;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.rForm_backend_url;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.currentOrganizationId;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.router;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.vTokenStateService;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.injector;
}
