// @verdocs/essentials/esm2015/lib/account/account.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable, Injector } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Router } from '@angular/router';
import { ReplaySubject } from 'rxjs';
import { VerdocsStateService } from '@verdocs/tokens';
import { ProfileModel, ProfileCollection } from '@verdocs/profiles';
import { EssentialsConfigToken } from '../essentials.module';
import { findIndex } from 'lodash';
export class AccountService {
    /**
     * @param {?} http
     * @param {?} router
     * @param {?} vTokenStateService
     * @param {?} injector
     */
    constructor(http, router, vTokenStateService, injector) {
        this.http = http;
        this.router = router;
        this.vTokenStateService = vTokenStateService;
        this.injector = injector;
        this.userDataSubject = new ReplaySubject();
        this.userProfileSubject = new ReplaySubject();
        this.profileLoadingSubject = new ReplaySubject();
        this.currentProfileSubject = new ReplaySubject();
        this.personalAccountSubject = new ReplaySubject();
        this.profilesSubject = new ReplaySubject();
        this.userFullname = '';
        this.userInitial = '';
        this.userEmail = '';
        this.userPhone = '';
        this.userData = {
            name: this.userFullname,
            initial: null,
            email: this.userEmail,
            phone: this.userPhone
        };
        this.currentOrganizationId = null;
        this._config_token = this.injector.get(EssentialsConfigToken);
        this.backend_url = this._config_token.rAccount_backend_url;
        this.rForm_backend_url = this._config_token.rForm_backend_url;
        this.rSecure_backend_url = this._config_token.rSecure_backend_url;
    }
    /**
     * @return {?}
     */
    getAccountData() {
        /***********
        * Refactor *
        ***********/
        // const headerService = this.injector.get(HeaderService);
        // headerService.noHeader.subscribe(status => {
        // if (status === false) {
        this.getProfiles();
        // }
        // });
    }
    /**
     * @return {?}
     */
    getProfiles() {
        return this.http.get(this.backend_url + '/profiles').toPromise().then((/**
         * @param {?} res
         * @return {?}
         */
        (res) => {
            /** @type {?} */
            const personal = res.splice(findIndex(res, (/**
             * @param {?} p
             * @return {?}
             */
            p => p.organization.name === 'realster')), 1);
            this.personalAccountSubject.next(personal[0]);
            this.profileData = res.sort((/**
             * @param {?} a
             * @param {?} b
             * @return {?}
             */
            (a, b) => {
                if (a.organization.name < b.organization.name) {
                    return -1;
                }
                if (a.organization.name > b.organization.name) {
                    return 1;
                }
                return 0;
            }));
            this.profileData.unshift(personal[0]);
            /** @type {?} */
            const currentIndex = findIndex(this.profileData, { current: true });
            if (this.profileData[currentIndex]['first_name']) {
                this.userFullname = this.profileData[currentIndex]['first_name'] + ' ' + this.profileData[currentIndex]['last_name'];
                this.userInitial = this.profileData[currentIndex]['first_name'].charAt(0) +
                    this.profileData[currentIndex]['last_name'].charAt(0);
            }
            else {
                this.userFullname = '';
            }
            if (this.profileData[currentIndex]['phone']) {
                this.userPhone = this.profileData[currentIndex]['phone'];
            }
            else {
                this.userPhone = '';
            }
            if (this.profileData[currentIndex]['email']) {
                this.userEmail = this.profileData[currentIndex]['email'];
            }
            this.userData = {
                name: this.userFullname,
                initial: this.userInitial,
                email: this.userEmail,
                phone: this.userPhone
            };
            this.userProfileSubject.next(this.profileData);
            /** @type {?} */
            const currentProfile = this.profileData[currentIndex];
            this.currentProfileSubject.next(currentProfile);
            this.currentOrganizationId = currentProfile.organization_id;
            this.userDataSubject.next(this.userData);
            return this.userData;
        }));
    }
    /**
     * @param {?} profiles
     * @return {?}
     */
    getProfileCollection(profiles) {
        /** @type {?} */
        const profileModels = [];
        profiles.forEach((/**
         * @param {?} profile
         * @return {?}
         */
        profile => {
            profileModels.push(new ProfileModel(profile));
        }));
        /** @type {?} */
        const profileCollection = new ProfileCollection(profileModels);
        return profileCollection;
    }
    /**
     * @param {?} id
     * @return {?}
     */
    removeFromProfileData(id) {
        /** @type {?} */
        const removeIndex = findIndex(this.profileData, { organization_id: name });
        if (removeIndex >= 0) {
            this.profileData.splice(removeIndex, 1);
            this.userProfileSubject.next(this.profileData);
        }
    }
    /**
     * @return {?}
     */
    getAllProfiles() {
        return this.http.get(this.backend_url + '/profiles/').toPromise();
    }
    /**
     * @param {?} profile_id
     * @return {?}
     */
    switchProfile(profile_id) {
        this.profileLoadingSubject.next(true);
        return this.http.post(this.backend_url + '/profiles/' + profile_id + '/switch', null).toPromise().then((/**
         * @param {?} res
         * @return {?}
         */
        (res) => tslib_1.__awaiter(this, void 0, void 0, function* () {
            this.vTokenStateService.setTokens(res.tokens);
            this.profileLoadingSubject.next(false);
            return Promise.resolve(res);
        })));
    }
    /**
     * @param {?=} sendMessage
     * @return {?}
     */
    updateTokens(sendMessage) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return new Promise((/**
             * @param {?} resolve
             * @param {?} reject
             * @return {?}
             */
            (resolve, reject) => {
                this.vTokenStateService.updateTokens(this.rSecure_backend_url).subscribe((/**
                 * @param {?} tokens
                 * @return {?}
                 */
                tokens => {
                    if (tokens && tokens.length > 0) {
                        if (sendMessage) {
                            // this.snackbar.dismiss();
                            this.createSnackbar('Status updated');
                        }
                    }
                    this.getAccountData();
                    return resolve();
                }));
            }));
        });
    }
    /**
     * @return {?}
     */
    resendEmailVerification() {
        this.http.post(this.rSecure_backend_url + '/user/email_verification', null).toPromise().then((/**
         * @return {?}
         */
        () => {
            this.createSnackbar('Verification email sent');
        }));
    }
    /**
     * *****************************
     * replace with snackbar service *
     * ******************************
     * @param {?} message
     * @return {?}
     */
    createSnackbar(message) {
        // let snackbarConfig: MatSnackBarConfig
        // if (window.innerWidth >= 920) {
        //   snackbarConfig = {
        //     verticalPosition: 'bottom',
        //     horizontalPosition: 'left',
        //     duration: 5000
        //   }
        // } else {
        //   snackbarConfig = {
        //     verticalPosition: 'top',
        //     duration: 5000
        //   }
        // }
        // this.snackbar.open(message, null, snackbarConfig);
    }
    /**
     * @param {?} userData
     * @return {?}
     */
    putAccountData(userData) {
        if (userData.email) {
            this.userData.email = userData.email;
        }
        if (userData.first_name && userData.last_name) {
            this.userData.name = userData.first_name + ' ' + userData.last_name;
            this.userData.initial = userData.first_name.charAt(0) + userData.last_name.charAt(0);
        }
        if (userData.phone) {
            this.userData.phone = userData.phone || '';
        }
        this.http.put(this.backend_url + '/profiles/' + userData.id, userData).toPromise().then((/**
         * @return {?}
         */
        () => {
            this.userDataSubject.next(this.userData);
        }));
    }
    /**
     * @return {?}
     */
    deleteAccount() {
        return this.http.delete(this.backend_url).toPromise();
    }
    /**
     * @param {?} email
     * @param {?} oldPassword
     * @param {?} newPassword
     * @return {?}
     */
    changePassword(email, oldPassword, newPassword) {
        return this.http.put(this.backend_url + 'updatepassword', {
            email,
            oldPassword,
            newPassword
        }).toPromise();
    }
    /**
     * @param {?} name
     * @return {?}
     */
    getNameValidity(name) {
        return this.http.get(this.backend_url + '/organizations/is_valid?name=' + name, this.getOptions())
            .toPromise()
            .then((/**
         * @param {?} res
         * @return {?}
         */
        res => {
            return Promise.resolve(res);
        }));
    }
    /**
     * @param {?} body
     * @return {?}
     */
    createOrganization(body) {
        return this.http.post(this.backend_url + '/organizations', body, this.getOptions())
            .toPromise()
            .then((/**
         * @param {?} res
         * @return {?}
         */
        res => {
            return Promise.resolve(res);
        }));
    }
    /**
     * @private
     * @return {?}
     */
    getOptions() {
        /** @type {?} */
        const rTokenCookie = this.vTokenStateService.getOtherCookieObject('rAccount_Verdocs');
        /** @type {?} */
        const authValue = 'Bearer ' + rTokenCookie['accessToken'];
        /** @type {?} */
        const headers = new Headers({
            'Cache-Control': 'no-cache',
            'Pragma': 'no-cache',
            'Expires': 'Sat, 01 Jan 2000 00:00:00 GMT',
            'If-Modified-Since': '0',
            'Content-Type': 'application/json',
            'Authorization': authValue,
            'Accept': 'application/json'
        });
        return { 'headers': headers };
    }
    /**
     * @param {?} id
     * @param {?} roleName
     * @param {?} invite
     * @param {?=} redirectReq
     * @return {?}
     */
    fetchToken(id, roleName, invite, redirectReq) {
        /** @type {?} */
        const req = this.rForm_backend_url + `/envelopes/${id}/recipients/${roleName}/invitation/${invite}`;
        return this.http.get(req, { observe: 'response' }).subscribe((/**
         * @param {?} res
         * @return {?}
         */
        res => {
            this.vTokenStateService.storeOtherCookie('signer_token', res['headers'].get('signer_token'));
            if (redirectReq) {
                this.router.navigate([`/view/sign/${id}/role/${roleName}`], { queryParams: { redirectReq: redirectReq } });
            }
            else {
                this.router.navigate([`/view/sign/${id}/role/${roleName}`]);
            }
        }), (/**
         * @param {?} err
         * @return {?}
         */
        err => {
            this.router.navigate([`/view/sign/${id}/role/${roleName}`, { error: err }]);
        }));
    }
}
AccountService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
AccountService.ctorParameters = () => [
    { type: HttpClient },
    { type: Router },
    { type: VerdocsStateService },
    { type: Injector }
];
if (false) {
    /** @type {?} */
    AccountService.prototype.userDataSubject;
    /** @type {?} */
    AccountService.prototype.userProfileSubject;
    /** @type {?} */
    AccountService.prototype.profileLoadingSubject;
    /** @type {?} */
    AccountService.prototype.currentProfileSubject;
    /** @type {?} */
    AccountService.prototype.personalAccountSubject;
    /** @type {?} */
    AccountService.prototype.profilesSubject;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype._config_token;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.userFullname;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.userInitial;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.userEmail;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.userPhone;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.userData;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.profileData;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.backend_url;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.rSecure_backend_url;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.rForm_backend_url;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.currentOrganizationId;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.router;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.vTokenStateService;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.injector;
}
