// @verdocs/essentials/esm5/lib/builder/builder-settings.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ReplaySubject } from 'rxjs';
var BuilderService = /** @class */ (function () {
    function BuilderService() {
        this.sideNavStateSubject = new ReplaySubject();
        this.recipientSideNavStateSubject = new ReplaySubject(1, 100);
        this.initiateAutofocus = new ReplaySubject();
        this.recipientAutoFocus = new ReplaySubject();
        this.deleteRoleSubject = new ReplaySubject(1, 100);
        this.hasActiveRoleSubject = new ReplaySubject();
    }
    // constructor() {}
    // constructor() {}
    /**
     * @return {?}
     */
    BuilderService.prototype.toggleSideNav = 
    // constructor() {}
    /**
     * @return {?}
     */
    function () {
        this.sideNavState = !this.sideNavState;
        this.sideNavStateSubject.next(this.sideNavState);
        if (this.sideNavState === false) {
            this.initiateAutofocus.next(false);
        }
    };
    /**
     * @return {?}
     */
    BuilderService.prototype.toggleRecipientSideNav = /**
     * @return {?}
     */
    function () {
        this.recipientSideNavState = !this.recipientSideNavState;
        this.recipientSideNavStateSubject.next(this.recipientSideNavState);
        if (this.recipientSideNavState === false) {
            this.recipientAutoFocus.next(false);
        }
    };
    /**
     * @return {?}
     */
    BuilderService.prototype.openSideNav = /**
     * @return {?}
     */
    function () {
        this.sideNavState = true;
        this.sideNavStateSubject.next(this.sideNavState);
    };
    /**
     * @return {?}
     */
    BuilderService.prototype.openRecipientSideNav = /**
     * @return {?}
     */
    function () {
        this.recipientSideNavState = true;
        this.recipientSideNavStateSubject.next(this.recipientSideNavState);
    };
    /**
     * @return {?}
     */
    BuilderService.prototype.closeSideNav = /**
     * @return {?}
     */
    function () {
        this.sideNavState = false;
        this.sideNavStateSubject.next(this.sideNavState);
        this.initiateAutofocus.next(false);
    };
    /**
     * @return {?}
     */
    BuilderService.prototype.closeRecipientSideNav = /**
     * @return {?}
     */
    function () {
        this.recipientSideNavState = false;
        this.recipientSideNavStateSubject.next(this.recipientSideNavState);
        this.recipientAutoFocus.next(false);
    };
    /**
     * @param {?} role
     * @return {?}
     */
    BuilderService.prototype.setActiveRole = /**
     * @param {?} role
     * @return {?}
     */
    function (role) {
        this.activeRole = role;
        this.hasActiveRoleSubject.next(true);
    };
    /**
     * @return {?}
     */
    BuilderService.prototype.deleteRole = /**
     * @return {?}
     */
    function () {
        if (!!this.activeRole) {
            this.deleteRoleSubject.next(this.activeRole);
            this.activeRole = null;
            this.hasActiveRoleSubject.next(false);
        }
    };
    /**
     * @return {?}
     */
    BuilderService.prototype.getSideNavState = /**
     * @return {?}
     */
    function () {
        return this.sideNavState;
    };
    /**
     * @return {?}
     */
    BuilderService.prototype.getRecipientSideNavState = /**
     * @return {?}
     */
    function () {
        return this.recipientSideNavState;
    };
    BuilderService.decorators = [
        { type: Injectable }
    ];
    return BuilderService;
}());
export { BuilderService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    BuilderService.prototype.sideNavState;
    /**
     * @type {?}
     * @private
     */
    BuilderService.prototype.recipientSideNavState;
    /**
     * @type {?}
     * @private
     */
    BuilderService.prototype.activeRole;
    /** @type {?} */
    BuilderService.prototype.sideNavStateSubject;
    /** @type {?} */
    BuilderService.prototype.recipientSideNavStateSubject;
    /** @type {?} */
    BuilderService.prototype.initiateAutofocus;
    /** @type {?} */
    BuilderService.prototype.recipientAutoFocus;
    /** @type {?} */
    BuilderService.prototype.deleteRoleSubject;
    /** @type {?} */
    BuilderService.prototype.hasActiveRoleSubject;
}
