// @verdocs/essentials/esm5/lib/builder/builder-data.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable, Injector } from '@angular/core';
import { HttpClient, HttpRequest } from '@angular/common/http';
import { Router } from '@angular/router';
import { TemplatesService, TemplatesGuardService, TemplateActions } from '@verdocs/sdk';
// import { MatSnackBar, MatSnackBarConfig } from '@angular/material/snack-bar';
import { throwError as observableThrowError, ReplaySubject, from, forkJoin, Subject, of } from 'rxjs';
import { catchError, map, delay, mergeMap } from 'rxjs/operators';
import { filter } from 'lodash';
import { findIndex, remove } from 'lodash';
import { nameToRGBA, getRGBA } from '@verdocs/sdk';
import { SnackbarService } from '../snackbar/snackbar.service';
import { EventTrackerService } from '@verdocs/event-tracker';
import { EssentialsConfigToken } from '../essentials.module';
var BuilderDataService = /** @class */ (function () {
    function BuilderDataService(http, router, templatesService, templateGuard, snackbarService, eventTracker, injector) {
        var _this = this;
        this.http = http;
        this.router = router;
        this.templatesService = templatesService;
        this.templateGuard = templateGuard;
        this.snackbarService = snackbarService;
        this.eventTracker = eventTracker;
        this.injector = injector;
        this.recipients = [];
        this.roles = [];
        this.templateSubject = new ReplaySubject(1, 500);
        this.hasFields = false;
        this.templateNameSubject = new ReplaySubject(1, 500);
        this.templatePagesSubject = new ReplaySubject(1, 500);
        this.recipientsSubject = new ReplaySubject();
        this.activeRecipientSubject = new ReplaySubject();
        this.scrollInfoSubject = new ReplaySubject();
        this.screenInfoSubject = new ReplaySubject();
        this.newTypeSubject = new Subject();
        this.newOptionSubject = new ReplaySubject(1, 100);
        this.duplicateFieldSubject = new ReplaySubject();
        this.saveStatusSubject = new ReplaySubject();
        this.renderedSubject = new ReplaySubject();
        this.activeFieldIndexSubject = new ReplaySubject(1, 100);
        this.numberOfRolesSubject = new ReplaySubject();
        this.numberOfSequenceSubject = new ReplaySubject();
        this.hasFieldsSubject = new ReplaySubject();
        this._config_token = this.injector.get(EssentialsConfigToken);
        this.rForm_backend_url = this._config_token.rForm_backend_url;
        /* Use NGRX to do the subscriptions */
        this.templateSubject.subscribe((/**
         * @param {?} template
         * @return {?}
         */
        function (template) {
            _this.template = template;
            if (template && template.name) {
                _this.templateNameSubject.next(template.name);
            }
        }));
        this.templateNameSubject.subscribe((/**
         * @param {?} name
         * @return {?}
         */
        function (name) {
            _this.templateName = name;
        }));
        /* Remove above after refactoring */
    }
    /**
     * @param {?} template_body
     * @return {?}
     */
    BuilderDataService.prototype.createTemplate = /**
     * @param {?} template_body
     * @return {?}
     */
    function (template_body) {
        var _this = this;
        return this.http
            .post(this.rForm_backend_url + '/templates', template_body)
            .pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            _this.templateSubject.next(res);
            _this.recipientsSubject.next([]);
            return res;
        })));
    };
    /**
     * @return {?}
     */
    BuilderDataService.prototype.updateLocalTemplateData = /**
     * @return {?}
     */
    function () {
        this.templateName = this.template.name;
        this.templateNameSubject.next(this.templateName);
        this.templatePages = this.template.pages.length;
        this.templatePagesSubject.next(this.templatePages);
    };
    /**
     * @param {?} template_body
     * @return {?}
     */
    BuilderDataService.prototype.updateTemplate = /**
     * @param {?} template_body
     * @return {?}
     */
    function (template_body) {
        return this.templatesService.updateTemplateObservable(template_body.id, template_body).pipe(map((/**
         * @param {?} template
         * @return {?}
         */
        function (template) { return template; })));
    };
    /**
     * @param {?} template
     * @param {?} name
     * @return {?}
     */
    BuilderDataService.prototype.updateTemplateName = /**
     * @param {?} template
     * @param {?} name
     * @return {?}
     */
    function (template, name) {
        var _this = this;
        if (template && template.id) {
            return this.templatesService.updateTemplateObservable(template.id, { name: name })
                .pipe(map((/**
             * @param {?} res
             * @return {?}
             */
            function (res) {
                template.name = res.name;
                _this.updateLocalTemplate(template);
                _this.saveStatusSubject.next('saved');
                _this.createSnackBar('Name Changed to: ' + name);
                return res.name;
            }), (/**
             * @param {?} err
             * @return {?}
             */
            function (err) {
                _this.saveStatusSubject.next('Failed to save');
                _this.createSnackBar('Failed to change Template name');
                return err;
            })));
        }
        else {
            this.templateNameSubject.next(name);
            this.saveStatusSubject.next('Name will be updated');
        }
    };
    /**
     * @param {?} templateId
     * @return {?}
     */
    BuilderDataService.prototype.getTemplateObservable = /**
     * @param {?} templateId
     * @return {?}
     */
    function (templateId) {
        return this.templatesService.getTemplateObservable(templateId);
    };
    /**
     * @param {?} templateId
     * @return {?}
     */
    BuilderDataService.prototype.getTemplate = /**
     * @param {?} templateId
     * @return {?}
     */
    function (templateId) {
        var _this = this;
        return this.templatesService.getTemplate(templateId).toPromise().then((/**
         * @param {?} template
         * @return {?}
         */
        function (template) {
            _this.updateLocalTemplate(template);
            _this.recipients = _this.sortRoles(_this.template);
            _this.template.roles = _this.recipients;
            _this.roles = _this.recipients;
            _this.recipientsSubject.next(_this.recipients);
            _this.activeRecipientSubject.next(_this.recipients[0]);
            return _this.template;
        }));
    };
    /**
     * @return {?}
     */
    BuilderDataService.prototype.watchForUpdatedTemplate = /**
     * @return {?}
     */
    function () {
        return this.templateSubject.pipe(map((/**
         * @param {?} template
         * @return {?}
         */
        function (template) { return template; })));
    };
    /**
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.updateLocalTemplate = /**
     * @param {?} template
     * @return {?}
     */
    function (template) {
        if (template) {
            this.template = template;
            if (this.template && this.template.pages) {
                this.template.pages = this.template.pages.sort((/**
                 * @param {?} a
                 * @param {?} b
                 * @return {?}
                 */
                function (a, b) {
                    return a.sequence - b.sequence;
                }));
            }
            if (this.template) {
                this.updateLocalTemplateData();
                this.checkForFields(this.template.roles);
            }
            this.recipientsSubject.next(this.template.roles.sort((/**
             * @param {?} a
             * @param {?} b
             * @return {?}
             */
            function (a, b) {
                return a.sequence - b.sequence;
            })));
            this.templateSubject.next(this.template);
        }
        else {
            this.templateNameSubject.next(null);
            this.templatePagesSubject.next(null);
            this.numberOfRolesSubject.next(null);
            this.numberOfSequenceSubject.next(null);
            this.hasFieldsSubject.next(false);
        }
    };
    /**
     * @param {?} file
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.uploadTemplateDocument = /**
     * @param {?} file
     * @param {?} template
     * @return {?}
     */
    function (file, template) {
        /** @type {?} */
        var formdata = new FormData();
        formdata.append('document', file, file.name);
        /** @type {?} */
        var req = new HttpRequest('POST', this.rForm_backend_url + '/templates/' + template.id + '/documents', formdata, {
            reportProgress: true
        });
        return this.http.request(req)
            .pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        function (res) { return res; })), catchError((/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            return observableThrowError(err);
        })));
    };
    /**
     * @param {?} templateId
     * @return {?}
     */
    BuilderDataService.prototype.getTemplateDocument = /**
     * @param {?} templateId
     * @return {?}
     */
    function (templateId) {
        var _this = this;
        this.templateDoc = this.templatesService.getAllTemplateDocumentsObservable(templateId).pipe(map((/**
         * @param {?} docs
         * @return {?}
         */
        function (docs) {
            _this.templateDoc = docs[0];
            return _this.templateDoc;
        })));
        return this.templateDoc;
    };
    /**
     * @param {?} templateId
     * @param {?} templateDoc
     * @return {?}
     */
    BuilderDataService.prototype.getTemplateDocumentFile = /**
     * @param {?} templateId
     * @param {?} templateDoc
     * @return {?}
     */
    function (templateId, templateDoc) {
        if (templateId && templateDoc) {
            return this.templatesService.getTemplateDocumentPDFObservable(templateId, templateDoc).pipe(map((/**
             * @param {?} pdf
             * @return {?}
             */
            function (pdf) {
                /** @type {?} */
                var pdfUrl = URL.createObjectURL(pdf);
                return pdfUrl;
            })));
        }
    };
    /**
     * @param {?} templateDocument
     * @return {?}
     */
    BuilderDataService.prototype.setTemplateDocument = /**
     * @param {?} templateDocument
     * @return {?}
     */
    function (templateDocument) {
        this.templateDoc = templateDocument;
    };
    /**
     * @param {?} document
     * @param {?} pageNum
     * @param {?} templateId
     * @return {?}
     */
    BuilderDataService.prototype.addTemplatePage = /**
     * @param {?} document
     * @param {?} pageNum
     * @param {?} templateId
     * @return {?}
     */
    function (document, pageNum, templateId) {
        /** @type {?} */
        var body = {
            sequence: pageNum,
            page_number: pageNum,
            document_id: document.id
        };
        return this.http.post(this.rForm_backend_url + ("/templates/" + templateId + "/pages"), body)
            .pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        function (res) { return res; })), catchError((/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            return observableThrowError(err);
        })));
    };
    /**
     * @param {?} document
     * @param {?} pageNums
     * @param {?} templateId
     * @return {?}
     */
    BuilderDataService.prototype.addTemplatePages = /**
     * @param {?} document
     * @param {?} pageNums
     * @param {?} templateId
     * @return {?}
     */
    function (document, pageNums, templateId) {
        var _this = this;
        return from(pageNums)
            .pipe(mergeMap((/**
         * @param {?} pageNum
         * @return {?}
         */
        function (pageNum) {
            /** @type {?} */
            var page = {
                sequence: pageNum,
                page_number: pageNum,
                document_id: document.id
            };
            return (/** @type {?} */ (_this.http.post(_this.rForm_backend_url + ("/templates/" + templateId + "/pages"), page).pipe(delay(200))));
        })));
    };
    /**
     * @param {?} template
     * @param {?} fieldName
     * @param {?} i
     * @param {?} j
     * @return {?}
     */
    BuilderDataService.prototype.deleteTemplateField = /**
     * @param {?} template
     * @param {?} fieldName
     * @param {?} i
     * @param {?} j
     * @return {?}
     */
    function (template, fieldName, i, j) {
        var _this = this;
        return this.http
            .delete(this.rForm_backend_url + '/templates/' + template.id + '/fields/' + fieldName.trim())
            .pipe(map((/**
         * @param {?} field
         * @return {?}
         */
        function (field) {
            /** @type {?} */
            var fieldIndex = findIndex(template.pages[i].fields, { name: fieldName });
            /** @type {?} */
            var roleIndex = findIndex(template.roles, { name: template.pages[i].fields[fieldIndex].role_name });
            if (roleIndex > -1) {
                /** @type {?} */
                var roleFieldIndex = findIndex(template.roles[roleIndex].fields, { name: fieldName });
                if (roleFieldIndex > -1) {
                    template.roles[roleIndex].fields.splice(roleFieldIndex, 1);
                }
            }
            template.pages[i].fields.splice(fieldIndex, 1);
            _this.updateLocalTemplate(template);
            _this.saveStatusSubject.next('saved');
            _this.checkForFields(template.roles);
            return template;
        })));
    };
    /**
     * @param {?} fieldNames
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.deleteTemplateFields = /**
     * @param {?} fieldNames
     * @param {?} template
     * @return {?}
     */
    function (fieldNames, template) {
        var e_1, _a;
        if (fieldNames && fieldNames.length > 0) {
            /** @type {?} */
            var pages = template.pages;
            try {
                for (var fieldNames_1 = tslib_1.__values(fieldNames), fieldNames_1_1 = fieldNames_1.next(); !fieldNames_1_1.done; fieldNames_1_1 = fieldNames_1.next()) {
                    var fieldName = fieldNames_1_1.value;
                    for (var x = 0; x < pages.length; x++) {
                        /** @type {?} */
                        var fieldIndex = findIndex(pages[x].fields, { name: fieldName });
                        if (fieldIndex >= 0) {
                            template.pages[x].fields.splice(fieldIndex, 1);
                        }
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (fieldNames_1_1 && !fieldNames_1_1.done && (_a = fieldNames_1.return)) _a.call(fieldNames_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            this.checkForFields(template.roles);
        }
    };
    /**
     * @param {?} newField
     * @param {?} i
     * @param {?} j
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.addTemplateField = /**
     * @param {?} newField
     * @param {?} i
     * @param {?} j
     * @param {?} template
     * @return {?}
     */
    function (newField, i, j, template) {
        var _this = this;
        newField.name = newField.name.trim();
        return this.http
            .post(this.rForm_backend_url + '/templates/' + newField.template_id + '/fields', newField)
            .pipe(map((/**
         * @param {?} field
         * @return {?}
         */
        function (field) {
            _this.eventTracker.createEvent({
                category: 'document',
                action: "document " + field.type + " field added",
                label: "document id: " + newField.template_id
            });
            template.pages[i].fields[j] = field;
            _this.checkForFields(template.roles);
            /** @type {?} */
            var roleIndex = findIndex(template.roles, { name: field.role_name });
            if (roleIndex >= 0) {
                template.roles[roleIndex]['fields'].push(field);
            }
            _this.templateSubject.next(template);
            return template;
        })));
    };
    /**
     * @param {?} id
     * @return {?}
     */
    BuilderDataService.prototype.getTemplateOwnerInfo = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        return this.templatesService.getTemplateOwnerInfo(id);
    };
    /**
     * @param {?} template
     * @param {?} body
     * @param {?} oldName
     * @param {?} i
     * @param {?} j
     * @return {?}
     */
    BuilderDataService.prototype.updateTemplateField = /**
     * @param {?} template
     * @param {?} body
     * @param {?} oldName
     * @param {?} i
     * @param {?} j
     * @return {?}
     */
    function (template, body, oldName, i, j) {
        var _this = this;
        oldName = oldName.trim();
        return this.http
            .put(this.rForm_backend_url + '/templates/' + template.id + '/fields/' + oldName, body)
            .pipe(map((/**
         * @param {?} field
         * @return {?}
         */
        function (field) {
            _this.eventTracker.createEvent({
                category: 'document',
                action: "document " + field.type + " field updated",
                label: "document id: " + template.id
            });
            /** @type {?} */
            var fieldIndex = findIndex(template.pages[i].fields, { name: oldName });
            if (field.page_sequence - 1 !== i) {
                template.pages[i].fields.splice(fieldIndex, 1);
                template.pages[field.page_sequence - 1].fields.push(field);
            }
            else {
                template.pages[i].fields[fieldIndex] = field;
            }
            /** @type {?} */
            var roleIndex = findIndex(template.roles, { name: field.role_name });
            if (roleIndex >= 0) {
                template = _this.updateRoleField(template, field, roleIndex, oldName);
            }
            _this.templateSubject.next(template);
            _this.saveStatusSubject.next('saved');
            return template;
        })));
    };
    /**
     * @param {?} template
     * @param {?} updated_field
     * @param {?} role_index
     * @param {?} old_name
     * @return {?}
     */
    BuilderDataService.prototype.updateRoleField = /**
     * @param {?} template
     * @param {?} updated_field
     * @param {?} role_index
     * @param {?} old_name
     * @return {?}
     */
    function (template, updated_field, role_index, old_name) {
        /** @type {?} */
        var field_index = findIndex(template.roles[role_index].fields, { name: old_name });
        if (field_index >= 0) {
            template.roles[role_index].fields[field_index] = updated_field;
        }
        else {
            template.roles[role_index].fields.push(updated_field);
        }
        return template;
    };
    /**
     * @param {?} template
     * @param {?} body
     * @param {?} oldName
     * @param {?} i
     * @param {?} j
     * @return {?}
     */
    BuilderDataService.prototype.updateDropdownField = /**
     * @param {?} template
     * @param {?} body
     * @param {?} oldName
     * @param {?} i
     * @param {?} j
     * @return {?}
     */
    function (template, body, oldName, i, j) {
        var _this = this;
        oldName = oldName.trim();
        return this.http
            .put(this.rForm_backend_url + '/templates/' + template.id + '/fields/' + oldName, body)
            .pipe(map((/**
         * @param {?} field
         * @return {?}
         */
        function (field) {
            /** @type {?} */
            var fieldIndex = findIndex(template.pages[i].fields, { name: oldName });
            template.pages[i].fields[fieldIndex] = field;
            _this.templateSubject.next(template);
            _this.saveStatusSubject.next('saved');
            return template;
        })));
    };
    /**
     * @param {?} template_id
     * @param {?} sequence_number
     * @return {?}
     */
    BuilderDataService.prototype.deleteSequence = /**
     * @param {?} template_id
     * @param {?} sequence_number
     * @return {?}
     */
    function (template_id, sequence_number) {
        return this.templatesService.deleteSequence(template_id, sequence_number);
    };
    /**
     * @param {?} field
     * @return {?}
     */
    BuilderDataService.prototype.prepareFieldDuplication = /**
     * @param {?} field
     * @return {?}
     */
    function (field) {
        this.duplicateFieldSubject.next(field);
    };
    ;
    /**
     * @param {?} roles
     * @return {?}
     */
    BuilderDataService.prototype.checkForFields = /**
     * @param {?} roles
     * @return {?}
     */
    function (roles) {
        var e_2, _a;
        /** @type {?} */
        var hasFields = true;
        if (roles && roles.length > 0) {
            try {
                for (var roles_1 = tslib_1.__values(roles), roles_1_1 = roles_1.next(); !roles_1_1.done; roles_1_1 = roles_1.next()) {
                    var role = roles_1_1.value;
                    if (role && role.type === 'signer' && role.fields.length === 0) {
                        hasFields = false;
                    }
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (roles_1_1 && !roles_1_1.done && (_a = roles_1.return)) _a.call(roles_1);
                }
                finally { if (e_2) throw e_2.error; }
            }
        }
        else {
            hasFields = false;
        }
        this.hasFields = hasFields;
        this.hasFieldsSubject.next(this.hasFields);
    };
    /**
     * @param {?} title
     * @param {?=} buttonTitle
     * @return {?}
     */
    BuilderDataService.prototype.createSnackBar = /**
     * @param {?} title
     * @param {?=} buttonTitle
     * @return {?}
     */
    function (title, buttonTitle) {
        if (buttonTitle === void 0) { buttonTitle = 'OK'; }
        //   let snackbarConfig: MatSnackBarConfig
        //   if (window.innerWidth >= 920) {
        //     snackbarConfig = {
        //       verticalPosition: 'bottom',
        //       horizontalPosition: 'left',
        //       duration: 5000
        //     }
        //   } else {
        //     snackbarConfig = {
        //       verticalPosition: 'top',
        //       duration: 5000
        //     }
        //   }
        //   this.snackbar.open(title, buttonTitle, snackbarConfig);
    };
    /**
     * @param {?} templateId
     * @return {?}
     */
    BuilderDataService.prototype.editDocsUrl = /**
     * @param {?} templateId
     * @return {?}
     */
    function (templateId) {
        /** @type {?} */
        var editDocsPageUrl = "/builder/" + templateId + "/docs";
        return editDocsPageUrl;
    };
    /**
     * @param {?} templateId
     * @return {?}
     */
    BuilderDataService.prototype.editRolesUrl = /**
     * @param {?} templateId
     * @return {?}
     */
    function (templateId) {
        /** @type {?} */
        var editRolesPageUrl = "/builder/" + templateId + "/roles";
        return editRolesPageUrl;
    };
    /**
     * @param {?} templateId
     * @return {?}
     */
    BuilderDataService.prototype.editTemplateUrl = /**
     * @param {?} templateId
     * @return {?}
     */
    function (templateId) {
        /** @type {?} */
        var editPageUrl = "/builder/" + templateId + "/fields";
        return editPageUrl;
    };
    /**
     * @param {?} templateId
     * @return {?}
     */
    BuilderDataService.prototype.previewTemplateUrl = /**
     * @param {?} templateId
     * @return {?}
     */
    function (templateId) {
        /** @type {?} */
        var reviewTemplateUrl = "/document/" + templateId;
        return reviewTemplateUrl;
    };
    /**
     * @param {?} templateId
     * @param {?} fieldId
     * @param {?} roleId
     * @param {?} i
     * @param {?} j
     * @return {?}
     */
    BuilderDataService.prototype.addFieldRole = /**
     * @param {?} templateId
     * @param {?} fieldId
     * @param {?} roleId
     * @param {?} i
     * @param {?} j
     * @return {?}
     */
    function (templateId, fieldId, roleId, i, j) {
        /** @type {?} */
        var fieldRoleInfo = JSON.stringify({
            field_id: fieldId,
            role_id: roleId
        });
        return this.http
            .post(this.rForm_backend_url + '/template/' + templateId + '/field_role', fieldRoleInfo).toPromise().then((/**
         * @param {?} fieldRole
         * @return {?}
         */
        function (fieldRole) {
            return fieldRole;
        }));
    };
    /**
     * @param {?} role
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.addRole = /**
     * @param {?} role
     * @param {?} template
     * @return {?}
     */
    function (role, template) {
        var _this = this;
        /** @type {?} */
        var templateBackend = this.rForm_backend_url + '/templates/' + template.id + '/roles';
        this.roles = [];
        this.roles = this.roles.concat(role);
        /** @type {?} */
        var roleRequests = [];
        this.roles.forEach((/**
         * @param {?} role
         * @return {?}
         */
        function (role) {
            roleRequests.push(_this.http.post(templateBackend, role));
        }));
        if (this.roles && this.roles.length > 0) {
            return forkJoin(roleRequests).pipe(map((/**
             * @param {?} savedRoles
             * @return {?}
             */
            function (savedRoles) {
                savedRoles.concat((/** @type {?} */ (savedRoles)));
                if (!template['roles']) {
                    template['roles'] = (/** @type {?} */ ([]));
                }
                /** @type {?} */
                var savedRolesWithFields = savedRoles.map((/**
                 * @param {?} newRole
                 * @return {?}
                 */
                function (newRole) {
                    newRole['fields'] = [];
                    return newRole;
                }));
                template['roles'] = template['roles'].concat(savedRolesWithFields);
                _this.recipients = _this.sortRoles(template);
                _this.roles = _this.recipients;
                _this.templateSubject.next(template);
                _this.recipientsSubject.next(_this.recipients);
                _this.recipients;
                return template;
            }), (/**
             * @param {?} err
             * @return {?}
             */
            function (err) {
                console.error(err);
                console.error('Couldn\'t save all the roles');
                return err;
            })));
        }
        else {
            return of(template);
        }
    };
    /**
     * @param {?} field
     * @return {?}
     */
    BuilderDataService.prototype.addCheckboxGroup = /**
     * @param {?} field
     * @return {?}
     */
    function (field) {
        this.newOptionSubject.next({
            type: 'checkbox_group',
            page_sequence: field.page_sequence,
            field_name: field.name,
        });
    };
    /**
     * @param {?} field
     * @return {?}
     */
    BuilderDataService.prototype.addRadioGroup = /**
     * @param {?} field
     * @return {?}
     */
    function (field) {
        this.newOptionSubject.next({
            type: 'radio_button_group',
            page_sequence: field.page_sequence,
            field_name: field.name,
        });
    };
    /**
     * @param {?} roles
     * @return {?}
     */
    BuilderDataService.prototype.getRolesInSequence = /**
     * @param {?} roles
     * @return {?}
     */
    function (roles) {
        /** @type {?} */
        var rolesInSequence = [];
        for (var roleIndex = 0; roleIndex < roles.length; roleIndex++) {
            /** @type {?} */
            var sequenceIndex = roles[roleIndex]['sequence'] - 1;
            /** @type {?} */
            var role = tslib_1.__assign({}, roles[roleIndex]);
            role['style'] = {
                backgroundColor: "" + this.getRoleColor(role.name, roleIndex)
            };
            if (!rolesInSequence[sequenceIndex]) {
                rolesInSequence[sequenceIndex] = [role];
            }
            else {
                rolesInSequence[sequenceIndex].push(role);
            }
        }
        this.numberOfSequenceSubject.next(rolesInSequence.length);
        this.numberOfRolesSubject.next(roles.length);
        return rolesInSequence;
    };
    /**
     * @param {?} roleName
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.deleteRole = /**
     * @param {?} roleName
     * @param {?} template
     * @return {?}
     */
    function (roleName, template) {
        var _this = this;
        return this.http.delete(this.rForm_backend_url + '/templates/' + template.id + '/roles/' + roleName)
            .toPromise()
            .then((/**
         * @return {?}
         */
        function () {
            remove(template.roles, (/**
             * @param {?} role
             * @return {?}
             */
            function (role) {
                var e_3, _a;
                _this.recipients = _this.sortRoles(template);
                if (role.name === roleName) {
                    /** @type {?} */
                    var fields = role.fields;
                    /** @type {?} */
                    var fieldNames = [];
                    try {
                        for (var fields_1 = tslib_1.__values(fields), fields_1_1 = fields_1.next(); !fields_1_1.done; fields_1_1 = fields_1.next()) {
                            var field = fields_1_1.value;
                            fieldNames.push(field.name);
                        }
                    }
                    catch (e_3_1) { e_3 = { error: e_3_1 }; }
                    finally {
                        try {
                            if (fields_1_1 && !fields_1_1.done && (_a = fields_1.return)) _a.call(fields_1);
                        }
                        finally { if (e_3) throw e_3.error; }
                    }
                    _this.deleteTemplateFields(fieldNames, template);
                }
                _this.templateSubject.next(template);
                _this.recipientsSubject.next(_this.recipients);
                return role.name === roleName;
            }));
        }));
    };
    /**
     * @param {?} roleNames
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.deleteRoles = /**
     * @param {?} roleNames
     * @param {?} template
     * @return {?}
     */
    function (roleNames, template) {
        var _this = this;
        /** @type {?} */
        var deleteCalls = [];
        roleNames.forEach((/**
         * @param {?} role_name
         * @return {?}
         */
        function (role_name) {
            template = tslib_1.__assign({}, template, { roles: template.roles.filter((/**
                 * @param {?} role
                 * @return {?}
                 */
                function (role) { return role.name !== role_name; })) });
            deleteCalls.push(_this.http.delete(_this.rForm_backend_url + '/templates/' + template.id + '/roles/' + role_name));
        }));
        return forkJoin(deleteCalls).pipe(map((/**
         * @return {?}
         */
        function () {
            return template;
        }), (/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            console.error(err);
            console.error('Couldn\'t delete all the roles');
            return err;
        })));
    };
    /**
     * @param {?} roles
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.updateRoles = /**
     * @param {?} roles
     * @param {?} template
     * @return {?}
     */
    function (roles, template) {
        var _this = this;
        /** @type {?} */
        var updateCalls = [];
        roles.forEach((/**
         * @param {?} role
         * @return {?}
         */
        function (role) {
            /** @type {?} */
            var body = {
                template_id: template.id,
                name: role.name.trim(),
                full_name: role.full_name,
                email: role.email,
                sequence: role.sequence,
                type: role.type,
                delegator: role.delegator,
                message: role.message,
                phone: role.phone
            };
            /** @type {?} */
            var role_index = findIndex(template.roles, { name: role.old_name });
            if (role_index >= 0) {
                template.roles[role_index] = tslib_1.__assign({}, template.roles[role_index], { template_id: template.id, name: role.name.trim(), full_name: role.full_name, email: role.email, sequence: role.sequence, type: role.type, delegator: role.delegator, message: role.message, phone: role.phone });
            }
            _this.eventTracker.createEvent({
                category: 'document',
                action: 'document role updated',
                label: "document id: " + template.id
            });
            updateCalls.push(_this.http.put(_this.rForm_backend_url + '/templates/' + template.id + '/roles/' + role.old_name, body));
        }));
        return forkJoin(updateCalls)
            .pipe(map((/**
         * @return {?}
         */
        function () {
            return template;
        }), (/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            console.error(err);
            console.error('Couldn\'t save all the roles');
            return err;
        })));
    };
    /**
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.sortRoles = /**
     * @param {?} template
     * @return {?}
     */
    function (template) {
        if (template.roles.length < 1) {
            return [];
        }
        /** @type {?} */
        var roles = template.roles.sort((/**
         * @param {?} a
         * @param {?} b
         * @return {?}
         */
        function (a, b) {
            if (a.sequence === b.sequence) {
                return a.name.toLowerCase() < b.name.toLowerCase() ? -1 : a.name.toLowerCase() > b.name.toLowerCase() ? 1 : 0;
            }
            return a.sequence - b.sequence;
        }));
        for (var x = 0; x < roles.length; x++) {
            roles[x]['rgba'] = getRGBA(x);
        }
        template.roles = roles;
        return roles;
    };
    /**
     * @param {?} name
     * @param {?=} index
     * @param {?=} template
     * @return {?}
     */
    BuilderDataService.prototype.getRoleColor = /**
     * @param {?} name
     * @param {?=} index
     * @param {?=} template
     * @return {?}
     */
    function (name, index, template) {
        if (index !== null && index > -1) {
            return getRGBA(index);
        }
        else if (template) {
            if (template.roles && template.roles.length > 0) {
                index = findIndex(template.roles, { name: name });
                if (index >= 0) {
                    return template.roles[index].rgba ? template.roles[index].rgba : getRGBA(index);
                }
            }
            else {
                return nameToRGBA(name);
            }
        }
        else {
            return nameToRGBA(name);
        }
    };
    /**
     * @param {?} role
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.updateFullName = /**
     * @param {?} role
     * @param {?} template
     * @return {?}
     */
    function (role, template) {
        var _this = this;
        for (var i = 0; i < template.pages.length; i++) {
            if (template.pages[i].fields.length > 0) {
                for (var j = 0; j < template.pages[i].fields.length; j++) {
                    if (template.pages[i].fields[j].type === 'placeholder' && template.pages[i].fields[j].setting['type'] === 'full_name') {
                        if (template.pages[i].fields[j].setting['result'] !== role.full_name &&
                            template.pages[i].fields[j].role_name === role.name) {
                            template.pages[i].fields[j].setting['result'] = role.full_name;
                            /** @type {?} */
                            var updatedField = tslib_1.__assign({}, template.pages[i].fields[j]);
                            this.updateTemplateField(template, updatedField, updatedField.name, i, j).toPromise().then((/**
                             * @return {?}
                             */
                            function () {
                                _this.templateSubject.next(template);
                            }));
                        }
                    }
                }
            }
        }
    };
    /**
     * @param {?} activeFieldIndex
     * @return {?}
     */
    BuilderDataService.prototype.setActiveFieldIndex = /**
     * @param {?} activeFieldIndex
     * @return {?}
     */
    function (activeFieldIndex) {
        this.activeFieldIndex = activeFieldIndex;
        this.activeFieldIndexSubject.next(this.activeFieldIndex);
    };
    /**
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.autoAddSigner = /**
     * @param {?} template
     * @return {?}
     */
    function (template) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                return [2 /*return*/, new Promise((/**
                     * @param {?} resolve
                     * @param {?} reject
                     * @return {?}
                     */
                    function (resolve, reject) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                        var recipients;
                        return tslib_1.__generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, this.addRole((/** @type {?} */ ({ name: 'Signer 1', type: 'signer', sequence: 1 })), template).toPromise()];
                                case 1:
                                    recipients = _a.sent();
                                    this.activeRecipientSubject.next(recipients[0]);
                                    setTimeout((/**
                                     * @return {?}
                                     */
                                    function () {
                                        resolve();
                                    }), 500);
                                    return [2 /*return*/];
                            }
                        });
                    }); }))];
            });
        });
    };
    /**
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.openTemplate = /**
     * @param {?} template
     * @return {?}
     */
    function (template) {
        if (this.canUserPreview(template)) {
            this.router.navigate(["document/" + template.id]);
        }
        else if (this.canUserEdit(template)) {
            this.router.navigate(["builder/" + template.id + "/fields"]);
        }
        else {
            this.snackbarService.open("Template is in build mode, and not ready for use.  Please check back soon.");
        }
    };
    /**
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.canUserEdit = /**
     * @param {?} template
     * @return {?}
     */
    function (template) {
        /** @type {?} */
        var response = this.templateGuard.canPerformAction(TemplateActions.WRITE, template);
        return response['canPerform'];
    };
    /**
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.canUserPreview = /**
     * @param {?} template
     * @return {?}
     */
    function (template) {
        var e_4, _a;
        /** @type {?} */
        var hasPermission = (this.templateGuard.canPerformAction(TemplateActions.READ, template)).canPerform;
        /** @type {?} */
        var canPreview;
        /** @type {?} */
        var signers = filter(template.roles, { type: 'signer' });
        canPreview = signers && signers.length > 0;
        try {
            for (var signers_1 = tslib_1.__values(signers), signers_1_1 = signers_1.next(); !signers_1_1.done; signers_1_1 = signers_1.next()) {
                var signer = signers_1_1.value;
                canPreview = signer['fields'] && signer['fields'].length > 0;
            }
        }
        catch (e_4_1) { e_4 = { error: e_4_1 }; }
        finally {
            try {
                if (signers_1_1 && !signers_1_1.done && (_a = signers_1.return)) _a.call(signers_1);
            }
            finally { if (e_4) throw e_4.error; }
        }
        return hasPermission && canPreview;
    };
    BuilderDataService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    BuilderDataService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: Router },
        { type: TemplatesService },
        { type: TemplatesGuardService },
        { type: SnackbarService },
        { type: EventTrackerService },
        { type: Injector }
    ]; };
    return BuilderDataService;
}());
export { BuilderDataService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype._config_token;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.recipients;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.template;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.templateDoc;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.templatePages;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.templateName;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.activeFieldIndex;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.roles;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.templateSubject;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.templateId;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.hasFields;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.rForm_backend_url;
    /** @type {?} */
    BuilderDataService.prototype.templateNameSubject;
    /** @type {?} */
    BuilderDataService.prototype.templatePagesSubject;
    /** @type {?} */
    BuilderDataService.prototype.recipientsSubject;
    /** @type {?} */
    BuilderDataService.prototype.activeRecipientSubject;
    /** @type {?} */
    BuilderDataService.prototype.scrollInfoSubject;
    /** @type {?} */
    BuilderDataService.prototype.screenInfoSubject;
    /** @type {?} */
    BuilderDataService.prototype.newTypeSubject;
    /** @type {?} */
    BuilderDataService.prototype.newOptionSubject;
    /** @type {?} */
    BuilderDataService.prototype.duplicateFieldSubject;
    /** @type {?} */
    BuilderDataService.prototype.saveStatusSubject;
    /** @type {?} */
    BuilderDataService.prototype.renderedSubject;
    /** @type {?} */
    BuilderDataService.prototype.activeFieldIndexSubject;
    /** @type {?} */
    BuilderDataService.prototype.numberOfRolesSubject;
    /** @type {?} */
    BuilderDataService.prototype.numberOfSequenceSubject;
    /** @type {?} */
    BuilderDataService.prototype.hasFieldsSubject;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.router;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.templatesService;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.templateGuard;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.snackbarService;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.eventTracker;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.injector;
    /* Skipping unhandled member: ;*/
}
