// @verdocs/tokens/esm2015/lib/services/authguard.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Inject, forwardRef } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { VerdocsAuthService } from './auth.service';
import { VerdocsStateService } from './state.service';
export class VerdocsAuthGuardService {
    // handle redirects based on authentication state
    /**
     * @param {?} authService
     * @param {?} stateService
     * @param {?} route
     */
    constructor(authService, stateService, route) {
        this.authService = authService;
        this.stateService = stateService;
        this.route = route;
        this.signOutRequestSubmitted = false;
        /** @type {?} */
        const snapshot = (/** @type {?} */ (this.route.snapshot));
    }
    /**
     * @return {?}
     */
    checkLogin() {
        if (!this.authService.isAuthenticated) {
            if (this.stateService.getLoginURL()) {
                this.signOut();
            }
        }
        this.stateService.watchAuthentication();
    }
    /**
     * @return {?}
     */
    checkAuthentication() {
        /** @type {?} */
        const authenticated = this.authService.isAuthenticated();
        if (!authenticated) {
            this.signOut();
        }
    }
    /**
     * @param {?=} redirectUrl
     * @return {?}
     */
    signOut(redirectUrl) {
        if (!this.signOutRequestSubmitted) {
            this.signOutRequestSubmitted = true;
            if (!this.authService.isAuthenticated()) {
                redirectUrl = window.location.href;
            }
            this.stateService.removeRCookies();
            this.stateService.removeIntercomCookie();
            /** @type {?} */
            const redirectUrlParam = (this.stateService.getLoginURL() && this.stateService.getLoginURL().includes('?') ? '&' : '?') + 'redirect_url';
            window.location.href = this.stateService.getLoginURL() + (redirectUrl ? `${redirectUrlParam}=${redirectUrl}` : '');
        }
    }
}
VerdocsAuthGuardService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
VerdocsAuthGuardService.ctorParameters = () => [
    { type: VerdocsAuthService, decorators: [{ type: Inject, args: [forwardRef((/**
                     * @return {?}
                     */
                    () => VerdocsAuthService)),] }] },
    { type: VerdocsStateService, decorators: [{ type: Inject, args: [forwardRef((/**
                     * @return {?}
                     */
                    () => VerdocsStateService)),] }] },
    { type: ActivatedRoute }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    VerdocsAuthGuardService.prototype.signOutRequestSubmitted;
    /**
     * @type {?}
     * @private
     */
    VerdocsAuthGuardService.prototype.authService;
    /**
     * @type {?}
     * @private
     */
    VerdocsAuthGuardService.prototype.stateService;
    /**
     * @type {?}
     * @private
     */
    VerdocsAuthGuardService.prototype.route;
}
