// @verdocs/essentials/esm5/lib/dialogs/error-dialogs/four-oh-one.dialog.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Injector } from '@angular/core';
import { VerdocsTokenObjectService } from '@verdocs/tokens';
import { Router } from '@angular/router';
import { EssentialsConfigToken } from '../../essentials.module';
var FourOhOneDialog = /** @class */ (function () {
    function FourOhOneDialog(router, injector, vTokenObjectService) {
        this.router = router;
        this.injector = injector;
        this.vTokenObjectService = vTokenObjectService;
        this.error = null;
        this.currentProfile = null;
        this.error_code = null;
        this.error_title = null;
        this.error_message = null;
        this._config_token = this.injector.get(EssentialsConfigToken);
        this.rSecureUrl = this._config_token.rSecure_frontend_url;
        this.rFormUrl = this._config_token.rForm_frontend_url;
    }
    /**
     * @return {?}
     */
    FourOhOneDialog.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.currentProfile = this.vTokenObjectService.getProfile();
        if (this.error && this.error.hasOwnProperty('error')) {
            this.error_code = this.error.error.code || null;
            if (this.error_code) {
                this.prepareErroContent();
            }
        }
    };
    /**
     * @return {?}
     */
    FourOhOneDialog.prototype.prepareErroContent = /**
     * @return {?}
     */
    function () {
        switch (this.error_code) {
            case 'AUTH0002':
                this.error_title = 'Authorization key is missing';
                this.error_message = 'Try after ' + (this.currentProfile ? 'signing in again.' : 'signing in or signing up.');
                break;
            case 'T000001':
                this.error_title = 'This document is not accessible';
                this.error_message = (this.currentProfile ? 'Try using a different organization/account, ' : 'Try signing in, signing up, ') + 'or requesting access from the owner of this document.';
                break;
            case 'AUTH0003':
                this.error_title = 'Cannot complete your request';
                this.error_message = 'Try after ' + (this.currentProfile ? 'signing in again.' : 'signing in or signing up.');
                break;
            default:
                this.error_title = 'Your request has been denied.';
                this.error_message = null;
                break;
        }
    };
    /**
     * @return {?}
     */
    FourOhOneDialog.prototype.goToDashboard = /**
     * @return {?}
     */
    function () {
        this.router.navigate(['/dashboard']);
    };
    /**
     * @return {?}
     */
    FourOhOneDialog.prototype.signIn = /**
     * @return {?}
     */
    function () {
        if (window) {
            window.location.href = this.rSecureUrl + "/login?redirect_url=" + (this.rFormUrl + this.router.url);
        }
    };
    /**
     * @return {?}
     */
    FourOhOneDialog.prototype.signUp = /**
     * @return {?}
     */
    function () {
        if (window) {
            window.location.href = this.rSecureUrl + "/signup?redirect_url=" + (this.rFormUrl + this.router.url);
        }
    };
    FourOhOneDialog.decorators = [
        { type: Component, args: [{
                    selector: 'four-oh-one-dialog',
                    template: "<div class=\"error__container\">\n  <div class=\"error__title\">\n    {{error_title}}\n  </div>\n  <div class=\"error__body\" *ngIf=\"error_message\">\n    {{error_message}}\n  </div>\n  <div class=\"error__action\">\n    <button class=\"error__action-button left\" mat-button *ngIf=\"currentProfile\" (click)=\"goToDashboard()\">go to dashboard</button>\n    <button class=\"error__action-button left\" mat-button *ngIf=\"!currentProfile\" (click)=\"signUp()\">sign up</button>\n    <button class=\"error__action-button\" mat-button (click)=\"signIn()\">sign in</button>\n  </div>\n</div>",
                    styles: [".error__container{padding-top:8px;padding-left:8px;padding-right:8px;max-width:640px}.error__title{padding:16px 16px 8px;font-size:20px;line-height:24px;font-weight:500;color:rgba(0,0,0,.87)}.error__body{padding:16px 16px 8px;font-size:16px;line-height:24px;color:rgba(0,0,0,.54)}.error__action{padding:8px;text-align:right}.error__action-button{color:#37474f;text-transform:uppercase}.error__action-button.left{margin-right:8px}"]
                }] }
    ];
    /** @nocollapse */
    FourOhOneDialog.ctorParameters = function () { return [
        { type: Router },
        { type: Injector },
        { type: VerdocsTokenObjectService }
    ]; };
    return FourOhOneDialog;
}());
export { FourOhOneDialog };
if (false) {
    /** @type {?} */
    FourOhOneDialog.prototype.error;
    /** @type {?} */
    FourOhOneDialog.prototype.currentProfile;
    /** @type {?} */
    FourOhOneDialog.prototype.error_code;
    /** @type {?} */
    FourOhOneDialog.prototype.error_title;
    /** @type {?} */
    FourOhOneDialog.prototype.error_message;
    /** @type {?} */
    FourOhOneDialog.prototype.rSecureUrl;
    /** @type {?} */
    FourOhOneDialog.prototype.rFormUrl;
    /**
     * @type {?}
     * @private
     */
    FourOhOneDialog.prototype._config_token;
    /**
     * @type {?}
     * @private
     */
    FourOhOneDialog.prototype.router;
    /**
     * @type {?}
     * @private
     */
    FourOhOneDialog.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    FourOhOneDialog.prototype.vTokenObjectService;
}
