// @verdocs/essentials/esm5/lib/modal/modal.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { find } from 'lodash';
var ModalService = /** @class */ (function () {
    function ModalService() {
        this.modals = [];
        this.subject = new Subject();
    }
    /**
     * @param {?} modal
     * @return {?}
     */
    ModalService.prototype.add = /**
     * @param {?} modal
     * @return {?}
     */
    function (modal) {
        this.modals.push(modal);
    };
    /**
     * @param {?} id
     * @return {?}
     */
    ModalService.prototype.open = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        /** @type {?} */
        var modal = find(this.modals, { id: id });
    };
    /**
     * @param {?} id
     * @return {?}
     */
    ModalService.prototype.close = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        /** @type {?} */
        var modal = find(this.modals, { id: id });
        modal.close();
    };
    /**
     * @param {?} componentData
     * @return {?}
     */
    ModalService.prototype.setComponentData = /**
     * @param {?} componentData
     * @return {?}
     */
    function (componentData) {
        this.componentData = componentData;
        this.subject.next(componentData);
    };
    /**
     * @return {?}
     */
    ModalService.prototype.getComponentData = /**
     * @return {?}
     */
    function () {
        return this.subject.asObservable();
    };
    ModalService.decorators = [
        { type: Injectable }
    ];
    return ModalService;
}());
export { ModalService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ModalService.prototype.modals;
    /**
     * @type {?}
     * @private
     */
    ModalService.prototype.componentData;
    /**
     * @type {?}
     * @private
     */
    ModalService.prototype.subject;
}
