// @verdocs/event-tracker/esm5/lib/event-tracker.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostListener, Input } from '@angular/core';
var EventTrackerDirective = /** @class */ (function () {
    function EventTrackerDirective() {
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    EventTrackerDirective.prototype.onClick = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if (navigator && navigator.userAgent) {
            this.analytics = window['analytics'];
            if (this.event.action) {
                this.analytics.track(this.event.action, {
                    category: this.event.category,
                    label: this.event.label,
                    value: this.event.value
                });
            }
        }
    };
    EventTrackerDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[VerdocsEventTracker]'
                },] }
    ];
    /** @nocollapse */
    EventTrackerDirective.ctorParameters = function () { return []; };
    EventTrackerDirective.propDecorators = {
        event: [{ type: Input, args: ['VerdocsEventTracker',] }],
        onClick: [{ type: HostListener, args: ['click', ['$event'],] }]
    };
    return EventTrackerDirective;
}());
export { EventTrackerDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    EventTrackerDirective.prototype.analytics;
    /** @type {?} */
    EventTrackerDirective.prototype.event;
}
