// @verdocs/essentials/esm5/lib/functions/conversion.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @param {?} field
 * @return {?}
 */
export function getDefaultHeight(field) {
    /** @type {?} */
    var height = 0;
    switch (field.type) {
        case 'signature':
        case 'initial':
            height = 36;
            break;
        case 'checkbox':
        case 'checkbox_group':
        case 'radio_button_group':
            height = 13.5;
            break;
        case 'attachment':
        case 'payment':
            height = 24;
            break;
        default:
            height = field.setting['height'] || 0;
            break;
    }
    return height;
}
/**
 * @param {?} field
 * @return {?}
 */
export function getDefaultWidth(field) {
    /** @type {?} */
    var width = 0;
    switch (field.type) {
        case 'signature':
        case 'initial':
            width = 82.63636363636;
            break;
        case 'checkbox':
        case 'checkbox_group':
        case 'radio_button_group':
            width = 13.5;
            break;
        case 'attachment':
        case 'payment':
            width = 24;
            break;
        case 'date':
            width = 64;
            break;
        case 'dropdown':
            width = field.setting['width'] || 64;
            break;
        default:
            width = field.setting['width'] || 0;
            break;
    }
    return width;
}
/**
 * @param {?} y
 * @param {?} fieldHeight
 * @param {?} iTextHeight
 * @param {?} yRatio
 * @return {?}
 */
export function getRTop(y, fieldHeight, iTextHeight, yRatio) {
    return iTextHeight - ((y + fieldHeight) * yRatio);
}
/**
 * @param {?} x
 * @param {?} ratio
 * @return {?}
 */
export function getRLeft(x, ratio) {
    return x * ratio;
}
/**
 * @param {?} y
 * @param {?} ratio
 * @return {?}
 */
export function getRValue(y, ratio) {
    return y * ratio;
}
/**
 * @param {?} field
 * @return {?}
 */
export function getUIFieldStyling(field) {
    /** @type {?} */
    var height = getDefaultHeight(field);
    /** @type {?} */
    var style = {
        left: field.setting.x + 'px',
        bottom: field.setting.y + 'px',
        height: '100%',
        width: '100%'
    };
    switch (field.type) {
        case 'textbox':
        case 'date':
        case 'timestamp':
        case 'placeholder':
        case 'dropdown':
            style['maxHeight'] = height + 'px';
            style['maxWidth'] = field.setting.width + 'px';
            break;
        case 'checkbox':
        case 'checkbox_group':
        case 'radio_button_group':
            style['maxHeight'] = 13.5 + 'px';
            style['maxWidth'] = 13.5 + 'px';
            break;
        case 'attachment':
        case 'payment':
            style['maxHeight'] = 24 + 'px';
            style['maxWidth'] = 24 + 'px';
            break;
        case 'signature':
        case 'initial':
            style['maxHeight'] = 36 + 'px';
            style['maxWidth'] = 82.63636363636 + 'px';
            style['line-height'] = 36 + 'px';
            break;
        default:
            break;
    }
    return style;
}
/**
 * @param {?} image
 * @return {?}
 */
export function blobToBase64(image) {
    /** @type {?} */
    var fileReader = new FileReader();
    return new Promise((/**
     * @param {?} resolve
     * @param {?} reject
     * @return {?}
     */
    function (resolve, reject) {
        fileReader.onerror = (/**
         * @return {?}
         */
        function () {
            reject(new DOMException("Problem reading blob."));
        });
        fileReader.onload = (/**
         * @return {?}
         */
        function () {
            resolve(fileReader.result);
        });
        fileReader.readAsDataURL(image);
    }));
}
