// @verdocs/tokens/esm2015/lib/dialogs/verify-email-dialog.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Inject, forwardRef } from '@angular/core';
import { MatDialogRef, MAT_DIALOG_DATA } from '@angular/material';
import { VerdocsAuthGuardService } from '../services/authguard.service';
import { VerdocsStateService } from '../services/state.service';
export class VerifyEmailDialogComponent {
    /**
     * @param {?} dialogRef
     * @param {?} vTokenAuthGuard
     * @param {?} vTokenStateService
     * @param {?} data
     */
    constructor(dialogRef, vTokenAuthGuard, vTokenStateService, data) {
        this.dialogRef = dialogRef;
        this.vTokenAuthGuard = vTokenAuthGuard;
        this.vTokenStateService = vTokenStateService;
        this.data = data;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        setInterval((/**
         * @return {?}
         */
        () => {
            if (this.vTokenStateService.getEmailVerification()) {
                this.closeDialog();
            }
        }), 10000);
    }
    /**
     * @return {?}
     */
    signOut() {
        this.vTokenAuthGuard.signOut(window.location.href);
    }
    /**
     * @return {?}
     */
    resendEmailVerification() {
        this.data.resendEmailVerification();
    }
    /**
     * @return {?}
     */
    closeDialog() {
        this.dialogRef.close();
    }
}
VerifyEmailDialogComponent.decorators = [
    { type: Component, args: [{
                selector: 'verdocs-verify-email-dialog.component',
                template: "<div class=\"verify-email__container\">\n    <div class=\"verify-email__title\">\n        Confirm your email address\n    </div>\n    <div class=\"verify-email__content\">\n        We've sent a confirmation email to you at:\n        <br>\n        <span class=\"verify-email__content-email\">\n        {{data.emailAddress}}\n      </span>\n    </div>\n    <div class=\"verify-email__action\">\n        <button mat-button class=\"verify-email__action-button\" (click)=\"signOut()\">LOGOUT</button>\n        <button mat-button class=\"verify-email__action-button\" (click)=\"resendEmailVerification()\">RESEND EMAIL</button>\n    </div>\n</div>\n",
                styles: [".verify-email__container{width:640px}@media only screen and (max-width:960px){.verify-email__container{width:auto}}.verify-email__title{padding:0 0 14px;color:rgba(0,0,0,.87);font-size:20px;font-weight:500}.verify-email__content{color:rgba(0,0,0,.54)}.verify-email__content-email{padding-top:32px;display:block}.verify-email__action{height:52px;padding-top:10.5px;margin:24px -24px -24px 0;text-align:right}.verify-email__action-button{color:#3f5894;text-transform:uppercase;margin-left:-8px}"]
            }] }
];
/** @nocollapse */
VerifyEmailDialogComponent.ctorParameters = () => [
    { type: MatDialogRef },
    { type: VerdocsAuthGuardService, decorators: [{ type: Inject, args: [forwardRef((/**
                     * @return {?}
                     */
                    () => VerdocsAuthGuardService)),] }] },
    { type: VerdocsStateService, decorators: [{ type: Inject, args: [forwardRef((/**
                     * @return {?}
                     */
                    () => VerdocsStateService)),] }] },
    { type: undefined, decorators: [{ type: Inject, args: [MAT_DIALOG_DATA,] }] }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    VerifyEmailDialogComponent.prototype.dialogRef;
    /**
     * @type {?}
     * @private
     */
    VerifyEmailDialogComponent.prototype.vTokenAuthGuard;
    /**
     * @type {?}
     * @private
     */
    VerifyEmailDialogComponent.prototype.vTokenStateService;
    /** @type {?} */
    VerifyEmailDialogComponent.prototype.data;
}
