// @verdocs/sdk/esm2015/lib/validators/validator.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable, Injector } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { regParse } from '../functions/regex';
import { VSDKConfigToken } from '../angular-api.module';
export class ValidatorService {
    /**
     * @param {?} http
     * @param {?} injector
     */
    constructor(http, injector) {
        this.http = http;
        this.injector = injector;
        this.validators = [];
        this.validatorsObject = {};
        this._config_token = this.injector.get(VSDKConfigToken);
        this._rForm_backend_url = this._config_token.rForm_backend_url;
    }
    /**
     * @return {?}
     */
    getValidatorsArray() {
        if (this.validators && this.validators.length > 0) {
            return this.validators;
        }
        else {
            return this.http.get(this._rForm_backend_url + '/validators').toPromise().then((/**
             * @param {?} response
             * @return {?}
             */
            (response) => {
                this.validators = response;
                return (/** @type {?} */ (this.validators));
            }), (/**
             * @param {?} err
             * @return {?}
             */
            (err) => {
                return null;
            }));
        }
    }
    /**
     * @return {?}
     */
    getValidatorsObject() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            this.validatorsObject = {};
            yield this.getValidatorsArray();
            for (let i = 0; i < this.validators['length']; i++) {
                /** @type {?} */
                const name = this.validators[i]['name'];
                this.validatorsObject[name] = regParse(this.validators[i]['regex']);
            }
            return this.validatorsObject;
        });
    }
}
ValidatorService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ValidatorService.ctorParameters = () => [
    { type: HttpClient },
    { type: Injector }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    ValidatorService.prototype._config_token;
    /**
     * @type {?}
     * @private
     */
    ValidatorService.prototype._rForm_backend_url;
    /**
     * @type {?}
     * @private
     */
    ValidatorService.prototype.validators;
    /**
     * @type {?}
     * @private
     */
    ValidatorService.prototype.validatorsObject;
    /**
     * @type {?}
     * @private
     */
    ValidatorService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    ValidatorService.prototype.injector;
}
