// @verdocs/sdk/esm2015/lib/guards/auth-guard.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Inject } from '@angular/core';
import { DOCUMENT } from "@angular/common";
import { VerdocsAuthService, VerdocsAuthGuardService, VerdocsStateService } from '@verdocs/tokens';
export class AuthGuardService {
    /**
     * @param {?} vTokenAuth
     * @param {?} vTokenAuthGuard
     * @param {?} vTokenStateService
     * @param {?} document
     */
    constructor(vTokenAuth, vTokenAuthGuard, vTokenStateService, document) {
        this.vTokenAuth = vTokenAuth;
        this.vTokenAuthGuard = vTokenAuthGuard;
        this.vTokenStateService = vTokenStateService;
        this.document = document;
    }
    /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    canActivate(route, state) {
        /** @type {?} */
        const signerToken = this.vTokenStateService.getOtherCookieObject('signer_token');
        /** @type {?} */
        const currentPath = route && route.routeConfig && route.routeConfig.path ? route.routeConfig.path : route.path;
        /** @type {?} */
        const matchesPath = currentPath ? currentPath.match(/(:id\/role\/:role)/g) || currentPath.includes('envelope') : false;
        if (route.routeConfig && !currentPath) {
            return true;
        }
        else if (signerToken && !!matchesPath) {
            return true;
        }
        else {
            return this.checkLogin(route, state.url);
        }
    }
    /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    canActivateChild(route, state) {
        return this.canActivate(route, state);
    }
    /**
     * @param {?} route
     * @param {?} url
     * @return {?}
     */
    checkLogin(route, url) {
        if (this.vTokenAuth.isAuthenticated() || this.isGuestLink(route)) {
            return true;
        }
        this.vTokenAuthGuard.signOut();
        return false;
    }
    /**
     * @param {?} route
     * @return {?}
     */
    canLoad(route) {
        /** @type {?} */
        const url = `/${route.path}`;
        return this.checkLogin(route, url);
    }
    /**
     * @param {?} route
     * @param {?} state
     * @return {?}
     */
    canLoadBuilder(route, state) {
    }
    /**
     * @param {?} route
     * @return {?}
     */
    isGuestLink(route) {
        /** @type {?} */
        const signer_token = this.vTokenStateService.getOtherCookie('signer_token');
        /** @type {?} */
        const currentPath = route && route.routeConfig && route.routeConfig.path ? route.routeConfig.path : route.path;
        /** @type {?} */
        const matchesPath = currentPath ? currentPath.match(/(:id\/role\/:role)/g) || currentPath.includes('envelope') : false;
        if (route && !!matchesPath && !!signer_token) {
            if (signer_token) {
                return true;
            }
            this.vTokenStateService.setTimer(true);
        }
        return route._routeConfig && route._routeConfig.path.match(/(:id\/role\/:role)/g) && !!signer_token;
    }
}
AuthGuardService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
AuthGuardService.ctorParameters = () => [
    { type: VerdocsAuthService },
    { type: VerdocsAuthGuardService },
    { type: VerdocsStateService },
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    AuthGuardService.prototype.vTokenAuth;
    /**
     * @type {?}
     * @private
     */
    AuthGuardService.prototype.vTokenAuthGuard;
    /**
     * @type {?}
     * @private
     */
    AuthGuardService.prototype.vTokenStateService;
    /**
     * @type {?}
     * @private
     */
    AuthGuardService.prototype.document;
}
