// @verdocs/essentials/esm5/lib/models/dynamic-field.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @template T
 */
var /**
 * @template T
 */
DynamicField = /** @class */ (function () {
    function DynamicField(field_params) {
        if (field_params === void 0) { field_params = {}; }
        this.value = field_params.value;
        this.key = field_params.key || '';
        this.required = !!field_params.required;
        this.order = field_params.order === undefined ? 1 : field_params.order;
        this.controlType = field_params.controlType || '';
        this.validator = field_params.validator === undefined ? '' : field_params.validator;
        this.maxLength = field_params.maxLength == null ? null : field_params.maxLength;
        this.showCounter = field_params.showCounter === false ? false : field_params.showCounter;
        this.prepared = field_params.prepared;
        this.recipientRole = field_params.recipientRole;
        this.options = field_params.options;
        this.field_name = field_params.field_name;
        this.min_checked = field_params.min_checked;
        this.max_checked = field_params.max_checked;
        this.temp_sig = field_params.temp_sig;
        this.sig_id = field_params.sig_id;
        this.temp_int = field_params.temp_int;
        this.int_id = field_params.int_id;
        this.dirty = field_params.dirty;
        this.full_name = field_params.full_name;
        if (field_params.url) {
            this.url = field_params.url;
        }
        if (field_params.name) {
            this.name = field_params.name;
        }
    }
    return DynamicField;
}());
/**
 * @template T
 */
export { DynamicField };
if (false) {
    /** @type {?} */
    DynamicField.prototype.value;
    /** @type {?} */
    DynamicField.prototype.key;
    /** @type {?} */
    DynamicField.prototype.required;
    /** @type {?} */
    DynamicField.prototype.order;
    /** @type {?} */
    DynamicField.prototype.controlType;
    /** @type {?} */
    DynamicField.prototype.validator;
    /** @type {?} */
    DynamicField.prototype.maxLength;
    /** @type {?} */
    DynamicField.prototype.showCounter;
    /** @type {?} */
    DynamicField.prototype.url;
    /** @type {?} */
    DynamicField.prototype.name;
    /** @type {?} */
    DynamicField.prototype.prepared;
    /** @type {?} */
    DynamicField.prototype.recipientRole;
    /** @type {?} */
    DynamicField.prototype.options;
    /** @type {?} */
    DynamicField.prototype.field_name;
    /** @type {?} */
    DynamicField.prototype.min_checked;
    /** @type {?} */
    DynamicField.prototype.max_checked;
    /** @type {?} */
    DynamicField.prototype.temp_sig;
    /** @type {?} */
    DynamicField.prototype.sig_id;
    /** @type {?} */
    DynamicField.prototype.temp_int;
    /** @type {?} */
    DynamicField.prototype.int_id;
    /** @type {?} */
    DynamicField.prototype.dirty;
    /** @type {?} */
    DynamicField.prototype.full_name;
}
