// @verdocs/essentials/fesm5/verdocs-essentials.js
import { __assign, __values, __awaiter, __generator } from 'tslib';
import { getRGBA, nameToRGBA, TemplateActions, TemplatesService, TemplatesGuardService, FieldData, DocumentsService, AngularApiModule } from '@verdocs/sdk';
import { findIndex, remove, filter, find } from 'lodash';
import printJS from 'print-js';
import { Injectable, Inject, PLATFORM_ID, Injector, ViewEncapsulation, RendererFactory2, Component, InjectionToken, NgModule } from '@angular/core';
import { isPlatformBrowser, DOCUMENT, CommonModule } from '@angular/common';
import { MatSnackBar } from '@angular/material/snack-bar';
import { HttpRequest, HttpClient } from '@angular/common/http';
import { Router, NavigationEnd, ActivatedRoute } from '@angular/router';
import { ReplaySubject, Subject, throwError, from, forkJoin, of, BehaviorSubject } from 'rxjs';
import { VerdocsStateService, VerdocsTokenObjectService } from '@verdocs/tokens';
import { ProfileModel, ProfileCollection } from '@verdocs/profiles';
import { MatDialogRef, MatDialog, MatDialogModule } from '@angular/material/dialog';
import { map, catchError, mergeMap, delay, filter as filter$1, switchMap } from 'rxjs/operators';
import { EventTrackerService } from '@verdocs/event-tracker';
import { v4 } from 'uuid';
import { Title, Meta } from '@angular/platform-browser';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @template T
 */
var  /**
 * @template T
 */
DynamicField = /** @class */ (function () {
    function DynamicField(field_params) {
        if (field_params === void 0) { field_params = {}; }
        this.value = field_params.value;
        this.key = field_params.key || '';
        this.required = !!field_params.required;
        this.order = field_params.order === undefined ? 1 : field_params.order;
        this.controlType = field_params.controlType || '';
        this.validator = field_params.validator === undefined ? '' : field_params.validator;
        this.maxLength = field_params.maxLength == null ? null : field_params.maxLength;
        this.showCounter = field_params.showCounter === false ? false : field_params.showCounter;
        this.prepared = field_params.prepared;
        this.recipientRole = field_params.recipientRole;
        this.options = field_params.options;
        this.field_name = field_params.field_name;
        this.min_checked = field_params.min_checked;
        this.max_checked = field_params.max_checked;
        this.temp_sig = field_params.temp_sig;
        this.sig_id = field_params.sig_id;
        this.temp_int = field_params.temp_int;
        this.int_id = field_params.int_id;
        this.dirty = field_params.dirty;
        this.full_name = field_params.full_name;
        if (field_params.url) {
            this.url = field_params.url;
        }
        if (field_params.name) {
            this.name = field_params.name;
        }
    }
    return DynamicField;
}());
if (false) {
    /** @type {?} */
    DynamicField.prototype.value;
    /** @type {?} */
    DynamicField.prototype.key;
    /** @type {?} */
    DynamicField.prototype.required;
    /** @type {?} */
    DynamicField.prototype.order;
    /** @type {?} */
    DynamicField.prototype.controlType;
    /** @type {?} */
    DynamicField.prototype.validator;
    /** @type {?} */
    DynamicField.prototype.maxLength;
    /** @type {?} */
    DynamicField.prototype.showCounter;
    /** @type {?} */
    DynamicField.prototype.url;
    /** @type {?} */
    DynamicField.prototype.name;
    /** @type {?} */
    DynamicField.prototype.prepared;
    /** @type {?} */
    DynamicField.prototype.recipientRole;
    /** @type {?} */
    DynamicField.prototype.options;
    /** @type {?} */
    DynamicField.prototype.field_name;
    /** @type {?} */
    DynamicField.prototype.min_checked;
    /** @type {?} */
    DynamicField.prototype.max_checked;
    /** @type {?} */
    DynamicField.prototype.temp_sig;
    /** @type {?} */
    DynamicField.prototype.sig_id;
    /** @type {?} */
    DynamicField.prototype.temp_int;
    /** @type {?} */
    DynamicField.prototype.int_id;
    /** @type {?} */
    DynamicField.prototype.dirty;
    /** @type {?} */
    DynamicField.prototype.full_name;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @param {?} fields
 * @param {?} mode
 * @return {?}
 */
function buildFields(fields, mode) {
    /*
       * PDF 72 DPI
       * Width: 612px
       * Height: 792px
       * */
    /*
         * PDF 72 DPI
         * Width: 612px
         * Height: 792px
         * */
    /** @type {?} */
    var _fields = [];
    /** @type {?} */
    var order = 0;
    /** @type {?} */
    var role;
    switch (mode) {
        case 'signerview':
        case 'prepareview':
        case 'preview':
            role = 'recipient_role';
            break;
        case 'liveview':
            role = 'role_name';
            break;
        default:
            role = 'recipient_role';
            break;
    }
    /** @type {?} */
    var setting = 'setting';
    for (var fieldIndex in fields) {
        if (fields.hasOwnProperty(fieldIndex) && fields[fieldIndex]) {
            /** @type {?} */
            var field = fields[fieldIndex];
            /** @type {?} */
            var keyNameSuffix = '';
            if (!!field['settings']) {
                setting = 'settings';
            }
            ;
            switch (field.type.toLowerCase()) {
                case 'attachment':
                case 'payment':
                    field[setting]['width'] = 24;
                    field[setting]['height'] = 24;
                    break;
                case 'checkbox':
                case 'checkbox_group':
                case 'radio_button_group':
                    field[setting]['width'] = 13.5;
                    field[setting]['height'] = 13.5;
                    keyNameSuffix = field['optionId'] ? '-' + field['optionId'] : '';
                    break;
                case 'signature':
                    field[setting]['width'] = 82.63636363636;
                    field[setting]['height'] = 36;
                    break;
                case 'initial':
                    field[setting]['width'] = 82.63636363636;
                    field[setting]['height'] = 36;
                    break;
                default:
                    break;
            }
            /** @type {?} */
            var fieldValues = new DynamicField({
                key: field.name + keyNameSuffix,
                value: field[setting]['result'] || '',
                required: field.required,
                controlType: field.type,
                order: order++,
                validator: field.validator,
                showCounter: false,
                prepared: field.prepared || false,
                recipientRole: field[role],
                field_name: field.name,
                dirty: false
            });
            switch (field.type) {
                case 'attachment':
                    fieldValues.value = field[setting]['name'] ? field[setting]['name'] : '';
                    break;
                case 'checkbox':
                    fieldValues.value = field[setting]['result'] === false ? false : true;
                    break;
                case 'payment':
                    fieldValues.value = field[setting]['payment_id'] ? field[setting]['payment_id'] : '';
                    break;
                case 'checkbox_group':
                    fieldValues.value = field[setting]['checked'];
                    if (field.required || field[setting]['maximum_checked'] > 0) {
                        fieldValues['min_checked'] = field[setting]['minimum_checked'];
                        fieldValues['max_checked'] = field[setting]['maximum_checked'];
                    }
                    break;
                case 'radio_button_group':
                    fieldValues.value = field[setting]['selected'];
                    break;
                case 'dropdown':
                    fieldValues.value = field[setting]['value'];
                    fieldValues['options'] = field[setting]['options'];
                    break;
                default:
                    break;
            }
            ;
            if (field[setting]['url']) {
                fieldValues['url'] = field[setting]['url'];
                fieldValues['name'] = field[setting]['name'];
                fieldValues['value'] = field[setting]['name'];
            }
            if (fields[fieldIndex]) {
                _fields[fieldIndex] = fieldValues;
            }
        }
    }
    return _fields;
}
/**
 * @param {?} fields
 * @param {?} pageNum
 * @return {?}
 */
function getFieldsMap(fields, pageNum) {
    /** @type {?} */
    var fieldsMap = {};
    for (var fieldIndex in fields) {
        if (fields.hasOwnProperty(fieldIndex) && fields[fieldIndex]) {
            /** @type {?} */
            var field = fields[fieldIndex];
            /** @type {?} */
            var keyNameSuffix = '';
            switch (field.type.toLowerCase()) {
                case 'checkbox':
                case 'checkbox_group':
                case 'radio_button_group':
                    keyNameSuffix = field['optionId'] ? '-' + field['optionId'] : '';
                    break;
                default:
                    break;
            }
            fieldsMap[field.name + keyNameSuffix] = {
                field_name: field.name,
                keyNameSuffix: field['optionId'] ? '-' + field['optionId'] : null,
                option_id: field['optionId'],
                field_pageNum: pageNum,
                field_field_index: fieldIndex
            };
        }
    }
    return fieldsMap;
}
/**
 * @param {?} fields
 * @param {?} _fields
 * @param {?} mode
 * @param {?} browserType
 * @return {?}
 */
function updateElementStyles(fields, _fields, mode, browserType) {
    if (fields && _fields && fields.length === _fields.length) {
        for (var x = 0; x < _fields.length; x++) {
            /** @type {?} */
            var fieldIndex = findIndex(fields, { name: _fields[x].field_name });
            if (fieldIndex >= 0) {
                _fields[x]['initialStyle'] = getInputStyle(fields[fieldIndex], mode, browserType);
                if (_fields[x].controlType.toLowerCase() === 'checkbox' || _fields[x].controlType.toLowerCase() === 'checkbox_group' || _fields[x].controlType.toLowerCase() === 'radio_button_group') {
                    _fields[x]['checkboxStyle'] = generateCheckboxLabel(fields[fieldIndex].required);
                }
            }
        }
    }
    return _fields;
}
/**
 * @param {?} field
 * @param {?} mode
 * @param {?} browserType
 * @return {?}
 */
function getInputStyle(field, mode, browserType) {
    /** @type {?} */
    var fontSize = 11;
    /** @type {?} */
    var setting = 'setting';
    if (!!field['settings']) {
        setting = 'settings';
    }
    /** @type {?} */
    var inputStyle = {
        'height': '100%',
        'width': '100%',
        'background': 'none'
    };
    if ((field[setting] && field[setting].font_size) || field.type === 'date' || field.type === 'signature' || field.type === 'initial' || field.type === 'timestamp') {
        inputStyle['fontSize'] = fontSize + 'px';
        inputStyle['letterSpacing'] = '.3px !important';
    }
    if (field.type === 'dropdown') {
        inputStyle['fontSize'] = '10.8px';
        delete inputStyle.background;
    }
    if (field.type === 'textbox') {
        inputStyle['fontSize'] = fontSize + 'px';
        inputStyle['letterSpacing'] = getLetterSpacing(browserType) + 'px';
    }
    if (field['required']) {
        inputStyle['border'] = '1px solid #cc0000';
    }
    if (field['prepared'] && field['prepared'] === true && mode !== 'prepareview') {
        inputStyle['visibility'] = 'hidden';
    }
    if (field[setting] && field[setting].color) {
        inputStyle['color'] = field[setting].color;
    }
    if (field[setting] && field[setting].upperCase) {
        inputStyle['textTransform'] = 'uppercase';
    }
    if (field[setting] && field[setting].leading > 0) {
        inputStyle['lineHeight'] = rescale(1, field[setting].leading + .5) + "px";
    }
    else if (field.leading && field.leading > 0) {
        inputStyle['lineHeight'] = rescale(1, field.leading + .5) + "px";
    }
    return inputStyle;
}
/**
 * @param {?} required
 * @return {?}
 */
function generateCheckboxLabel(required) {
    /** @type {?} */
    var labelStyle = {
        position: 'absolute',
        top: 0,
        left: 0,
        cursor: 'pointer',
        height: "13.5px",
        width: "13.5px",
        backgroundColor: 'transparent',
        border: '1px solid #777'
    };
    if (required === true) {
        labelStyle['boxShadow'] = '0 0 0 1px #cc0000';
    }
    return labelStyle;
}
;
/**
 * @param {?} browserType
 * @return {?}
 */
function getLetterSpacing(browserType) {
    switch (browserType) {
        case 'opera':
            return -0.0018;
        case 'firefox':
            return -0.23594210526315787;
        case 'ie':
            return -0.0019;
        case 'edge':
            return -0.0019;
        case 'chrome':
            return -0.0018;
        case 'safari':
            return -0.0018;
        case 'blink':
            return -0.0018;
        default:
            return -0.0018;
    }
}
/**
 * @param {?} r
 * @param {?} n
 * @return {?}
 */
function rescale(r, n) {
    return r * n;
}
/**
 * @param {?} fields
 * @return {?}
 */
function prepareFieldsForSigner(fields) {
    /** @type {?} */
    var fieldsForCurrentSigner = new Array(fields.length);
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @param {?} field
 * @return {?}
 */
function getDefaultHeight(field) {
    /** @type {?} */
    var height = 0;
    switch (field.type) {
        case 'signature':
        case 'initial':
            height = 36;
            break;
        case 'checkbox':
        case 'checkbox_group':
        case 'radio_button_group':
            height = 13.5;
            break;
        case 'attachment':
        case 'payment':
            height = 24;
            break;
        default:
            height = field.setting['height'] || 0;
            break;
    }
    return height;
}
/**
 * @param {?} field
 * @return {?}
 */
function getDefaultWidth(field) {
    /** @type {?} */
    var width = 0;
    switch (field.type) {
        case 'signature':
        case 'initial':
            width = 82.63636363636;
            break;
        case 'checkbox':
        case 'checkbox_group':
        case 'radio_button_group':
            width = 13.5;
            break;
        case 'attachment':
        case 'payment':
            width = 24;
            break;
        case 'date':
            width = 64;
            break;
        case 'dropdown':
            width = field.setting['width'] || 64;
            break;
        default:
            width = field.setting['width'] || 0;
            break;
    }
    return width;
}
/**
 * @param {?} y
 * @param {?} fieldHeight
 * @param {?} iTextHeight
 * @param {?} yRatio
 * @return {?}
 */
function getRTop(y, fieldHeight, iTextHeight, yRatio) {
    return iTextHeight - ((y + fieldHeight) * yRatio);
}
/**
 * @param {?} x
 * @param {?} ratio
 * @return {?}
 */
function getRLeft(x, ratio) {
    return x * ratio;
}
/**
 * @param {?} y
 * @param {?} ratio
 * @return {?}
 */
function getRValue(y, ratio) {
    return y * ratio;
}
/**
 * @param {?} field
 * @return {?}
 */
function getUIFieldStyling(field) {
    /** @type {?} */
    var height = getDefaultHeight(field);
    /** @type {?} */
    var style = {
        left: field.setting.x + 'px',
        bottom: field.setting.y + 'px',
        height: '100%',
        width: '100%'
    };
    switch (field.type) {
        case 'textbox':
        case 'date':
        case 'timestamp':
        case 'placeholder':
        case 'dropdown':
            style['maxHeight'] = height + 'px';
            style['maxWidth'] = field.setting.width + 'px';
            break;
        case 'checkbox':
        case 'checkbox_group':
        case 'radio_button_group':
            style['maxHeight'] = 13.5 + 'px';
            style['maxWidth'] = 13.5 + 'px';
            break;
        case 'attachment':
        case 'payment':
            style['maxHeight'] = 24 + 'px';
            style['maxWidth'] = 24 + 'px';
            break;
        case 'signature':
        case 'initial':
            style['maxHeight'] = 36 + 'px';
            style['maxWidth'] = 82.63636363636 + 'px';
            style['line-height'] = 36 + 'px';
            break;
        default:
            break;
    }
    return style;
}
/**
 * @param {?} image
 * @return {?}
 */
function blobToBase64(image) {
    /** @type {?} */
    var fileReader = new FileReader();
    return new Promise((/**
     * @param {?} resolve
     * @param {?} reject
     * @return {?}
     */
    function (resolve, reject) {
        fileReader.onerror = (/**
         * @return {?}
         */
        function () {
            reject(new DOMException("Problem reading blob."));
        });
        fileReader.onload = (/**
         * @return {?}
         */
        function () {
            resolve(fileReader.result);
        });
        fileReader.readAsDataURL(image);
    }));
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @param {?} pages
 * @param {?} pdfPages
 * @return {?}
 */
function setRatio(pages, pdfPages) {
    for (var i = 0; i < pages.length; i++) {
        for (var j = 0; j < pages[i].fields.length; j++) {
            if (pages[i].fields[j]) {
                pages[i].fields[j]['ratio'] = {
                    x: pdfPages[i].xRatio,
                    y: pdfPages[i].yRatio
                };
            }
        }
    }
    return pages;
}
/**
 * @param {?} pages
 * @param {?} roles
 * @param {?} pdfPages
 * @param {?} browserType
 * @return {?}
 */
function constructPages(pages, roles, pdfPages, browserType) {
    if (pages && pages.length > 0) {
        /** @type {?} */
        var fieldPages = new Array(pages.length);
        for (var i = 0; i < pages.length; i++) {
            fieldPages[i] = { fields: [] };
            for (var j = 0; j < pages[i].fields.length; j++) {
                if (pages[i].fields[j]['setting'] && pages[i].fields[j]['setting']['options'] &&
                    pages[i].fields[j]['setting']['options'].length > 0 && (pages[i].fields[j]['type'] === 'checkbox_group') ||
                    pages[i].fields[j]['type'] === 'radio_button_group') {
                    /** @type {?} */
                    var groupFields = constructGroupFields(pages, i, j, roles, pdfPages, fieldPages, browserType);
                    fieldPages[i].fields.concat(groupFields);
                }
                else {
                    /** @type {?} */
                    var fieldCopy = __assign({}, pages[i].fields[j]);
                    /** @type {?} */
                    var copyIndex = findIndex(fieldPages[i].fields, { name: fieldCopy.name });
                    fieldCopy['originalIndex'] = j;
                    if (fieldCopy.type !== 'checkbox_group' && fieldCopy.type !== 'radio_button_group') {
                        fieldCopy['initialStyle'] = generateInitialStyle(fieldCopy, roles, browserType);
                        fieldCopy['style'] = generateStyle(fieldCopy, pdfPages);
                    }
                    if (fieldCopy.type === 'checkbox' || fieldCopy.type === 'checkbox_group' || fieldCopy.type === 'radio_button_group') {
                        fieldCopy['checkboxStyle'] = generateGroupLabel(i, fieldCopy.role_name, roles, pdfPages);
                    }
                    if (copyIndex < 0) {
                        fieldPages[i].fields[fieldPages[i].fields.length] = fieldCopy;
                    }
                }
            }
        }
        return fieldPages;
    }
    else {
        return [];
    }
}
/**
 * @param {?} pages
 * @param {?} i
 * @param {?} j
 * @param {?} roles
 * @param {?} pdfPages
 * @param {?} fieldPages
 * @param {?} browserType
 * @return {?}
 */
function constructGroupFields(pages, i, j, roles, pdfPages, fieldPages, browserType) {
    /** @type {?} */
    var k = 0;
    /** @type {?} */
    var optionsLength = pages[i].fields[j]['setting']['options'].length;
    while (k < optionsLength) {
        /** @type {?} */
        var fieldCopy = __assign({}, pages[i].fields[j]);
        /** @type {?} */
        var options = pages[i].fields[j]['setting']['options'][k];
        fieldCopy.setting['x'] = options.x;
        fieldCopy.setting['y'] = options.y;
        if (fieldCopy.type === 'checkbox_group') {
            fieldCopy.setting['checked'] = options.checked;
        }
        else {
            fieldCopy.setting['selected'] = options.selected;
        }
        fieldCopy['initialStyle'] = generateInitialStyle(fieldCopy, roles, browserType);
        fieldCopy['style'] = generateStyle(fieldCopy, pdfPages);
        fieldCopy['checkboxStyle'] = generateGroupLabel(i, fieldCopy.role_name, roles, pdfPages);
        fieldCopy['originalIndex'] = j;
        fieldCopy['optionIndex'] = k;
        fieldCopy['optionId'] = options.id;
        /** @type {?} */
        var copyIndex = findIndex(fieldPages[i].fields, { optionIndex: fieldCopy.optionIndex, name: fieldCopy.name });
        if (copyIndex < 0) {
            fieldPages[i].fields[fieldPages[i].fields.length] = fieldCopy;
        }
        k++;
    }
    return fieldPages;
}
/**
 * @param {?} field
 * @param {?} roles
 * @param {?} browserType
 * @return {?}
 */
function generateInitialStyle(field, roles, browserType) {
    /** @type {?} */
    var fontSize = 9.5;
    /** @type {?} */
    var setting = field.setting;
    /** @type {?} */
    var role_name = field.role_name;
    /** @type {?} */
    var initialStyle = {
        height: '100%',
        width: '100%',
        backgroundColor: getRoleColor(role_name, roles),
        letterSpacing: null,
        fontSize: null
    };
    field['initialStyle'] = initialStyle;
    switch (field.type) {
        case 'textbox':
        case 'textarea':
            field['initialStyle']['letterSpacing'] = convertToUIRatio(getLetterSpacing(browserType), field['ratio'].x) + 'px';
        case 'signature':
        case 'initial':
        case 'date':
        case 'dropdown':
            field['initialStyle']['fontSize'] = convertToUIRatio(fontSize, field['ratio'].x) + 'px';
            break;
        case 'timestamp':
            field['initialStyle']['fontSize'] = convertToUIRatio(8.5, field['ratio'].x) + 'px';
            break;
        case 'attachment':
        case 'payment':
            field['initialStyle']['height'] = '24px';
            field['initialStyle']['mat-icon'] = {
                position: 'absolute',
                top: 0,
                left: 0,
                height: '24px',
                width: '24px',
                lineHeight: '24px',
                fontSize: '18px'
            };
            break;
        default:
            break;
    }
    if (setting.leading && setting.leading > 0) {
        field.initialStyle['line-height'] = convertToUIRatio(setting.leading + .11395, field['ratio'].y) + 'px';
    }
    return field.initialStyle;
}
/**
 * @param {?} field
 * @param {?} pdfPages
 * @return {?}
 */
function generateStyle(field, pdfPages) {
    return getUIFieldStyling(field);
}
/**
 * @param {?} i
 * @param {?} role_name
 * @param {?} roles
 * @param {?} pdfPages
 * @return {?}
 */
function generateGroupLabel(i, role_name, roles, pdfPages) {
    return {
        position: 'absolute',
        top: 0,
        left: 0,
        height: '13.5px',
        width: '13.5px',
        backgroundColor: getRoleColor(role_name, roles),
        border: '2px solid #777'
    };
}
/**
 * @param {?} name
 * @param {?} roles
 * @param {?=} index
 * @return {?}
 */
function getRoleColor(name, roles, index) {
    if (index) {
        return getRGBA(index);
    }
    else if (roles && roles.length > 0) {
        index = findIndex(roles, { name: name });
        if (index >= 0) {
            return roles[index].rgba ? roles[index].rgba : getRGBA(index);
        }
        else {
            return nameToRGBA(name);
        }
    }
    else {
        return nameToRGBA(name);
    }
}
/**
 * @param {?} value
 * @param {?} ratio
 * @return {?}
 */
function convertToUIRatio(value, ratio) {
    return value * ratio;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @param {?} scrollTop
 * @param {?} pageTop
 * @param {?} pageHeight
 * @return {?}
 */
function isInView(scrollTop, pageTop, pageHeight) {
    if (typeof scrollTop === 'undefined') {
        return null;
    }
    if (window) {
        /** @type {?} */
        var headerHeight = 112;
        /** @type {?} */
        var viewportWidth = window.innerWidth;
        if (viewportWidth > 600) {
            headerHeight = 128;
        }
        /** @type {?} */
        var viewportHeight = window.innerHeight - headerHeight;
        return topOfPageIsInView(scrollTop, pageTop, viewportHeight) ||
            middleOfPageIsInView(scrollTop, pageTop, pageHeight, viewportHeight) ||
            bottomOfPageIsInView(scrollTop, pageTop, pageHeight, viewportHeight);
    }
}
/**
 * @param {?} scrollTop
 * @param {?} pageTop
 * @param {?} viewportHeight
 * @return {?}
 */
function topOfPageIsInView(scrollTop, pageTop, viewportHeight) {
    return pageTop > scrollTop && pageTop < scrollTop + viewportHeight;
}
/**
 * @param {?} scrollTop
 * @param {?} pageTop
 * @param {?} pageHeight
 * @param {?} viewportHeight
 * @return {?}
 */
function middleOfPageIsInView(scrollTop, pageTop, pageHeight, viewportHeight) {
    /** @type {?} */
    var viewportBottom = scrollTop + viewportHeight;
    /** @type {?} */
    var pageBottom = pageTop + pageHeight;
    return scrollTop > pageTop && viewportBottom < pageBottom;
}
/**
 * @param {?} scrollTop
 * @param {?} pageTop
 * @param {?} pageHeight
 * @param {?} viewportHeight
 * @return {?}
 */
function bottomOfPageIsInView(scrollTop, pageTop, pageHeight, viewportHeight) {
    /** @type {?} */
    var viewportBottom = scrollTop + viewportHeight;
    /** @type {?} */
    var pageBottom = pageTop + pageHeight;
    return viewportBottom > pageBottom && scrollTop < pageBottom;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @param {?} type
 * @return {?}
 */
function timePeriod(type) {
    /** @type {?} */
    var endDate = new Date().getTime();
    /** @type {?} */
    var today = new Date();
    /** @type {?} */
    var month = today.getMonth();
    /** @type {?} */
    var year = today.getFullYear();
    /** @type {?} */
    var startDate = null;
    switch (type) {
        case '30d':
            startDate = endDate - (60 * 60 * 24 * 30 * 1000);
            break;
        case '60d':
            startDate = endDate - (60 * 60 * 24 * 60 * 1000);
            break;
        case '6m':
            startDate = endDate - (60 * 60 * 24 * 30 * 6 * 1000);
            break;
        case 'this_month':
            startDate = new Date(year, month, 1).getTime();
            break;
        case 'last_month':
            startDate = new Date(year, month - 1, 1).getTime();
            endDate = new Date(year, month, 0).getTime();
            break;
        case 'this_year':
            startDate = new Date(year, 0, 1);
            break;
        case 'all_time':
            startDate = null;
            endDate = null;
            break;
        default:
            startDate = null;
            endDate = null;
            break;
    }
    if (startDate == null && endDate == null) {
        return null;
    }
    return (/** @type {?} */ ({
        start_time: new Date(startDate).toISOString(),
        end_time: new Date(endDate).toISOString()
    }));
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @param {?} string
 * @return {?}
 */
function regParse(string) {
    return JSON.parse(string, (/**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    function (key, value) {
        /** @type {?} */
        var prefix;
        if (typeof value !== 'string') {
            return value;
        }
        if (value.length < 8) {
            return value;
        }
        prefix = value.substring(0, 8);
        if (prefix === 'function') {
            return eval('(' + value + ')');
        }
        if (prefix === '_PxEgEr_') {
            return eval(value.slice(8));
        }
        if (prefix === '_NuFrRa_') {
            return eval(value.slice(8));
        }
        return value;
    }));
}
/**
 * @param {?} dataUrl
 * @param {?=} type
 * @return {?}
 */
function dataURLtoBlob(dataUrl, type) {
    if (type === void 0) { type = 'image/png'; }
    /** @type {?} */
    var arr = dataUrl.split(',');
    /** @type {?} */
    var 
    // mime = arr[0].match(/:(.*?);/)[1],
    bstr = atob(arr[1]);
    /** @type {?} */
    var n = bstr.length;
    /** @type {?} */
    var u8arr = new Uint8Array(n);
    while (n--) {
        u8arr[n] = bstr.charCodeAt(n);
    }
    return new Blob([u8arr], { type: type });
}
/**
 * @param {?} pdfUrl
 * @return {?}
 */
function printPdfUrl(pdfUrl) {
    if (typeof (pdfUrl) === 'string') {
        printJS(pdfUrl);
    }
    else {
        console.error('pdfUrl should be provided, instead received: ', pdfUrl);
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var SnackbarService = /** @class */ (function () {
    function SnackbarService(snackbar, platform) {
        this.snackbar = snackbar;
        this.platform = platform;
    }
    /**
     * @param {?=} duration
     * @param {?=} position
     * @return {?}
     */
    SnackbarService.prototype.initConfig = /**
     * @param {?=} duration
     * @param {?=} position
     * @return {?}
     */
    function (duration, position) {
        if ((isPlatformBrowser(this.platform))) {
            /** @type {?} */
            var snackbarConfig = void 0;
            if (window && window.innerWidth >= 920) {
                snackbarConfig = {
                    verticalPosition: position ? position : 'bottom',
                    horizontalPosition: 'left',
                    duration: duration ? duration : 5000
                };
            }
            else {
                snackbarConfig = {
                    verticalPosition: position ? position : 'top',
                    duration: duration ? duration : 5000
                };
            }
            return snackbarConfig;
        }
    };
    /**
     * @param {?} message
     * @param {?=} action
     * @param {?=} config
     * @return {?}
     */
    SnackbarService.prototype.open = /**
     * @param {?} message
     * @param {?=} action
     * @param {?=} config
     * @return {?}
     */
    function (message, action, config) {
        return this.snackbar.open(message, action ? action : undefined, config ? config : this.initConfig());
    };
    /**
     * @return {?}
     */
    SnackbarService.prototype.dismiss = /**
     * @return {?}
     */
    function () {
        this.snackbar.dismiss();
    };
    SnackbarService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    SnackbarService.ctorParameters = function () { return [
        { type: MatSnackBar },
        { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
    ]; };
    return SnackbarService;
}());
if (false) {
    /**
     * @type {?}
     * @private
     */
    SnackbarService.prototype.snackbar;
    /**
     * @type {?}
     * @private
     */
    SnackbarService.prototype.platform;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var BrowserToiTextService = /** @class */ (function () {
    function BrowserToiTextService() {
    }
    /**
     * @return {?}
     */
    BrowserToiTextService.prototype.detectBrowser = /**
     * @return {?}
     */
    function () {
        //noinspection TypeScriptUnresolvedVariable
        /** @type {?} */
        var isFirefox = false;
        if ('InstallTrigger' in window) {
            isFirefox = true;
        }
        /** @type {?} */
        var isOpera = (!!window['opr'] && !!window['opr'].addons) || !!window['opera'] || navigator.userAgent.indexOf(' OPR/') >= 0;
        /** @type {?} */
        var isSafari = /constructor/i.test(window['HTMLElement']) || ((/**
         * @param {?} p
         * @return {?}
         */
        function (p) { return p.toString() === "[object SafariRemoteNotification]"; }))(!window['safari'] || (window['safari'] && window['safari'].pushNotification));
        /** @type {?} */
        var isIE = /*@cc_on!@*/ false || !!document['documentMode'];
        /** @type {?} */
        var isEdge = !isIE && !!window['StyleMedia'];
        /** @type {?} */
        var isChrome = !!window['chrome'] && !!window['chrome'].webstore;
        /** @type {?} */
        var isBlink = (isChrome || isOpera) && !!window['CSS'];
        if (isOpera) {
            return 'opera';
        }
        else if (isFirefox) {
            return 'firefox';
        }
        else if (isIE) {
            return 'ie';
        }
        else if (isEdge) {
            return 'edge';
        }
        else if (isChrome) {
            return 'chrome';
        }
        else if (isBlink) {
            return 'blink';
        }
        else if (isSafari) {
            return 'safari';
        }
        return 'unknown';
    };
    /**
     * @return {?}
     */
    BrowserToiTextService.prototype.getLetterSpacing = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var browserType = this.detectBrowser();
        switch (browserType) {
            case 'opera':
                return -0.0018;
            case 'firefox':
                return -0.23594210526315787;
            case 'ie':
                return -0.0019;
            case 'edge':
                return -0.0019;
            case 'chrome':
                return -0.0018;
            case 'safari':
                return -0.0018;
            case 'blink':
                return -0.0018;
            default:
                return -0.0018;
        }
    };
    BrowserToiTextService.decorators = [
        { type: Injectable }
    ];
    return BrowserToiTextService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var BuilderDataService = /** @class */ (function () {
    function BuilderDataService(http, router, templatesService, templateGuard, snackbarService, eventTracker, injector) {
        var _this = this;
        this.http = http;
        this.router = router;
        this.templatesService = templatesService;
        this.templateGuard = templateGuard;
        this.snackbarService = snackbarService;
        this.eventTracker = eventTracker;
        this.injector = injector;
        this.recipients = [];
        this.roles = [];
        this.templateSubject = new ReplaySubject(1, 500);
        this.hasFields = false;
        this.templateNameSubject = new ReplaySubject(1, 500);
        this.templatePagesSubject = new ReplaySubject(1, 500);
        this.recipientsSubject = new ReplaySubject();
        this.activeRecipientSubject = new ReplaySubject();
        this.scrollInfoSubject = new ReplaySubject();
        this.screenInfoSubject = new ReplaySubject();
        this.newTypeSubject = new Subject();
        this.newOptionSubject = new ReplaySubject(1, 100);
        this.duplicateFieldSubject = new ReplaySubject();
        this.saveStatusSubject = new ReplaySubject();
        this.renderedSubject = new ReplaySubject();
        this.activeFieldIndexSubject = new ReplaySubject(1, 100);
        this.numberOfRolesSubject = new ReplaySubject();
        this.numberOfSequenceSubject = new ReplaySubject();
        this.hasFieldsSubject = new ReplaySubject();
        this._config_token = this.injector.get(EssentialsConfigToken);
        this.rForm_backend_url = this._config_token.rForm_backend_url;
        /* Use NGRX to do the subscriptions */
        this.templateSubject.subscribe((/**
         * @param {?} template
         * @return {?}
         */
        function (template) {
            _this.template = template;
            if (template && template.name) {
                _this.templateNameSubject.next(template.name);
            }
        }));
        this.templateNameSubject.subscribe((/**
         * @param {?} name
         * @return {?}
         */
        function (name) {
            _this.templateName = name;
        }));
        /* Remove above after refactoring */
    }
    /**
     * @param {?} template_body
     * @return {?}
     */
    BuilderDataService.prototype.createTemplate = /**
     * @param {?} template_body
     * @return {?}
     */
    function (template_body) {
        var _this = this;
        return this.http
            .post(this.rForm_backend_url + '/templates', template_body)
            .pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            _this.templateSubject.next(res);
            _this.recipientsSubject.next([]);
            return res;
        })));
    };
    /**
     * @return {?}
     */
    BuilderDataService.prototype.updateLocalTemplateData = /**
     * @return {?}
     */
    function () {
        this.templateName = this.template.name;
        this.templateNameSubject.next(this.templateName);
        this.templatePages = this.template.pages.length;
        this.templatePagesSubject.next(this.templatePages);
    };
    /**
     * @param {?} template_body
     * @return {?}
     */
    BuilderDataService.prototype.updateTemplate = /**
     * @param {?} template_body
     * @return {?}
     */
    function (template_body) {
        return this.templatesService.updateTemplateObservable(template_body.id, template_body).pipe(map((/**
         * @param {?} template
         * @return {?}
         */
        function (template) { return template; })));
    };
    /**
     * @param {?} template
     * @param {?} name
     * @return {?}
     */
    BuilderDataService.prototype.updateTemplateName = /**
     * @param {?} template
     * @param {?} name
     * @return {?}
     */
    function (template, name) {
        var _this = this;
        if (template && template.id) {
            return this.templatesService.updateTemplateObservable(template.id, { name: name })
                .pipe(map((/**
             * @param {?} res
             * @return {?}
             */
            function (res) {
                template.name = res.name;
                _this.updateLocalTemplate(template);
                _this.saveStatusSubject.next('saved');
                _this.createSnackBar('Name Changed to: ' + name);
                return res.name;
            }), (/**
             * @param {?} err
             * @return {?}
             */
            function (err) {
                _this.saveStatusSubject.next('Failed to save');
                _this.createSnackBar('Failed to change Template name');
                return err;
            })));
        }
        else {
            this.templateNameSubject.next(name);
            this.saveStatusSubject.next('Name will be updated');
        }
    };
    /**
     * @param {?} templateId
     * @return {?}
     */
    BuilderDataService.prototype.getTemplateObservable = /**
     * @param {?} templateId
     * @return {?}
     */
    function (templateId) {
        return this.templatesService.getTemplateObservable(templateId);
    };
    /**
     * @param {?} templateId
     * @return {?}
     */
    BuilderDataService.prototype.getTemplate = /**
     * @param {?} templateId
     * @return {?}
     */
    function (templateId) {
        var _this = this;
        return this.templatesService.getTemplate(templateId).toPromise().then((/**
         * @param {?} template
         * @return {?}
         */
        function (template) {
            _this.updateLocalTemplate(template);
            _this.recipients = _this.sortRoles(_this.template);
            _this.template.roles = _this.recipients;
            _this.roles = _this.recipients;
            _this.recipientsSubject.next(_this.recipients);
            _this.activeRecipientSubject.next(_this.recipients[0]);
            return _this.template;
        }));
    };
    /**
     * @return {?}
     */
    BuilderDataService.prototype.watchForUpdatedTemplate = /**
     * @return {?}
     */
    function () {
        return this.templateSubject.pipe(map((/**
         * @param {?} template
         * @return {?}
         */
        function (template) { return template; })));
    };
    /**
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.updateLocalTemplate = /**
     * @param {?} template
     * @return {?}
     */
    function (template) {
        if (template) {
            this.template = template;
            if (this.template && this.template.pages) {
                this.template.pages = this.template.pages.sort((/**
                 * @param {?} a
                 * @param {?} b
                 * @return {?}
                 */
                function (a, b) {
                    return a.sequence - b.sequence;
                }));
            }
            if (this.template) {
                this.updateLocalTemplateData();
                this.checkForFields(this.template.roles);
            }
            this.recipientsSubject.next(this.template.roles.sort((/**
             * @param {?} a
             * @param {?} b
             * @return {?}
             */
            function (a, b) {
                return a.sequence - b.sequence;
            })));
            this.templateSubject.next(this.template);
        }
        else {
            this.templateNameSubject.next(null);
            this.templatePagesSubject.next(null);
            this.numberOfRolesSubject.next(null);
            this.numberOfSequenceSubject.next(null);
            this.hasFieldsSubject.next(false);
        }
    };
    /**
     * @param {?} file
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.uploadTemplateDocument = /**
     * @param {?} file
     * @param {?} template
     * @return {?}
     */
    function (file, template) {
        /** @type {?} */
        var formdata = new FormData();
        formdata.append('document', file, file.name);
        /** @type {?} */
        var req = new HttpRequest('POST', this.rForm_backend_url + '/templates/' + template.id + '/documents', formdata, {
            reportProgress: true
        });
        return this.http.request(req)
            .pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        function (res) { return res; })), catchError((/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            return throwError(err);
        })));
    };
    /**
     * @param {?} templateId
     * @return {?}
     */
    BuilderDataService.prototype.getTemplateDocument = /**
     * @param {?} templateId
     * @return {?}
     */
    function (templateId) {
        var _this = this;
        this.templateDoc = this.templatesService.getAllTemplateDocumentsObservable(templateId).pipe(map((/**
         * @param {?} docs
         * @return {?}
         */
        function (docs) {
            _this.templateDoc = docs[0];
            return _this.templateDoc;
        })));
        return this.templateDoc;
    };
    /**
     * @param {?} templateId
     * @param {?} templateDoc
     * @return {?}
     */
    BuilderDataService.prototype.getTemplateDocumentFile = /**
     * @param {?} templateId
     * @param {?} templateDoc
     * @return {?}
     */
    function (templateId, templateDoc) {
        if (templateId && templateDoc) {
            return this.templatesService.getTemplateDocumentPDFObservable(templateId, templateDoc).pipe(map((/**
             * @param {?} pdf
             * @return {?}
             */
            function (pdf) {
                /** @type {?} */
                var pdfUrl = URL.createObjectURL(pdf);
                return pdfUrl;
            })));
        }
    };
    /**
     * @param {?} templateDocument
     * @return {?}
     */
    BuilderDataService.prototype.setTemplateDocument = /**
     * @param {?} templateDocument
     * @return {?}
     */
    function (templateDocument) {
        this.templateDoc = templateDocument;
    };
    /**
     * @param {?} document
     * @param {?} pageNum
     * @param {?} templateId
     * @return {?}
     */
    BuilderDataService.prototype.addTemplatePage = /**
     * @param {?} document
     * @param {?} pageNum
     * @param {?} templateId
     * @return {?}
     */
    function (document, pageNum, templateId) {
        /** @type {?} */
        var body = {
            sequence: pageNum,
            page_number: pageNum,
            document_id: document.id
        };
        return this.http.post(this.rForm_backend_url + ("/templates/" + templateId + "/pages"), body)
            .pipe(map((/**
         * @param {?} res
         * @return {?}
         */
        function (res) { return res; })), catchError((/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            return throwError(err);
        })));
    };
    /**
     * @param {?} document
     * @param {?} pageNums
     * @param {?} templateId
     * @return {?}
     */
    BuilderDataService.prototype.addTemplatePages = /**
     * @param {?} document
     * @param {?} pageNums
     * @param {?} templateId
     * @return {?}
     */
    function (document, pageNums, templateId) {
        var _this = this;
        return from(pageNums)
            .pipe(mergeMap((/**
         * @param {?} pageNum
         * @return {?}
         */
        function (pageNum) {
            /** @type {?} */
            var page = {
                sequence: pageNum,
                page_number: pageNum,
                document_id: document.id
            };
            return (/** @type {?} */ (_this.http.post(_this.rForm_backend_url + ("/templates/" + templateId + "/pages"), page).pipe(delay(200))));
        })));
    };
    /**
     * @param {?} template
     * @param {?} fieldName
     * @param {?} i
     * @param {?} j
     * @return {?}
     */
    BuilderDataService.prototype.deleteTemplateField = /**
     * @param {?} template
     * @param {?} fieldName
     * @param {?} i
     * @param {?} j
     * @return {?}
     */
    function (template, fieldName, i, j) {
        var _this = this;
        return this.http
            .delete(this.rForm_backend_url + '/templates/' + template.id + '/fields/' + fieldName.trim())
            .pipe(map((/**
         * @param {?} field
         * @return {?}
         */
        function (field) {
            /** @type {?} */
            var fieldIndex = findIndex(template.pages[i].fields, { name: fieldName });
            /** @type {?} */
            var roleIndex = findIndex(template.roles, { name: template.pages[i].fields[fieldIndex].role_name });
            if (roleIndex > -1) {
                /** @type {?} */
                var roleFieldIndex = findIndex(template.roles[roleIndex].fields, { name: fieldName });
                if (roleFieldIndex > -1) {
                    template.roles[roleIndex].fields.splice(roleFieldIndex, 1);
                }
            }
            template.pages[i].fields.splice(fieldIndex, 1);
            _this.updateLocalTemplate(template);
            _this.saveStatusSubject.next('saved');
            _this.checkForFields(template.roles);
            return template;
        })));
    };
    /**
     * @param {?} fieldNames
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.deleteTemplateFields = /**
     * @param {?} fieldNames
     * @param {?} template
     * @return {?}
     */
    function (fieldNames, template) {
        var e_1, _a;
        if (fieldNames && fieldNames.length > 0) {
            /** @type {?} */
            var pages = template.pages;
            try {
                for (var fieldNames_1 = __values(fieldNames), fieldNames_1_1 = fieldNames_1.next(); !fieldNames_1_1.done; fieldNames_1_1 = fieldNames_1.next()) {
                    var fieldName = fieldNames_1_1.value;
                    for (var x = 0; x < pages.length; x++) {
                        /** @type {?} */
                        var fieldIndex = findIndex(pages[x].fields, { name: fieldName });
                        if (fieldIndex >= 0) {
                            template.pages[x].fields.splice(fieldIndex, 1);
                        }
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (fieldNames_1_1 && !fieldNames_1_1.done && (_a = fieldNames_1.return)) _a.call(fieldNames_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
            this.checkForFields(template.roles);
        }
    };
    /**
     * @param {?} newField
     * @param {?} i
     * @param {?} j
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.addTemplateField = /**
     * @param {?} newField
     * @param {?} i
     * @param {?} j
     * @param {?} template
     * @return {?}
     */
    function (newField, i, j, template) {
        var _this = this;
        newField.name = newField.name.trim();
        return this.http
            .post(this.rForm_backend_url + '/templates/' + newField.template_id + '/fields', newField)
            .pipe(map((/**
         * @param {?} field
         * @return {?}
         */
        function (field) {
            _this.eventTracker.createEvent({
                category: 'document',
                action: "document " + field.type + " field added",
                label: "document id: " + newField.template_id
            });
            template.pages[i].fields[j] = field;
            _this.checkForFields(template.roles);
            /** @type {?} */
            var roleIndex = findIndex(template.roles, { name: field.role_name });
            if (roleIndex >= 0) {
                template.roles[roleIndex]['fields'].push(field);
            }
            _this.templateSubject.next(template);
            return template;
        })));
    };
    /**
     * @param {?} id
     * @return {?}
     */
    BuilderDataService.prototype.getTemplateOwnerInfo = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        return this.templatesService.getTemplateOwnerInfo(id);
    };
    /**
     * @param {?} template
     * @param {?} body
     * @param {?} oldName
     * @param {?} i
     * @param {?} j
     * @return {?}
     */
    BuilderDataService.prototype.updateTemplateField = /**
     * @param {?} template
     * @param {?} body
     * @param {?} oldName
     * @param {?} i
     * @param {?} j
     * @return {?}
     */
    function (template, body, oldName, i, j) {
        var _this = this;
        oldName = oldName.trim();
        return this.http
            .put(this.rForm_backend_url + '/templates/' + template.id + '/fields/' + oldName, body)
            .pipe(map((/**
         * @param {?} field
         * @return {?}
         */
        function (field) {
            _this.eventTracker.createEvent({
                category: 'document',
                action: "document " + field.type + " field updated",
                label: "document id: " + template.id
            });
            /** @type {?} */
            var fieldIndex = findIndex(template.pages[i].fields, { name: oldName });
            if (field.page_sequence - 1 !== i) {
                template.pages[i].fields.splice(fieldIndex, 1);
                template.pages[field.page_sequence - 1].fields.push(field);
            }
            else {
                template.pages[i].fields[fieldIndex] = field;
            }
            /** @type {?} */
            var roleIndex = findIndex(template.roles, { name: field.role_name });
            if (roleIndex >= 0) {
                template = _this.updateRoleField(template, field, roleIndex, oldName);
            }
            _this.templateSubject.next(template);
            _this.saveStatusSubject.next('saved');
            return template;
        })));
    };
    /**
     * @param {?} template
     * @param {?} updated_field
     * @param {?} role_index
     * @param {?} old_name
     * @return {?}
     */
    BuilderDataService.prototype.updateRoleField = /**
     * @param {?} template
     * @param {?} updated_field
     * @param {?} role_index
     * @param {?} old_name
     * @return {?}
     */
    function (template, updated_field, role_index, old_name) {
        /** @type {?} */
        var field_index = findIndex(template.roles[role_index].fields, { name: old_name });
        if (field_index >= 0) {
            template.roles[role_index].fields[field_index] = updated_field;
        }
        else {
            template.roles[role_index].fields.push(updated_field);
        }
        return template;
    };
    /**
     * @param {?} template
     * @param {?} body
     * @param {?} oldName
     * @param {?} i
     * @param {?} j
     * @return {?}
     */
    BuilderDataService.prototype.updateDropdownField = /**
     * @param {?} template
     * @param {?} body
     * @param {?} oldName
     * @param {?} i
     * @param {?} j
     * @return {?}
     */
    function (template, body, oldName, i, j) {
        var _this = this;
        oldName = oldName.trim();
        return this.http
            .put(this.rForm_backend_url + '/templates/' + template.id + '/fields/' + oldName, body)
            .pipe(map((/**
         * @param {?} field
         * @return {?}
         */
        function (field) {
            /** @type {?} */
            var fieldIndex = findIndex(template.pages[i].fields, { name: oldName });
            template.pages[i].fields[fieldIndex] = field;
            _this.templateSubject.next(template);
            _this.saveStatusSubject.next('saved');
            return template;
        })));
    };
    /**
     * @param {?} template_id
     * @param {?} sequence_number
     * @return {?}
     */
    BuilderDataService.prototype.deleteSequence = /**
     * @param {?} template_id
     * @param {?} sequence_number
     * @return {?}
     */
    function (template_id, sequence_number) {
        return this.templatesService.deleteSequence(template_id, sequence_number);
    };
    /**
     * @param {?} field
     * @return {?}
     */
    BuilderDataService.prototype.prepareFieldDuplication = /**
     * @param {?} field
     * @return {?}
     */
    function (field) {
        this.duplicateFieldSubject.next(field);
    };
    ;
    /**
     * @param {?} roles
     * @return {?}
     */
    BuilderDataService.prototype.checkForFields = /**
     * @param {?} roles
     * @return {?}
     */
    function (roles) {
        var e_2, _a;
        /** @type {?} */
        var hasFields = true;
        if (roles && roles.length > 0) {
            try {
                for (var roles_1 = __values(roles), roles_1_1 = roles_1.next(); !roles_1_1.done; roles_1_1 = roles_1.next()) {
                    var role = roles_1_1.value;
                    if (role && role.type === 'signer' && role.fields.length === 0) {
                        hasFields = false;
                    }
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (roles_1_1 && !roles_1_1.done && (_a = roles_1.return)) _a.call(roles_1);
                }
                finally { if (e_2) throw e_2.error; }
            }
        }
        else {
            hasFields = false;
        }
        this.hasFields = hasFields;
        this.hasFieldsSubject.next(this.hasFields);
    };
    /**
     * @param {?} title
     * @param {?=} buttonTitle
     * @return {?}
     */
    BuilderDataService.prototype.createSnackBar = /**
     * @param {?} title
     * @param {?=} buttonTitle
     * @return {?}
     */
    function (title, buttonTitle) {
        if (buttonTitle === void 0) { buttonTitle = 'OK'; }
        //   let snackbarConfig: MatSnackBarConfig
        //   if (window.innerWidth >= 920) {
        //     snackbarConfig = {
        //       verticalPosition: 'bottom',
        //       horizontalPosition: 'left',
        //       duration: 5000
        //     }
        //   } else {
        //     snackbarConfig = {
        //       verticalPosition: 'top',
        //       duration: 5000
        //     }
        //   }
        //   this.snackbar.open(title, buttonTitle, snackbarConfig);
    };
    /**
     * @param {?} templateId
     * @return {?}
     */
    BuilderDataService.prototype.editDocsUrl = /**
     * @param {?} templateId
     * @return {?}
     */
    function (templateId) {
        /** @type {?} */
        var editDocsPageUrl = "/builder/" + templateId + "/docs";
        return editDocsPageUrl;
    };
    /**
     * @param {?} templateId
     * @return {?}
     */
    BuilderDataService.prototype.editRolesUrl = /**
     * @param {?} templateId
     * @return {?}
     */
    function (templateId) {
        /** @type {?} */
        var editRolesPageUrl = "/builder/" + templateId + "/roles";
        return editRolesPageUrl;
    };
    /**
     * @param {?} templateId
     * @return {?}
     */
    BuilderDataService.prototype.editTemplateUrl = /**
     * @param {?} templateId
     * @return {?}
     */
    function (templateId) {
        /** @type {?} */
        var editPageUrl = "/builder/" + templateId + "/fields";
        return editPageUrl;
    };
    /**
     * @param {?} templateId
     * @return {?}
     */
    BuilderDataService.prototype.previewTemplateUrl = /**
     * @param {?} templateId
     * @return {?}
     */
    function (templateId) {
        /** @type {?} */
        var reviewTemplateUrl = "/document/" + templateId;
        return reviewTemplateUrl;
    };
    /**
     * @param {?} templateId
     * @param {?} fieldId
     * @param {?} roleId
     * @param {?} i
     * @param {?} j
     * @return {?}
     */
    BuilderDataService.prototype.addFieldRole = /**
     * @param {?} templateId
     * @param {?} fieldId
     * @param {?} roleId
     * @param {?} i
     * @param {?} j
     * @return {?}
     */
    function (templateId, fieldId, roleId, i, j) {
        /** @type {?} */
        var fieldRoleInfo = JSON.stringify({
            field_id: fieldId,
            role_id: roleId
        });
        return this.http
            .post(this.rForm_backend_url + '/template/' + templateId + '/field_role', fieldRoleInfo).toPromise().then((/**
         * @param {?} fieldRole
         * @return {?}
         */
        function (fieldRole) {
            return fieldRole;
        }));
    };
    /**
     * @param {?} role
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.addRole = /**
     * @param {?} role
     * @param {?} template
     * @return {?}
     */
    function (role, template) {
        var _this = this;
        /** @type {?} */
        var templateBackend = this.rForm_backend_url + '/templates/' + template.id + '/roles';
        this.roles = [];
        this.roles = this.roles.concat(role);
        /** @type {?} */
        var roleRequests = [];
        this.roles.forEach((/**
         * @param {?} role
         * @return {?}
         */
        function (role) {
            roleRequests.push(_this.http.post(templateBackend, role));
        }));
        if (this.roles && this.roles.length > 0) {
            return forkJoin(roleRequests).pipe(map((/**
             * @param {?} savedRoles
             * @return {?}
             */
            function (savedRoles) {
                savedRoles.concat((/** @type {?} */ (savedRoles)));
                if (!template['roles']) {
                    template['roles'] = (/** @type {?} */ ([]));
                }
                /** @type {?} */
                var savedRolesWithFields = savedRoles.map((/**
                 * @param {?} newRole
                 * @return {?}
                 */
                function (newRole) {
                    newRole['fields'] = [];
                    return newRole;
                }));
                template['roles'] = template['roles'].concat(savedRolesWithFields);
                _this.recipients = _this.sortRoles(template);
                _this.roles = _this.recipients;
                _this.templateSubject.next(template);
                _this.recipientsSubject.next(_this.recipients);
                _this.recipients;
                return template;
            }), (/**
             * @param {?} err
             * @return {?}
             */
            function (err) {
                console.error(err);
                console.error('Couldn\'t save all the roles');
                return err;
            })));
        }
        else {
            return of(template);
        }
    };
    /**
     * @param {?} field
     * @return {?}
     */
    BuilderDataService.prototype.addCheckboxGroup = /**
     * @param {?} field
     * @return {?}
     */
    function (field) {
        this.newOptionSubject.next({
            type: 'checkbox_group',
            page_sequence: field.page_sequence,
            field_name: field.name,
        });
    };
    /**
     * @param {?} field
     * @return {?}
     */
    BuilderDataService.prototype.addRadioGroup = /**
     * @param {?} field
     * @return {?}
     */
    function (field) {
        this.newOptionSubject.next({
            type: 'radio_button_group',
            page_sequence: field.page_sequence,
            field_name: field.name,
        });
    };
    /**
     * @param {?} roles
     * @return {?}
     */
    BuilderDataService.prototype.getRolesInSequence = /**
     * @param {?} roles
     * @return {?}
     */
    function (roles) {
        /** @type {?} */
        var rolesInSequence = [];
        for (var roleIndex = 0; roleIndex < roles.length; roleIndex++) {
            /** @type {?} */
            var sequenceIndex = roles[roleIndex]['sequence'] - 1;
            /** @type {?} */
            var role = __assign({}, roles[roleIndex]);
            role['style'] = {
                backgroundColor: "" + this.getRoleColor(role.name, roleIndex)
            };
            if (!rolesInSequence[sequenceIndex]) {
                rolesInSequence[sequenceIndex] = [role];
            }
            else {
                rolesInSequence[sequenceIndex].push(role);
            }
        }
        this.numberOfSequenceSubject.next(rolesInSequence.length);
        this.numberOfRolesSubject.next(roles.length);
        return rolesInSequence;
    };
    /**
     * @param {?} roleName
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.deleteRole = /**
     * @param {?} roleName
     * @param {?} template
     * @return {?}
     */
    function (roleName, template) {
        var _this = this;
        return this.http.delete(this.rForm_backend_url + '/templates/' + template.id + '/roles/' + roleName)
            .toPromise()
            .then((/**
         * @return {?}
         */
        function () {
            remove(template.roles, (/**
             * @param {?} role
             * @return {?}
             */
            function (role) {
                var e_3, _a;
                _this.recipients = _this.sortRoles(template);
                if (role.name === roleName) {
                    /** @type {?} */
                    var fields = role.fields;
                    /** @type {?} */
                    var fieldNames = [];
                    try {
                        for (var fields_1 = __values(fields), fields_1_1 = fields_1.next(); !fields_1_1.done; fields_1_1 = fields_1.next()) {
                            var field = fields_1_1.value;
                            fieldNames.push(field.name);
                        }
                    }
                    catch (e_3_1) { e_3 = { error: e_3_1 }; }
                    finally {
                        try {
                            if (fields_1_1 && !fields_1_1.done && (_a = fields_1.return)) _a.call(fields_1);
                        }
                        finally { if (e_3) throw e_3.error; }
                    }
                    _this.deleteTemplateFields(fieldNames, template);
                }
                _this.templateSubject.next(template);
                _this.recipientsSubject.next(_this.recipients);
                return role.name === roleName;
            }));
        }));
    };
    /**
     * @param {?} roleNames
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.deleteRoles = /**
     * @param {?} roleNames
     * @param {?} template
     * @return {?}
     */
    function (roleNames, template) {
        var _this = this;
        /** @type {?} */
        var deleteCalls = [];
        roleNames.forEach((/**
         * @param {?} role_name
         * @return {?}
         */
        function (role_name) {
            template = __assign({}, template, { roles: template.roles.filter((/**
                 * @param {?} role
                 * @return {?}
                 */
                function (role) { return role.name !== role_name; })) });
            deleteCalls.push(_this.http.delete(_this.rForm_backend_url + '/templates/' + template.id + '/roles/' + role_name));
        }));
        return forkJoin(deleteCalls).pipe(map((/**
         * @return {?}
         */
        function () {
            return template;
        }), (/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            console.error(err);
            console.error('Couldn\'t delete all the roles');
            return err;
        })));
    };
    /**
     * @param {?} roles
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.updateRoles = /**
     * @param {?} roles
     * @param {?} template
     * @return {?}
     */
    function (roles, template) {
        var _this = this;
        /** @type {?} */
        var updateCalls = [];
        roles.forEach((/**
         * @param {?} role
         * @return {?}
         */
        function (role) {
            /** @type {?} */
            var body = {
                template_id: template.id,
                name: role.name.trim(),
                full_name: role.full_name,
                email: role.email,
                sequence: role.sequence,
                type: role.type,
                delegator: role.delegator,
                message: role.message,
                phone: role.phone
            };
            /** @type {?} */
            var role_index = findIndex(template.roles, { name: role.old_name });
            if (role_index >= 0) {
                template.roles[role_index] = __assign({}, template.roles[role_index], { template_id: template.id, name: role.name.trim(), full_name: role.full_name, email: role.email, sequence: role.sequence, type: role.type, delegator: role.delegator, message: role.message, phone: role.phone });
            }
            _this.eventTracker.createEvent({
                category: 'document',
                action: 'document role updated',
                label: "document id: " + template.id
            });
            updateCalls.push(_this.http.put(_this.rForm_backend_url + '/templates/' + template.id + '/roles/' + role.old_name, body));
        }));
        return forkJoin(updateCalls)
            .pipe(map((/**
         * @return {?}
         */
        function () {
            return template;
        }), (/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            console.error(err);
            console.error('Couldn\'t save all the roles');
            return err;
        })));
    };
    /**
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.sortRoles = /**
     * @param {?} template
     * @return {?}
     */
    function (template) {
        if (template.roles.length < 1) {
            return [];
        }
        /** @type {?} */
        var roles = template.roles.sort((/**
         * @param {?} a
         * @param {?} b
         * @return {?}
         */
        function (a, b) {
            if (a.sequence === b.sequence) {
                return a.name.toLowerCase() < b.name.toLowerCase() ? -1 : a.name.toLowerCase() > b.name.toLowerCase() ? 1 : 0;
            }
            return a.sequence - b.sequence;
        }));
        for (var x = 0; x < roles.length; x++) {
            roles[x]['rgba'] = getRGBA(x);
        }
        template.roles = roles;
        return roles;
    };
    /**
     * @param {?} name
     * @param {?=} index
     * @param {?=} template
     * @return {?}
     */
    BuilderDataService.prototype.getRoleColor = /**
     * @param {?} name
     * @param {?=} index
     * @param {?=} template
     * @return {?}
     */
    function (name, index, template) {
        if (index !== null && index > -1) {
            return getRGBA(index);
        }
        else if (template) {
            if (template.roles && template.roles.length > 0) {
                index = findIndex(template.roles, { name: name });
                if (index >= 0) {
                    return template.roles[index].rgba ? template.roles[index].rgba : getRGBA(index);
                }
            }
            else {
                return nameToRGBA(name);
            }
        }
        else {
            return nameToRGBA(name);
        }
    };
    /**
     * @param {?} role
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.updateFullName = /**
     * @param {?} role
     * @param {?} template
     * @return {?}
     */
    function (role, template) {
        var _this = this;
        for (var i = 0; i < template.pages.length; i++) {
            if (template.pages[i].fields.length > 0) {
                for (var j = 0; j < template.pages[i].fields.length; j++) {
                    if (template.pages[i].fields[j].type === 'placeholder' && template.pages[i].fields[j].setting['type'] === 'full_name') {
                        if (template.pages[i].fields[j].setting['result'] !== role.full_name &&
                            template.pages[i].fields[j].role_name === role.name) {
                            template.pages[i].fields[j].setting['result'] = role.full_name;
                            /** @type {?} */
                            var updatedField = __assign({}, template.pages[i].fields[j]);
                            this.updateTemplateField(template, updatedField, updatedField.name, i, j).toPromise().then((/**
                             * @return {?}
                             */
                            function () {
                                _this.templateSubject.next(template);
                            }));
                        }
                    }
                }
            }
        }
    };
    /**
     * @param {?} activeFieldIndex
     * @return {?}
     */
    BuilderDataService.prototype.setActiveFieldIndex = /**
     * @param {?} activeFieldIndex
     * @return {?}
     */
    function (activeFieldIndex) {
        this.activeFieldIndex = activeFieldIndex;
        this.activeFieldIndexSubject.next(this.activeFieldIndex);
    };
    /**
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.autoAddSigner = /**
     * @param {?} template
     * @return {?}
     */
    function (template) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, new Promise((/**
                     * @param {?} resolve
                     * @param {?} reject
                     * @return {?}
                     */
                    function (resolve, reject) { return __awaiter(_this, void 0, void 0, function () {
                        var recipients;
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, this.addRole((/** @type {?} */ ({ name: 'Signer 1', type: 'signer', sequence: 1 })), template).toPromise()];
                                case 1:
                                    recipients = _a.sent();
                                    this.activeRecipientSubject.next(recipients[0]);
                                    setTimeout((/**
                                     * @return {?}
                                     */
                                    function () {
                                        resolve();
                                    }), 500);
                                    return [2 /*return*/];
                            }
                        });
                    }); }))];
            });
        });
    };
    /**
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.openTemplate = /**
     * @param {?} template
     * @return {?}
     */
    function (template) {
        if (this.canUserPreview(template)) {
            this.router.navigate(["document/" + template.id]);
        }
        else if (this.canUserEdit(template)) {
            this.router.navigate(["builder/" + template.id + "/fields"]);
        }
        else {
            this.snackbarService.open("Template is in build mode, and not ready for use.  Please check back soon.");
        }
    };
    /**
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.canUserEdit = /**
     * @param {?} template
     * @return {?}
     */
    function (template) {
        /** @type {?} */
        var response = this.templateGuard.canPerformAction(TemplateActions.WRITE, template);
        return response['canPerform'];
    };
    /**
     * @param {?} template
     * @return {?}
     */
    BuilderDataService.prototype.canUserPreview = /**
     * @param {?} template
     * @return {?}
     */
    function (template) {
        var e_4, _a;
        /** @type {?} */
        var hasPermission = (this.templateGuard.canPerformAction(TemplateActions.READ, template)).canPerform;
        /** @type {?} */
        var canPreview;
        /** @type {?} */
        var signers = filter(template.roles, { type: 'signer' });
        canPreview = signers && signers.length > 0;
        try {
            for (var signers_1 = __values(signers), signers_1_1 = signers_1.next(); !signers_1_1.done; signers_1_1 = signers_1.next()) {
                var signer = signers_1_1.value;
                canPreview = signer['fields'] && signer['fields'].length > 0;
            }
        }
        catch (e_4_1) { e_4 = { error: e_4_1 }; }
        finally {
            try {
                if (signers_1_1 && !signers_1_1.done && (_a = signers_1.return)) _a.call(signers_1);
            }
            finally { if (e_4) throw e_4.error; }
        }
        return hasPermission && canPreview;
    };
    BuilderDataService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    BuilderDataService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: Router },
        { type: TemplatesService },
        { type: TemplatesGuardService },
        { type: SnackbarService },
        { type: EventTrackerService },
        { type: Injector }
    ]; };
    return BuilderDataService;
}());
if (false) {
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype._config_token;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.recipients;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.template;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.templateDoc;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.templatePages;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.templateName;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.activeFieldIndex;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.roles;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.templateSubject;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.templateId;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.hasFields;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.rForm_backend_url;
    /** @type {?} */
    BuilderDataService.prototype.templateNameSubject;
    /** @type {?} */
    BuilderDataService.prototype.templatePagesSubject;
    /** @type {?} */
    BuilderDataService.prototype.recipientsSubject;
    /** @type {?} */
    BuilderDataService.prototype.activeRecipientSubject;
    /** @type {?} */
    BuilderDataService.prototype.scrollInfoSubject;
    /** @type {?} */
    BuilderDataService.prototype.screenInfoSubject;
    /** @type {?} */
    BuilderDataService.prototype.newTypeSubject;
    /** @type {?} */
    BuilderDataService.prototype.newOptionSubject;
    /** @type {?} */
    BuilderDataService.prototype.duplicateFieldSubject;
    /** @type {?} */
    BuilderDataService.prototype.saveStatusSubject;
    /** @type {?} */
    BuilderDataService.prototype.renderedSubject;
    /** @type {?} */
    BuilderDataService.prototype.activeFieldIndexSubject;
    /** @type {?} */
    BuilderDataService.prototype.numberOfRolesSubject;
    /** @type {?} */
    BuilderDataService.prototype.numberOfSequenceSubject;
    /** @type {?} */
    BuilderDataService.prototype.hasFieldsSubject;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.router;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.templatesService;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.templateGuard;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.snackbarService;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.eventTracker;
    /**
     * @type {?}
     * @private
     */
    BuilderDataService.prototype.injector;
    /* Skipping unhandled member: ;*/
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var BuilderService = /** @class */ (function () {
    function BuilderService() {
        this.sideNavStateSubject = new ReplaySubject();
        this.recipientSideNavStateSubject = new ReplaySubject(1, 100);
        this.initiateAutofocus = new ReplaySubject();
        this.recipientAutoFocus = new ReplaySubject();
        this.deleteRoleSubject = new ReplaySubject(1, 100);
        this.hasActiveRoleSubject = new ReplaySubject();
    }
    // constructor() {}
    // constructor() {}
    /**
     * @return {?}
     */
    BuilderService.prototype.toggleSideNav = 
    // constructor() {}
    /**
     * @return {?}
     */
    function () {
        this.sideNavState = !this.sideNavState;
        this.sideNavStateSubject.next(this.sideNavState);
        if (this.sideNavState === false) {
            this.initiateAutofocus.next(false);
        }
    };
    /**
     * @return {?}
     */
    BuilderService.prototype.toggleRecipientSideNav = /**
     * @return {?}
     */
    function () {
        this.recipientSideNavState = !this.recipientSideNavState;
        this.recipientSideNavStateSubject.next(this.recipientSideNavState);
        if (this.recipientSideNavState === false) {
            this.recipientAutoFocus.next(false);
        }
    };
    /**
     * @return {?}
     */
    BuilderService.prototype.openSideNav = /**
     * @return {?}
     */
    function () {
        this.sideNavState = true;
        this.sideNavStateSubject.next(this.sideNavState);
    };
    /**
     * @return {?}
     */
    BuilderService.prototype.openRecipientSideNav = /**
     * @return {?}
     */
    function () {
        this.recipientSideNavState = true;
        this.recipientSideNavStateSubject.next(this.recipientSideNavState);
    };
    /**
     * @return {?}
     */
    BuilderService.prototype.closeSideNav = /**
     * @return {?}
     */
    function () {
        this.sideNavState = false;
        this.sideNavStateSubject.next(this.sideNavState);
        this.initiateAutofocus.next(false);
    };
    /**
     * @return {?}
     */
    BuilderService.prototype.closeRecipientSideNav = /**
     * @return {?}
     */
    function () {
        this.recipientSideNavState = false;
        this.recipientSideNavStateSubject.next(this.recipientSideNavState);
        this.recipientAutoFocus.next(false);
    };
    /**
     * @param {?} role
     * @return {?}
     */
    BuilderService.prototype.setActiveRole = /**
     * @param {?} role
     * @return {?}
     */
    function (role) {
        this.activeRole = role;
        this.hasActiveRoleSubject.next(true);
    };
    /**
     * @return {?}
     */
    BuilderService.prototype.deleteRole = /**
     * @return {?}
     */
    function () {
        if (!!this.activeRole) {
            this.deleteRoleSubject.next(this.activeRole);
            this.activeRole = null;
            this.hasActiveRoleSubject.next(false);
        }
    };
    /**
     * @return {?}
     */
    BuilderService.prototype.getSideNavState = /**
     * @return {?}
     */
    function () {
        return this.sideNavState;
    };
    /**
     * @return {?}
     */
    BuilderService.prototype.getRecipientSideNavState = /**
     * @return {?}
     */
    function () {
        return this.recipientSideNavState;
    };
    BuilderService.decorators = [
        { type: Injectable }
    ];
    return BuilderService;
}());
if (false) {
    /**
     * @type {?}
     * @private
     */
    BuilderService.prototype.sideNavState;
    /**
     * @type {?}
     * @private
     */
    BuilderService.prototype.recipientSideNavState;
    /**
     * @type {?}
     * @private
     */
    BuilderService.prototype.activeRole;
    /** @type {?} */
    BuilderService.prototype.sideNavStateSubject;
    /** @type {?} */
    BuilderService.prototype.recipientSideNavStateSubject;
    /** @type {?} */
    BuilderService.prototype.initiateAutofocus;
    /** @type {?} */
    BuilderService.prototype.recipientAutoFocus;
    /** @type {?} */
    BuilderService.prototype.deleteRoleSubject;
    /** @type {?} */
    BuilderService.prototype.hasActiveRoleSubject;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var ConversionService = /** @class */ (function () {
    function ConversionService() {
        this.ITextWidth = 612;
        this.ITextHeight = 792;
        this.XRatio = 1;
        this.YRatio = 1;
        this.a3Portrait = {
            x: 842,
            y: 1191
        };
        this.a3Landscape = {
            x: this.a3Portrait.y,
            y: this.a3Portrait.x
        };
        this.a4Portrait = {
            x: 595,
            y: 842
        };
        this.a4Landscape = {
            x: this.a4Portrait.y,
            y: this.a4Portrait.x
        };
        this.a5Portrait = {
            x: 420,
            y: 595
        };
        this.a5Landscape = {
            x: this.a5Portrait.y,
            y: this.a5Portrait.x
        };
        this.letterPortrait = {
            x: 612,
            y: 792
        };
        this.letterLandscape = {
            x: this.letterPortrait.y,
            y: this.letterPortrait.x
        };
        this.legalPortrait = {
            x: 612,
            y: 1008
        };
        this.legalLandscape = {
            x: this.legalPortrait.y,
            y: this.legalPortrait.x
        };
        this.portraitSizes = [
            this.a3Portrait, this.a4Portrait, this.a5Portrait, this.letterPortrait, this.legalPortrait
        ];
        this.landscapeSizes = [
            this.a3Landscape, this.a4Landscape, this.a5Landscape, this.letterLandscape, this.legalLandscape
        ];
        this.rFormWidth = 0;
        this.rFormHeight = 0;
        this.pageCache = [];
    }
    /**
     * @param {?} width
     * @param {?} height
     * @return {?}
     */
    ConversionService.prototype.setRValue = /**
     * @param {?} width
     * @param {?} height
     * @return {?}
     */
    function (width, height) {
        this.rFormWidth = width;
        this.rFormHeight = height;
    };
    /**
     * @param {?} pageCache
     * @return {?}
     */
    ConversionService.prototype.setPageCache = /**
     * @param {?} pageCache
     * @return {?}
     */
    function (pageCache) {
        this.pageCache = pageCache;
    };
    Object.defineProperty(ConversionService.prototype, "xRatio", {
        get: /**
         * @return {?}
         */
        function () {
            return this.XRatio;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} value
     * @return {?}
     */
    ConversionService.prototype.setXRatio = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.XRatio = value;
    };
    Object.defineProperty(ConversionService.prototype, "yRatio", {
        get: /**
         * @return {?}
         */
        function () {
            return this.YRatio;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} value
     * @return {?}
     */
    ConversionService.prototype.setYRatio = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.YRatio = value;
    };
    /**
     * @param {?} y
     * @param {?} fieldHeight
     * @param {?=} iTextHeight
     * @param {?=} yRatio
     * @return {?}
     */
    ConversionService.prototype.getRTop = /**
     * @param {?} y
     * @param {?} fieldHeight
     * @param {?=} iTextHeight
     * @param {?=} yRatio
     * @return {?}
     */
    function (y, fieldHeight, iTextHeight, yRatio) {
        if (!iTextHeight) {
            iTextHeight = this.ITextHeight;
        }
        if (!yRatio) {
            yRatio = this.yRatio;
        }
        return iTextHeight - ((y + fieldHeight) * yRatio);
    };
    /**
     * @param {?} y
     * @param {?=} ratio
     * @return {?}
     */
    ConversionService.prototype.getRValue = /**
     * @param {?} y
     * @param {?=} ratio
     * @return {?}
     */
    function (y, ratio) {
        if (!ratio) {
            ratio = this.yRatio;
        }
        return y * ratio;
    };
    /**
     * @param {?} x
     * @param {?=} ratio
     * @return {?}
     */
    ConversionService.prototype.getRLeft = /**
     * @param {?} x
     * @param {?=} ratio
     * @return {?}
     */
    function (x, ratio) {
        if (!ratio) {
            ratio = this.xRatio;
        }
        return x * ratio;
    };
    /**
     * @param {?} x
     * @param {?} xRatio
     * @return {?}
     */
    ConversionService.prototype.getITextLeft = /**
     * @param {?} x
     * @param {?} xRatio
     * @return {?}
     */
    function (x, xRatio) {
        if (!xRatio) {
            xRatio = this.xRatio;
        }
        return x / xRatio;
    };
    /**
     * @param {?} y
     * @param {?} fieldHeight
     * @param {?=} rFormHeight
     * @param {?=} yRatio
     * @return {?}
     */
    ConversionService.prototype.getITextBottomFromRTop = /**
     * @param {?} y
     * @param {?} fieldHeight
     * @param {?=} rFormHeight
     * @param {?=} yRatio
     * @return {?}
     */
    function (y, fieldHeight, rFormHeight, yRatio) {
        if (!rFormHeight) {
            rFormHeight = this.rFormHeight;
        }
        /** @type {?} */
        var uiHeight = fieldHeight;
        return (rFormHeight - y - uiHeight);
    };
    /**
     * @param {?} y
     * @return {?}
     */
    ConversionService.prototype.getITextBottom = /**
     * @param {?} y
     * @return {?}
     */
    function (y) {
        return y;
    };
    /**
     * @param {?} field
     * @return {?}
     */
    ConversionService.prototype.getDefaultHeight = /**
     * @param {?} field
     * @return {?}
     */
    function (field) {
        return getDefaultHeight(field);
    };
    /**
     * @param {?} field
     * @return {?}
     */
    ConversionService.prototype.getDefaultWidth = /**
     * @param {?} field
     * @return {?}
     */
    function (field) {
        return getDefaultWidth(field);
    };
    /**
     * @param {?} field
     * @param {?} pageHeight
     * @param {?=} xRatio
     * @param {?=} yRatio
     * @return {?}
     */
    ConversionService.prototype.getUIFieldStyling = /**
     * @param {?} field
     * @param {?} pageHeight
     * @param {?=} xRatio
     * @param {?=} yRatio
     * @return {?}
     */
    function (field, pageHeight, xRatio, yRatio) {
        if (!xRatio) {
            xRatio = this.xRatio;
        }
        if (!yRatio) {
            yRatio = this.yRatio;
        }
        return getUIFieldStyling(field);
    };
    /**
     * @param {?} field
     * @param {?} x
     * @param {?} y
     * @param {?} pageHeight
     * @return {?}
     */
    ConversionService.prototype.getUIFieldSetting = /**
     * @param {?} field
     * @param {?} x
     * @param {?} y
     * @param {?} pageHeight
     * @return {?}
     */
    function (field, x, y, pageHeight) {
        /** @type {?} */
        var setting;
        if (field.type === 'textbox' || field.type === 'date' || field.type === 'placeholder') {
            setting = {
                width: field.setting.width ? field.setting.width : 150,
                height: field.setting.height ? field.setting.height : 15,
                x: field.setting.x ? field.setting.x : x,
                y: this.getITextBottomFromRTop(y, field.setting.height ? field.setting.height : 15, pageHeight)
            };
            if (field.type === 'date') {
                setting.width = 64;
            }
        }
        else if (field.type === 'timestamp') {
            setting = {
                width: field.setting.width ? field.setting.width : 64,
                height: field.setting.height ? field.setting.height : 15,
                x: field.setting.x ? field.setting.x : x,
                y: this.getITextBottomFromRTop(y, field.setting.height ? field.setting.height : 15, pageHeight)
            };
        }
        else if (field.type === 'signature' || field.type === 'initial') {
            setting = {
                x: field.setting.x ? field.setting.x : x,
                y: this.getITextBottomFromRTop(y, 36, pageHeight)
            };
        }
        else if (field.type === 'checkbox_group') {
            setting = {
                options: [{
                        id: v4(),
                        x: field.setting.x ? field.setting.x : x,
                        y: this.getITextBottomFromRTop(y, 13.5, pageHeight),
                        checked: false,
                        value: 'Option 1'
                    }]
            };
        }
        else if (field.type === 'radio_button_group') {
            setting = {
                options: [{
                        id: v4(),
                        x: field.setting.x ? field.setting.x : x,
                        y: this.getITextBottomFromRTop(y, 13.5, pageHeight),
                        selected: false,
                        value: 'Option 1'
                    }]
            };
        }
        else if (field.type === 'dropdown') {
            setting = {
                x: field.setting.x ? field.setting.x : x,
                y: this.getITextBottomFromRTop(y, 15, pageHeight),
                width: field.setting.width ? field.setting.width : 64,
                height: field.setting.height ? field.setting.height : 15,
                value: null,
                placeholder: 'Choose',
                options: [{
                        id: v4(),
                        value: 'Option 1'
                    }]
            };
        }
        else if (field.type === 'attachment') {
            setting = {
                x: field.setting.x ? field.setting.x : x,
                y: this.getITextBottomFromRTop(y, 24, pageHeight),
                label: ''
            };
        }
        else if (field.type === 'payment') {
            setting = {
                x: field.setting.x ? field.setting.x : x,
                y: this.getITextBottomFromRTop(y, 24, pageHeight),
                account_id: '',
                currency: 'usd',
                payment_fee_type: 'static',
                amount: 1,
                if_transfer_fee: false
            };
        }
        return setting;
    };
    /**
     * @param {?} newHeight
     * @param {?} newWidth
     * @return {?}
     */
    ConversionService.prototype.updateITextValues = /**
     * @param {?} newHeight
     * @param {?} newWidth
     * @return {?}
     */
    function (newHeight, newWidth) {
        this.ITextHeight = newHeight;
        this.ITextWidth = newWidth;
    };
    ConversionService.decorators = [
        { type: Injectable }
    ];
    return ConversionService;
}());
if (false) {
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.ITextWidth;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.ITextHeight;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.XRatio;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.YRatio;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.a3Portrait;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.a3Landscape;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.a4Portrait;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.a4Landscape;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.a5Portrait;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.a5Landscape;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.letterPortrait;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.letterLandscape;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.legalPortrait;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.legalLandscape;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.portraitSizes;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.landscapeSizes;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.rFormWidth;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.rFormHeight;
    /**
     * @type {?}
     * @private
     */
    ConversionService.prototype.pageCache;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var SignatureService = /** @class */ (function () {
    function SignatureService(httpClient, injector, envelopeSvc, vTokenStateService) {
        this.httpClient = httpClient;
        this.injector = injector;
        this.envelopeSvc = envelopeSvc;
        this.vTokenStateService = vTokenStateService;
        this.workingField = new FieldData({ order: 0, fName: '' });
        this.fields = [];
        this._total = -1;
        this._fields = new ReplaySubject();
        this._recipient = new ReplaySubject();
        this._showSig = new ReplaySubject();
        this._signedFields = new BehaviorSubject({});
        this._envId = new ReplaySubject();
        this._rName = new ReplaySubject();
        this.signImgSubject = new ReplaySubject();
        this.initialImgSubject = new ReplaySubject();
        this.initialIdSubject = new ReplaySubject();
        this.signatureIdSubject = new ReplaySubject();
        this.signedFields = {};
        this.mode = '';
        this._config_token = this.injector.get(EssentialsConfigToken);
        this._rForm_backend_url = this._config_token.rForm_backend_url;
        this.envUrl = this._rForm_backend_url + "/envelopes";
    }
    /**
     * @param {?} payment
     * @return {?}
     */
    SignatureService.prototype.setWorkingPayment = /**
     * @param {?} payment
     * @return {?}
     */
    function (payment) {
        this.workingPayment = payment;
    };
    /**
     * @param {?} fields
     * @return {?}
     */
    SignatureService.prototype.updateFields = /**
     * @param {?} fields
     * @return {?}
     */
    function (fields) {
        this.fields = fields;
        this._fields.next(this.fields);
    };
    /**
     * @return {?}
     */
    SignatureService.prototype.postSignatureBlob = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var blobFile = this.signatureBlob;
        /** @type {?} */
        var formData = new FormData();
        formData.append('signature', blobFile, blobFile['name']);
        /** @type {?} */
        var request = new HttpRequest('POST', this._rForm_backend_url + '/signatures', formData);
        return this.httpClient.request(request)
            .toPromise()
            .then((/**
         * @param {?} response
         * @return {?}
         */
        function (response) {
            if (response && response['body']) {
                return response['body'];
            }
            else {
                console.error('Failed to upload signature image');
            }
        }));
    };
    /**
     * @return {?}
     */
    SignatureService.prototype.postInitialBlob = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var blobFile = this.initialBlob;
        /** @type {?} */
        var formData = new FormData();
        formData.append('initial', blobFile, blobFile['name']);
        /** @type {?} */
        var request = new HttpRequest('POST', this._rForm_backend_url + '/initials', formData);
        return this.httpClient.request(request)
            .toPromise()
            .then((/**
         * @param {?} response
         * @return {?}
         */
        function (response) {
            if (response && response['body']) {
                return response['body'];
            }
            else {
                console.error('Failed to upload Initial');
            }
        }));
    };
    /**
     * @param {?} bool
     * @return {?}
     */
    SignatureService.prototype.toggleSig = /**
     * @param {?} bool
     * @return {?}
     */
    function (bool) {
        this._showSig.next(bool);
    };
    /**
     * @param {?} id
     * @return {?}
     */
    SignatureService.prototype.setSigId = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        this.signatureId = id;
        this.signatureIdSubject.next(id);
    };
    /**
     * @param {?} id
     * @return {?}
     */
    SignatureService.prototype.setInitialId = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        this.initialId = id;
        this.initialIdSubject.next(id);
    };
    /**
     * @param {?} url
     * @return {?}
     */
    SignatureService.prototype.setSignImg = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        this.signImg = url;
        this.signImgSubject.next(this.signImg);
    };
    /**
     * @return {?}
     */
    SignatureService.prototype.getSignImg = /**
     * @return {?}
     */
    function () {
        return this.signImg;
    };
    /**
     * @param {?} url
     * @return {?}
     */
    SignatureService.prototype.setInitialImg = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        this.initialImg = url;
        this.initialImgSubject.next(this.initialImg);
    };
    /**
     * @return {?}
     */
    SignatureService.prototype.getInitialImg = /**
     * @return {?}
     */
    function () {
        return this.initialImg;
    };
    /**
     * @param {?} envelopeId
     * @param {?} fieldName
     * @param {?} signatureId
     * @return {?}
     */
    SignatureService.prototype.putSignatureField = /**
     * @param {?} envelopeId
     * @param {?} fieldName
     * @param {?} signatureId
     * @return {?}
     */
    function (envelopeId, fieldName, signatureId) {
        var _this = this;
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        function (resolve, reject) { return __awaiter(_this, void 0, void 0, function () {
            var ipAddress;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getPublicIp()];
                    case 1:
                        ipAddress = _a.sent();
                        this.httpClient
                            .put(this.envUrl + ("/" + envelopeId + "/fields/" + fieldName + "/signature/" + signatureId), {
                            ip_address: ipAddress
                        }).toPromise().then((/**
                         * @param {?} res
                         * @return {?}
                         */
                        function (res) {
                            return resolve(res);
                        }));
                        return [2 /*return*/];
                }
            });
        }); }));
    };
    /**
     * @param {?} envelopeId
     * @param {?} fieldName
     * @param {?} initialId
     * @return {?}
     */
    SignatureService.prototype.putInitialField = /**
     * @param {?} envelopeId
     * @param {?} fieldName
     * @param {?} initialId
     * @return {?}
     */
    function (envelopeId, fieldName, initialId) {
        var _this = this;
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        function (resolve, reject) { return __awaiter(_this, void 0, void 0, function () {
            var ipAddress;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getPublicIp()];
                    case 1:
                        ipAddress = _a.sent();
                        this.httpClient
                            .put(this.envUrl + ("/" + envelopeId + "/fields/" + fieldName + "/initial/" + initialId), {
                            ip_address: ipAddress
                        }).toPromise().then((/**
                         * @param {?} res
                         * @return {?}
                         */
                        function (res) {
                            return resolve(res);
                        }));
                        return [2 /*return*/];
                }
            });
        }); }));
    };
    // refactor it to 4 functions
    // refactor it to 4 functions
    /**
     * @param {?} envelopeId
     * @param {?=} ifPrepared
     * @return {?}
     */
    SignatureService.prototype.updateEnvelopeField = 
    // refactor it to 4 functions
    /**
     * @param {?} envelopeId
     * @param {?=} ifPrepared
     * @return {?}
     */
    function (envelopeId, ifPrepared) {
        if (this.workingField && this.workingField.type) {
            /** @type {?} */
            var result = this.getEnvelopeFieldValue(ifPrepared);
            this.putEnvelopeField(envelopeId, result);
        }
    };
    /**
     * @param {?=} isPrepared
     * @return {?}
     */
    SignatureService.prototype.getEnvelopeFieldValue = /**
     * @param {?=} isPrepared
     * @return {?}
     */
    function (isPrepared) {
        if (this.workingField && this.workingField.type) {
            /** @type {?} */
            var result = {
                value: ''
            };
            if (typeof (isPrepared) === 'boolean') {
                result['prepared'] = isPrepared;
            }
            switch (this.workingField.type.toLowerCase()) {
                case 'textbox':
                case 'checkbox':
                case 'date':
                    result.value = this.workingField.value;
                    break;
                default:
                    break;
            }
            return result;
        }
        return null;
    };
    /**
     * @param {?} envelopeId
     * @param {?} result
     * @return {?}
     */
    SignatureService.prototype.putEnvelopeField = /**
     * @param {?} envelopeId
     * @param {?} result
     * @return {?}
     */
    function (envelopeId, result) {
        var _this = this;
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        function (resolve, reject) {
            _this.httpClient
                .put(_this.envUrl + ("/" + envelopeId + "/fields/" + _this.workingField.fName), result).toPromise().then((/**
             * @param {?} res
             * @return {?}
             */
            function (res) {
                return resolve(res);
            }));
        }));
    };
    /**
     * @param {?} envelopeId
     * @param {?} body
     * @param {?} fieldName
     * @param {?=} ifPrepared
     * @return {?}
     */
    SignatureService.prototype.updateGroupedField = /**
     * @param {?} envelopeId
     * @param {?} body
     * @param {?} fieldName
     * @param {?=} ifPrepared
     * @return {?}
     */
    function (envelopeId, body, fieldName, ifPrepared) {
        var _this = this;
        if (typeof (ifPrepared) === 'boolean') {
            body['prepared'] = ifPrepared;
        }
        return new Promise((/**
         * @param {?} resolve
         * @param {?} reject
         * @return {?}
         */
        function (resolve, reject) {
            _this.httpClient
                .put(_this.envUrl + ("/" + envelopeId + "/fields/" + fieldName), body).toPromise().then((/**
             * @param {?} res
             * @return {?}
             */
            function (res) {
                return resolve(res);
            }));
        }));
    };
    /**
     * @param {?} envelopeId
     * @param {?} setPrepared
     * @return {?}
     */
    SignatureService.prototype.prepareEnvelopeField = /**
     * @param {?} envelopeId
     * @param {?} setPrepared
     * @return {?}
     */
    function (envelopeId, setPrepared) {
        return __awaiter(this, void 0, void 0, function () {
            var fieldType, fieldResponse;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!(this.workingField && this.workingField.type)) return [3 /*break*/, 3];
                        fieldType = this.workingField.type.toLowerCase();
                        if (!(fieldType !== 'textbox' && fieldType !== 'date' && fieldType !== 'checkbox')) return [3 /*break*/, 1];
                        return [2 /*return*/, false];
                    case 1: return [4 /*yield*/, this.updateEnvelopeField(envelopeId, setPrepared)];
                    case 2:
                        fieldResponse = _a.sent();
                        if (fieldResponse) {
                            return [2 /*return*/, fieldResponse.prepared];
                        }
                        _a.label = 3;
                    case 3: return [2 /*return*/, false];
                }
            });
        });
    };
    /**
     * @private
     * @return {?}
     */
    SignatureService.prototype.getPublicIp = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var apiUrl = 'https://api.ipify.org?format=jsonp';
        /** @type {?} */
        var callback = 'callback=JSONP_CALLBACK';
        return this.httpClient
            .jsonp(apiUrl, callback)
            .toPromise()
            .then((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            return res['ip'];
        })).catch((/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            console.error('Failed to get ip', err);
            return 'ip_unavailable';
        }));
    };
    /**
     * @param {?} fName
     * @param {?} result
     * @param {?} vName
     * @param {?} pageNum
     * @param {?} id
     * @param {?} required
     * @param {?} order
     * @param {?} type
     * @return {?}
     */
    SignatureService.prototype.updateCurrentField = /**
     * @param {?} fName
     * @param {?} result
     * @param {?} vName
     * @param {?} pageNum
     * @param {?} id
     * @param {?} required
     * @param {?} order
     * @param {?} type
     * @return {?}
     */
    function (fName, result, vName, pageNum, id, required, order, type) {
        this.workingField.fName = fName;
        this.workingField.value = result;
        this.workingField.pageNum = pageNum;
        this.workingField.id = id;
        this.workingField.vName = vName;
        this.workingField.required = required;
        this.workingField.order = order;
        this.workingField.type = type;
    };
    /**
     * @param {?} fields
     * @return {?}
     */
    SignatureService.prototype.setCurrentFields = /**
     * @param {?} fields
     * @return {?}
     */
    function (fields) {
        this.currentFields = fields;
    };
    Object.defineProperty(SignatureService.prototype, "currField", {
        get: /**
         * @return {?}
         */
        function () {
            return this.workingField;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} pageNum
     * @param {?} id
     * @return {?}
     */
    SignatureService.prototype.errorMessages = /**
     * @param {?} pageNum
     * @param {?} id
     * @return {?}
     */
    function (pageNum, id) {
        return this.currentFields[pageNum][id]['error'];
    };
    /**
     * @param {?} id
     * @return {?}
     */
    SignatureService.prototype.setEnvId = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        this._envId.next(id);
    };
    /**
     * @param {?} role
     * @return {?}
     */
    SignatureService.prototype.setrName = /**
     * @param {?} role
     * @return {?}
     */
    function (role) {
        this.rName = role;
        this._rName.next(role);
    };
    /**
     * @param {?} data
     * @return {?}
     */
    SignatureService.prototype.getSignatureUrl = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        if (data) {
            /** @type {?} */
            var blob = dataURLtoBlob(data);
            /** @type {?} */
            var url = URL.createObjectURL(blob);
            return url;
        }
        return null;
    };
    /**
     * @param {?} data
     * @return {?}
     */
    SignatureService.prototype.setSignatureData = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        /** @type {?} */
        var blob = dataURLtoBlob(data);
        /** @type {?} */
        var url = URL.createObjectURL(blob);
        this.signatureBlob = blob;
        this.setSignImg(url);
    };
    /**
     * @param {?} data
     * @return {?}
     */
    SignatureService.prototype.setInitialData = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        /** @type {?} */
        var blob = dataURLtoBlob(data);
        /** @type {?} */
        var url = URL.createObjectURL(blob);
        this.initialBlob = blob;
        this.setInitialImg(url);
    };
    /**
     * @param {?} fName
     * @param {?} result
     * @return {?}
     */
    SignatureService.prototype.updateSigned = /**
     * @param {?} fName
     * @param {?} result
     * @return {?}
     */
    function (fName, result) {
        this.workingField['value'] = 'signed';
        this.signedFields[fName] = result;
        this._signedFields.next(this.signedFields);
    };
    /**
     * @param {?} fName
     * @param {?} result
     * @return {?}
     */
    SignatureService.prototype.updateInitialed = /**
     * @param {?} fName
     * @param {?} result
     * @return {?}
     */
    function (fName, result) {
        this.workingField['value'] = 'initialed';
        this.signedFields[fName] = result;
        this._signedFields.next(this.signedFields);
    };
    /**
     * @param {?} id
     * @return {?}
     */
    SignatureService.prototype.setSignatureId = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        this.signatureId = id;
        this.signatureIdSubject.next(id);
    };
    /**
     * @param {?} recipient
     * @return {?}
     */
    SignatureService.prototype.setRecipient = /**
     * @param {?} recipient
     * @return {?}
     */
    function (recipient) {
        this._recipient.next(recipient);
    };
    /**
     * @param {?} t
     * @return {?}
     */
    SignatureService.prototype.setTotal = /**
     * @param {?} t
     * @return {?}
     */
    function (t) {
        this._total = t;
    };
    SignatureService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    SignatureService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: Injector },
        { type: DocumentsService },
        { type: VerdocsStateService }
    ]; };
    return SignatureService;
}());
if (false) {
    /**
     * @type {?}
     * @private
     */
    SignatureService.prototype._config_token;
    /**
     * @type {?}
     * @private
     */
    SignatureService.prototype.signatureId;
    /**
     * @type {?}
     * @private
     */
    SignatureService.prototype.initialId;
    /**
     * @type {?}
     * @private
     */
    SignatureService.prototype._rForm_backend_url;
    /**
     * @type {?}
     * @private
     */
    SignatureService.prototype.envUrl;
    /**
     * @type {?}
     * @private
     */
    SignatureService.prototype.currentFields;
    /**
     * @type {?}
     * @private
     */
    SignatureService.prototype.workingField;
    /**
     * @type {?}
     * @private
     */
    SignatureService.prototype.signatureBlob;
    /**
     * @type {?}
     * @private
     */
    SignatureService.prototype.initialBlob;
    /**
     * @type {?}
     * @private
     */
    SignatureService.prototype.fields;
    /**
     * @type {?}
     * @private
     */
    SignatureService.prototype._total;
    /**
     * @type {?}
     * @private
     */
    SignatureService.prototype.initialImg;
    /**
     * @type {?}
     * @private
     */
    SignatureService.prototype.workingPayment;
    /**
     * @type {?}
     * @private
     */
    SignatureService.prototype.rName;
    /**
     * @type {?}
     * @private
     */
    SignatureService.prototype.recipients;
    /** @type {?} */
    SignatureService.prototype._fields;
    /** @type {?} */
    SignatureService.prototype._recipient;
    /** @type {?} */
    SignatureService.prototype._showSig;
    /** @type {?} */
    SignatureService.prototype._signedFields;
    /** @type {?} */
    SignatureService.prototype._envId;
    /** @type {?} */
    SignatureService.prototype._rName;
    /** @type {?} */
    SignatureService.prototype.signImgSubject;
    /** @type {?} */
    SignatureService.prototype.initialImgSubject;
    /** @type {?} */
    SignatureService.prototype.initialIdSubject;
    /** @type {?} */
    SignatureService.prototype.signatureIdSubject;
    /** @type {?} */
    SignatureService.prototype.signedFields;
    /** @type {?} */
    SignatureService.prototype.signImg;
    /** @type {?} */
    SignatureService.prototype.mode;
    /**
     * @type {?}
     * @private
     */
    SignatureService.prototype.httpClient;
    /**
     * @type {?}
     * @private
     */
    SignatureService.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    SignatureService.prototype.envelopeSvc;
    /**
     * @type {?}
     * @private
     */
    SignatureService.prototype.vTokenStateService;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var EnvelopeViewService = /** @class */ (function () {
    function EnvelopeViewService(signatureService) {
        var _this = this;
        this.signatureService = signatureService;
        this.recipientSubject = new ReplaySubject();
        this.jumpCoordinateSubject = new ReplaySubject();
        this.fieldTypeSubject = new ReplaySubject();
        this.pdfUrlSubject = new ReplaySubject();
        this.pdfBlobSubject = new ReplaySubject();
        this.toggleNextSubject = new ReplaySubject();
        this.viewModeSubject = new ReplaySubject();
        this.agreedSubject = new ReplaySubject(1, 500);
        this.attachmentsSubject = new ReplaySubject();
        this.formValiditySubject = new ReplaySubject();
        this.requestFormValiditySubject = new ReplaySubject();
        this.signatureService._rName.subscribe((/**
         * @param {?} role
         * @return {?}
         */
        function (role) {
            _this.rName = role;
        }));
    }
    /**
     * @param {?} mode
     * @return {?}
     */
    EnvelopeViewService.prototype.setMode = /**
     * @param {?} mode
     * @return {?}
     */
    function (mode) {
        this.mode = mode;
        this.viewModeSubject.next(this.mode);
    };
    /**
     * @param {?} signatureService
     * @return {?}
     */
    EnvelopeViewService.prototype.setSignatureService = /**
     * @param {?} signatureService
     * @return {?}
     */
    function (signatureService) {
        this.signatureService = signatureService;
    };
    /**
     * @param {?} value
     * @return {?}
     */
    EnvelopeViewService.prototype.requestFormValidity = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.requestFormValiditySubject.next(value);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    EnvelopeViewService.prototype.setFormValidityData = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.formValiditySubject.next(value);
    };
    /**
     * @param {?} type
     * @return {?}
     */
    EnvelopeViewService.prototype.ifBelongsToPreparer = /**
     * @param {?} type
     * @return {?}
     */
    function (type) {
        return type === 'textbox' || type === 'date' || type === 'checkbox';
    };
    /**
     * @param {?} field
     * @return {?}
     */
    EnvelopeViewService.prototype.ifBelongsToCurrentRecipient = /**
     * @param {?} field
     * @return {?}
     */
    function (field) {
        if (this.mode === 'prepareview') {
            return this.ifBelongsToPreparer(field.controlType);
        }
        else {
            return field.recipientRole === this.rName;
        }
    };
    EnvelopeViewService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    EnvelopeViewService.ctorParameters = function () { return [
        { type: SignatureService }
    ]; };
    return EnvelopeViewService;
}());
if (false) {
    /** @type {?} */
    EnvelopeViewService.prototype.recipientSubject;
    /** @type {?} */
    EnvelopeViewService.prototype.jumpCoordinateSubject;
    /** @type {?} */
    EnvelopeViewService.prototype.fieldTypeSubject;
    /** @type {?} */
    EnvelopeViewService.prototype.pdfUrlSubject;
    /** @type {?} */
    EnvelopeViewService.prototype.pdfBlobSubject;
    /** @type {?} */
    EnvelopeViewService.prototype.toggleNextSubject;
    /** @type {?} */
    EnvelopeViewService.prototype.viewModeSubject;
    /** @type {?} */
    EnvelopeViewService.prototype.agreedSubject;
    /** @type {?} */
    EnvelopeViewService.prototype.attachmentsSubject;
    /** @type {?} */
    EnvelopeViewService.prototype.formValiditySubject;
    /** @type {?} */
    EnvelopeViewService.prototype.requestFormValiditySubject;
    /** @type {?} */
    EnvelopeViewService.prototype.mode;
    /**
     * @type {?}
     * @private
     */
    EnvelopeViewService.prototype.rName;
    /**
     * @type {?}
     * @private
     */
    EnvelopeViewService.prototype.signatureService;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
var Robots = {
    Noindex: 'Noindex',
    Index: 'Index',
    Follow: 'Follow',
    Noimageindex: 'Noimageindex',
    None: 'None',
    Noarchive: 'Noarchive',
    Nocache: 'Nocache',
    Nosnippet: 'Nosnippet',
};
var PageService = /** @class */ (function () {
    function PageService(titleService, injector, meta, rendererFactory, activatedRoute, router, vTokenStateService, tokenObject, platform, document) {
        this.titleService = titleService;
        this.injector = injector;
        this.meta = meta;
        this.rendererFactory = rendererFactory;
        this.activatedRoute = activatedRoute;
        this.router = router;
        this.vTokenStateService = vTokenStateService;
        this.tokenObject = tokenObject;
        this.platform = platform;
        this.document = document;
        this.segmentIsInitialized = false;
        this.lazyTitle = false;
        this._config_token = this.injector.get(EssentialsConfigToken);
    }
    /**
     * @return {?}
     */
    PageService.prototype.inti = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.initializeSegment();
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.additionalSetupForWidgets();
        }), 500);
        this.router.events.pipe(filter$1((/**
         * @param {?} event
         * @return {?}
         */
        function (event) { return event instanceof NavigationEnd; })), map((/**
         * @return {?}
         */
        function () { return _this.activatedRoute; })), map((/**
         * @param {?} route
         * @return {?}
         */
        function (route) {
            /** @type {?} */
            var targetRoute = route;
            while (targetRoute.firstChild) {
                targetRoute = targetRoute.firstChild;
            }
            return targetRoute;
        })), switchMap((/**
         * @param {?} route
         * @return {?}
         */
        function (route) { return route.data; })), map((/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            _this.lazyTitle = data.lazyTitle;
            return data.title;
        }))).subscribe((/**
         * @param {?} title
         * @return {?}
         */
        function (title) {
            if (_this.lazyTitle) {
                _this.lazyTitle = false;
            }
            else {
                _this.setTitleAndRecord(title);
            }
        }));
    };
    /**
     * @param {?} title
     * @param {?=} pageName
     * @return {?}
     */
    PageService.prototype.setTitleAndRecord = /**
     * @param {?} title
     * @param {?=} pageName
     * @return {?}
     */
    function (title, pageName) {
        /** @type {?} */
        var newTitle = 'Verdocs';
        if (isPlatformBrowser(this.platform)) {
            if (this.document.cookie.includes(this._config_token.rForm_cookie_name)) {
                newTitle = title ? title : 'Verdocs';
                window['analytics'].page(pageName || newTitle);
            }
        }
        else {
            newTitle = 'Verdocs' + (title ? " | " + title : '');
        }
        this.titleService.setTitle(newTitle);
    };
    /**
     * @param {?} property
     * @param {?} content
     * @return {?}
     */
    PageService.prototype.setMetaProperty = /**
     * @param {?} property
     * @param {?} content
     * @return {?}
     */
    function (property, content) {
        if (!!content) {
            if (!!this.getMetaTagProperty(property)) {
                this.meta.updateTag({ property: property, content: content }, "property=\"" + property + "\"");
            }
            else {
                this.meta.addTag({ property: property, content: content }, true);
            }
        }
        else {
            if (!!this.getMetaTagProperty(property)) {
                this.meta.removeTag("property=\u201C" + property + "\u201D");
            }
        }
    };
    /**
     * @param {?} property
     * @param {?} content
     * @return {?}
     */
    PageService.prototype.setMetaName = /**
     * @param {?} property
     * @param {?} content
     * @return {?}
     */
    function (property, content) {
        if (!!content) {
            if (!!this.getMetaTagProperty(property)) {
                this.meta.updateTag({ property: property, content: content }, "name=\"" + property + "\"");
            }
            else {
                this.meta.addTag({ property: property, content: content }, true);
            }
        }
        else {
            if (!!this.getMetaTagProperty(property)) {
                this.meta.removeTag("name=\u201C" + property + "\u201D");
            }
        }
    };
    /**
     * @param {?} type
     * @param {?} title
     * @param {?} description
     * @param {?} image
     * @param {?} url
     * @return {?}
     */
    PageService.prototype.setOpenGraphMeta = /**
     * @param {?} type
     * @param {?} title
     * @param {?} description
     * @param {?} image
     * @param {?} url
     * @return {?}
     */
    function (type, title, description, image, url) {
        if (!!type) {
            this.setMetaProperty('og:type', type);
        }
        if (!!title) {
            this.setMetaProperty('og:title', title);
        }
        if (!!description) {
            this.setMetaProperty('og:description', description);
        }
        if (!!image) {
            this.setMetaProperty('og:image', image);
        }
        if (!!url) {
            this.setMetaProperty('og:url', url);
        }
        if (!!this.getMetaTagProperty('og:site_name')) {
            return;
        }
        else {
            this.meta.addTag({ property: 'og:site_name', content: 'Verdocs' });
        }
    };
    /**
     * @param {...?} robot_values
     * @return {?}
     */
    PageService.prototype.setRobotMeta = /**
     * @param {...?} robot_values
     * @return {?}
     */
    function () {
        var robot_values = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            robot_values[_i] = arguments[_i];
        }
        /** @type {?} */
        var parameters = '';
        for (var x = 0; x < robot_values.length; x++) {
            if (robot_values[x]) {
                if (x === robot_values.length - 1) {
                    parameters += robot_values[x];
                }
                else {
                    parameters += robot_values[x] + ", ";
                }
            }
        }
        /** @type {?} */
        var robot_parameters = parameters;
        this.setMetaName('robots', robot_parameters);
    };
    /**
     * @param {?} tag
     * @return {?}
     */
    PageService.prototype.setCanonicalUrl = /**
     * @param {?} tag
     * @return {?}
     */
    function (tag) {
        if (!tag.href) {
            tag['href'] = this.document.url;
        }
        this.addTag(tag);
    };
    /**
     * @param {?} content
     * @return {?}
     */
    PageService.prototype.setDescriptionMeta = /**
     * @param {?} content
     * @return {?}
     */
    function (content) {
        if (!!this.getMetaTag('description')) {
            if (content) {
                this.meta.updateTag({ name: 'description', content: content }, "name='description'");
            }
            else {
                this.meta.removeTag("name='description'");
            }
        }
        else {
            if (content) {
                this.meta.addTag({ name: 'description', content: content }, true);
            }
        }
    };
    /**
     * @param {?} name
     * @return {?}
     */
    PageService.prototype.getMetaTag = /**
     * @param {?} name
     * @return {?}
     */
    function (name) {
        return this.meta.getTag("name='" + name + "'");
    };
    /**
     * @param {?} property
     * @return {?}
     */
    PageService.prototype.getMetaTagProperty = /**
     * @param {?} property
     * @return {?}
     */
    function (property) {
        return this.meta.getTag("property='" + property + "'");
    };
    /**
     * @private
     * @return {?}
     */
    PageService.prototype.initializeSegment = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        if (isPlatformBrowser(this.platform)) {
            /** @type {?} */
            var analytics_1 = window['analytics'];
            if (document.cookie.includes(this._config_token.rForm_cookie_name)) {
                /** @type {?} */
                var idTokenSubscription_1 = this.vTokenStateService.decodedIdTokenSubject.subscribe((/**
                 * @param {?} decodedIdToken
                 * @return {?}
                 */
                function (decodedIdToken) {
                    idTokenSubscription_1.unsubscribe();
                    if (!_this.segmentIsInitialized) {
                        _this.segmentIsInitialized = true;
                        analytics_1.reset();
                        /** @type {?} */
                        var profile = _this.tokenObject.getProfile();
                        analytics_1.identify(decodedIdToken.sub, {
                            userId: decodedIdToken.sub,
                            email: decodedIdToken.email,
                            name: decodedIdToken.name,
                            'Verdoc plan': _this.planType('verdoc'),
                            'Organization plan': _this.planType('organization'),
                            'Current Profile Id': _this.tokenObject.getProfileID(),
                            'Current Company': profile.organization.name,
                            company: {
                                id: profile.organization_id,
                                name: profile.organization.name
                            }
                        }, {
                            Intercom: { hideDefaultLauncher: true }
                        });
                    }
                }));
            }
        }
    };
    /**
     * @return {?}
     */
    PageService.prototype.robotEnum = /**
     * @return {?}
     */
    function () {
        return Robots;
    };
    /**
     * @private
     * @param {?} product
     * @return {?}
     */
    PageService.prototype.planType = /**
     * @private
     * @param {?} product
     * @return {?}
     */
    function (product) {
        /** @type {?} */
        var productName = 'undefined';
        switch (product) {
            case 'verdoc':
                productName = 'env';
                break;
            case 'organization':
                productName = 'org';
                break;
        }
        /** @type {?} */
        var plan = this.tokenObject.getPlans().find((/**
         * @param {?} p
         * @return {?}
         */
        function (p) {
            return p.includes(productName);
        }));
        /** @type {?} */
        var planType = 'None';
        if (plan) {
            if (productName === 'env') {
                planType = plan.includes('pro') ? 'Pro' : 'Essential';
            }
            else {
                planType = plan.includes('premium') ? 'Premium' : 'Standard';
            }
        }
        return planType;
    };
    /**
     * @param {?} tag
     * @param {?=} forceCreation
     * @return {?}
     */
    PageService.prototype.addTag = /**
     * @param {?} tag
     * @param {?=} forceCreation
     * @return {?}
     */
    function (tag, forceCreation) {
        /** @type {?} */
        var renderer = this.rendererFactory.createRenderer(this.document, {
            id: '-1',
            encapsulation: ViewEncapsulation.None,
            styles: [],
            data: {}
        });
        /** @type {?} */
        var link = renderer.createElement('link');
        /** @type {?} */
        var head = this.document.head;
        /** @type {?} */
        var selector = this._parseSelector(tag);
        Object.keys(tag).forEach((/**
         * @param {?} prop
         * @return {?}
         */
        function (prop) {
            return renderer.setAttribute(link, prop, tag[prop]);
        }));
        try {
            if (head === null) {
                throw new Error('<head> not found within DOCUMENT.');
            }
            /** @type {?} */
            var existingLink = renderer.selectRootElement(selector);
            if (!!existingLink) {
                renderer.removeAttribute(existingLink, 'href');
                renderer.setAttribute(existingLink, 'href', tag['href']);
            }
            else {
                renderer.appendChild(head, link);
            }
        }
        catch (e) {
            // console.error('Error within linkService : ', e);
            renderer.appendChild(head, link);
        }
    };
    /**
     * @private
     * @param {?} tag
     * @return {?}
     */
    PageService.prototype._parseSelector = /**
     * @private
     * @param {?} tag
     * @return {?}
     */
    function (tag) {
        /** @type {?} */
        var attr = tag.rel ? 'rel' : 'hreflang';
        return "link[" + attr + "=\"" + tag[attr] + "\"]";
    };
    /**
     * @private
     * @return {?}
     */
    PageService.prototype.additionalSetupForWidgets = /**
     * @private
     * @return {?}
     */
    function () {
        if (isPlatformBrowser(this.platform)) {
            if (window['Intercom']) {
                window['Intercom']('onHide', (/**
                 * @return {?}
                 */
                function () {
                    window['intercomIsOpen'] = false;
                }));
                window['Intercom']('onShow', (/**
                 * @return {?}
                 */
                function () {
                    window['intercomIsOpen'] = true;
                }));
            }
            window['lunchIntercom'] = (/**
             * @return {?}
             */
            function () {
                if (window['Intercom']) {
                    if (window['intercomIsOpen'] && window['intercomIsOpen'] === true) {
                        window['Intercom']('hide');
                    }
                    else {
                        window['Intercom']('show');
                    }
                }
                else {
                    console.error('Intercom is not loaded!');
                }
            });
            window['lunchElevio'] = (/**
             * @return {?}
             */
            function () {
                if (window['_elev']) {
                    if (window['_elev'].widgetIsOpen()) {
                        window['_elev'].close();
                    }
                    else {
                        window['_elev'].open();
                    }
                }
                else {
                    console.error('Elevio is not loaded!');
                }
            });
        }
    };
    /** @nocollapse */
    PageService.ctorParameters = function () { return [
        { type: Title },
        { type: Injector },
        { type: Meta },
        { type: RendererFactory2 },
        { type: ActivatedRoute },
        { type: Router },
        { type: VerdocsStateService },
        { type: VerdocsTokenObjectService },
        { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
    ]; };
    return PageService;
}());
if (false) {
    /**
     * @type {?}
     * @private
     */
    PageService.prototype._config_token;
    /**
     * @type {?}
     * @private
     */
    PageService.prototype.segmentIsInitialized;
    /**
     * @type {?}
     * @private
     */
    PageService.prototype.lazyTitle;
    /**
     * @type {?}
     * @private
     */
    PageService.prototype.titleService;
    /**
     * @type {?}
     * @private
     */
    PageService.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    PageService.prototype.meta;
    /**
     * @type {?}
     * @private
     */
    PageService.prototype.rendererFactory;
    /**
     * @type {?}
     * @private
     */
    PageService.prototype.activatedRoute;
    /**
     * @type {?}
     * @private
     */
    PageService.prototype.router;
    /**
     * @type {?}
     * @private
     */
    PageService.prototype.vTokenStateService;
    /**
     * @type {?}
     * @private
     */
    PageService.prototype.tokenObject;
    /**
     * @type {?}
     * @private
     */
    PageService.prototype.platform;
    /**
     * @type {?}
     * @private
     */
    PageService.prototype.document;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var ModalService = /** @class */ (function () {
    function ModalService() {
        this.modals = [];
        this.subject = new Subject();
    }
    /**
     * @param {?} modal
     * @return {?}
     */
    ModalService.prototype.add = /**
     * @param {?} modal
     * @return {?}
     */
    function (modal) {
        this.modals.push(modal);
    };
    /**
     * @param {?} id
     * @return {?}
     */
    ModalService.prototype.open = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        /** @type {?} */
        var modal = find(this.modals, { id: id });
    };
    /**
     * @param {?} id
     * @return {?}
     */
    ModalService.prototype.close = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        /** @type {?} */
        var modal = find(this.modals, { id: id });
        modal.close();
    };
    /**
     * @param {?} componentData
     * @return {?}
     */
    ModalService.prototype.setComponentData = /**
     * @param {?} componentData
     * @return {?}
     */
    function (componentData) {
        this.componentData = componentData;
        this.subject.next(componentData);
    };
    /**
     * @return {?}
     */
    ModalService.prototype.getComponentData = /**
     * @return {?}
     */
    function () {
        return this.subject.asObservable();
    };
    ModalService.decorators = [
        { type: Injectable }
    ];
    return ModalService;
}());
if (false) {
    /**
     * @type {?}
     * @private
     */
    ModalService.prototype.modals;
    /**
     * @type {?}
     * @private
     */
    ModalService.prototype.componentData;
    /**
     * @type {?}
     * @private
     */
    ModalService.prototype.subject;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var FourOhOneDialog = /** @class */ (function () {
    function FourOhOneDialog(router, injector, vTokenObjectService) {
        this.router = router;
        this.injector = injector;
        this.vTokenObjectService = vTokenObjectService;
        this.error = null;
        this.currentProfile = null;
        this.error_code = null;
        this.error_title = null;
        this.error_message = null;
        this._config_token = this.injector.get(EssentialsConfigToken);
        this.rSecureUrl = this._config_token.rSecure_frontend_url;
        this.rFormUrl = this._config_token.rForm_frontend_url;
    }
    /**
     * @return {?}
     */
    FourOhOneDialog.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.currentProfile = this.vTokenObjectService.getProfile();
        if (this.error && this.error.hasOwnProperty('error')) {
            this.error_code = this.error.error.code || null;
            if (this.error_code) {
                this.prepareErroContent();
            }
        }
    };
    /**
     * @return {?}
     */
    FourOhOneDialog.prototype.prepareErroContent = /**
     * @return {?}
     */
    function () {
        switch (this.error_code) {
            case 'AUTH0002':
                this.error_title = 'Authorization key is missing';
                this.error_message = 'Try after ' + (this.currentProfile ? 'signing in again.' : 'signing in or signing up.');
                break;
            case 'T000001':
                this.error_title = 'This document is not accessible';
                this.error_message = (this.currentProfile ? 'Try using a different organization/account, ' : 'Try signing in, signing up, ') + 'or requesting access from the owner of this document.';
                break;
            case 'AUTH0003':
                this.error_title = 'Cannot complete your request';
                this.error_message = 'Try after ' + (this.currentProfile ? 'signing in again.' : 'signing in or signing up.');
                break;
            default:
                this.error_title = 'Your request has been denied.';
                this.error_message = null;
                break;
        }
    };
    /**
     * @return {?}
     */
    FourOhOneDialog.prototype.goToDashboard = /**
     * @return {?}
     */
    function () {
        this.router.navigate(['/dashboard']);
    };
    /**
     * @return {?}
     */
    FourOhOneDialog.prototype.signIn = /**
     * @return {?}
     */
    function () {
        if (window) {
            window.location.href = this.rSecureUrl + "/login?redirect_url=" + (this.rFormUrl + this.router.url);
        }
    };
    /**
     * @return {?}
     */
    FourOhOneDialog.prototype.signUp = /**
     * @return {?}
     */
    function () {
        if (window) {
            window.location.href = this.rSecureUrl + "/signup?redirect_url=" + (this.rFormUrl + this.router.url);
        }
    };
    FourOhOneDialog.decorators = [
        { type: Component, args: [{
                    selector: 'four-oh-one-dialog',
                    template: "<div class=\"error__container\">\n  <div class=\"error__title\">\n    {{error_title}}\n  </div>\n  <div class=\"error__body\" *ngIf=\"error_message\">\n    {{error_message}}\n  </div>\n  <div class=\"error__action\">\n    <button class=\"error__action-button left\" mat-button *ngIf=\"currentProfile\" (click)=\"goToDashboard()\">go to dashboard</button>\n    <button class=\"error__action-button left\" mat-button *ngIf=\"!currentProfile\" (click)=\"signUp()\">sign up</button>\n    <button class=\"error__action-button\" mat-button (click)=\"signIn()\">sign in</button>\n  </div>\n</div>",
                    styles: [".error__container{padding-top:8px;padding-left:8px;padding-right:8px;max-width:640px}.error__title{padding:16px 16px 8px;font-size:20px;line-height:24px;font-weight:500;color:rgba(0,0,0,.87)}.error__body{padding:16px 16px 8px;font-size:16px;line-height:24px;color:rgba(0,0,0,.54)}.error__action{padding:8px;text-align:right}.error__action-button{color:#37474f;text-transform:uppercase}.error__action-button.left{margin-right:8px}"]
                }] }
    ];
    /** @nocollapse */
    FourOhOneDialog.ctorParameters = function () { return [
        { type: Router },
        { type: Injector },
        { type: VerdocsTokenObjectService }
    ]; };
    return FourOhOneDialog;
}());
if (false) {
    /** @type {?} */
    FourOhOneDialog.prototype.error;
    /** @type {?} */
    FourOhOneDialog.prototype.currentProfile;
    /** @type {?} */
    FourOhOneDialog.prototype.error_code;
    /** @type {?} */
    FourOhOneDialog.prototype.error_title;
    /** @type {?} */
    FourOhOneDialog.prototype.error_message;
    /** @type {?} */
    FourOhOneDialog.prototype.rSecureUrl;
    /** @type {?} */
    FourOhOneDialog.prototype.rFormUrl;
    /**
     * @type {?}
     * @private
     */
    FourOhOneDialog.prototype._config_token;
    /**
     * @type {?}
     * @private
     */
    FourOhOneDialog.prototype.router;
    /**
     * @type {?}
     * @private
     */
    FourOhOneDialog.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    FourOhOneDialog.prototype.vTokenObjectService;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var PlansDialog = /** @class */ (function () {
    function PlansDialog(router, dialog) {
        this.router = router;
        this.dialog = dialog;
        this.redirectUrl = null;
        this.type = null;
        this.dialogTitle = null;
        this.dialogMessage = null;
    }
    /**
     * @return {?}
     */
    PlansDialog.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        switch (this.type) {
            case 'free':
                this.dialogTitle = 'Upgrade to create another document';
                this.dialogMessage = 'We are glad to see you are drawing value from Verdocs. You have reached your limit of three documents per month with your free account. Please upgrade to an Essentials or Pro subscription to create more documents.';
                break;
            case 'essential':
                this.dialogTitle = 'Upgrade to access this feature';
                this.dialogMessage = 'This feature requires an essentials subscription or above. Please upgrade to access this feature.';
                break;
            case 'pro':
                this.dialogTitle = 'Upgrade to access this feature';
                this.dialogMessage = 'This feature requires a pro subscription. Please upgrade to access this feature.';
                break;
            default:
                this.dialog.close();
                break;
        }
    };
    /**
     * @return {?}
     */
    PlansDialog.prototype.goToBilling = /**
     * @return {?}
     */
    function () {
        if (window && !!this.redirectUrl && !!this.type) {
            window.location.href = (this.redirectUrl);
        }
        else {
            this.dialog.close();
        }
    };
    /**
     * @return {?}
     */
    PlansDialog.prototype.close = /**
     * @return {?}
     */
    function () {
        this.dialog.close();
    };
    PlansDialog.decorators = [
        { type: Component, args: [{
                    selector: 'plans-dialog',
                    template: "<div class=\"plans-dialog__container\">\n  <div class=\"plans-dialog__title\">{{dialogTitle}}</div>\n  <div class=\"plans-dialog__message\">\n      {{dialogMessage}}\n  </div>\n  <div class=\"plans-dialog__action\">\n    <button class=\"plans-dialog__button\" mat-button (click)=\"goToBilling()\">let's go</button>\n    <button class=\"plans-dialog__button\" mat-button (click)=\"close()\">not now</button>\n  </div>\n</div>",
                    styles: [".plans-dialog__title{padding:16px;font-size:24px;font-weight:500;color:rgba(0,0,0,.87)}.plans-dialog__message{padding:0 16px 16px;font-size:16px;line-height:24px;color:rgba(0,0,0,.54)}.plans-dialog__action{padding-top:8px;text-align:right}.plans-dialog__button{color:#37474f;text-transform:uppercase}"]
                }] }
    ];
    /** @nocollapse */
    PlansDialog.ctorParameters = function () { return [
        { type: Router },
        { type: MatDialogRef }
    ]; };
    return PlansDialog;
}());
if (false) {
    /** @type {?} */
    PlansDialog.prototype.redirectUrl;
    /** @type {?} */
    PlansDialog.prototype.type;
    /** @type {?} */
    PlansDialog.prototype.dialogTitle;
    /** @type {?} */
    PlansDialog.prototype.dialogMessage;
    /**
     * @type {?}
     * @private
     */
    PlansDialog.prototype.router;
    /**
     * @type {?}
     * @private
     */
    PlansDialog.prototype.dialog;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
Injectable();
var ErrorDialogService = /** @class */ (function () {
    function ErrorDialogService(dialog) {
        this.dialog = dialog;
    }
    /**
     * @param {?} err
     * @return {?}
     */
    ErrorDialogService.prototype.show401ResponseError = /**
     * @param {?} err
     * @return {?}
     */
    function (err) {
        /** @type {?} */
        var errorDialog = this.dialog.open(FourOhOneDialog, {
            panelClass: 'error_dialog',
            disableClose: true
        });
        errorDialog.componentInstance.error = err;
    };
    /**
     * @param {?} type
     * @param {?} redirect_url
     * @param {?=} dialog_title
     * @param {?=} dialog_message
     * @return {?}
     */
    ErrorDialogService.prototype.showPlansError = /**
     * @param {?} type
     * @param {?} redirect_url
     * @param {?=} dialog_title
     * @param {?=} dialog_message
     * @return {?}
     */
    function (type, redirect_url, dialog_title, dialog_message) {
        /** @type {?} */
        var planErrorDialog = this.dialog.open(PlansDialog, {
            panelClass: 'error_dialog',
            disableClose: true
        });
        if (type) {
            planErrorDialog.componentInstance.type = type;
        }
        if (redirect_url) {
            planErrorDialog.componentInstance.type = redirect_url;
        }
        if (dialog_title) {
            planErrorDialog.componentInstance.dialogTitle = dialog_title;
        }
        if (dialog_message) {
            planErrorDialog.componentInstance.dialogMessage = dialog_message;
        }
    };
    return ErrorDialogService;
}());
if (false) {
    /**
     * @type {?}
     * @private
     */
    ErrorDialogService.prototype.dialog;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var GuardService = /** @class */ (function () {
    function GuardService(injector, stateService, dialog, platform) {
        this.injector = injector;
        this.stateService = stateService;
        this.dialog = dialog;
        this.platform = platform;
        this._redirectUrl = null;
        this._config_token = this.injector.get(EssentialsConfigToken);
        this._plans = this._config_token.plans;
        this._rAccount_frontend_url = this._config_token.rAccount_frontend_url;
    }
    /**
     * @param {?} type
     * @param {?} subscriptionType
     * @param {?} openDialog
     * @return {?}
     */
    GuardService.prototype.checkSubscription = /**
     * @param {?} type
     * @param {?} subscriptionType
     * @param {?} openDialog
     * @return {?}
     */
    function (type, subscriptionType, openDialog) {
        /** @type {?} */
        var idToken = this.stateService.getIDToken();
        /** @type {?} */
        var userId = this.toTokenObject(idToken)['sub'].slice(6);
        this._redirectUrl = this._rAccount_frontend_url + "/rAccount/user/" + userId + "/billing";
        switch (type) {
            case 'reminders':
                if (subscriptionType === this._plans['level-2'] || subscriptionType === this._plans['level-3']) {
                    return true;
                }
                else {
                    if (openDialog === true) {
                        /** @type {?} */
                        var essentialDialog = this.dialog.open(PlansDialog, {
                            panelClass: 'confirmation'
                        });
                        essentialDialog.componentInstance.type = 'essential';
                        essentialDialog.componentInstance.redirectUrl = this._redirectUrl;
                    }
                    return false;
                }
            case 'live-link':
            case 'attachment':
            case 'payment':
                if (subscriptionType === this._plans['level-3']) {
                    return true;
                }
                else {
                    if (openDialog === true) {
                        /** @type {?} */
                        var proDialog = this.dialog.open(PlansDialog, {
                            panelClass: 'confirmation'
                        });
                        proDialog.componentInstance.type = 'pro';
                        proDialog.componentInstance.redirectUrl = this._redirectUrl;
                    }
                    return false;
                }
            case 'open-free-envelopes':
                /** @type {?} */
                var freeDialog = this.dialog.open(PlansDialog, {
                    panelClass: 'confirmation'
                });
                freeDialog.componentInstance.type = 'free';
                freeDialog.componentInstance.redirectUrl = this._redirectUrl;
                return false;
            default:
                return false;
        }
    };
    /**
     * @param {?} accessToken
     * @return {?}
     */
    GuardService.prototype.toTokenObject = /**
     * @param {?} accessToken
     * @return {?}
     */
    function (accessToken) {
        if (isPlatformBrowser(this.platform)) {
            /** @type {?} */
            var base64Url = accessToken.split('.')[1];
            /** @type {?} */
            var base64 = base64Url.replace('-', '+').replace('_', '/');
            return JSON.parse(window.atob(base64));
        }
    };
    /**
     * @param {?} plans
     * @return {?}
     */
    GuardService.prototype.getHighestEnvelopePlan = /**
     * @param {?} plans
     * @return {?}
     */
    function (plans) {
        /** @type {?} */
        var plan = null;
        for (var x = 0; x < plans.length; x++) {
            if (plans[x] === this._plans['level-3']) {
                plan = plans[x];
            }
            else if (plans[x] === this._plans['level-2'] && (plan === null || plan === 'env:free')) {
                plan = plans[x];
            }
            else if (plans[x] === this._plans['level-1'] && plan === null) {
                plan = plans[x];
            }
        }
        return plan;
    };
    GuardService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    GuardService.ctorParameters = function () { return [
        { type: Injector },
        { type: VerdocsStateService },
        { type: MatDialog },
        { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
    ]; };
    return GuardService;
}());
if (false) {
    /**
     * @type {?}
     * @private
     */
    GuardService.prototype._config_token;
    /**
     * @type {?}
     * @private
     */
    GuardService.prototype._plans;
    /**
     * @type {?}
     * @private
     */
    GuardService.prototype._rAccount_frontend_url;
    /**
     * @type {?}
     * @private
     */
    GuardService.prototype._redirectUrl;
    /**
     * @type {?}
     * @private
     */
    GuardService.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    GuardService.prototype.stateService;
    /**
     * @type {?}
     * @private
     */
    GuardService.prototype.dialog;
    /**
     * @type {?}
     * @private
     */
    GuardService.prototype.platform;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
function EssentialsConfig() { }
if (false) {
    /** @type {?} */
    EssentialsConfig.prototype.production;
    /** @type {?} */
    EssentialsConfig.prototype.domain;
    /** @type {?} */
    EssentialsConfig.prototype.rForm_backend_url;
    /** @type {?} */
    EssentialsConfig.prototype.rForm_frontend_url;
    /** @type {?} */
    EssentialsConfig.prototype.rSecure_frontend_url;
    /** @type {?} */
    EssentialsConfig.prototype.rSecure_backend_url;
    /** @type {?} */
    EssentialsConfig.prototype.rAccount_frontend_url;
    /** @type {?} */
    EssentialsConfig.prototype.rAccount_backend_url;
    /** @type {?} */
    EssentialsConfig.prototype.rNotification_backend_url;
    /** @type {?} */
    EssentialsConfig.prototype.rForm_cookie_name;
    /** @type {?} */
    EssentialsConfig.prototype.loginURL;
    /** @type {?} */
    EssentialsConfig.prototype.origin;
    /** @type {?} */
    EssentialsConfig.prototype.plans;
}
/** @type {?} */
var EssentialsConfigToken = new InjectionToken('EssentialsConfig');
var EssentialsModule = /** @class */ (function () {
    function EssentialsModule() {
    }
    /**
     * @param {?} config
     * @return {?}
     */
    EssentialsModule.initializeEssentials = /**
     * @param {?} config
     * @return {?}
     */
    function (config) {
        return {
            ngModule: EssentialsModule,
            import: [
                CommonModule,
                AngularApiModule.initSDK({
                    production: config.production,
                    domain: config.domain,
                    origin: config.origin,
                    loginURL: config.loginURL,
                    rForm_backend_url: config.rForm_backend_url,
                    rForm_frontend_url: config.rForm_frontend_url,
                    rSecure_backend_url: config.rSecure_backend_url,
                    rSecure_frontend_url: config.rSecure_frontend_url,
                    rAccount_backend_url: config.rAccount_backend_url,
                    rAccount_frontend_url: config.rAccount_frontend_url,
                    rNotification_backend_url: config.rNotification_backend_url,
                    plans: config.plans
                }),
                MatDialogModule
            ],
            providers: [
                {
                    provide: EssentialsConfigToken,
                    useValue: config
                },
                SnackbarService,
                AccountService,
                BrowserToiTextService,
                BuilderDataService,
                BuilderService,
                ConversionService,
                SignatureService,
                EnvelopeViewService,
                PageService,
                ModalService,
                ErrorDialogService,
                GuardService
            ],
            declarations: [
                FourOhOneDialog,
                PlansDialog
            ],
            EntryComponents: [
                FourOhOneDialog,
                PlansDialog
            ],
            exports: [
                FourOhOneDialog,
                PlansDialog
            ]
        };
    };
    EssentialsModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule
                    ],
                    providers: [],
                    declarations: [
                        FourOhOneDialog,
                        PlansDialog
                    ]
                },] }
    ];
    return EssentialsModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var AccountService = /** @class */ (function () {
    function AccountService(http, router, vTokenStateService, injector) {
        this.http = http;
        this.router = router;
        this.vTokenStateService = vTokenStateService;
        this.injector = injector;
        this.userDataSubject = new ReplaySubject();
        this.userProfileSubject = new ReplaySubject();
        this.profileLoadingSubject = new ReplaySubject();
        this.currentProfileSubject = new ReplaySubject();
        this.personalAccountSubject = new ReplaySubject();
        this.profilesSubject = new ReplaySubject();
        this.userFullname = '';
        this.userInitial = '';
        this.userEmail = '';
        this.userPhone = '';
        this.userData = {
            name: this.userFullname,
            initial: null,
            email: this.userEmail,
            phone: this.userPhone
        };
        this.currentOrganizationId = null;
        this._config_token = this.injector.get(EssentialsConfigToken);
        this.backend_url = this._config_token.rAccount_backend_url;
        this.rForm_backend_url = this._config_token.rForm_backend_url;
        this.rSecure_backend_url = this._config_token.rSecure_backend_url;
    }
    /**
     * @return {?}
     */
    AccountService.prototype.getAccountData = /**
     * @return {?}
     */
    function () {
        /***********
        * Refactor *
        ***********/
        // const headerService = this.injector.get(HeaderService);
        // headerService.noHeader.subscribe(status => {
        // if (status === false) {
        this.getProfiles();
        // }
        // });
    };
    /**
     * @return {?}
     */
    AccountService.prototype.getProfiles = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return this.http.get(this.backend_url + '/profiles').toPromise().then((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            /** @type {?} */
            var personal = res.splice(findIndex(res, (/**
             * @param {?} p
             * @return {?}
             */
            function (p) { return p.organization.name === 'realster'; })), 1);
            _this.personalAccountSubject.next(personal[0]);
            _this.profileData = res.sort((/**
             * @param {?} a
             * @param {?} b
             * @return {?}
             */
            function (a, b) {
                if (a.organization.name < b.organization.name) {
                    return -1;
                }
                if (a.organization.name > b.organization.name) {
                    return 1;
                }
                return 0;
            }));
            _this.profileData.unshift(personal[0]);
            /** @type {?} */
            var currentIndex = findIndex(_this.profileData, { current: true });
            if (_this.profileData[currentIndex]['first_name']) {
                _this.userFullname = _this.profileData[currentIndex]['first_name'] + ' ' + _this.profileData[currentIndex]['last_name'];
                _this.userInitial = _this.profileData[currentIndex]['first_name'].charAt(0) +
                    _this.profileData[currentIndex]['last_name'].charAt(0);
            }
            else {
                _this.userFullname = '';
            }
            if (_this.profileData[currentIndex]['phone']) {
                _this.userPhone = _this.profileData[currentIndex]['phone'];
            }
            else {
                _this.userPhone = '';
            }
            if (_this.profileData[currentIndex]['email']) {
                _this.userEmail = _this.profileData[currentIndex]['email'];
            }
            _this.userData = {
                name: _this.userFullname,
                initial: _this.userInitial,
                email: _this.userEmail,
                phone: _this.userPhone
            };
            _this.userProfileSubject.next(_this.profileData);
            /** @type {?} */
            var currentProfile = _this.profileData[currentIndex];
            _this.currentProfileSubject.next(currentProfile);
            _this.currentOrganizationId = currentProfile.organization_id;
            _this.userDataSubject.next(_this.userData);
            return _this.userData;
        }));
    };
    /**
     * @param {?} profiles
     * @return {?}
     */
    AccountService.prototype.getProfileCollection = /**
     * @param {?} profiles
     * @return {?}
     */
    function (profiles) {
        /** @type {?} */
        var profileModels = [];
        profiles.forEach((/**
         * @param {?} profile
         * @return {?}
         */
        function (profile) {
            profileModels.push(new ProfileModel(profile));
        }));
        /** @type {?} */
        var profileCollection = new ProfileCollection(profileModels);
        return profileCollection;
    };
    /**
     * @param {?} id
     * @return {?}
     */
    AccountService.prototype.removeFromProfileData = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        /** @type {?} */
        var removeIndex = findIndex(this.profileData, { organization_id: name });
        if (removeIndex >= 0) {
            this.profileData.splice(removeIndex, 1);
            this.userProfileSubject.next(this.profileData);
        }
    };
    /**
     * @return {?}
     */
    AccountService.prototype.getAllProfiles = /**
     * @return {?}
     */
    function () {
        return this.http.get(this.backend_url + '/profiles/').toPromise();
    };
    /**
     * @param {?} profile_id
     * @return {?}
     */
    AccountService.prototype.switchProfile = /**
     * @param {?} profile_id
     * @return {?}
     */
    function (profile_id) {
        var _this = this;
        this.profileLoadingSubject.next(true);
        return this.http.post(this.backend_url + '/profiles/' + profile_id + '/switch', null).toPromise().then((/**
         * @param {?} res
         * @return {?}
         */
        function (res) { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.vTokenStateService.setTokens(res.tokens);
                this.profileLoadingSubject.next(false);
                return [2 /*return*/, Promise.resolve(res)];
            });
        }); }));
    };
    /**
     * @param {?=} sendMessage
     * @return {?}
     */
    AccountService.prototype.updateTokens = /**
     * @param {?=} sendMessage
     * @return {?}
     */
    function (sendMessage) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, new Promise((/**
                     * @param {?} resolve
                     * @param {?} reject
                     * @return {?}
                     */
                    function (resolve, reject) {
                        _this.vTokenStateService.updateTokens(_this.rSecure_backend_url).subscribe((/**
                         * @param {?} tokens
                         * @return {?}
                         */
                        function (tokens) {
                            if (tokens && tokens.length > 0) {
                                if (sendMessage) {
                                    // this.snackbar.dismiss();
                                    _this.createSnackbar('Status updated');
                                }
                            }
                            _this.getAccountData();
                            return resolve();
                        }));
                    }))];
            });
        });
    };
    /**
     * @return {?}
     */
    AccountService.prototype.resendEmailVerification = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.http.post(this.rSecure_backend_url + '/user/email_verification', null).toPromise().then((/**
         * @return {?}
         */
        function () {
            _this.createSnackbar('Verification email sent');
        }));
    };
    /********************************
    * replace with snackbar service *
    ********************************/
    /**
     * *****************************
     * replace with snackbar service *
     * ******************************
     * @param {?} message
     * @return {?}
     */
    AccountService.prototype.createSnackbar = /**
     * *****************************
     * replace with snackbar service *
     * ******************************
     * @param {?} message
     * @return {?}
     */
    function (message) {
        // let snackbarConfig: MatSnackBarConfig
        // if (window.innerWidth >= 920) {
        //   snackbarConfig = {
        //     verticalPosition: 'bottom',
        //     horizontalPosition: 'left',
        //     duration: 5000
        //   }
        // } else {
        //   snackbarConfig = {
        //     verticalPosition: 'top',
        //     duration: 5000
        //   }
        // }
        // this.snackbar.open(message, null, snackbarConfig);
    };
    /**
     * @param {?} userData
     * @return {?}
     */
    AccountService.prototype.putAccountData = /**
     * @param {?} userData
     * @return {?}
     */
    function (userData) {
        var _this = this;
        if (userData.email) {
            this.userData.email = userData.email;
        }
        if (userData.first_name && userData.last_name) {
            this.userData.name = userData.first_name + ' ' + userData.last_name;
            this.userData.initial = userData.first_name.charAt(0) + userData.last_name.charAt(0);
        }
        if (userData.phone) {
            this.userData.phone = userData.phone || '';
        }
        this.http.put(this.backend_url + '/profiles/' + userData.id, userData).toPromise().then((/**
         * @return {?}
         */
        function () {
            _this.userDataSubject.next(_this.userData);
        }));
    };
    /**
     * @return {?}
     */
    AccountService.prototype.deleteAccount = /**
     * @return {?}
     */
    function () {
        return this.http.delete(this.backend_url).toPromise();
    };
    /**
     * @param {?} email
     * @param {?} oldPassword
     * @param {?} newPassword
     * @return {?}
     */
    AccountService.prototype.changePassword = /**
     * @param {?} email
     * @param {?} oldPassword
     * @param {?} newPassword
     * @return {?}
     */
    function (email, oldPassword, newPassword) {
        return this.http.put(this.backend_url + 'updatepassword', {
            email: email,
            oldPassword: oldPassword,
            newPassword: newPassword
        }).toPromise();
    };
    /**
     * @param {?} name
     * @return {?}
     */
    AccountService.prototype.getNameValidity = /**
     * @param {?} name
     * @return {?}
     */
    function (name) {
        return this.http.get(this.backend_url + '/organizations/is_valid?name=' + name, this.getOptions())
            .toPromise()
            .then((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            return Promise.resolve(res);
        }));
    };
    /**
     * @param {?} body
     * @return {?}
     */
    AccountService.prototype.createOrganization = /**
     * @param {?} body
     * @return {?}
     */
    function (body) {
        return this.http.post(this.backend_url + '/organizations', body, this.getOptions())
            .toPromise()
            .then((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            return Promise.resolve(res);
        }));
    };
    /**
     * @private
     * @return {?}
     */
    AccountService.prototype.getOptions = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var rTokenCookie = this.vTokenStateService.getOtherCookieObject('rAccount_Verdocs');
        /** @type {?} */
        var authValue = 'Bearer ' + rTokenCookie['accessToken'];
        /** @type {?} */
        var headers = new Headers({
            'Cache-Control': 'no-cache',
            'Pragma': 'no-cache',
            'Expires': 'Sat, 01 Jan 2000 00:00:00 GMT',
            'If-Modified-Since': '0',
            'Content-Type': 'application/json',
            'Authorization': authValue,
            'Accept': 'application/json'
        });
        return { 'headers': headers };
    };
    /**
     * @param {?} id
     * @param {?} roleName
     * @param {?} invite
     * @param {?=} redirectReq
     * @return {?}
     */
    AccountService.prototype.fetchToken = /**
     * @param {?} id
     * @param {?} roleName
     * @param {?} invite
     * @param {?=} redirectReq
     * @return {?}
     */
    function (id, roleName, invite, redirectReq) {
        var _this = this;
        /** @type {?} */
        var req = this.rForm_backend_url + ("/envelopes/" + id + "/recipients/" + roleName + "/invitation/" + invite);
        return this.http.get(req, { observe: 'response' }).subscribe((/**
         * @param {?} res
         * @return {?}
         */
        function (res) {
            _this.vTokenStateService.storeOtherCookie('signer_token', res['headers'].get('signer_token'));
            if (redirectReq) {
                _this.router.navigate(["/view/sign/" + id + "/role/" + roleName], { queryParams: { redirectReq: redirectReq } });
            }
            else {
                _this.router.navigate(["/view/sign/" + id + "/role/" + roleName]);
            }
        }), (/**
         * @param {?} err
         * @return {?}
         */
        function (err) {
            _this.router.navigate(["/view/sign/" + id + "/role/" + roleName, { error: err }]);
        }));
    };
    AccountService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    AccountService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: Router },
        { type: VerdocsStateService },
        { type: Injector }
    ]; };
    return AccountService;
}());
if (false) {
    /** @type {?} */
    AccountService.prototype.userDataSubject;
    /** @type {?} */
    AccountService.prototype.userProfileSubject;
    /** @type {?} */
    AccountService.prototype.profileLoadingSubject;
    /** @type {?} */
    AccountService.prototype.currentProfileSubject;
    /** @type {?} */
    AccountService.prototype.personalAccountSubject;
    /** @type {?} */
    AccountService.prototype.profilesSubject;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype._config_token;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.userFullname;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.userInitial;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.userEmail;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.userPhone;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.userData;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.profileData;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.backend_url;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.rSecure_backend_url;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.rForm_backend_url;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.currentOrganizationId;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.router;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.vTokenStateService;
    /**
     * @type {?}
     * @private
     */
    AccountService.prototype.injector;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { AccountService, BrowserToiTextService, BuilderDataService, BuilderService, ConversionService, DynamicField, EnvelopeViewService, ErrorDialogService, EssentialsConfigToken, EssentialsModule, FourOhOneDialog, GuardService, ModalService, PageService, PlansDialog, SignatureService, SnackbarService, blobToBase64, bottomOfPageIsInView, buildFields, constructGroupFields, constructPages, convertToUIRatio, dataURLtoBlob, generateCheckboxLabel, generateGroupLabel, generateInitialStyle, generateStyle, getDefaultHeight, getDefaultWidth, getFieldsMap, getInputStyle, getLetterSpacing, getRLeft, getRTop, getRValue, getRoleColor, getUIFieldStyling, isInView, middleOfPageIsInView, prepareFieldsForSigner, printPdfUrl, regParse, rescale, setRatio, timePeriod, topOfPageIsInView, updateElementStyles };
