// @verdocs/header/esm2015/lib/components/header-tools.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, Injector, Inject, PLATFORM_ID } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
import { Router } from '@angular/router';
import { MatSnackBar } from '@angular/material/snack-bar';
import { NotificationService } from '@verdocs/notifications';
import { OrganizationDialogService } from '@verdocs/profiles';
import { VerdocsStateService, VerdocsAuthGuardService, VerdocsTokenObjectService } from '@verdocs/tokens';
import { Subscription } from 'rxjs';
import { findIndex, includes } from 'lodash';
import { connectionConfigToken } from '../header.module';
import { VerdocsHeaderService } from '../services/header.service';
export class HeaderToolsComponent {
    /**
     * @param {?} verdocsStateService
     * @param {?} verdocsObjectService
     * @param {?} verdocsAuthGuardService
     * @param {?} organizationDialogService
     * @param {?} notificationsService
     * @param {?} verdocsHeaderService
     * @param {?} snackbar
     * @param {?} injector
     * @param {?} router
     * @param {?} platform
     */
    constructor(verdocsStateService, verdocsObjectService, verdocsAuthGuardService, organizationDialogService, notificationsService, verdocsHeaderService, snackbar, injector, router, platform) {
        this.verdocsStateService = verdocsStateService;
        this.verdocsObjectService = verdocsObjectService;
        this.verdocsAuthGuardService = verdocsAuthGuardService;
        this.organizationDialogService = organizationDialogService;
        this.notificationsService = notificationsService;
        this.verdocsHeaderService = verdocsHeaderService;
        this.snackbar = snackbar;
        this.injector = injector;
        this.router = router;
        this.platform = platform;
        this.apps = [];
        this.notifications = [];
        this.profileLoading = false;
        this.hasProfile = false;
        this.profileOverride = {
            top: 56,
            right: 20
        };
        this.sidenav = false;
        this.type = 'default';
        this.notificationsSubscription = new Subscription();
        this.profileLoadingSubscription = new Subscription();
        this.tokenConfig = this.injector.get(connectionConfigToken);
        this.myAccountUrl = this.tokenConfig.rAccount_frontend_url;
        this.rSecureUrl = this.tokenConfig.rSecure_frontend_url;
        this.rFormUrl = this.tokenConfig.rForm_frontend_url;
        this.termsUrl = this.tokenConfig.rSecure_frontend_url + '/terms';
        this.privacyUrl = this.tokenConfig.rSecure_frontend_url + '/privacy';
        this.loginUrl = this.tokenConfig.rSecure_frontend_url + '/login';
        this.signupUrl = this.tokenConfig.rSecure_frontend_url + '/signup';
        this.organizationImgSrc = `${this.tokenConfig.rAccount_frontend_url}/assets/organization/select-profile.png`;
        this.verdocs_logo_url = `${this.rFormUrl}/assets/blue-logo.svg`;
        this.origin = this.tokenConfig.origin;
        this.isAtOriginalOrigin = this.verdocsHeaderService.getAtOrigin(this.origin);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.checkEmailVerification();
        this.hasProfile = this.verdocsObjectService.getProfile();
        this.verdocsHeaderService.userProfileSubject.subscribe((/**
         * @param {?} profiles
         * @return {?}
         */
        profiles => {
            this.profiles = this.verdocsHeaderService.getProfileCollection(profiles);
            this.verdocsHeaderService.profilesSubject.next(profiles);
            /** @type {?} */
            const index = findIndex(profiles, { organization_id: this.organization_id });
            if (index >= 0) {
                this.newProfile = profiles[index];
                this.organizationDialogService.setNewProfile(this.newProfile);
            }
        }));
        this.notificationsSubscription = this.notificationsService.Notificatations().subscribe((/**
         * @param {?} notifications
         * @return {?}
         */
        notifications => {
            this.notifications = notifications;
        }));
        this.profileLoadingSubscription = this.verdocsHeaderService.profileLoadingSubject.subscribe((/**
         * @param {?} status
         * @return {?}
         */
        status => {
            this.profileLoading = status;
        }));
        if (!this.profiles && this.hasProfile) {
            this.verdocsHeaderService.getProfiles();
        }
    }
    /**
     * @return {?}
     */
    ngAfterContentChecked() {
        if (isPlatformBrowser(this.platform)) {
            if (typeof window !== 'undefined') {
                if (window.innerWidth > 600) {
                    this.profileOverride.top = 60;
                }
                else {
                    this.profileOverride.top = 56;
                }
            }
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.notificationsSubscription.unsubscribe();
        this.profileLoadingSubscription.unsubscribe();
    }
    /**
     * @private
     * @return {?}
     */
    checkEmailVerification() {
        this.emailVerified = this.verdocsStateService.getEmailVerification();
        this.organizationDialogService.setEmailVerified(this.emailVerified);
    }
    /**
     * @return {?}
     */
    get admin_url() {
        if (this.hasAccessToAdmin()) {
            return `${this.tokenConfig.rAccount_frontend_url}/admin`;
        }
        return undefined;
    }
    /**
     * @param {?} notification
     * @return {?}
     */
    removeNotification(notification) {
        this.notificationsService.dismissNotification(notification);
    }
    /**
     * @return {?}
     */
    signOut() {
        this.verdocsAuthGuardService.signOut();
    }
    /**
     * @return {?}
     */
    signIn() {
        if (window) {
            window.location.href = `${this.rSecureUrl}/login?redirect_url=${this.rFormUrl + this.router.url}`;
        }
    }
    /**
     * @return {?}
     */
    signUp() {
        if (window) {
            window.location.href = `${this.rSecureUrl}/signup?redirect_url=${this.rFormUrl + this.router.url}`;
        }
    }
    /**
     * @param {?} profile_id
     * @return {?}
     */
    switch(profile_id) {
        this.verdocsHeaderService.switchProfile(profile_id).then((/**
         * @return {?}
         */
        () => {
            window.location.href = `${this.tokenConfig.rForm_frontend_url}`;
        }));
    }
    /**
     * @return {?}
     */
    switchAndGoToAdmin() {
        if (this.newProfile) {
            this.verdocsHeaderService.switchProfile(this.newProfile.id).then((/**
             * @param {?} profile
             * @return {?}
             */
            profile => {
                window.location.href = `${this.myAccountUrl}/admin/profile/${this.newProfile.id}`;
            }));
        }
    }
    /**
     * @param {?=} sendMessage
     * @return {?}
     */
    updateTokenStatus(sendMessage) {
        this.createSnackbar('Checking email verification', 'OK');
        this.verdocsHeaderService.updateTokens(sendMessage).then((/**
         * @return {?}
         */
        () => {
            this.emailVerified = this.verdocsStateService.getEmailVerification();
            this.organizationDialogService.setEmailVerified(this.emailVerified);
        }));
    }
    /**
     * @return {?}
     */
    resendEmailVerification() {
        this.verdocsHeaderService.resendEmailVerification();
    }
    /**
     * @param {?} name
     * @return {?}
     */
    checkNameValidity(name) {
        return this.verdocsHeaderService.getNameValidity(name)
            .then((/**
         * @param {?} result
         * @return {?}
         */
        result => {
            this.organizationDialogService.setNameValidation(result);
        }));
    }
    /**
     * @param {?} organizationGroup
     * @return {?}
     */
    createOrganization(organizationGroup) {
        this.verdocsHeaderService.createOrganization(organizationGroup)
            .then((/**
         * @param {?} organization
         * @return {?}
         */
        (organization) => tslib_1.__awaiter(this, void 0, void 0, function* () {
            this.organization_id = organization['id'];
            yield this.verdocsHeaderService.getAccountData();
            this.createSnackbar('Organization created', null);
        })));
    }
    /**
     * @return {?}
     */
    switchProfile() {
        this.switch(this.newProfile.id);
    }
    /**
     * @param {?} clientName
     * @return {?}
     */
    checkAccess(clientName) {
        return this.verdocsStateService.hasAccessTo(clientName + '_Realster');
    }
    /**
     * @param {?} clientName
     * @return {?}
     */
    jumpTo(clientName) {
        window.location.href = this.tokenConfig[clientName + '_frontend_url'];
    }
    /**
     * @private
     * @return {?}
     */
    hasAccessToAdmin() {
        /** @type {?} */
        const roles = this.verdocsStateService.getRAccountRoles();
        return includes(roles, 'owner') || includes(roles, 'admin');
    }
    /**
     * @param {?} message
     * @param {?=} buttonTitle
     * @return {?}
     */
    createSnackbar(message, buttonTitle) {
        if (!buttonTitle) {
            buttonTitle = null;
        }
        /** @type {?} */
        let snackbarConfig;
        if (window.innerWidth >= 920) {
            snackbarConfig = {
                verticalPosition: 'bottom',
                horizontalPosition: 'left',
                duration: 5000
            };
        }
        else {
            snackbarConfig = {
                verticalPosition: 'top',
                duration: 5000
            };
        }
        this.snackbar.open(message, buttonTitle, snackbarConfig);
    }
    /**
     * @return {?}
     */
    lunchElevio() {
        window['lunchElevio']();
    }
    /**
     * @return {?}
     */
    lunchIntercom() {
        window['lunchIntercom']();
    }
}
HeaderToolsComponent.decorators = [
    { type: Component, args: [{
                selector: 'verdocs-header-tools',
                template: "<div class=\"verdocs__wrapper\" *ngIf=\"hasProfile\" [ngClass]=\"{sidenav: sidenav}\">\n  <div class=\"verdocs__chat\">\n    <button mat-icon-button *ngIf=\"!sidenav\" (click)=\"lunchIntercom()\" matTooltip=\"Live Chat\">\n      <mat-icon>forum</mat-icon>\n    </button>\n  </div>\n  <div class=\"verdocs__help\">\n    <button mat-icon-button *ngIf=\"!sidenav\" (click)=\"lunchElevio()\" matTooltip=\"Help Center\">\n      <mat-icon>help</mat-icon>\n    </button>\n  </div>\n  <div class=\"verdocs__notification\" *ngIf=\"!sidenav\" matTooltip=\"Notifications\">\n    <verdocs-notifications [notifications]=\"notifications\" (removeNotification)=\"removeNotification($event)\"\n      [verdocs_logo_url]=\"verdocs_logo_url\"></verdocs-notifications>\n  </div>\n  <div class=\"verdocs__account\" *ngIf=\"profiles\"\n    [ngClass]=\"{header: !sidenav, sidenav: sidenav, default: type === 'default'}\" matTooltip=\"My Account\">\n    <verdocs-profiles [currentProfile]=\"profiles.currentProfile()\" [override]=\"profileOverride\" [admin_url]=\"admin_url\">\n      <verdocs-account-dialog [profiles]=\"profiles\" [termsUrl]=\"termsUrl\" [privacyUrl]=\"privacyUrl\"\n        [accountUrl]=\"myAccountUrl\" (signOutEvent)=\"signOut()\" [admin_url]=\"admin_url\">\n        <verdocs-account-profile [profiles]=\"profiles\" [loading]=\"profileLoading\"\n          [organizationImgSrc]=\"organizationImgSrc\" (switchEvent)=\"switch($event)\"\n          (switchAndRedirectEvent)=\"switchAndGoToAdmin()\" (getEmailVerificationEvent)=\"updateTokenStatus()\"\n          (resendEmailVerificationEvent)=\"resendEmailVerification()\" (getNameValidityEvent)=\"checkNameValidity($event)\"\n          (createOrganizationEvent)=\"createOrganization($event)\" (switchProfileEvent)=\"switchProfile()\">\n        </verdocs-account-profile>\n      </verdocs-account-dialog>\n    </verdocs-profiles>\n  </div>\n</div>\n<div class=\"verdocs__wrapper\" *ngIf=\"!hasProfile\">\n  <a class=\"verdocs__signin-button\" mat-button [routerLink]=\"'/login'\"\n    *ngIf=\"origin === 'rSecure' && isAtOriginalOrigin\" [ngClass]=\"{preview: mode === 'preview'}\">Sign in</a>\n  <a class=\"verdocs__signup-button\" mat-button [routerLink]=\"'/signup'\"\n    *ngIf=\"origin === 'rSecure' && isAtOriginalOrigin\" [ngClass]=\"{preview: mode === 'preview'}\">Sign up</a>\n  <button class=\"verdocs__signin-button\" mat-button *ngIf=\"origin !== 'rSecure'\"\n    [ngClass]=\"{preview: mode === 'preview'}\" (click)=\"signIn()\">Sign in</button>\n  <button class=\"verdocs__signup-button\" mat-button *ngIf=\"origin !== 'rSecure'\"\n    [ngClass]=\"{preview: mode === 'preview'}\" (click)=\"signUp()\">Sign up</button>\n</div>",
                styles: [".verdocs__wrapper{height:64px;flex-direction:row;box-sizing:border-box;display:flex;justify-content:flex-end;align-items:center}.verdocs__notification{margin-top:0;line-height:64px}@media only screen and (max-width:600px){.verdocs__wrapper{height:56px}.verdocs__notification{line-height:56px}}.verdocs__chat{line-height:64px;margin-right:7px}.verdocs__chat button{height:40px;min-width:40px}.verdocs__help{line-height:64px;margin-right:7px}.verdocs__help button{height:40px;min-width:40px}.verdocs__app{line-height:64px}.verdocs__app-button{padding:0;height:48px;min-width:48px;border-radius:24px}.verdocs__app.sidenav{margin-right:4px}@media only screen and (max-width:600px){.verdocs__app{line-height:56px}}.verdocs__account{line-height:64px}.verdocs__account-button{background-color:#9b9b9b;color:#fff}@media only screen and (max-width:854px){.verdocs__wrapper{display:none}.verdocs__wrapper.sidenav{display:flex}.verdocs__account.header.default,.verdocs__app.header.default,.verdocs__chat,.verdocs__help,.verdocs__notification{display:none}}@media only screen and (min-width:855px){.verdocs__account.sidenav,.verdocs__app.sidenav{display:none}}@media only screen and (max-width:600px){.verdocs__account{line-height:56px}}.verdocs__account,.verdocs__app,.verdocs__chat,.verdocs__notification{float:left}.verdocs__signin-button,.verdocs__signup-button{padding-left:8px;padding-right:8px;border-radius:2px;height:36px;background-color:#33354c;color:#fff;min-width:60px}.verdocs__signin-button.preview,.verdocs__signup-button.preview{background-color:transparent;color:rgba(0,0,0,.87)}.verdocs__signup-button{margin-left:8px;border:.5px solid #fff}.verdocs__signup-button.preview{border:.5px solid rgba(0,0,0,.87)}.menu__container{width:320px;height:248px}.menu__container .mat-menu-content{padding:0}.menu__container.mat-menu-panel{padding:8px}.menu__item{padding:0!important;margin-bottom:8px!important;border:none;min-width:64px!important;border-radius:50%!important}.menu__item-wrapper{padding:12px;float:left;text-align:center}.menu__item-wrapper.my-account{padding-left:6.5px;padding-right:6.5px}.menu__item img{vertical-align:middle}.menu__icon{width:40px}.menu__icon.rForm{width:30px}.menu__background{background-color:#495057;width:64px;height:64px;line-height:64px;border-radius:50%}.menu__title{font-size:14px;text-align:center;line-height:20px}.menu__account-container{width:270px;height:228px}"]
            }] }
];
/** @nocollapse */
HeaderToolsComponent.ctorParameters = () => [
    { type: VerdocsStateService },
    { type: VerdocsTokenObjectService },
    { type: VerdocsAuthGuardService },
    { type: OrganizationDialogService },
    { type: NotificationService },
    { type: VerdocsHeaderService },
    { type: MatSnackBar },
    { type: Injector },
    { type: Router },
    { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
];
HeaderToolsComponent.propDecorators = {
    profileOverride: [{ type: Input }],
    sidenav: [{ type: Input }],
    type: [{ type: Input }],
    mode: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    HeaderToolsComponent.prototype.tokenConfig;
    /** @type {?} */
    HeaderToolsComponent.prototype.apps;
    /** @type {?} */
    HeaderToolsComponent.prototype.emailVerified;
    /** @type {?} */
    HeaderToolsComponent.prototype.notifications;
    /** @type {?} */
    HeaderToolsComponent.prototype.profiles;
    /** @type {?} */
    HeaderToolsComponent.prototype.newProfile;
    /** @type {?} */
    HeaderToolsComponent.prototype.organization_id;
    /** @type {?} */
    HeaderToolsComponent.prototype.myAccountUrl;
    /** @type {?} */
    HeaderToolsComponent.prototype.rSecureUrl;
    /** @type {?} */
    HeaderToolsComponent.prototype.rFormUrl;
    /** @type {?} */
    HeaderToolsComponent.prototype.termsUrl;
    /** @type {?} */
    HeaderToolsComponent.prototype.privacyUrl;
    /** @type {?} */
    HeaderToolsComponent.prototype.loginUrl;
    /** @type {?} */
    HeaderToolsComponent.prototype.signupUrl;
    /** @type {?} */
    HeaderToolsComponent.prototype.organizationImgSrc;
    /** @type {?} */
    HeaderToolsComponent.prototype.profileLoading;
    /** @type {?} */
    HeaderToolsComponent.prototype.hasProfile;
    /** @type {?} */
    HeaderToolsComponent.prototype.origin;
    /** @type {?} */
    HeaderToolsComponent.prototype.isAtOriginalOrigin;
    /** @type {?} */
    HeaderToolsComponent.prototype.verdocs_logo_url;
    /** @type {?} */
    HeaderToolsComponent.prototype.profileOverride;
    /** @type {?} */
    HeaderToolsComponent.prototype.sidenav;
    /** @type {?} */
    HeaderToolsComponent.prototype.type;
    /** @type {?} */
    HeaderToolsComponent.prototype.mode;
    /**
     * @type {?}
     * @private
     */
    HeaderToolsComponent.prototype.notificationsSubscription;
    /**
     * @type {?}
     * @private
     */
    HeaderToolsComponent.prototype.profileLoadingSubscription;
    /**
     * @type {?}
     * @private
     */
    HeaderToolsComponent.prototype.verdocsStateService;
    /**
     * @type {?}
     * @private
     */
    HeaderToolsComponent.prototype.verdocsObjectService;
    /**
     * @type {?}
     * @private
     */
    HeaderToolsComponent.prototype.verdocsAuthGuardService;
    /**
     * @type {?}
     * @private
     */
    HeaderToolsComponent.prototype.organizationDialogService;
    /**
     * @type {?}
     * @private
     */
    HeaderToolsComponent.prototype.notificationsService;
    /**
     * @type {?}
     * @private
     */
    HeaderToolsComponent.prototype.verdocsHeaderService;
    /**
     * @type {?}
     * @private
     */
    HeaderToolsComponent.prototype.snackbar;
    /**
     * @type {?}
     * @private
     */
    HeaderToolsComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    HeaderToolsComponent.prototype.router;
    /**
     * @type {?}
     * @private
     */
    HeaderToolsComponent.prototype.platform;
}
