// @verdocs/tokens/esm5/lib/services/authguard.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Inject, forwardRef } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { VerdocsAuthService } from './auth.service';
import { VerdocsStateService } from './state.service';
var VerdocsAuthGuardService = /** @class */ (function () {
    // handle redirects based on authentication state
    function VerdocsAuthGuardService(authService, stateService, route) {
        this.authService = authService;
        this.stateService = stateService;
        this.route = route;
        this.signOutRequestSubmitted = false;
        /** @type {?} */
        var snapshot = (/** @type {?} */ (this.route.snapshot));
    }
    /**
     * @return {?}
     */
    VerdocsAuthGuardService.prototype.checkLogin = /**
     * @return {?}
     */
    function () {
        if (!this.authService.isAuthenticated) {
            if (this.stateService.getLoginURL()) {
                this.signOut();
            }
        }
        this.stateService.watchAuthentication();
    };
    /**
     * @return {?}
     */
    VerdocsAuthGuardService.prototype.checkAuthentication = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var authenticated = this.authService.isAuthenticated();
        if (!authenticated) {
            this.signOut();
        }
    };
    /**
     * @param {?=} redirectUrl
     * @return {?}
     */
    VerdocsAuthGuardService.prototype.signOut = /**
     * @param {?=} redirectUrl
     * @return {?}
     */
    function (redirectUrl) {
        if (!this.signOutRequestSubmitted) {
            this.signOutRequestSubmitted = true;
            if (!this.authService.isAuthenticated()) {
                redirectUrl = window.location.href;
            }
            this.stateService.removeRCookies();
            this.stateService.removeIntercomCookie();
            /** @type {?} */
            var redirectUrlParam = (this.stateService.getLoginURL() && this.stateService.getLoginURL().includes('?') ? '&' : '?') + 'redirect_url';
            window.location.href = this.stateService.getLoginURL() + (redirectUrl ? redirectUrlParam + "=" + redirectUrl : '');
        }
    };
    VerdocsAuthGuardService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    VerdocsAuthGuardService.ctorParameters = function () { return [
        { type: VerdocsAuthService, decorators: [{ type: Inject, args: [forwardRef((/**
                         * @return {?}
                         */
                        function () { return VerdocsAuthService; })),] }] },
        { type: VerdocsStateService, decorators: [{ type: Inject, args: [forwardRef((/**
                         * @return {?}
                         */
                        function () { return VerdocsStateService; })),] }] },
        { type: ActivatedRoute }
    ]; };
    return VerdocsAuthGuardService;
}());
export { VerdocsAuthGuardService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    VerdocsAuthGuardService.prototype.signOutRequestSubmitted;
    /**
     * @type {?}
     * @private
     */
    VerdocsAuthGuardService.prototype.authService;
    /**
     * @type {?}
     * @private
     */
    VerdocsAuthGuardService.prototype.stateService;
    /**
     * @type {?}
     * @private
     */
    VerdocsAuthGuardService.prototype.route;
}
