// @verdocs/tokens/esm5/lib/services/token.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { VerdocsStateService } from './state.service';
import * as i0 from "@angular/core";
import * as i1 from "./state.service";
var VerdocsTokenObjectService = /** @class */ (function () {
    function VerdocsTokenObjectService(stateService) {
        this.stateService = stateService;
    }
    /**
     * @return {?}
     */
    VerdocsTokenObjectService.prototype.getAccessToken = /**
     * @return {?}
     */
    function () {
        return this.tokenToObject(this.stateService.getAccessToken());
    };
    /**
     * @return {?}
     */
    VerdocsTokenObjectService.prototype.getIDToken = /**
     * @return {?}
     */
    function () {
        return this.tokenToObject(this.stateService.getIDToken());
    };
    /**
     * @return {?}
     */
    VerdocsTokenObjectService.prototype.getAccountAccessToken = /**
     * @return {?}
     */
    function () {
        return this.tokenToObject(this.stateService.getAccountAccessToken());
    };
    /**
     * @return {?}
     */
    VerdocsTokenObjectService.prototype.getVerdocsAccessToken = /**
     * @return {?}
     */
    function () {
        return this.tokenToObject(this.stateService.getVerdocsAccessToken());
    };
    /**
     * @return {?}
     */
    VerdocsTokenObjectService.prototype.getProfile = /**
     * @return {?}
     */
    function () {
        return this.getIDToken() ? this.getIDToken()[VerdocsTokenObjectService.TOKEN_NAMESPACE + "/profile"] : null;
    };
    /**
     * @return {?}
     */
    VerdocsTokenObjectService.prototype.getPermissions = /**
     * @return {?}
     */
    function () {
        return this.getAccessToken() ? this.getAccessToken()[VerdocsTokenObjectService.TOKEN_NAMESPACE + "/permissions"] : null;
    };
    /**
     * @return {?}
     */
    VerdocsTokenObjectService.prototype.getAccountPermissions = /**
     * @return {?}
     */
    function () {
        return this.getAccountAccessToken() ? this.getAccountAccessToken()[VerdocsTokenObjectService.TOKEN_NAMESPACE + "/permissions"] : null;
    };
    /**
     * @return {?}
     */
    VerdocsTokenObjectService.prototype.getVerdocsPermissions = /**
     * @return {?}
     */
    function () {
        return this.getVerdocsAccessToken() ? this.getVerdocsAccessToken()[VerdocsTokenObjectService.TOKEN_NAMESPACE + "/permissions"] : null;
    };
    /**
     * @return {?}
     */
    VerdocsTokenObjectService.prototype.getPlans = /**
     * @return {?}
     */
    function () {
        return this.getAccessToken() ? this.getAccessToken()[VerdocsTokenObjectService.TOKEN_NAMESPACE + "/plans"] : this.getAccountPlans() ? this.getAccountPlans() : this.getVerdocsPlans() ? this.getVerdocsPlans() : null;
    };
    /**
     * @return {?}
     */
    VerdocsTokenObjectService.prototype.getOrganizationID = /**
     * @return {?}
     */
    function () {
        return this.getAccountAccessToken() ? this.getAccountAccessToken()[VerdocsTokenObjectService.TOKEN_NAMESPACE + "/organization_id"] : null;
    };
    /**
     * @return {?}
     */
    VerdocsTokenObjectService.prototype.getAccountPlans = /**
     * @return {?}
     */
    function () {
        return this.getAccountAccessToken() ? this.getAccountAccessToken()[VerdocsTokenObjectService.TOKEN_NAMESPACE + "/plans"] : null;
    };
    /**
     * @return {?}
     */
    VerdocsTokenObjectService.prototype.getVerdocsPlans = /**
     * @return {?}
     */
    function () {
        return this.getVerdocsAccessToken() ? this.getVerdocsAccessToken()[VerdocsTokenObjectService.TOKEN_NAMESPACE + "/plans"] : null;
    };
    /**
     * @return {?}
     */
    VerdocsTokenObjectService.prototype.getRoles = /**
     * @return {?}
     */
    function () {
        return this.getAccessToken() ? this.getAccessToken()[VerdocsTokenObjectService.TOKEN_NAMESPACE + "/roles"] : this.getAccountRoles() ? this.getAccountRoles() : this.getVerdocsRoles() ? this.getVerdocsRoles() : null;
    };
    /**
     * @return {?}
     */
    VerdocsTokenObjectService.prototype.getAccountRoles = /**
     * @return {?}
     */
    function () {
        return this.getAccountAccessToken() ? this.getAccountAccessToken()[VerdocsTokenObjectService.TOKEN_NAMESPACE + "/roles"] : null;
    };
    /**
     * @return {?}
     */
    VerdocsTokenObjectService.prototype.getVerdocsRoles = /**
     * @return {?}
     */
    function () {
        return this.getVerdocsAccessToken() ? this.getVerdocsAccessToken()[VerdocsTokenObjectService.TOKEN_NAMESPACE + "/roles"] : null;
    };
    /**
     * @return {?}
     */
    VerdocsTokenObjectService.prototype.getProfileID = /**
     * @return {?}
     */
    function () {
        return this.getAccessToken() ? this.getAccessToken()[VerdocsTokenObjectService.TOKEN_NAMESPACE + "/profile_id"] : this.getAccountAccessToken() ? this.getAccountAccessToken()[VerdocsTokenObjectService.TOKEN_NAMESPACE + "/profile_id"] : this.getVerdocsAccessToken() ? this.getVerdocsAccessToken()[VerdocsTokenObjectService.TOKEN_NAMESPACE + "/profile_id"] : null;
    };
    /**
     * @return {?}
     */
    VerdocsTokenObjectService.prototype.getUserID = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var sub = this.getAccessToken()['sub'] || this.getAccountAccessToken()['sub'];
        /** @type {?} */
        var idStart = sub.indexOf('|') + 1;
        return this.getAccessToken()['sub'].slice(idStart);
    };
    /**
     * @param {?} token
     * @return {?}
     */
    VerdocsTokenObjectService.prototype.tokenToObject = /**
     * @param {?} token
     * @return {?}
     */
    function (token) {
        if (token) {
            /** @type {?} */
            var base64Url = token.split('.')[1];
            /** @type {?} */
            var base64 = base64Url.replace('-', '+').replace('_', '/');
            return JSON.parse(window.atob(base64));
        }
        return null;
    };
    VerdocsTokenObjectService.TOKEN_NAMESPACE = 'https://verdocs.com';
    VerdocsTokenObjectService.decorators = [
        { type: Injectable, args: [{
                    providedIn: "root"
                },] }
    ];
    /** @nocollapse */
    VerdocsTokenObjectService.ctorParameters = function () { return [
        { type: VerdocsStateService }
    ]; };
    /** @nocollapse */ VerdocsTokenObjectService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function VerdocsTokenObjectService_Factory() { return new VerdocsTokenObjectService(i0.ɵɵinject(i1.VerdocsStateService)); }, token: VerdocsTokenObjectService, providedIn: "root" });
    return VerdocsTokenObjectService;
}());
export { VerdocsTokenObjectService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    VerdocsTokenObjectService.TOKEN_NAMESPACE;
    /**
     * @type {?}
     * @private
     */
    VerdocsTokenObjectService.prototype.stateService;
}
