// @verdocs/tokens/esm5/lib/tokens.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule, InjectionToken } from '@angular/core';
import { VerdocsRequestInterceptor } from './interceptors/request.interceptor';
import { VerdocsErrorInterceptor } from './interceptors/error.interceptor';
import { VerdocsResponseInterceptor } from './interceptors/response.interceptor';
import { CookieModule } from 'ngx-cookie';
import { VerdocsAuthService } from './services/auth.service';
import { VerdocsAuthGuardService } from './services/authguard.service';
import { VerdocsStateService } from './services/state.service';
import { VerdocsTokenObjectService } from './services/token.service';
import { VerdocsTokenGuardDirective } from './directives/realster-guard.directive';
import { VerifyEmailDialogComponent } from './dialogs/verify-email-dialog.component';
import { MatButtonModule, MatDialogModule } from '@angular/material';
/**
 * @record
 */
export function VTokenConfig() { }
if (false) {
    /** @type {?} */
    VTokenConfig.prototype.production;
    /** @type {?} */
    VTokenConfig.prototype.domain;
    /** @type {?} */
    VTokenConfig.prototype.cookieName;
    /** @type {?} */
    VTokenConfig.prototype.loginURL;
    /** @type {?|undefined} */
    VTokenConfig.prototype.verdocs_name;
    /** @type {?|undefined} */
    VTokenConfig.prototype.account_name;
    /** @type {?|undefined} */
    VTokenConfig.prototype.origin;
}
/** @type {?} */
export var VTokenConfigToken = new InjectionToken('VTokenConfig');
var TokensModule = /** @class */ (function () {
    function TokensModule() {
    }
    /**
     * @param {?} config
     * @return {?}
     */
    TokensModule.initInterceptor = /**
     * @param {?} config
     * @return {?}
     */
    function (config) {
        return {
            ngModule: TokensModule,
            providers: [
                {
                    provide: VTokenConfigToken,
                    useValue: config
                }
            ]
        };
    };
    TokensModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CookieModule.forRoot(),
                        MatButtonModule,
                        MatDialogModule,
                    ],
                    declarations: [
                        VerdocsTokenGuardDirective,
                        VerifyEmailDialogComponent
                    ],
                    providers: [
                        VerdocsResponseInterceptor,
                        VerdocsRequestInterceptor,
                        VerdocsErrorInterceptor,
                        VerdocsAuthService,
                        VerdocsAuthGuardService,
                        VerdocsStateService,
                        VerdocsTokenObjectService
                    ],
                    entryComponents: [
                        VerifyEmailDialogComponent
                    ],
                    exports: [
                        VerdocsTokenGuardDirective,
                        VerifyEmailDialogComponent
                    ]
                },] }
    ];
    return TokensModule;
}());
export { TokensModule };
