// @verdocs/essentials/esm5/lib/snackbar/snackbar.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Inject, PLATFORM_ID } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
import { MatSnackBar } from '@angular/material/snack-bar';
var SnackbarService = /** @class */ (function () {
    function SnackbarService(snackbar, platform) {
        this.snackbar = snackbar;
        this.platform = platform;
    }
    /**
     * @param {?=} duration
     * @param {?=} position
     * @return {?}
     */
    SnackbarService.prototype.initConfig = /**
     * @param {?=} duration
     * @param {?=} position
     * @return {?}
     */
    function (duration, position) {
        if ((isPlatformBrowser(this.platform))) {
            /** @type {?} */
            var snackbarConfig = void 0;
            if (window && window.innerWidth >= 920) {
                snackbarConfig = {
                    verticalPosition: position ? position : 'bottom',
                    horizontalPosition: 'left',
                    duration: duration ? duration : 5000
                };
            }
            else {
                snackbarConfig = {
                    verticalPosition: position ? position : 'top',
                    duration: duration ? duration : 5000
                };
            }
            return snackbarConfig;
        }
    };
    /**
     * @param {?} message
     * @param {?=} action
     * @param {?=} config
     * @return {?}
     */
    SnackbarService.prototype.open = /**
     * @param {?} message
     * @param {?=} action
     * @param {?=} config
     * @return {?}
     */
    function (message, action, config) {
        return this.snackbar.open(message, action ? action : undefined, config ? config : this.initConfig());
    };
    /**
     * @return {?}
     */
    SnackbarService.prototype.dismiss = /**
     * @return {?}
     */
    function () {
        this.snackbar.dismiss();
    };
    SnackbarService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    SnackbarService.ctorParameters = function () { return [
        { type: MatSnackBar },
        { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
    ]; };
    return SnackbarService;
}());
export { SnackbarService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    SnackbarService.prototype.snackbar;
    /**
     * @type {?}
     * @private
     */
    SnackbarService.prototype.platform;
}
