// @verdocs/tokens/esm2015/lib/directives/realster-guard.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Output, Input, EventEmitter, Inject, forwardRef, PLATFORM_ID } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
import { MatDialog } from '@angular/material';
import { Subscription } from 'rxjs';
import { VerifyEmailDialogComponent } from '../dialogs/verify-email-dialog.component';
import { VerdocsAuthGuardService } from '../services/authguard.service';
import { VerdocsStateService } from '../services/state.service';
import { VerdocsAuthService } from '../services/auth.service';
export class VerdocsTokenGuardDirective {
    /**
     * @param {?} dialog
     * @param {?} verdocsAuthGuard
     * @param {?} verdocsStateService
     * @param {?} verdocsAuthService
     * @param {?} platform
     */
    constructor(dialog, verdocsAuthGuard, verdocsStateService, verdocsAuthService, platform) {
        this.dialog = dialog;
        this.verdocsAuthGuard = verdocsAuthGuard;
        this.verdocsStateService = verdocsStateService;
        this.verdocsAuthService = verdocsAuthService;
        this.platform = platform;
        this.verifyEmailDialogIsOpen = false;
        this.idTokenSubscription = new Subscription();
        this.resendEmailVerificationEvent = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.idTokenSubscription = this.verdocsStateService.decodedIdTokenSubject.subscribe((/**
         * @param {?} decodedIdToken
         * @return {?}
         */
        decodedIdToken => {
            this.emailAddress = decodedIdToken.email;
        }));
        setInterval((/**
         * @return {?}
         */
        () => {
            if (this.enforceAuthentication) {
                this.verdocsAuthGuard.checkAuthentication();
            }
            if (this.verdocsAuthService.isAuthenticated()) {
                this.checkEmailVerification();
            }
        }), 5000);
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.idTokenSubscription.unsubscribe();
    }
    /**
     * @private
     * @return {?}
     */
    checkEmailVerification() {
        /** @type {?} */
        const emailVerified = this.verdocsStateService.getEmailVerification();
        if (!emailVerified && !this.verifyEmailDialogIsOpen && isPlatformBrowser(this.platform)) {
            this.verifyEmailDialogIsOpen = true;
            /** @type {?} */
            const email_data = {
                resendEmailVerification: (/**
                 * @return {?}
                 */
                () => this.resendEmailVerification()),
                emailAddress: this.emailAddress
            };
            /** @type {?} */
            const verifyEmailDialog = this.dialog.open(VerifyEmailDialogComponent, {
                data: email_data,
                disableClose: true
            });
            verifyEmailDialog.afterClosed().subscribe((/**
             * @return {?}
             */
            () => {
                this.verifyEmailDialogIsOpen = false;
                location.reload();
            }));
        }
    }
    /**
     * @return {?}
     */
    resendEmailVerification() {
        this.resendEmailVerificationEvent.emit();
    }
}
VerdocsTokenGuardDirective.decorators = [
    { type: Directive, args: [{
                selector: '[verdocs-guard]'
            },] }
];
/** @nocollapse */
VerdocsTokenGuardDirective.ctorParameters = () => [
    { type: MatDialog },
    { type: VerdocsAuthGuardService, decorators: [{ type: Inject, args: [forwardRef((/**
                     * @return {?}
                     */
                    () => VerdocsAuthGuardService)),] }] },
    { type: VerdocsStateService, decorators: [{ type: Inject, args: [forwardRef((/**
                     * @return {?}
                     */
                    () => VerdocsStateService)),] }] },
    { type: VerdocsAuthService, decorators: [{ type: Inject, args: [forwardRef((/**
                     * @return {?}
                     */
                    () => VerdocsAuthService)),] }] },
    { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
];
VerdocsTokenGuardDirective.propDecorators = {
    resendEmailVerificationEvent: [{ type: Output }],
    enforceAuthentication: [{ type: Input }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    VerdocsTokenGuardDirective.prototype.verifyEmailDialogIsOpen;
    /**
     * @type {?}
     * @private
     */
    VerdocsTokenGuardDirective.prototype.emailAddress;
    /**
     * @type {?}
     * @private
     */
    VerdocsTokenGuardDirective.prototype.idTokenSubscription;
    /**
     * @type {?}
     * @private
     */
    VerdocsTokenGuardDirective.prototype.resendEmailVerificationEvent;
    /**
     * @type {?}
     * @private
     */
    VerdocsTokenGuardDirective.prototype.enforceAuthentication;
    /**
     * @type {?}
     * @private
     */
    VerdocsTokenGuardDirective.prototype.dialog;
    /**
     * @type {?}
     * @private
     */
    VerdocsTokenGuardDirective.prototype.verdocsAuthGuard;
    /**
     * @type {?}
     * @private
     */
    VerdocsTokenGuardDirective.prototype.verdocsStateService;
    /**
     * @type {?}
     * @private
     */
    VerdocsTokenGuardDirective.prototype.verdocsAuthService;
    /**
     * @type {?}
     * @private
     */
    VerdocsTokenGuardDirective.prototype.platform;
}
