// @verdocs/sdk/esm2015/lib/angular-api.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule, InjectionToken } from '@angular/core';
import { TokensModule } from '@verdocs/tokens';
import { ValidatorService } from './validators/validator.service';
import { AuthGuardService } from './guards/auth-guard.service';
import { DocumentsService } from './documents/document.service';
import { EnvelopRemindersService } from './documents/envelope-reminders.service';
import { RecipientService } from './recipients/recipients.service';
import { BuilderService } from './templates/builder.service';
import { TemplatesRemindersService } from './templates/template-reminders.service';
import { TemplatesGuardService } from './templates/templates.guard';
import { TemplatesService } from './templates/templates.service';
/**
 * @record
 */
export function VSDKConfig() { }
if (false) {
    /** @type {?} */
    VSDKConfig.prototype.production;
    /** @type {?} */
    VSDKConfig.prototype.domain;
    /** @type {?} */
    VSDKConfig.prototype.rForm_backend_url;
    /** @type {?} */
    VSDKConfig.prototype.rForm_frontend_url;
    /** @type {?} */
    VSDKConfig.prototype.rSecure_frontend_url;
    /** @type {?} */
    VSDKConfig.prototype.rSecure_backend_url;
    /** @type {?} */
    VSDKConfig.prototype.rAccount_frontend_url;
    /** @type {?} */
    VSDKConfig.prototype.rAccount_backend_url;
    /** @type {?} */
    VSDKConfig.prototype.rNotification_backend_url;
    /** @type {?} */
    VSDKConfig.prototype.loginURL;
    /** @type {?} */
    VSDKConfig.prototype.origin;
    /** @type {?} */
    VSDKConfig.prototype.plans;
}
/** @type {?} */
export const VSDKConfigToken = new InjectionToken('VSDKConfig');
export class AngularApiModule {
    /**
     * @param {?} config
     * @return {?}
     */
    static initSDK(config) {
        return {
            ngModule: AngularApiModule,
            imports: [
                TokensModule.initInterceptor({
                    cookieName: '',
                    domain: config.domain,
                    origin: config.origin,
                    verdocs_name: 'rForm',
                    account_name: 'rAccount',
                    loginURL: config.loginURL,
                    production: config.production
                })
            ],
            providers: [
                {
                    provide: VSDKConfigToken,
                    useValue: config
                },
                ValidatorService,
                AuthGuardService,
                DocumentsService,
                EnvelopRemindersService,
                RecipientService,
                BuilderService,
                TemplatesGuardService,
                TemplatesRemindersService,
                TemplatesService
            ]
        };
    }
}
AngularApiModule.decorators = [
    { type: NgModule, args: [{
                imports: [],
                providers: []
            },] }
];
