// @verdocs/sdk/esm5/lib/validators/validator.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable, Injector } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { regParse } from '../functions/regex';
import { VSDKConfigToken } from '../angular-api.module';
var ValidatorService = /** @class */ (function () {
    function ValidatorService(http, injector) {
        this.http = http;
        this.injector = injector;
        this.validators = [];
        this.validatorsObject = {};
        this._config_token = this.injector.get(VSDKConfigToken);
        this._rForm_backend_url = this._config_token.rForm_backend_url;
    }
    /**
     * @return {?}
     */
    ValidatorService.prototype.getValidatorsArray = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.validators && this.validators.length > 0) {
            return this.validators;
        }
        else {
            return this.http.get(this._rForm_backend_url + '/validators').toPromise().then((/**
             * @param {?} response
             * @return {?}
             */
            function (response) {
                _this.validators = response;
                return (/** @type {?} */ (_this.validators));
            }), (/**
             * @param {?} err
             * @return {?}
             */
            function (err) {
                return null;
            }));
        }
    };
    /**
     * @return {?}
     */
    ValidatorService.prototype.getValidatorsObject = /**
     * @return {?}
     */
    function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var i, name_1;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.validatorsObject = {};
                        return [4 /*yield*/, this.getValidatorsArray()];
                    case 1:
                        _a.sent();
                        for (i = 0; i < this.validators['length']; i++) {
                            name_1 = this.validators[i]['name'];
                            this.validatorsObject[name_1] = regParse(this.validators[i]['regex']);
                        }
                        return [2 /*return*/, this.validatorsObject];
                }
            });
        });
    };
    ValidatorService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ValidatorService.ctorParameters = function () { return [
        { type: HttpClient },
        { type: Injector }
    ]; };
    return ValidatorService;
}());
export { ValidatorService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ValidatorService.prototype._config_token;
    /**
     * @type {?}
     * @private
     */
    ValidatorService.prototype._rForm_backend_url;
    /**
     * @type {?}
     * @private
     */
    ValidatorService.prototype.validators;
    /**
     * @type {?}
     * @private
     */
    ValidatorService.prototype.validatorsObject;
    /**
     * @type {?}
     * @private
     */
    ValidatorService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    ValidatorService.prototype.injector;
}
