// @verdocs/tokens/esm2015/lib/interceptors/error.interceptor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector, Inject, PLATFORM_ID } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
import { HttpErrorResponse } from '@angular/common/http';
import { catchError } from 'rxjs/operators';
import { VerdocsAuthGuardService } from '../services/authguard.service';
import { VerdocsStateService } from '../services/state.service';
import { throwError } from 'rxjs';
export class VerdocsErrorInterceptor {
    /**
     * @param {?} injector
     * @param {?} platform
     */
    constructor(injector, platform) {
        this.injector = injector;
        this.platform = platform;
    }
    /**
     * @param {?} request
     * @param {?} next
     * @return {?}
     */
    intercept(request, next) {
        return next.handle(request).pipe(catchError((/**
         * @param {?} err
         * @return {?}
         */
        err => {
            if (err instanceof HttpErrorResponse) {
                if (err['status'] === 403) {
                    this.stateService = this.injector.get(VerdocsStateService);
                    /** @type {?} */
                    let redirectUrl;
                    if (isPlatformBrowser(this.platform) && err['error'] && err['error']['reason'] === 'expired') {
                        redirectUrl = window.location.href;
                    }
                    this.authGuardService = this.injector.get(VerdocsAuthGuardService);
                    this.authGuardService.signOut(redirectUrl);
                }
                return throwError(err);
            }
        })));
    }
}
VerdocsErrorInterceptor.decorators = [
    { type: Injectable }
];
/** @nocollapse */
VerdocsErrorInterceptor.ctorParameters = () => [
    { type: Injector },
    { type: undefined, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    VerdocsErrorInterceptor.prototype.authGuardService;
    /**
     * @type {?}
     * @private
     */
    VerdocsErrorInterceptor.prototype.stateService;
    /**
     * @type {?}
     * @private
     */
    VerdocsErrorInterceptor.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    VerdocsErrorInterceptor.prototype.platform;
}
