// @verdocs/tokens/esm2015/lib/interceptors/request.interceptor.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector } from '@angular/core';
import { VerdocsStateService } from '../services/state.service';
export class VerdocsRequestInterceptor {
    /**
     * @param {?} injector
     */
    constructor(injector) {
        this.injector = injector;
    }
    /**
     * @param {?} request
     * @param {?} next
     * @return {?}
     */
    intercept(request, next) {
        this.stateService = this.injector.get(VerdocsStateService);
        /** @type {?} */
        const accessToken = this.stateService.getAccessToken();
        /** @type {?} */
        const idToken = this.stateService.getIDToken();
        /** @type {?} */
        const signerToken = this.stateService.getOtherCookieObject('signer_token');
        /** @type {?} */
        const setHeaders = {
            Pragma: 'no-cache',
            'Cache-Control': 'no-cache',
            Expires: 'Sat, 01 Jan 2000 00:00:00 GMT',
            'If-Modified-Since': '0'
        };
        if (request.url.startsWith('assets/')) {
            return next.handle(request);
        }
        if (!request.headers.get('Authorization') && accessToken) {
            setHeaders['Authorization'] = `Bearer ${accessToken}`;
        }
        if (idToken) {
            setHeaders['Authentication'] = `Bearer ${idToken}`;
        }
        if (signerToken) {
            setHeaders['Signer'] = `Bearer ${signerToken}`;
        }
        request = request.clone({ setHeaders });
        return next.handle(request);
    }
}
VerdocsRequestInterceptor.decorators = [
    { type: Injectable }
];
/** @nocollapse */
VerdocsRequestInterceptor.ctorParameters = () => [
    { type: Injector }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    VerdocsRequestInterceptor.prototype.stateService;
    /**
     * @type {?}
     * @private
     */
    VerdocsRequestInterceptor.prototype.injector;
}
