// @verdocs/essentials/esm2015/lib/builder/builder-settings.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ReplaySubject } from 'rxjs';
export class BuilderService {
    constructor() {
        this.sideNavStateSubject = new ReplaySubject();
        this.recipientSideNavStateSubject = new ReplaySubject(1, 100);
        this.initiateAutofocus = new ReplaySubject();
        this.recipientAutoFocus = new ReplaySubject();
        this.deleteRoleSubject = new ReplaySubject(1, 100);
        this.hasActiveRoleSubject = new ReplaySubject();
    }
    // constructor() {}
    /**
     * @return {?}
     */
    toggleSideNav() {
        this.sideNavState = !this.sideNavState;
        this.sideNavStateSubject.next(this.sideNavState);
        if (this.sideNavState === false) {
            this.initiateAutofocus.next(false);
        }
    }
    /**
     * @return {?}
     */
    toggleRecipientSideNav() {
        this.recipientSideNavState = !this.recipientSideNavState;
        this.recipientSideNavStateSubject.next(this.recipientSideNavState);
        if (this.recipientSideNavState === false) {
            this.recipientAutoFocus.next(false);
        }
    }
    /**
     * @return {?}
     */
    openSideNav() {
        this.sideNavState = true;
        this.sideNavStateSubject.next(this.sideNavState);
    }
    /**
     * @return {?}
     */
    openRecipientSideNav() {
        this.recipientSideNavState = true;
        this.recipientSideNavStateSubject.next(this.recipientSideNavState);
    }
    /**
     * @return {?}
     */
    closeSideNav() {
        this.sideNavState = false;
        this.sideNavStateSubject.next(this.sideNavState);
        this.initiateAutofocus.next(false);
    }
    /**
     * @return {?}
     */
    closeRecipientSideNav() {
        this.recipientSideNavState = false;
        this.recipientSideNavStateSubject.next(this.recipientSideNavState);
        this.recipientAutoFocus.next(false);
    }
    /**
     * @param {?} role
     * @return {?}
     */
    setActiveRole(role) {
        this.activeRole = role;
        this.hasActiveRoleSubject.next(true);
    }
    /**
     * @return {?}
     */
    deleteRole() {
        if (!!this.activeRole) {
            this.deleteRoleSubject.next(this.activeRole);
            this.activeRole = null;
            this.hasActiveRoleSubject.next(false);
        }
    }
    /**
     * @return {?}
     */
    getSideNavState() {
        return this.sideNavState;
    }
    /**
     * @return {?}
     */
    getRecipientSideNavState() {
        return this.recipientSideNavState;
    }
}
BuilderService.decorators = [
    { type: Injectable }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    BuilderService.prototype.sideNavState;
    /**
     * @type {?}
     * @private
     */
    BuilderService.prototype.recipientSideNavState;
    /**
     * @type {?}
     * @private
     */
    BuilderService.prototype.activeRole;
    /** @type {?} */
    BuilderService.prototype.sideNavStateSubject;
    /** @type {?} */
    BuilderService.prototype.recipientSideNavStateSubject;
    /** @type {?} */
    BuilderService.prototype.initiateAutofocus;
    /** @type {?} */
    BuilderService.prototype.recipientAutoFocus;
    /** @type {?} */
    BuilderService.prototype.deleteRoleSubject;
    /** @type {?} */
    BuilderService.prototype.hasActiveRoleSubject;
}
