// @verdocs/header/esm5/lib/template/source-file/source-file.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { VerdocsHeaderService } from '../../services/header.service';
var TemplateSourceComponent = /** @class */ (function () {
    function TemplateSourceComponent(templatesService) {
        this.templatesService = templatesService;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    TemplateSourceComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes['template']) {
        }
    };
    /**
     * @return {?}
     */
    TemplateSourceComponent.prototype.downloadTemplateDocument = /**
     * @return {?}
     */
    function () {
        this.templatesService.downloadTemplateDocument(this.template);
    };
    TemplateSourceComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-template-source',
                    template: "<div class=\"template-source__container\" *ngIf=\"template\">\n  <div class=\"template-source__download\" (click)=\"downloadTemplateDocument()\">\n    <mat-icon class=\"template-source__download-icon\">cloud_download</mat-icon>\n    <div class=\"template-source__download-text\">{{template.template_documents[0].name}}</div>\n  </div>\n</div>",
                    styles: [".template-source__download{display:flex;align-items:center}.template-source__download:hover{cursor:pointer}.template-source__download-icon{margin-right:4px;color:rgba(0,0,0,.54)}.template-source__download-text{color:rgba(0,0,0,.87)}"]
                }] }
    ];
    /** @nocollapse */
    TemplateSourceComponent.ctorParameters = function () { return [
        { type: VerdocsHeaderService }
    ]; };
    TemplateSourceComponent.propDecorators = {
        template: [{ type: Input }]
    };
    return TemplateSourceComponent;
}());
export { TemplateSourceComponent };
if (false) {
    /** @type {?} */
    TemplateSourceComponent.prototype.template;
    /**
     * @type {?}
     * @private
     */
    TemplateSourceComponent.prototype.templatesService;
}
