// @verdocs/header/esm5/lib/template/visibility/visibility-editor.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { VerdocsHeaderService } from '../../services/header.service';
import { TemplatesGuardService } from '../../services/templates.guard';
import { TemplateActions } from '../../models/verdocs.model';
var VisibilityEditorComponent = /** @class */ (function () {
    function VisibilityEditorComponent(templateGuard, verdocsHeaderService) {
        this.templateGuard = templateGuard;
        this.verdocsHeaderService = verdocsHeaderService;
        this.updateVisibility = false;
        this.isSaving = false;
        this.isEditMode = false;
        this._canShareWithOrg = null;
        this._canMakeItPrivate = null;
        this._canShareWithPublic = null;
    }
    /**
     * @return {?}
     */
    VisibilityEditorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.template_visibility = this.getTemplateVisibility();
    };
    /**
     * @return {?}
     */
    VisibilityEditorComponent.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        this.template_visibility = this.getTemplateVisibility();
    };
    /**
     * @return {?}
     */
    VisibilityEditorComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.template_visibility = undefined;
    };
    /**
     * @return {?}
     */
    VisibilityEditorComponent.prototype.getTemplateVisibility = /**
     * @return {?}
     */
    function () {
        if (this.visibility) {
            if (this.visibility.is_personal) {
                if (this.visibility.is_public) {
                    return 'personal & public';
                }
                else {
                    return 'personal';
                }
            }
            else {
                if (this.visibility.is_public) {
                    return 'public & org';
                }
                else {
                    return 'org';
                }
            }
        }
        return undefined;
    };
    /**
     * @return {?}
     */
    VisibilityEditorComponent.prototype.setUpdateVisibility = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var original_visibility = this.getTemplateVisibility();
        this.updateVisibility = original_visibility !== this.template_visibility;
    };
    /**
     * @return {?}
     */
    VisibilityEditorComponent.prototype.saveChanges = /**
     * @return {?}
     */
    function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var body, template;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.isSaving = true;
                        body = {};
                        if (this.updateVisibility && this.template) {
                            switch (this.template_visibility) {
                                case 'personal':
                                    body['is_personal'] = true;
                                    body['is_public'] = false;
                                    break;
                                case 'org':
                                    body['is_personal'] = false;
                                    body['is_public'] = false;
                                    break;
                                case 'public & org':
                                    body['is_personal'] = false;
                                    body['is_public'] = true;
                                    break;
                                case 'personal & public':
                                    body['is_personal'] = true;
                                    body['is_public'] = true;
                                default:
                                    break;
                            }
                        }
                        return [4 /*yield*/, this.verdocsHeaderService.updateTemplate(this.template.id, body)];
                    case 1:
                        template = _a.sent();
                        this.updateVisibility = false;
                        this.isSaving = false;
                        return [2 /*return*/];
                }
            });
        });
    };
    Object.defineProperty(VisibilityEditorComponent.prototype, "canShareWithPublic", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.template && typeof (this._canShareWithPublic) !== 'boolean') {
                this._canShareWithPublic = (this.templateGuard.canPerformAction(TemplateActions.CHANGE_VISIBILITY_PUBLIC, this.template))['canPerform'];
            }
            return this._canShareWithPublic;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(VisibilityEditorComponent.prototype, "canChangeVisibility", {
        get: /**
         * @return {?}
         */
        function () {
            return this.canMakeItPrivate || this.canShareWithOrg || this.canShareWithPublic;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(VisibilityEditorComponent.prototype, "canShareWithOrg", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.template && typeof (this._canShareWithOrg) !== 'boolean') {
                this._canShareWithOrg = (this.templateGuard.canPerformAction(TemplateActions.CHANGE_VISIBILITY_ORG, this.template))['canPerform'];
            }
            return this._canShareWithOrg;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(VisibilityEditorComponent.prototype, "canMakeItPrivate", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.template && typeof (this._canMakeItPrivate) !== 'boolean') {
                this._canMakeItPrivate =
                    (this.templateGuard.canPerformAction(TemplateActions.CHANGE_VISIBILITY_PERSONAL, this.template))['canPerform'];
            }
            return this._canMakeItPrivate;
        },
        enumerable: true,
        configurable: true
    });
    VisibilityEditorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-preview-visibility',
                    template: "<div class=\"template-visiblity__container\" *ngIf=\"visibility\">\n  <ng-template #readView>\n    <div class=\"template-visibility__read\">\n      <div class=\"template-visibility__type\" *ngIf=\"visibility.is_personal && !visibility.is_public\">\n        <mat-icon class=\"template-visibility__type-icon\">lock</mat-icon>\n        <div class=\"template-visibility__type-text\">Personal</div>\n      </div>\n      <div class=\"template-visibility__type\" *ngIf=\"visibility.is_public\">\n        <mat-icon class=\"template-visibility__type-icon\">public</mat-icon>\n        <div class=\"template-visibility__type-text\">Public</div>\n      </div>\n      <div class=\"template-visibility__type\" *ngIf=\"!visibility.is_personal\">\n        <mat-icon class=\"template-visibility__type-icon\">group</mat-icon>\n        <div class=\"template-visibility__type-text\">Shared</div>\n      </div>\n    </div>\n  </ng-template>\n  <ng-template #editView>\n    <div class=\"template-visibility__edit\">\n      <mat-radio-group [(ngModel)]=\"template_visibility\" [disabled]=\"!canChangeVisibility\"\n        matTooltip=\"{{ !canChangeVisibility? 'You don\\'t have the permission to change the visibility of this document':''}}\">\n        <mat-radio-button class=\"template-visibility__radio-button\" [value]=\"'personal'\" (click)=\"setUpdateVisibility()\"\n          [disabled]=\"!canMakeItPrivate\"\n          matTooltip=\"{{ !canMakeItPrivate? 'You don\\'t have the permission to select this option':''}}\">\n          <div class=\"template-visibility__radio-text\">\n            Private\n          </div>\n          <div class=\"template-visibility__radio-description\">\n            Only you can see this document.\n          </div>\n        </mat-radio-button>\n        <mat-radio-button class=\"template-visibility__radio-button\" [value]=\"'org'\" (click)=\"setUpdateVisibility()\"\n          [disabled]=\"!canShareWithOrg\"\n          matTooltip=\"{{ !canShareWithOrg? 'You don\\'t have the permission to select this option':''}}\">\n          <div class=\"template-visibility__radio-text\">\n            Shared to your organization\n          </div>\n          <div class=\"template-visibility__radio-description\">\n            Give access to your organization to share this document.\n          </div>\n        </mat-radio-button>\n        <mat-radio-button class=\"template-visibility__radio-button\" [value]=\"'public & org'\" (click)=\"setUpdateVisibility()\"\n          [disabled]=\"!canShareWithOrg || !canShareWithPublic\"\n          matTooltip=\"{{ !canShareWithOrg || !canShareWithPublic ? 'You don\\'t have the permission to select this option':''}}\">\n          <div class=\"template-visibility__radio-text\">\n            Shared to your organization, and public on the web\n          </div>\n          <div class=\"template-visibility__radio-description\">\n            Give access to your organization to share, and anyone on the web can find and access this document.\n          </div>\n        </mat-radio-button>\n        <mat-radio-button class=\"template-visibility__radio-button\" [value]=\"'personal & public'\"\n          [disabled]=\"!canShareWithPublic\" (click)=\"setUpdateVisibility()\"\n          matTooltip=\"{{ !canShareWithPublic? 'You don\\'t have the permission to select this option':''}}\">\n          <div class=\"template-visibility__radio-text\">\n            Public on the web\n          </div>\n          <div class=\"template-visibility__radio-description\">\n            Anyone on the web can find and access this document.\n          </div>\n        </mat-radio-button>\n      </mat-radio-group>\n    </div>\n  </ng-template>\n  <ng-container *ngTemplateOutlet=\"isEditMode ? editView : readView\"></ng-container>\n</div>",
                    styles: [".template-visibility__read,.template-visibility__type{display:flex;align-items:center}.template-visibility__type-icon{margin-right:4px;color:rgba(0,0,0,.54)}.template-visibility__type-text{color:rgba(0,0,0,.87);margin-right:8px}"]
                }] }
    ];
    /** @nocollapse */
    VisibilityEditorComponent.ctorParameters = function () { return [
        { type: TemplatesGuardService },
        { type: VerdocsHeaderService }
    ]; };
    VisibilityEditorComponent.propDecorators = {
        visibility: [{ type: Input }],
        isEditMode: [{ type: Input }],
        template: [{ type: Input }]
    };
    return VisibilityEditorComponent;
}());
export { VisibilityEditorComponent };
if (false) {
    /** @type {?} */
    VisibilityEditorComponent.prototype.template_visibility;
    /** @type {?} */
    VisibilityEditorComponent.prototype.updateVisibility;
    /** @type {?} */
    VisibilityEditorComponent.prototype.isSaving;
    /** @type {?} */
    VisibilityEditorComponent.prototype.visibility;
    /** @type {?} */
    VisibilityEditorComponent.prototype.isEditMode;
    /** @type {?} */
    VisibilityEditorComponent.prototype.template;
    /**
     * @type {?}
     * @private
     */
    VisibilityEditorComponent.prototype._canShareWithOrg;
    /**
     * @type {?}
     * @private
     */
    VisibilityEditorComponent.prototype._canMakeItPrivate;
    /**
     * @type {?}
     * @private
     */
    VisibilityEditorComponent.prototype._canShareWithPublic;
    /**
     * @type {?}
     * @private
     */
    VisibilityEditorComponent.prototype.templateGuard;
    /**
     * @type {?}
     * @private
     */
    VisibilityEditorComponent.prototype.verdocsHeaderService;
}
